{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cfd49a4b-ae8f-4279-a940-d72726d29424",
   "metadata": {},
   "source": [
    "# Analysis of mouse empirical dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32bddeff-6a85-41c2-bae5-1f7160f9e179",
   "metadata": {},
   "source": [
    "TODO add here a brief description of your project or analysis. Explain the purpose, goals, or context of the notebook here.\n",
    "\n",
    "## Table of Contents\n",
    "- [1. Chapter 1 - Introduction](#chapter-1-introduction)\n",
    "- [2. Chapter 2 - R packages](#chapter-2-R-packages)\n",
    "- [3. Chapter 3 - Input specification](#chapter-3-Input-specification)\n",
    "- [4. Chapter 4 - RF distance summaries](#chapter-4-RF-distance-summaries)\n",
    "- [4. Chapter 5 - Running times](#chapter-5-Running-times)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8e366a2-65de-425e-9cf8-6a44c1d6fcbc",
   "metadata": {},
   "source": [
    "<a id=\"chapter-1-introduction\"></a>\n",
    "\n",
    "## Chapter 1 - Introduction\n",
    "\n",
    "TODO add description of notebook and how to use it.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c64433ad-f6bc-4891-a3cf-9165d3538d9e",
   "metadata": {},
   "source": [
    "<a id=\"chapter-2-R-packages\"></a>\n",
    "\n",
    "## Chapter 2 - R packages\n",
    "\n",
    "Here is the list of packages needed for this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a1a8e10a-6755-4ca8-8818-f267f292639b",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"phangorn\")\n",
    "library(\"rlist\")\n",
    "library(\"reshape2\")\n",
    "library(\"ggplot2\")\n",
    "library(\"geiger\")\n",
    "library(\"adephylo\")\n",
    "library(\"phytools\")\n",
    "library(\"hrbrthemes\")\n",
    "#library(\"ggtree\")\n",
    "#library(\"pheatmap\")\n",
    "library(\"gplots\")\n",
    "library(\"ape\")\n",
    "library(\"plotrix\")\n",
    "library(\"plyr\")\n",
    "library(\"dplyr\")\n",
    "library(\"rstatix\")\n",
    "library(\"grid\")\n",
    "library(\"gridExtra\")\n",
    "library(\"ggforce\")\n",
    "library(\"viridis\")\n",
    "library(\"tidyr\")\n",
    "library(\"stringr\")\n",
    "library(\"Biostrings\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de9ad2d2-464e-4838-86ff-1f4cb300b6bf",
   "metadata": {},
   "source": [
    "<a id=\"chapter-3-Input-specification\"></a>\n",
    "\n",
    "## Chapter 3 - Input specification\n",
    "\n",
    "Change the parent_of_all_dir variable to give the location of the input. All input files are expected to be under it. Otherwise change the single input location to you liking.<br>\n",
    "All input files or directory paths are listed here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3c6c3d6d-1571-44ab-9b01-8ca5097e22b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the location of the directory that contains all inputs for analyses \n",
    "parent_of_all_dir = \"/users/cn/avignoli/1_paralogs/MOUSE_above_40_entries_clean_run/rerun_with_struct_concat/\"\n",
    "\n",
    "# The paths to all PFAM families directories\n",
    "pfam_dirs_paths = paste0(parent_of_all_dir, \"PF*\")\n",
    "\n",
    "# here we get all PFAM directories paths into an iterable used later on in the analysis\n",
    "pfam_dirs = Sys.glob(pfam_dirs_paths)\n",
    "pfam_dirs = pfam_dirs[file.info(pfam_dirs)$isdir]\n",
    "\n",
    "# each run of main.nf has one output directory.\n",
    "run_dir = Sys.glob(parent_of_all_dir)\n",
    "run_dir = run_dir[file.info(run_dir)$isdir]\n",
    "\n",
    "# reference species keyword.\n",
    "ref_species = \"MOUSE\"\n",
    "\n",
    "# Create the 'figures' directory if it doesn't exist. Created where this notebook is executed.\n",
    "if (!dir.exists(\"figures\")) {\n",
    "    dir.create(\"figures\")\n",
    "}\n",
    "\n",
    "# Create the 'tables' directory if it doesn't exist. Created where this notebook is executed.\n",
    "if (!dir.exists(\"tables\")) {\n",
    "    dir.create(\"tables\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f0a3763-0bec-4c8c-954d-4188e860d17e",
   "metadata": {},
   "source": [
    "<a id=\"chapter-4-RF-distance-summaries\"></a>\n",
    "\n",
    "## Chapter 4 - RF distance summaries\n",
    "\n",
    "TODO add description "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "32acab53-5f25-4e79-9893-4eeef906efd4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"PF00004\"\n",
      "[1] \"PF00019\"\n",
      "[1] \"PF00022\"\n",
      "[1] \"PF00025\"\n",
      "[1] \"PF00041\"\n",
      "[1] \"PF00059\"\n",
      "[1] \"PF00061\"\n",
      "[1] \"PF00092\"\n",
      "[1] \"PF00102\"\n",
      "[1] \"PF00104\"\n",
      "[1] \"PF00106\"\n",
      "[1] \"PF00122\"\n",
      "[1] \"PF00134\"\n",
      "[1] \"PF00167\"\n",
      "[1] \"PF00168\"\n",
      "[1] \"PF00178\"\n",
      "[1] \"PF00179\"\n",
      "[1] \"PF00250\"\n",
      "[1] \"PF00307\"\n",
      "[1] \"PF00373\"\n",
      "[1] \"PF00531\"\n",
      "[1] \"PF00566\"\n",
      "[1] \"PF00595\"\n",
      "[1] \"PF00615\"\n",
      "[1] \"PF00620\"\n",
      "[1] \"PF00621\"\n",
      "[1] \"PF00622\"\n",
      "[1] \"PF00640\"\n",
      "[1] \"PF00777\"\n",
      "[1] \"PF00782\"\n",
      "[1] \"PF00787\"\n",
      "[1] \"PF00788\"\n",
      "[1] \"PF00822\"\n",
      "[1] \"PF01392\"\n",
      "[1] \"PF01412\"\n",
      "[1] \"PF01421\"\n",
      "[1] \"PF02214\"\n",
      "[1] \"PF02931\"\n",
      "[1] \"PF07679\"\n",
      "[1] \"PF08205\"\n",
      "[1] \"PF12796\"\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Compute RF distances between supermatrix and supertree trees and bigtree ML tree without Mouse sequences.\n",
    "#  supermatrix and supertree are also 6 units becouse Mouse MSAs has been excluded from the generation of the units.\n",
    "\n",
    "# create the dataframe that will hold RF values for each tree comparison. \n",
    "# there are (6) reference trees for a family in this case. because there are (7-1) (Mouse) species and each one of them might yeld a diffrent paralog tree.\n",
    "# for this reason the unit_sample trees are compared to all (6) reference trees.  \n",
    "# all columns (except the rf_val column) will work as identifier for that comparison if put toghether.\n",
    "df_all_rf = data.frame(\n",
    "    method    = c(),\n",
    "    family    = c(),\n",
    "    species   = c(),\n",
    "    unit      = c(),\n",
    "    sample    = c(),\n",
    "    rf_val    = c()\n",
    ")\n",
    "\n",
    "# create a dataframe with miscellaneuos informations about the gene family\n",
    "df_fam_info = data.frame(\n",
    "    family_name  = c(),\n",
    "    avg_len_seq  = c(),\n",
    "    sd_len_seq   = c(),\n",
    "    num_units    = c(),\n",
    "    num_paralogs = c(),\n",
    "    num_col_ML   = c(),\n",
    "    organisms    = c()\n",
    ")\n",
    "\n",
    "\n",
    "# the directory prefix where all analazyd files are. this is created by main.nf pipeline in empirical data mode. \n",
    "seq_prefix = \"results_without_mouse_10_samples/tcoffee\"\n",
    "\n",
    "# iterate over all family specific output directories\n",
    "for (family in pfam_dirs) {\n",
    "\n",
    "    print(basename(family))\n",
    "\n",
    "    # load the BIgtree MSA sequence names mapping scheme, aka seq 1 would be genename_BOVIN_1.\n",
    "    codefile_full_aln_path = file.path(family, seq_prefix, \"no_ref_full_aln.code_name\")\n",
    "    codefile_full_aln      = read.table(codefile_full_aln_path)\n",
    "\n",
    "    # load only the mouse MSA sequence names mapping scheme.\n",
    "    codefile_path = file.path(family, seq_prefix, \"file.code_name\")\n",
    "    codefile      = read.table(codefile_path)\n",
    "\n",
    "    # get the codename/acronym for the organism/species in the gene family. \n",
    "    orgs_path = file.path(family, seq_prefix, paste0(basename(family), \".orthologs_org_ids_to_concatenate\"))\n",
    "    orgs      = read.table(orgs_path)\n",
    "\n",
    "    # read the paralog gene names \n",
    "    paralogs_file_path = file.path(family, paste0(basename(family), \".intersecting_genes\"))\n",
    "    paralogs_file      = read.table(paralogs_file_path)\n",
    "\n",
    "    # read the fasta file that contains all sequences in the gene family, the one containing all paralogs from the (7) species, ref included.\n",
    "    gene_fam_fasta_path = file.path(family, seq_prefix, paste0(basename(family), \".domain_sequences_prior_after_intersection.fasta\"))\n",
    "    gene_fam_fasta      = readLines(gene_fam_fasta_path)\n",
    "    sequences           = gene_fam_fasta[!grepl(\"^>\", gene_fam_fasta)]     # Extract sequences \n",
    "\n",
    "    #  extract from the fasta aln Bigtree ML number of columns. since from this MSA the subMSA fed to SM and ST are extracted. \n",
    "    # This allows for SM ST and BigTree to be compared in comparison to the MSA matrix size.\n",
    "    no_ref_bigtree_ML_path = file.path(family, seq_prefix, paste0(basename(family), \".domain_sequences_prior_after_intersection_full.fasta_aln\"))\n",
    "    no_ref_bigtree_ML_aln  = readAAStringSet(no_ref_bigtree_ML_path)\n",
    "    \n",
    "    # calculate numbers of: average/sd len of sequence in fam, number of units (6) and number of paralog per species. \n",
    "    # All these three value are correlated: num_species * pralog_per_species = sequence_in_fam \n",
    "    avg_len_seq_in_fam = mean(nchar(sequences))\n",
    "    sd_len_seq_in_fam  = sd(nchar(sequences))\n",
    "    num_of_units       = length(orgs$V1) - 1\n",
    "    num_paralogs       = length(paralogs_file$V1)\n",
    "    num_columns_ML     = length(no_ref_bigtree_ML_aln[[1]])\n",
    "\n",
    "    # compile the dataframe with family name and organism present\n",
    "    df_fam_info = data.frame(\n",
    "        family_name  = c(df_fam_info$family_name, basename(family)),      \n",
    "        avg_len_seq  = c(df_fam_info$avg_len_seq, avg_len_seq_in_fam),\n",
    "        sd_len_seq   = c(df_fam_info$sd_len_seq, sd_len_seq_in_fam),\n",
    "        num_units    = c(df_fam_info$num_units, num_of_units),\n",
    "        num_paralogs = c(df_fam_info$num_paralogs, num_paralogs),\n",
    "        num_col_ML   = c(df_fam_info$num_col_ML, num_columns_ML),\n",
    "        organism     = c(df_fam_info$organism, paste(orgs[[1]], collapse = \", \"))     # this is just a string holding all species keywords\n",
    "    )\n",
    "\n",
    "    # Get all units trees (60 = 6 organism/units (7 - maouse) * 10 samples/statistical repeats) for the combination supermatrix method (SM) + Minimun Evolution (ME).\n",
    "    # make it to a binary tree using the multi2di package\n",
    "    all_units_trees_seq_path = file.path(family, seq_prefix, \"all_units.nwk\")\n",
    "    all_units_trees_seq      = read.tree(all_units_trees_seq_path, keep.multi=TRUE)\n",
    "    all_units_trees_seq      = multi2di(all_units_trees_seq)\n",
    "\n",
    "    # do the same as above for SM + Maximum Likelihood (ML).\n",
    "    all_units_raxml_trees_seq_path = file.path(family, seq_prefix, \"all_units_RAxML_trees.nwk\")\n",
    "    all_units_raxml_trees_seq      = read.tree(all_units_raxml_trees_seq_path, keep.multi=TRUE)\n",
    "    all_units_raxml_trees_seq      = multi2di(all_units_raxml_trees_seq)\n",
    "\n",
    "    # do the same as above for Super Tree (ST) method + ME.\n",
    "    all_units_superfine_trees_seq_path = file.path(family, seq_prefix, \"all_units_superfine_trees.nwk\")\n",
    "    all_units_superfine_trees_seq      = read.tree(all_units_superfine_trees_seq_path, keep.multi=TRUE)\n",
    "    all_units_superfine_trees_seq      = multi2di(all_units_superfine_trees_seq)\n",
    "\n",
    "    # do the same as above for ST + ML.\n",
    "    all_units_raxml_superfine_trees_seq_path = file.path(family, seq_prefix, \"all_units_RAxML_superfine_trees.nwk\")\n",
    "    all_units_raxml_superfine_trees_seq      = read.tree(all_units_raxml_superfine_trees_seq_path, keep.multi=TRUE)\n",
    "    all_units_raxml_superfine_trees_seq      = multi2di(all_units_raxml_superfine_trees_seq)\n",
    "\n",
    "    # it will store all the species refernce paralog trees (6)\n",
    "    all_species_raxml_subtrees_from_full_aln = c()\n",
    "    \n",
    "    # read the BigTree ML \n",
    "    species_raxml_tree_from_full_aln_path = file.path(family, seq_prefix, paste0(\"no_ref_RAxML_bestTree.\", basename(family), \".domain_sequences_prior_after_intersection_full_coded_raxml.nwk\"))\n",
    "    species_raxml_tree_from_full_aln = read.tree(species_raxml_tree_from_full_aln_path, keep.multi=FALSE)\n",
    "    \n",
    "    # for each tip in the above tree rename according to the mapping coding file for ML bigtree\n",
    "    for (j in 1:length(species_raxml_tree_from_full_aln$tip.label)) { \n",
    "        species_raxml_tree_from_full_aln$tip.label[j] = as.character(codefile_full_aln$V1[which(codefile_full_aln$V2 == species_raxml_tree_from_full_aln[[\"tip.label\"]][j])])\n",
    "    }\n",
    "\n",
    "    # get the list of species/organism that arepresent in the gene family excluding the reference species (MOUSE).\n",
    "    noref_species = orgs$V1[orgs$V1 != ref_species]\n",
    "\n",
    "    # do a for loop on the species (7 -1) \n",
    "    for (species in noref_species) {\n",
    "    \n",
    "        # get the list of indices referring to the organism in question\n",
    "        tips_to_keep = grep(species,species_raxml_tree_from_full_aln$tip.label)\n",
    "    \n",
    "        # extract the subtree (paralog tree) from the BigTree ML relating to the species in question\n",
    "        test_subtree = keep.tip(species_raxml_tree_from_full_aln,tips_to_keep)\n",
    "    \n",
    "        # rename each tip sustituting the specific species codename with MOUSE, effectively going from ACAP2_AILME_1 to ACAP2_MOUSE_1\n",
    "        # this makes all tips names from all species subtrees comparable altough they refer to a different sequence.\n",
    "        # this is fine since only the correctness of the topology of the tree is tested later on. \n",
    "        test_subtree$tip.label = gsub(paste0(\"_\", species, \"_\"), \"_MOUSE_\", test_subtree$tip.label)\n",
    "    \n",
    "        # the tips are further mapped to the C!, C2, C3, ecc.. nomenclature. going from  ACAP2_MOUSE_1 to C2\n",
    "        for (j in 1:length(test_subtree$tip.label)) { \n",
    "            test_subtree$tip.label[j] = as.character(codefile$V2[which(codefile$V1 == test_subtree[[\"tip.label\"]][j])])\n",
    "        }\n",
    "        \n",
    "        # trasform the subtree into the ape multiPhylo class and added it to the list with all reference paralog trees.\n",
    "        test_subtree = as.multiPhylo(test_subtree)\n",
    "        all_species_raxml_subtrees_from_full_aln = c(all_species_raxml_subtrees_from_full_aln, test_subtree)\n",
    "    }\n",
    "\n",
    "    #\n",
    "    # SM + ME rf computation block \n",
    "    # \n",
    "    \n",
    "    # iterate over all possible species reference trees\n",
    "    for (r in 1:length(all_species_raxml_subtrees_from_full_aln)) {\n",
    "    \n",
    "        # RF to work needs all trees it needs to compare into a single list (class) and then it will automaticly do all against all comparisons.\n",
    "        # here we are comparing all unit_sample (60) family SM + ME trees with the family (7) species paralog reference trees\n",
    "        trees_to_compare_ME_SM         = c(all_species_raxml_subtrees_from_full_aln[r], all_units_trees_seq)\n",
    "        class(trees_to_compare_ME_SM)  = \"multiPhylo\"\n",
    "        rf_trees_to_compare_ME_SM      = RF.dist(trees_to_compare_ME_SM, normalize=TRUE)\n",
    "    \n",
    "        # transform Rf.dist output format to a matrix and take the first row without the first element. \n",
    "        # Basically retrieve only the comparisons between reference tree and each unit tree\n",
    "        rf_ME_SM_comparisons                = data.matrix(rf_trees_to_compare_ME_SM)[1, -1]\n",
    "    \n",
    "        # for loop over the created vector to correctly update the dataframe. \n",
    "        # the (60) values in the above vector  are from different units and samples in an ordered manner, the for loop is just to \"count\" correctly.\n",
    "        # the dataframe will be updated in chuncks of 10 \n",
    "        for (u in seq(10, 60, by=10)) {\n",
    "        \n",
    "            # update the df with all relevant info. since the rf vector is chuncked in pieces of ten all other things have to be repeated accordingly.\n",
    "            df_all_rf = data.frame(\n",
    "                method    = c(df_all_rf$method, rep(\"SM_ME\", 10)),\n",
    "                family    = c(df_all_rf$family, rep(basename(family), 10)),           # use family name\n",
    "                species   = c(df_all_rf$reference, rep(noref_species[r] , 10)),       # for the reference tree get the species which it is from\n",
    "                unit      = c(df_all_rf$unit, rep((u/10), 10)),\n",
    "                sample    = c(df_all_rf$sample, c(1, 2, 3, 4, 5, 6 ,7 ,8, 9, 10)),\n",
    "                rf_val    = c(df_all_rf$rf_val, rf_ME_SM_comparisons[(u-9):u])        # get all sample from the same unit\n",
    "            )\n",
    "        }\n",
    "    }\n",
    "\n",
    "    #\n",
    "    # SM + ML rf computation block \n",
    "    # \n",
    "    \n",
    "    # repeat the previous block for ML as well\n",
    "    for (r in 1:length(all_species_raxml_subtrees_from_full_aln)) {\n",
    "        trees_to_compare_ML_SM         = c(all_species_raxml_subtrees_from_full_aln[r], all_units_raxml_trees_seq)\n",
    "        class(trees_to_compare_ML_SM)  = \"multiPhylo\"\n",
    "        rf_trees_to_compare_ML_SM      = RF.dist(trees_to_compare_ML_SM, normalize=TRUE)\n",
    "        rf_ML_SM_comparisons           = data.matrix(rf_trees_to_compare_ML_SM)[1, -1]\n",
    "        for (u in seq(10, 60, by=10)) {\n",
    "            df_all_rf = data.frame(\n",
    "                method    = c(df_all_rf$method, rep(\"SM_ML\", 10)),\n",
    "                family    = c(df_all_rf$family, rep(basename(family), 10)),           # use family name\n",
    "                species   = c(df_all_rf$reference, rep(noref_species[r] , 10)),       # for the reference tree get the species which it is from\n",
    "                unit      = c(df_all_rf$unit, rep((u/10), 10)),\n",
    "                sample    = c(df_all_rf$sample, c(1, 2, 3, 4, 5, 6 ,7 ,8, 9, 10)),\n",
    "                rf_val    = c(df_all_rf$rf_val, rf_ML_SM_comparisons[(u-9):u])        # get all sample from the same unit\n",
    "            )\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    #\n",
    "    # ST + ME rf computation block \n",
    "    # \n",
    "    \n",
    "    # repeat the previous block for super tree as well\n",
    "    for (r in 1:length(all_species_raxml_subtrees_from_full_aln)) {\n",
    "        trees_to_compare_ME_ST         = c(all_species_raxml_subtrees_from_full_aln[r], all_units_superfine_trees_seq)\n",
    "        class(trees_to_compare_ME_ST)  = \"multiPhylo\"\n",
    "        rf_trees_to_compare_ME_ST      = RF.dist(trees_to_compare_ME_ST, normalize=TRUE)\n",
    "        rf_ME_ST_comparisons           = data.matrix(rf_trees_to_compare_ME_ST)[1, -1]\n",
    "        for (u in seq(10, 60, by=10)) {\n",
    "            df_all_rf = data.frame(\n",
    "                method    = c(df_all_rf$method, rep(\"ST_ME\", 10)),\n",
    "                family    = c(df_all_rf$family, rep(basename(family), 10)),           # use family name\n",
    "                species   = c(df_all_rf$reference, rep(noref_species[r] , 10)),             # for the reference tree get the species which it is from\n",
    "                unit      = c(df_all_rf$unit, rep((u/10), 10)),\n",
    "                sample    = c(df_all_rf$sample, c(1, 2, 3, 4, 5, 6 ,7 ,8, 9, 10)),\n",
    "                rf_val    = c(df_all_rf$rf_val, rf_ME_ST_comparisons[(u-9):u])        # get all sample from the same unit\n",
    "            )\n",
    "        }\n",
    "    }\n",
    "\n",
    "    #\n",
    "    # ST + ML rf computation block \n",
    "    # \n",
    "    \n",
    "    # repeat the previous block for ML as well\n",
    "    for (r in 1:length(all_species_raxml_subtrees_from_full_aln)) {\n",
    "        trees_to_compare_ML_ST         = c(all_species_raxml_subtrees_from_full_aln[r], all_units_raxml_superfine_trees_seq)\n",
    "        class(trees_to_compare_ML_ST)  = \"multiPhylo\"\n",
    "        rf_trees_to_compare_ML_ST      = RF.dist(trees_to_compare_ML_ST, normalize=TRUE)\n",
    "        rf_ML_ST_comparisons           = data.matrix(rf_trees_to_compare_ML_ST)[1, -1]\n",
    "        for (u in seq(10, 60, by=10)) {\n",
    "            df_all_rf = data.frame(\n",
    "                method    = c(df_all_rf$method, rep(\"ST_ML\", 10)),\n",
    "                family    = c(df_all_rf$family, rep(basename(family), 10)),           # use family name\n",
    "                species   = c(df_all_rf$reference, rep(noref_species[r] , 10)),       # for the reference tree get the species which it is from\n",
    "                unit      = c(df_all_rf$unit, rep((u/10), 10)),\n",
    "                sample    = c(df_all_rf$sample, c(1, 2, 3, 4, 5, 6 ,7 ,8, 9, 10)),\n",
    "                rf_val    = c(df_all_rf$rf_val, rf_ML_ST_comparisons[(u-9):u])        # get all sample from the same unit\n",
    "            )\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "# save the dataframes\n",
    "write.table(df_fam_info, file = \"tables/RF_empirical_set_fam_info.tsv\", sep=\"\\t\", quote = FALSE, row.names = FALSE)\n",
    "write.table(df_all_rf, file = \"tables/RF_empirical_set_between_Bigtree_ML_no_ref_supermatrix_and_supertree.tsv\", quote = FALSE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "09c3d96f-4e70-49e2-90b6-1d8499d92a0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'method'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 24 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>method</th><th scope=col>unit</th><th scope=col>mean_rf_val</th><th scope=col>sd_rf_val</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>SM_ME</td><td>1</td><td>0.4813789</td><td>0.2188298</td></tr>\n",
       "\t<tr><td>SM_ME</td><td>2</td><td>0.4209160</td><td>0.2167914</td></tr>\n",
       "\t<tr><td>SM_ME</td><td>3</td><td>0.3859349</td><td>0.2183725</td></tr>\n",
       "\t<tr><td>SM_ME</td><td>4</td><td>0.3668104</td><td>0.2227733</td></tr>\n",
       "\t<tr><td>SM_ME</td><td>5</td><td>0.3525344</td><td>0.2257878</td></tr>\n",
       "\t<tr><td>SM_ME</td><td>6</td><td>0.3477010</td><td>0.2227963</td></tr>\n",
       "\t<tr><td>SM_ML</td><td>1</td><td>0.4606626</td><td>0.2216494</td></tr>\n",
       "\t<tr><td>SM_ML</td><td>2</td><td>0.3563474</td><td>0.2267008</td></tr>\n",
       "\t<tr><td>SM_ML</td><td>3</td><td>0.3208072</td><td>0.2243196</td></tr>\n",
       "\t<tr><td>SM_ML</td><td>4</td><td>0.2748926</td><td>0.2160326</td></tr>\n",
       "\t<tr><td>SM_ML</td><td>5</td><td>0.2608946</td><td>0.2176701</td></tr>\n",
       "\t<tr><td>SM_ML</td><td>6</td><td>0.2624915</td><td>0.2267215</td></tr>\n",
       "\t<tr><td>ST_ME</td><td>1</td><td>0.4813789</td><td>0.2188298</td></tr>\n",
       "\t<tr><td>ST_ME</td><td>2</td><td>0.4533785</td><td>0.2216242</td></tr>\n",
       "\t<tr><td>ST_ME</td><td>3</td><td>0.4251924</td><td>0.2154442</td></tr>\n",
       "\t<tr><td>ST_ME</td><td>4</td><td>0.3998402</td><td>0.2224377</td></tr>\n",
       "\t<tr><td>ST_ME</td><td>5</td><td>0.3907645</td><td>0.2227625</td></tr>\n",
       "\t<tr><td>ST_ME</td><td>6</td><td>0.3814559</td><td>0.2193258</td></tr>\n",
       "\t<tr><td>ST_ML</td><td>1</td><td>0.4606626</td><td>0.2216494</td></tr>\n",
       "\t<tr><td>ST_ML</td><td>2</td><td>0.4328437</td><td>0.2134253</td></tr>\n",
       "\t<tr><td>ST_ML</td><td>3</td><td>0.3731068</td><td>0.2250891</td></tr>\n",
       "\t<tr><td>ST_ML</td><td>4</td><td>0.3454786</td><td>0.2320328</td></tr>\n",
       "\t<tr><td>ST_ML</td><td>5</td><td>0.3262430</td><td>0.2187741</td></tr>\n",
       "\t<tr><td>ST_ML</td><td>6</td><td>0.3183386</td><td>0.2163969</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 24 × 4\n",
       "\\begin{tabular}{llll}\n",
       " method & unit & mean\\_rf\\_val & sd\\_rf\\_val\\\\\n",
       " <chr> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t SM\\_ME & 1 & 0.4813789 & 0.2188298\\\\\n",
       "\t SM\\_ME & 2 & 0.4209160 & 0.2167914\\\\\n",
       "\t SM\\_ME & 3 & 0.3859349 & 0.2183725\\\\\n",
       "\t SM\\_ME & 4 & 0.3668104 & 0.2227733\\\\\n",
       "\t SM\\_ME & 5 & 0.3525344 & 0.2257878\\\\\n",
       "\t SM\\_ME & 6 & 0.3477010 & 0.2227963\\\\\n",
       "\t SM\\_ML & 1 & 0.4606626 & 0.2216494\\\\\n",
       "\t SM\\_ML & 2 & 0.3563474 & 0.2267008\\\\\n",
       "\t SM\\_ML & 3 & 0.3208072 & 0.2243196\\\\\n",
       "\t SM\\_ML & 4 & 0.2748926 & 0.2160326\\\\\n",
       "\t SM\\_ML & 5 & 0.2608946 & 0.2176701\\\\\n",
       "\t SM\\_ML & 6 & 0.2624915 & 0.2267215\\\\\n",
       "\t ST\\_ME & 1 & 0.4813789 & 0.2188298\\\\\n",
       "\t ST\\_ME & 2 & 0.4533785 & 0.2216242\\\\\n",
       "\t ST\\_ME & 3 & 0.4251924 & 0.2154442\\\\\n",
       "\t ST\\_ME & 4 & 0.3998402 & 0.2224377\\\\\n",
       "\t ST\\_ME & 5 & 0.3907645 & 0.2227625\\\\\n",
       "\t ST\\_ME & 6 & 0.3814559 & 0.2193258\\\\\n",
       "\t ST\\_ML & 1 & 0.4606626 & 0.2216494\\\\\n",
       "\t ST\\_ML & 2 & 0.4328437 & 0.2134253\\\\\n",
       "\t ST\\_ML & 3 & 0.3731068 & 0.2250891\\\\\n",
       "\t ST\\_ML & 4 & 0.3454786 & 0.2320328\\\\\n",
       "\t ST\\_ML & 5 & 0.3262430 & 0.2187741\\\\\n",
       "\t ST\\_ML & 6 & 0.3183386 & 0.2163969\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 24 × 4\n",
       "\n",
       "| method &lt;chr&gt; | unit &lt;int&gt; | mean_rf_val &lt;dbl&gt; | sd_rf_val &lt;dbl&gt; |\n",
       "|---|---|---|---|\n",
       "| SM_ME | 1 | 0.4813789 | 0.2188298 |\n",
       "| SM_ME | 2 | 0.4209160 | 0.2167914 |\n",
       "| SM_ME | 3 | 0.3859349 | 0.2183725 |\n",
       "| SM_ME | 4 | 0.3668104 | 0.2227733 |\n",
       "| SM_ME | 5 | 0.3525344 | 0.2257878 |\n",
       "| SM_ME | 6 | 0.3477010 | 0.2227963 |\n",
       "| SM_ML | 1 | 0.4606626 | 0.2216494 |\n",
       "| SM_ML | 2 | 0.3563474 | 0.2267008 |\n",
       "| SM_ML | 3 | 0.3208072 | 0.2243196 |\n",
       "| SM_ML | 4 | 0.2748926 | 0.2160326 |\n",
       "| SM_ML | 5 | 0.2608946 | 0.2176701 |\n",
       "| SM_ML | 6 | 0.2624915 | 0.2267215 |\n",
       "| ST_ME | 1 | 0.4813789 | 0.2188298 |\n",
       "| ST_ME | 2 | 0.4533785 | 0.2216242 |\n",
       "| ST_ME | 3 | 0.4251924 | 0.2154442 |\n",
       "| ST_ME | 4 | 0.3998402 | 0.2224377 |\n",
       "| ST_ME | 5 | 0.3907645 | 0.2227625 |\n",
       "| ST_ME | 6 | 0.3814559 | 0.2193258 |\n",
       "| ST_ML | 1 | 0.4606626 | 0.2216494 |\n",
       "| ST_ML | 2 | 0.4328437 | 0.2134253 |\n",
       "| ST_ML | 3 | 0.3731068 | 0.2250891 |\n",
       "| ST_ML | 4 | 0.3454786 | 0.2320328 |\n",
       "| ST_ML | 5 | 0.3262430 | 0.2187741 |\n",
       "| ST_ML | 6 | 0.3183386 | 0.2163969 |\n",
       "\n"
      ],
      "text/plain": [
       "   method unit mean_rf_val sd_rf_val\n",
       "1  SM_ME  1    0.4813789   0.2188298\n",
       "2  SM_ME  2    0.4209160   0.2167914\n",
       "3  SM_ME  3    0.3859349   0.2183725\n",
       "4  SM_ME  4    0.3668104   0.2227733\n",
       "5  SM_ME  5    0.3525344   0.2257878\n",
       "6  SM_ME  6    0.3477010   0.2227963\n",
       "7  SM_ML  1    0.4606626   0.2216494\n",
       "8  SM_ML  2    0.3563474   0.2267008\n",
       "9  SM_ML  3    0.3208072   0.2243196\n",
       "10 SM_ML  4    0.2748926   0.2160326\n",
       "11 SM_ML  5    0.2608946   0.2176701\n",
       "12 SM_ML  6    0.2624915   0.2267215\n",
       "13 ST_ME  1    0.4813789   0.2188298\n",
       "14 ST_ME  2    0.4533785   0.2216242\n",
       "15 ST_ME  3    0.4251924   0.2154442\n",
       "16 ST_ME  4    0.3998402   0.2224377\n",
       "17 ST_ME  5    0.3907645   0.2227625\n",
       "18 ST_ME  6    0.3814559   0.2193258\n",
       "19 ST_ML  1    0.4606626   0.2216494\n",
       "20 ST_ML  2    0.4328437   0.2134253\n",
       "21 ST_ML  3    0.3731068   0.2250891\n",
       "22 ST_ML  4    0.3454786   0.2320328\n",
       "23 ST_ML  5    0.3262430   0.2187741\n",
       "24 ST_ML  6    0.3183386   0.2163969"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#\n",
    "# Table 2 SM and ST VS Big-Tree ML no ref. group  RF by method and unit.\n",
    "#\n",
    "\n",
    "# the directory prefix where all analazyd files are. this is created by main.nf pipeline in empirical data mode. \n",
    "seq_prefix = \"results_without_mouse_10_samples/tcoffee\"\n",
    "\n",
    "# define how many units there are \n",
    "orgs_path      = file.path(pfam_dirs[[1]], seq_prefix, paste0(basename(pfam_dirs[[1]]), \".orthologs_org_ids_to_concatenate\"))\n",
    "orgs           = read.table(orgs_path)\n",
    "num_of_species = length(orgs$V1)\n",
    "num_of_units   = length(orgs$V1) - 1\n",
    "\n",
    "# read the table compiled in chapter 4. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df_new = read.table(\"tables/RF_empirical_set_between_Bigtree_ML_no_ref_supermatrix_and_supertree.tsv\", header=T)\n",
    "\n",
    "# Group data by method and unit, then calculate mean and standard deviation for all families \n",
    "summary_table = df_new %>%\n",
    "  group_by(method, unit) %>%\n",
    "  summarise(\n",
    "    mean_rf_val = mean(rf_val, na.rm = TRUE),\n",
    "    sd_rf_val = sd(rf_val, na.rm = TRUE)\n",
    "  ) %>%\n",
    "  ungroup()\n",
    "\n",
    "# print the unit rf table and save it\n",
    "summary_table\n",
    "write.table(summary_table, file = \"tables/RF_empirical_set_per_unit_SM_ST_ME_ML_vs_BigTreeML_no_ref.tsv\", quote = FALSE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1dcf7904-9d4d-44fb-8f7e-b86da0ade9e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Fig 5a Supermatrix-ME and Supermatrix-ML VS Big-Tree ML without ref (no_ref) (aka MOUSE) paralog trees projections(aka species subtrees extracted).\n",
    "#\n",
    "\n",
    "# read the table compiled in chapter 4 cell above. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df = read.table(\"tables/RF_empirical_set_per_unit_SM_ST_ME_ML_vs_BigTreeML_no_ref.tsv\", header=T)\n",
    "\n",
    "# retrieve only Suspermatrix method rows.\n",
    "df_figure5a = subset(df, grepl(\"SM\", method))\n",
    "\n",
    "# Add prefix \"unit_\" to the \"unit\" column\n",
    "df_figure5a$unit = paste0(\"unit_\", df_figure5a$unit)\n",
    "\n",
    "# set limits for plot\n",
    "options(repr.plot.width = 5, repr.plot.height = 5, repr.plot.res = 300)\n",
    "\n",
    "# Define custom colors for each Method\n",
    "custom_colors = c(\"SM_ME\" = \"#2F739B\",   # Dark blue\n",
    "                  \"SM_ML\" = \"#D05159\")   # Red\n",
    "\n",
    "# plot the table \n",
    "unit_rf_5a = ggplot(data=df_figure5a, aes(x=unit, y=mean_rf_val, group=method, color=method)) +\n",
    "    geom_line() + \n",
    "    geom_point() + \n",
    "    theme_light() + \n",
    "    ylab(\"Normalized RF\") + xlab(\"Topology\") +\n",
    "    theme(\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size=8),\n",
    "        axis.text.y = element_text(size=8),\n",
    "        axis.title = element_text(face = \"bold\"),\n",
    "        legend.position=\"right\"\n",
    "        ) + \n",
    "    scale_y_continuous(limits = c(0, 0.51)) +\n",
    "    # Custom legend labels and apply custom colors\n",
    "    scale_color_manual(\n",
    "        name = \"Method\", \n",
    "        labels = c(\"SM_ME\" = \"Supermatrix-ME\", \"SM_ML\" = \"Supermatrix-ML\"), # Custom legend labels\n",
    "        values = custom_colors\n",
    "    )\n",
    "\n",
    "ggsave(filename=\"figures/RF_empirical_set_lineplot_SM_vs_ML_Bigtree_noref_figure5a.png\", plot=unit_rf_5a, dpi = \"retina\", width = 5, height = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fa359065-c348-4b1e-9c6c-99901ea2b2a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Fig 5b Supertree-ME and Supertree-ML VS Big-Tree ML without ref (no_ref) (aka MOUSE) paralog trees projections(aka species subtrees extracted).\n",
    "#\n",
    "\n",
    "# read the table compiled in chapter 4 two cell above. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df = read.table(\"tables/RF_empirical_set_per_unit_SM_ST_ME_ML_vs_BigTreeML_no_ref.tsv\", header=T)\n",
    "\n",
    "# retrieve only Supertree method rows.\n",
    "df_figure5b = subset(df, grepl(\"ST\", method))\n",
    "\n",
    "# Add prefix \"unit_\" to the \"unit\" column\n",
    "df_figure5b$unit = paste0(\"unit_\", df_figure5b$unit)\n",
    "\n",
    "# set limits for plot\n",
    "options(repr.plot.width = 5, repr.plot.height = 5, repr.plot.res = 300)\n",
    "\n",
    "# Define custom colors for each Method\n",
    "custom_colors = c(\"ST_ME\" = \"#7CB6E2\",     # Light blue \n",
    "                  \"ST_ML\" = \"#E5A23D\")     # Orange/yellow\n",
    "\n",
    "# plot the table \n",
    "unit_rf_5b = ggplot(data=df_figure5b, aes(x=unit, y=mean_rf_val, group=method, color=method)) +\n",
    "    geom_line() + \n",
    "    geom_point() + \n",
    "    theme_light() + \n",
    "    ylab(\"Normalized RF\") + xlab(\"Topology\") +\n",
    "    theme(\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size=8),\n",
    "        axis.text.y = element_text(size=8),\n",
    "        axis.title = element_text(face = \"bold\"),\n",
    "        legend.position=\"right\"\n",
    "        ) + \n",
    "    scale_y_continuous(limits = c(0, 0.51)) +\n",
    "    # Custom legend labels and apply custom colors\n",
    "    scale_color_manual(\n",
    "        name = \"Method\", \n",
    "        labels = c(\"ST_ME\" = \"Supertree-ME\", \"ST_ML\" = \"Supertree-ML\"), # Custom legend labels\n",
    "        values = custom_colors\n",
    "    )\n",
    "\n",
    "ggsave(filename=\"figures/RF_empirical_set_lineplot_ST_vs_ML_Bigtree_noref_figure5b.png\", plot=unit_rf_5b, dpi = \"retina\", width = 5, height = 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56345ec2-afcf-4934-a8dd-488ac3bd0bf1",
   "metadata": {},
   "source": [
    "### TODO decide which of this is in figure 5 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "4e55ab46-137d-4d70-b61b-8decc175e0fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'family'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 164 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>family</th><th scope=col>method</th><th scope=col>mean_rf_val</th><th scope=col>sd_rf_val</th><th scope=col>num_paralogs</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>PF00004</td><td>SM_ME</td><td>0.6197917</td><td>0.05018897</td><td>19</td></tr>\n",
       "\t<tr><td>PF00004</td><td>SM_ML</td><td>0.2572917</td><td>0.20347724</td><td>19</td></tr>\n",
       "\t<tr><td>PF00004</td><td>ST_ME</td><td>0.6270833</td><td>0.07576789</td><td>19</td></tr>\n",
       "\t<tr><td>PF00004</td><td>ST_ML</td><td>0.4354167</td><td>0.15071499</td><td>19</td></tr>\n",
       "\t<tr><td>PF00019</td><td>SM_ME</td><td>0.2912281</td><td>0.08134700</td><td>22</td></tr>\n",
       "\t<tr><td>PF00019</td><td>SM_ML</td><td>0.1649123</td><td>0.12710014</td><td>22</td></tr>\n",
       "\t<tr><td>PF00019</td><td>ST_ME</td><td>0.3245614</td><td>0.10099850</td><td>22</td></tr>\n",
       "\t<tr><td>PF00019</td><td>ST_ML</td><td>0.2570175</td><td>0.10412447</td><td>22</td></tr>\n",
       "\t<tr><td>PF00022</td><td>SM_ME</td><td>0.2400000</td><td>0.09708854</td><td>13</td></tr>\n",
       "\t<tr><td>PF00022</td><td>SM_ML</td><td>0.2033333</td><td>0.13799197</td><td>13</td></tr>\n",
       "\t<tr><td>PF00022</td><td>ST_ME</td><td>0.2783333</td><td>0.10830422</td><td>13</td></tr>\n",
       "\t<tr><td>PF00022</td><td>ST_ML</td><td>0.2900000</td><td>0.12889070</td><td>13</td></tr>\n",
       "\t<tr><td>PF00025</td><td>SM_ME</td><td>0.2277778</td><td>0.14019044</td><td>18</td></tr>\n",
       "\t<tr><td>PF00025</td><td>SM_ML</td><td>0.1944444</td><td>0.15527454</td><td>18</td></tr>\n",
       "\t<tr><td>PF00025</td><td>ST_ME</td><td>0.2977778</td><td>0.11720462</td><td>18</td></tr>\n",
       "\t<tr><td>PF00025</td><td>ST_ML</td><td>0.2222222</td><td>0.14144324</td><td>18</td></tr>\n",
       "\t<tr><td>PF00041</td><td>SM_ME</td><td>0.4000000</td><td>0.22776919</td><td>17</td></tr>\n",
       "\t<tr><td>PF00041</td><td>SM_ML</td><td>0.4416667</td><td>0.15312763</td><td>17</td></tr>\n",
       "\t<tr><td>PF00041</td><td>ST_ME</td><td>0.5404762</td><td>0.18439181</td><td>17</td></tr>\n",
       "\t<tr><td>PF00041</td><td>ST_ML</td><td>0.5238095</td><td>0.18499201</td><td>17</td></tr>\n",
       "\t<tr><td>PF00059</td><td>SM_ME</td><td>0.3217949</td><td>0.14716027</td><td>16</td></tr>\n",
       "\t<tr><td>PF00059</td><td>SM_ML</td><td>0.3243590</td><td>0.16586341</td><td>16</td></tr>\n",
       "\t<tr><td>PF00059</td><td>ST_ME</td><td>0.3551282</td><td>0.14070159</td><td>16</td></tr>\n",
       "\t<tr><td>PF00059</td><td>ST_ML</td><td>0.3717949</td><td>0.16045377</td><td>16</td></tr>\n",
       "\t<tr><td>PF00061</td><td>SM_ME</td><td>0.2233333</td><td>0.13480616</td><td>13</td></tr>\n",
       "\t<tr><td>PF00061</td><td>SM_ML</td><td>0.2566667</td><td>0.14205025</td><td>13</td></tr>\n",
       "\t<tr><td>PF00061</td><td>ST_ME</td><td>0.2416667</td><td>0.14313443</td><td>13</td></tr>\n",
       "\t<tr><td>PF00061</td><td>ST_ML</td><td>0.3333333</td><td>0.12884747</td><td>13</td></tr>\n",
       "\t<tr><td>PF00092</td><td>SM_ME</td><td>0.1761905</td><td>0.15750600</td><td>10</td></tr>\n",
       "\t<tr><td>PF00092</td><td>SM_ML</td><td>0.2714286</td><td>0.09667310</td><td>10</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>PF01392</td><td>ST_ME</td><td>0.30909091</td><td>0.17526819</td><td>14</td></tr>\n",
       "\t<tr><td>PF01392</td><td>ST_ML</td><td>0.46666667</td><td>0.15365585</td><td>14</td></tr>\n",
       "\t<tr><td>PF01412</td><td>SM_ME</td><td>0.64791667</td><td>0.06659627</td><td>11</td></tr>\n",
       "\t<tr><td>PF01412</td><td>SM_ML</td><td>0.17291667</td><td>0.21691807</td><td>11</td></tr>\n",
       "\t<tr><td>PF01412</td><td>ST_ME</td><td>0.64791667</td><td>0.13026850</td><td>11</td></tr>\n",
       "\t<tr><td>PF01412</td><td>ST_ML</td><td>0.33125000</td><td>0.21659679</td><td>11</td></tr>\n",
       "\t<tr><td>PF01421</td><td>SM_ME</td><td>0.45757576</td><td>0.06430034</td><td>14</td></tr>\n",
       "\t<tr><td>PF01421</td><td>SM_ML</td><td>0.35757576</td><td>0.08288186</td><td>14</td></tr>\n",
       "\t<tr><td>PF01421</td><td>ST_ME</td><td>0.45909091</td><td>0.08704103</td><td>14</td></tr>\n",
       "\t<tr><td>PF01421</td><td>ST_ML</td><td>0.34090909</td><td>0.11470127</td><td>14</td></tr>\n",
       "\t<tr><td>PF02214</td><td>SM_ME</td><td>0.37142857</td><td>0.07777241</td><td>24</td></tr>\n",
       "\t<tr><td>PF02214</td><td>SM_ML</td><td>0.25793651</td><td>0.14858950</td><td>24</td></tr>\n",
       "\t<tr><td>PF02214</td><td>ST_ME</td><td>0.40079365</td><td>0.07350950</td><td>24</td></tr>\n",
       "\t<tr><td>PF02214</td><td>ST_ML</td><td>0.32301587</td><td>0.13986566</td><td>24</td></tr>\n",
       "\t<tr><td>PF02931</td><td>SM_ME</td><td>0.10185185</td><td>0.04446765</td><td>21</td></tr>\n",
       "\t<tr><td>PF02931</td><td>SM_ML</td><td>0.06574074</td><td>0.09608646</td><td>21</td></tr>\n",
       "\t<tr><td>PF02931</td><td>ST_ME</td><td>0.10185185</td><td>0.06975711</td><td>21</td></tr>\n",
       "\t<tr><td>PF02931</td><td>ST_ML</td><td>0.10000000</td><td>0.09784641</td><td>21</td></tr>\n",
       "\t<tr><td>PF07679</td><td>SM_ME</td><td>0.63055556</td><td>0.05972443</td><td>15</td></tr>\n",
       "\t<tr><td>PF07679</td><td>SM_ML</td><td>0.57777778</td><td>0.10972518</td><td>15</td></tr>\n",
       "\t<tr><td>PF07679</td><td>ST_ME</td><td>0.63055556</td><td>0.11130052</td><td>15</td></tr>\n",
       "\t<tr><td>PF07679</td><td>ST_ML</td><td>0.63888889</td><td>0.10737289</td><td>15</td></tr>\n",
       "\t<tr><td>PF08205</td><td>SM_ME</td><td>0.62380952</td><td>0.11959418</td><td>10</td></tr>\n",
       "\t<tr><td>PF08205</td><td>SM_ML</td><td>0.46428571</td><td>0.23159314</td><td>10</td></tr>\n",
       "\t<tr><td>PF08205</td><td>ST_ME</td><td>0.66428571</td><td>0.11918946</td><td>10</td></tr>\n",
       "\t<tr><td>PF08205</td><td>ST_ML</td><td>0.63809524</td><td>0.16818797</td><td>10</td></tr>\n",
       "\t<tr><td>PF12796</td><td>SM_ME</td><td>0.80447531</td><td>0.07121249</td><td>21</td></tr>\n",
       "\t<tr><td>PF12796</td><td>SM_ML</td><td>0.84490741</td><td>0.08352011</td><td>21</td></tr>\n",
       "\t<tr><td>PF12796</td><td>ST_ME</td><td>0.84891975</td><td>0.04894638</td><td>21</td></tr>\n",
       "\t<tr><td>PF12796</td><td>ST_ML</td><td>0.82376543</td><td>0.07896702</td><td>21</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 164 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " family & method & mean\\_rf\\_val & sd\\_rf\\_val & num\\_paralogs\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t PF00004 & SM\\_ME & 0.6197917 & 0.05018897 & 19\\\\\n",
       "\t PF00004 & SM\\_ML & 0.2572917 & 0.20347724 & 19\\\\\n",
       "\t PF00004 & ST\\_ME & 0.6270833 & 0.07576789 & 19\\\\\n",
       "\t PF00004 & ST\\_ML & 0.4354167 & 0.15071499 & 19\\\\\n",
       "\t PF00019 & SM\\_ME & 0.2912281 & 0.08134700 & 22\\\\\n",
       "\t PF00019 & SM\\_ML & 0.1649123 & 0.12710014 & 22\\\\\n",
       "\t PF00019 & ST\\_ME & 0.3245614 & 0.10099850 & 22\\\\\n",
       "\t PF00019 & ST\\_ML & 0.2570175 & 0.10412447 & 22\\\\\n",
       "\t PF00022 & SM\\_ME & 0.2400000 & 0.09708854 & 13\\\\\n",
       "\t PF00022 & SM\\_ML & 0.2033333 & 0.13799197 & 13\\\\\n",
       "\t PF00022 & ST\\_ME & 0.2783333 & 0.10830422 & 13\\\\\n",
       "\t PF00022 & ST\\_ML & 0.2900000 & 0.12889070 & 13\\\\\n",
       "\t PF00025 & SM\\_ME & 0.2277778 & 0.14019044 & 18\\\\\n",
       "\t PF00025 & SM\\_ML & 0.1944444 & 0.15527454 & 18\\\\\n",
       "\t PF00025 & ST\\_ME & 0.2977778 & 0.11720462 & 18\\\\\n",
       "\t PF00025 & ST\\_ML & 0.2222222 & 0.14144324 & 18\\\\\n",
       "\t PF00041 & SM\\_ME & 0.4000000 & 0.22776919 & 17\\\\\n",
       "\t PF00041 & SM\\_ML & 0.4416667 & 0.15312763 & 17\\\\\n",
       "\t PF00041 & ST\\_ME & 0.5404762 & 0.18439181 & 17\\\\\n",
       "\t PF00041 & ST\\_ML & 0.5238095 & 0.18499201 & 17\\\\\n",
       "\t PF00059 & SM\\_ME & 0.3217949 & 0.14716027 & 16\\\\\n",
       "\t PF00059 & SM\\_ML & 0.3243590 & 0.16586341 & 16\\\\\n",
       "\t PF00059 & ST\\_ME & 0.3551282 & 0.14070159 & 16\\\\\n",
       "\t PF00059 & ST\\_ML & 0.3717949 & 0.16045377 & 16\\\\\n",
       "\t PF00061 & SM\\_ME & 0.2233333 & 0.13480616 & 13\\\\\n",
       "\t PF00061 & SM\\_ML & 0.2566667 & 0.14205025 & 13\\\\\n",
       "\t PF00061 & ST\\_ME & 0.2416667 & 0.14313443 & 13\\\\\n",
       "\t PF00061 & ST\\_ML & 0.3333333 & 0.12884747 & 13\\\\\n",
       "\t PF00092 & SM\\_ME & 0.1761905 & 0.15750600 & 10\\\\\n",
       "\t PF00092 & SM\\_ML & 0.2714286 & 0.09667310 & 10\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t PF01392 & ST\\_ME & 0.30909091 & 0.17526819 & 14\\\\\n",
       "\t PF01392 & ST\\_ML & 0.46666667 & 0.15365585 & 14\\\\\n",
       "\t PF01412 & SM\\_ME & 0.64791667 & 0.06659627 & 11\\\\\n",
       "\t PF01412 & SM\\_ML & 0.17291667 & 0.21691807 & 11\\\\\n",
       "\t PF01412 & ST\\_ME & 0.64791667 & 0.13026850 & 11\\\\\n",
       "\t PF01412 & ST\\_ML & 0.33125000 & 0.21659679 & 11\\\\\n",
       "\t PF01421 & SM\\_ME & 0.45757576 & 0.06430034 & 14\\\\\n",
       "\t PF01421 & SM\\_ML & 0.35757576 & 0.08288186 & 14\\\\\n",
       "\t PF01421 & ST\\_ME & 0.45909091 & 0.08704103 & 14\\\\\n",
       "\t PF01421 & ST\\_ML & 0.34090909 & 0.11470127 & 14\\\\\n",
       "\t PF02214 & SM\\_ME & 0.37142857 & 0.07777241 & 24\\\\\n",
       "\t PF02214 & SM\\_ML & 0.25793651 & 0.14858950 & 24\\\\\n",
       "\t PF02214 & ST\\_ME & 0.40079365 & 0.07350950 & 24\\\\\n",
       "\t PF02214 & ST\\_ML & 0.32301587 & 0.13986566 & 24\\\\\n",
       "\t PF02931 & SM\\_ME & 0.10185185 & 0.04446765 & 21\\\\\n",
       "\t PF02931 & SM\\_ML & 0.06574074 & 0.09608646 & 21\\\\\n",
       "\t PF02931 & ST\\_ME & 0.10185185 & 0.06975711 & 21\\\\\n",
       "\t PF02931 & ST\\_ML & 0.10000000 & 0.09784641 & 21\\\\\n",
       "\t PF07679 & SM\\_ME & 0.63055556 & 0.05972443 & 15\\\\\n",
       "\t PF07679 & SM\\_ML & 0.57777778 & 0.10972518 & 15\\\\\n",
       "\t PF07679 & ST\\_ME & 0.63055556 & 0.11130052 & 15\\\\\n",
       "\t PF07679 & ST\\_ML & 0.63888889 & 0.10737289 & 15\\\\\n",
       "\t PF08205 & SM\\_ME & 0.62380952 & 0.11959418 & 10\\\\\n",
       "\t PF08205 & SM\\_ML & 0.46428571 & 0.23159314 & 10\\\\\n",
       "\t PF08205 & ST\\_ME & 0.66428571 & 0.11918946 & 10\\\\\n",
       "\t PF08205 & ST\\_ML & 0.63809524 & 0.16818797 & 10\\\\\n",
       "\t PF12796 & SM\\_ME & 0.80447531 & 0.07121249 & 21\\\\\n",
       "\t PF12796 & SM\\_ML & 0.84490741 & 0.08352011 & 21\\\\\n",
       "\t PF12796 & ST\\_ME & 0.84891975 & 0.04894638 & 21\\\\\n",
       "\t PF12796 & ST\\_ML & 0.82376543 & 0.07896702 & 21\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 164 × 5\n",
       "\n",
       "| family &lt;chr&gt; | method &lt;chr&gt; | mean_rf_val &lt;dbl&gt; | sd_rf_val &lt;dbl&gt; | num_paralogs &lt;int&gt; |\n",
       "|---|---|---|---|---|\n",
       "| PF00004 | SM_ME | 0.6197917 | 0.05018897 | 19 |\n",
       "| PF00004 | SM_ML | 0.2572917 | 0.20347724 | 19 |\n",
       "| PF00004 | ST_ME | 0.6270833 | 0.07576789 | 19 |\n",
       "| PF00004 | ST_ML | 0.4354167 | 0.15071499 | 19 |\n",
       "| PF00019 | SM_ME | 0.2912281 | 0.08134700 | 22 |\n",
       "| PF00019 | SM_ML | 0.1649123 | 0.12710014 | 22 |\n",
       "| PF00019 | ST_ME | 0.3245614 | 0.10099850 | 22 |\n",
       "| PF00019 | ST_ML | 0.2570175 | 0.10412447 | 22 |\n",
       "| PF00022 | SM_ME | 0.2400000 | 0.09708854 | 13 |\n",
       "| PF00022 | SM_ML | 0.2033333 | 0.13799197 | 13 |\n",
       "| PF00022 | ST_ME | 0.2783333 | 0.10830422 | 13 |\n",
       "| PF00022 | ST_ML | 0.2900000 | 0.12889070 | 13 |\n",
       "| PF00025 | SM_ME | 0.2277778 | 0.14019044 | 18 |\n",
       "| PF00025 | SM_ML | 0.1944444 | 0.15527454 | 18 |\n",
       "| PF00025 | ST_ME | 0.2977778 | 0.11720462 | 18 |\n",
       "| PF00025 | ST_ML | 0.2222222 | 0.14144324 | 18 |\n",
       "| PF00041 | SM_ME | 0.4000000 | 0.22776919 | 17 |\n",
       "| PF00041 | SM_ML | 0.4416667 | 0.15312763 | 17 |\n",
       "| PF00041 | ST_ME | 0.5404762 | 0.18439181 | 17 |\n",
       "| PF00041 | ST_ML | 0.5238095 | 0.18499201 | 17 |\n",
       "| PF00059 | SM_ME | 0.3217949 | 0.14716027 | 16 |\n",
       "| PF00059 | SM_ML | 0.3243590 | 0.16586341 | 16 |\n",
       "| PF00059 | ST_ME | 0.3551282 | 0.14070159 | 16 |\n",
       "| PF00059 | ST_ML | 0.3717949 | 0.16045377 | 16 |\n",
       "| PF00061 | SM_ME | 0.2233333 | 0.13480616 | 13 |\n",
       "| PF00061 | SM_ML | 0.2566667 | 0.14205025 | 13 |\n",
       "| PF00061 | ST_ME | 0.2416667 | 0.14313443 | 13 |\n",
       "| PF00061 | ST_ML | 0.3333333 | 0.12884747 | 13 |\n",
       "| PF00092 | SM_ME | 0.1761905 | 0.15750600 | 10 |\n",
       "| PF00092 | SM_ML | 0.2714286 | 0.09667310 | 10 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| PF01392 | ST_ME | 0.30909091 | 0.17526819 | 14 |\n",
       "| PF01392 | ST_ML | 0.46666667 | 0.15365585 | 14 |\n",
       "| PF01412 | SM_ME | 0.64791667 | 0.06659627 | 11 |\n",
       "| PF01412 | SM_ML | 0.17291667 | 0.21691807 | 11 |\n",
       "| PF01412 | ST_ME | 0.64791667 | 0.13026850 | 11 |\n",
       "| PF01412 | ST_ML | 0.33125000 | 0.21659679 | 11 |\n",
       "| PF01421 | SM_ME | 0.45757576 | 0.06430034 | 14 |\n",
       "| PF01421 | SM_ML | 0.35757576 | 0.08288186 | 14 |\n",
       "| PF01421 | ST_ME | 0.45909091 | 0.08704103 | 14 |\n",
       "| PF01421 | ST_ML | 0.34090909 | 0.11470127 | 14 |\n",
       "| PF02214 | SM_ME | 0.37142857 | 0.07777241 | 24 |\n",
       "| PF02214 | SM_ML | 0.25793651 | 0.14858950 | 24 |\n",
       "| PF02214 | ST_ME | 0.40079365 | 0.07350950 | 24 |\n",
       "| PF02214 | ST_ML | 0.32301587 | 0.13986566 | 24 |\n",
       "| PF02931 | SM_ME | 0.10185185 | 0.04446765 | 21 |\n",
       "| PF02931 | SM_ML | 0.06574074 | 0.09608646 | 21 |\n",
       "| PF02931 | ST_ME | 0.10185185 | 0.06975711 | 21 |\n",
       "| PF02931 | ST_ML | 0.10000000 | 0.09784641 | 21 |\n",
       "| PF07679 | SM_ME | 0.63055556 | 0.05972443 | 15 |\n",
       "| PF07679 | SM_ML | 0.57777778 | 0.10972518 | 15 |\n",
       "| PF07679 | ST_ME | 0.63055556 | 0.11130052 | 15 |\n",
       "| PF07679 | ST_ML | 0.63888889 | 0.10737289 | 15 |\n",
       "| PF08205 | SM_ME | 0.62380952 | 0.11959418 | 10 |\n",
       "| PF08205 | SM_ML | 0.46428571 | 0.23159314 | 10 |\n",
       "| PF08205 | ST_ME | 0.66428571 | 0.11918946 | 10 |\n",
       "| PF08205 | ST_ML | 0.63809524 | 0.16818797 | 10 |\n",
       "| PF12796 | SM_ME | 0.80447531 | 0.07121249 | 21 |\n",
       "| PF12796 | SM_ML | 0.84490741 | 0.08352011 | 21 |\n",
       "| PF12796 | ST_ME | 0.84891975 | 0.04894638 | 21 |\n",
       "| PF12796 | ST_ML | 0.82376543 | 0.07896702 | 21 |\n",
       "\n"
      ],
      "text/plain": [
       "    family  method mean_rf_val sd_rf_val  num_paralogs\n",
       "1   PF00004 SM_ME  0.6197917   0.05018897 19          \n",
       "2   PF00004 SM_ML  0.2572917   0.20347724 19          \n",
       "3   PF00004 ST_ME  0.6270833   0.07576789 19          \n",
       "4   PF00004 ST_ML  0.4354167   0.15071499 19          \n",
       "5   PF00019 SM_ME  0.2912281   0.08134700 22          \n",
       "6   PF00019 SM_ML  0.1649123   0.12710014 22          \n",
       "7   PF00019 ST_ME  0.3245614   0.10099850 22          \n",
       "8   PF00019 ST_ML  0.2570175   0.10412447 22          \n",
       "9   PF00022 SM_ME  0.2400000   0.09708854 13          \n",
       "10  PF00022 SM_ML  0.2033333   0.13799197 13          \n",
       "11  PF00022 ST_ME  0.2783333   0.10830422 13          \n",
       "12  PF00022 ST_ML  0.2900000   0.12889070 13          \n",
       "13  PF00025 SM_ME  0.2277778   0.14019044 18          \n",
       "14  PF00025 SM_ML  0.1944444   0.15527454 18          \n",
       "15  PF00025 ST_ME  0.2977778   0.11720462 18          \n",
       "16  PF00025 ST_ML  0.2222222   0.14144324 18          \n",
       "17  PF00041 SM_ME  0.4000000   0.22776919 17          \n",
       "18  PF00041 SM_ML  0.4416667   0.15312763 17          \n",
       "19  PF00041 ST_ME  0.5404762   0.18439181 17          \n",
       "20  PF00041 ST_ML  0.5238095   0.18499201 17          \n",
       "21  PF00059 SM_ME  0.3217949   0.14716027 16          \n",
       "22  PF00059 SM_ML  0.3243590   0.16586341 16          \n",
       "23  PF00059 ST_ME  0.3551282   0.14070159 16          \n",
       "24  PF00059 ST_ML  0.3717949   0.16045377 16          \n",
       "25  PF00061 SM_ME  0.2233333   0.13480616 13          \n",
       "26  PF00061 SM_ML  0.2566667   0.14205025 13          \n",
       "27  PF00061 ST_ME  0.2416667   0.14313443 13          \n",
       "28  PF00061 ST_ML  0.3333333   0.12884747 13          \n",
       "29  PF00092 SM_ME  0.1761905   0.15750600 10          \n",
       "30  PF00092 SM_ML  0.2714286   0.09667310 10          \n",
       "⋮   ⋮       ⋮      ⋮           ⋮          ⋮           \n",
       "135 PF01392 ST_ME  0.30909091  0.17526819 14          \n",
       "136 PF01392 ST_ML  0.46666667  0.15365585 14          \n",
       "137 PF01412 SM_ME  0.64791667  0.06659627 11          \n",
       "138 PF01412 SM_ML  0.17291667  0.21691807 11          \n",
       "139 PF01412 ST_ME  0.64791667  0.13026850 11          \n",
       "140 PF01412 ST_ML  0.33125000  0.21659679 11          \n",
       "141 PF01421 SM_ME  0.45757576  0.06430034 14          \n",
       "142 PF01421 SM_ML  0.35757576  0.08288186 14          \n",
       "143 PF01421 ST_ME  0.45909091  0.08704103 14          \n",
       "144 PF01421 ST_ML  0.34090909  0.11470127 14          \n",
       "145 PF02214 SM_ME  0.37142857  0.07777241 24          \n",
       "146 PF02214 SM_ML  0.25793651  0.14858950 24          \n",
       "147 PF02214 ST_ME  0.40079365  0.07350950 24          \n",
       "148 PF02214 ST_ML  0.32301587  0.13986566 24          \n",
       "149 PF02931 SM_ME  0.10185185  0.04446765 21          \n",
       "150 PF02931 SM_ML  0.06574074  0.09608646 21          \n",
       "151 PF02931 ST_ME  0.10185185  0.06975711 21          \n",
       "152 PF02931 ST_ML  0.10000000  0.09784641 21          \n",
       "153 PF07679 SM_ME  0.63055556  0.05972443 15          \n",
       "154 PF07679 SM_ML  0.57777778  0.10972518 15          \n",
       "155 PF07679 ST_ME  0.63055556  0.11130052 15          \n",
       "156 PF07679 ST_ML  0.63888889  0.10737289 15          \n",
       "157 PF08205 SM_ME  0.62380952  0.11959418 10          \n",
       "158 PF08205 SM_ML  0.46428571  0.23159314 10          \n",
       "159 PF08205 ST_ME  0.66428571  0.11918946 10          \n",
       "160 PF08205 ST_ML  0.63809524  0.16818797 10          \n",
       "161 PF12796 SM_ME  0.80447531  0.07121249 21          \n",
       "162 PF12796 SM_ML  0.84490741  0.08352011 21          \n",
       "163 PF12796 ST_ME  0.84891975  0.04894638 21          \n",
       "164 PF12796 ST_ML  0.82376543  0.07896702 21          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#\n",
    "# SM and ST VS Big-Tree ML no ref. group by method and num paralogs per scpecies in gene family.\n",
    "#\n",
    "\n",
    "# the directory prefix where all analazyd files are. this is created by main.nf pipeline in empirical data mode. \n",
    "seq_prefix = \"results_without_mouse_10_samples/tcoffee\"\n",
    "\n",
    "# define how many units there are \n",
    "orgs_path      = file.path(pfam_dirs[[1]], seq_prefix, paste0(basename(pfam_dirs[[1]]), \".orthologs_org_ids_to_concatenate\"))\n",
    "orgs           = read.table(orgs_path)\n",
    "num_of_species = length(orgs$V1)\n",
    "num_of_units   = length(orgs$V1) - 1\n",
    "\n",
    "# read the tables compiled in chapter 4. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df_rf   = read.table(\"tables/RF_empirical_set_between_Bigtree_ML_no_ref_supermatrix_and_supertree.tsv\", header=T)\n",
    "df_info = read.table(\"tables/RF_empirical_set_fam_info.tsv\", sep=\"\\t\", header=T)\n",
    "\n",
    "# Group data by method, family. then calculate mean and standard deviation for all families.\n",
    "family_table = df_rf %>%\n",
    "  group_by(family, method) %>%\n",
    "  summarise(\n",
    "    mean_rf_val = mean(rf_val, na.rm = TRUE),\n",
    "    sd_rf_val = sd(rf_val, na.rm = TRUE)\n",
    "  ) %>%\n",
    "  ungroup()\n",
    "\n",
    "# Repeat each value in  df_info$num_paralogs 4 times and add it to the family df. 4 = SM/ST + ML/ME\n",
    "# this can be done like this becouse both df are ordered lexographically for the PFAM family ID column. \n",
    "repeated_values           = rep(df_info$num_paralogs, each = 4)\n",
    "family_table$num_paralogs = repeated_values\n",
    "\n",
    "# print the unit rf table and save it\n",
    "family_table\n",
    "write.table(family_table, file = \"tables/RF_empirical_set_per_family_SM_ST_ME_ML_vs_BigTreeML_no_ref.tsv\", quote = FALSE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "3d87c699-f11e-44de-9a2c-b370ccb2eb5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mNo shared levels found between `names(values)` of the manual scale and the\n",
      "data's \u001b[32mcolour\u001b[39m values.”\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Fig 5a Supermatrix-ME and Supermatrix-ML VS Big-Tree ML without ref (no_ref) (aka MOUSE) paralog trees projections(aka species subtrees extracted).\n",
    "#\n",
    "\n",
    "# read the table compiled in chapter 4 cell above. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df = read.table(\"tables/RF_empirical_set_per_family_SM_ST_ME_ML_vs_BigTreeML_no_ref.tsv\", header=T)\n",
    "\n",
    "# retrieve only Suspermatrix method rows.\n",
    "df_figure5a = subset(df, grepl(\"SM\", method))\n",
    "\n",
    "# Add prefix \"unit_\" to the \"unit\" column\n",
    "df_figure5a$unit = paste0(\"unit_\", df_figure5a$unit)\n",
    "\n",
    "# set limits for plot\n",
    "options(repr.plot.width = 5, repr.plot.height = 5, repr.plot.res = 300)\n",
    "\n",
    "# Define custom colors for each Method\n",
    "custom_colors = c(\"SM_ME\" = \"#2F739B\",   # Dark blue\n",
    "                  \"SM_ML\" = \"#D05159\")   # Red\n",
    "\n",
    "# plot the table \n",
    "unit_rf_5a = ggplot(data=df_figure5a, aes(x=factor(num_paralogs), y=mean_rf_val, fill=method)) +\n",
    "    geom_boxplot() + \n",
    "    theme_light() + \n",
    "    ylab(\"Normalized RF\") + xlab(\"Num. paralogs per species\") +\n",
    "    theme(\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size=8),\n",
    "        axis.text.y = element_text(size=8),\n",
    "        axis.title = element_text(face = \"bold\"),\n",
    "        legend.position=\"top\"\n",
    "        ) + \n",
    "    #scale_y_continuous(limits = c(0, 0.51)) +\n",
    "    # Custom legend labels and apply custom colors\n",
    "    scale_color_manual(\n",
    "        name = \"Method\", \n",
    "        labels = c(\"SM_ME\" = \"Supermatrix-ME\", \"SM_ML\" = \"Supermatrix-ML\"), # Custom legend labels\n",
    "        values = custom_colors\n",
    "    )\n",
    "\n",
    "ggsave(filename=\"figures/RF_empirical_set_num_paralogs_boxplot_SM_vs_ML_Bigtree_noref_figure5a.png\", plot=unit_rf_5a, dpi = \"retina\", width = 5, height = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "31abaa34-5ff2-4e36-b744-727eae45fcd8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mNo shared levels found between `names(values)` of the manual scale and the\n",
      "data's \u001b[32mcolour\u001b[39m values.”\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Fig 5b Supertree-ME and Supertree-ML VS Big-Tree ML without ref (no_ref) (aka MOUSE) paralog trees projections(aka species subtrees extracted).\n",
    "#\n",
    "\n",
    "# read the table compiled in chapter 4 cell above. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df = read.table(\"tables/RF_empirical_set_per_family_SM_ST_ME_ML_vs_BigTreeML_no_ref.tsv\", header=T)\n",
    "\n",
    "# retrieve only Suspertree method rows.\n",
    "df_figure5b = subset(df, grepl(\"ST\", method))\n",
    "\n",
    "# Add prefix \"unit_\" to the \"unit\" column\n",
    "df_figure5b$unit = paste0(\"unit_\", df_figure5b$unit)\n",
    "\n",
    "# set limits for plot\n",
    "options(repr.plot.width = 5, repr.plot.height = 5, repr.plot.res = 300)\n",
    "\n",
    "# Define custom colors for each Method\n",
    "custom_colors = c(\"ST_ME\" = \"#7CB6E2\",     # Light blue \n",
    "                  \"ST_ML\" = \"#E5A23D\")     # Orange/yellow\n",
    "\n",
    "# plot the table \n",
    "unit_rf_5b = ggplot(data=df_figure5b, aes(x=factor(num_paralogs), y=mean_rf_val, fill=method)) +\n",
    "    geom_boxplot() + \n",
    "    theme_light() + \n",
    "    ylab(\"Normalized RF\") + xlab(\"Num. paralogs per species\") +\n",
    "    theme(\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size=8),\n",
    "        axis.text.y = element_text(size=8),\n",
    "        axis.title = element_text(face = \"bold\"),\n",
    "        legend.position=\"top\"\n",
    "        ) + \n",
    "    #scale_y_continuous(limits = c(0, 0.51)) +\n",
    "    # Custom legend labels and apply custom colors\n",
    "    scale_color_manual(\n",
    "        name = \"Method\", \n",
    "        labels = c(\"ST_ME\" = \"Supertree-ME\", \"ST_ML\" = \"Supertree-ML\"), # Custom legend labels\n",
    "        values = custom_colors\n",
    "    )\n",
    "\n",
    "ggsave(filename=\"figures/RF_empirical_set_num_paralogs_boxplot_ST_vs_ML_Bigtree_noref_figure5b.png\", plot=unit_rf_5b, dpi = \"retina\", width = 5, height = 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc9b1c45-c2d7-49ac-9cf3-950c644a1d0c",
   "metadata": {},
   "source": [
    "<a id=\"chapter-5-Running-times\"></a>\n",
    "\n",
    "## Chapter 5 - Running times\n",
    "\n",
    "TODO add description "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cb1aad01-55f2-4ed7-902f-d6325f2b66d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO make this into the section of custom functions.\n",
    "# Function to convert to milliseconds\n",
    "convert_to_ms <- function(time_str) {\n",
    "    # Extract components using regular expressions\n",
    "    millis = \"0\"\n",
    "    seconds = \"0\"\n",
    "    minutes = \"0\"\n",
    "    if (str_detect(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=ms)\")) {\n",
    "        millis = str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=ms)\")\n",
    "    } else {\n",
    "        seconds = str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=s)\")\n",
    "        minutes = str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=m)\")\n",
    "    }\n",
    "    \n",
    "    # Convert each component to milliseconds (coercing to numeric where needed)\n",
    "    seconds_ms = as.numeric(seconds) * 1000\n",
    "    minutes_ms = as.numeric(minutes) * 60 * 1000\n",
    "    millis_ms = as.numeric(millis)\n",
    "    \n",
    "    # Sum all components and handle NAs by treating them as 0\n",
    "    total_ms = sum(c(seconds_ms, minutes_ms, millis_ms), na.rm = TRUE)\n",
    "    return(total_ms)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "88ec8178-4f22-497d-b4fd-9a74678152bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# each main.nf output run has a trace file repoprting various information on the run itself\n",
    "tracefile_path = file.path(run_dir, \"trace_cpu_time.txt\")\n",
    "tracefile      = read.table(tracefile_path, sep = \"\\t\", header = TRUE)\n",
    "\n",
    "# MSA computation step is not included in the running time only the step necessary to go from MSA to paralog tree.\n",
    "# the list of process names to extract from th trace file\n",
    "process_names = paste(\n",
    "    \"empirical_data:run_phylo_ML_full_aln_emp\",                     # BigTree ML with ref (41x)\n",
    "    \"empirical_data:extract_species_submsa_ML_emp\",\n",
    "    \"empirical_data:run_phylo_ME_full_aln_emp\",                     # BigTree ME with ref (41x)\n",
    "    \"empirical_data:extract_species_submsa_ME_emp\",\n",
    "    \"empirical_data:superfine_emp \\\\(\",                             # both BigTree ML and BigTree ME with ref (summation) (82x)\n",
    "    \"empirical_data:run_phylo_ML_full_aln_emp_no_ref\",              # BigTree ML without ref (41x)\n",
    "    \"empirical_data:extract_species_submsa_ML_emp_no_ref\",\n",
    "    \"empirical_data:run_phylo_ME_full_aln_emp_no_ref\",              # BigTree ME without ref (41x)\n",
    "    \"empirical_data:extract_species_submsa_ME_emp_no_ref\",\n",
    "    \"empirical_data:superfine_emp_no_ref\",                          # both BigTree ML and BigTree ME without ref (summation) (82x)\n",
    "    \"empirical_data:run_phylo_ML_supermatrix_aln_emp\",              # supermatrix ML (41x)\n",
    "    \"empirical_data:run_phylo_ME_supermatrix_aln_emp\",              # supermatrix ME (41x)\n",
    "    \"empirical_data:only_concatenate_aln_emp\",                      # same for supermatrix ML and supermatrix ME (41x)\n",
    "    \"empirical_data:run_phylo_ML_supertree_aln_emp\",                # supertree ML (246x)\n",
    "    \"empirical_data:run_phylo_ME_supertree_aln_emp\",                # supertree ME (246x)\n",
    "    \"empirical_data:superfine_supertree_emp\",                       # both supertree ML and supertree ME (summation) (82x)\n",
    "    sep=\"|\")\n",
    "\n",
    "# here we parse the trace file to retain only the processes that we want to take the running time (realtime) from\n",
    "tracefile_mod = tracefile[grep(process_names, tracefile$name), c(\"name\", \"realtime\", \"status\")]\n",
    "\n",
    "# among these filter out those that have not completed, meaning those processes that finished with an error and were retried.\n",
    "completed_tracefile = tracefile_mod[tracefile_mod$status %in% c(\"COMPLETED\", \"CACHED\"), ]\n",
    "\n",
    "# trasform from human readable running time to millisecond time\n",
    "completed_tracefile$realtime = sapply(completed_tracefile$realtime, convert_to_ms)\n",
    "\n",
    "# Use regex to extract PFAM family names\n",
    "family_names = str_extract(completed_tracefile$name, \"PF\\\\d{5}\")\n",
    "\n",
    "# Get unique family names and sort human-readably (41)\n",
    "unique_family_names = str_sort(unique(family_names), numeric = TRUE)\n",
    "\n",
    "# Create a new dataframe with row name the family name and the additional columns filled with zeroes for each method\n",
    "running_time_df = data.frame(\n",
    "    family_name       = unique_family_names,\n",
    "    BigTree_ML        = 0,                        # with reference specie as well (7)\n",
    "    BigTree_ME        = 0,\n",
    "    BigTree_ML_no_ref = 0,                        # without reference species (6)\n",
    "    BigTree_ME_no_ref = 0,\n",
    "    Supermatrix_ML    = 0,                        # on unit6 (not using ref MSA)\n",
    "    Supermatrix_ME    = 0,\n",
    "    Supertree_ML      = 0,                        # on unit6 (not using ref MSA)\n",
    "    Supertree_ME      = 0,\n",
    "    row.names         = unique_family_names\n",
    ")\n",
    "\n",
    "# now all processes running times have to be processed and added accordingly into the above df, that has as row the families id and as columns the methods names.\n",
    "# there are many operation to do based on which method and they reflect the nomenclature and the logic of the main.nf pipeline for each method this is why a for loop is necessary.\n",
    "for (line_num in 1:nrow(completed_tracefile)) {\n",
    "\n",
    "    # get the process name of the present row\n",
    "    process_name = str_extract(completed_tracefile[line_num, ]$name, \"(?<=:)[^ ]+\")\n",
    "\n",
    "    # get the family name on which the process run\n",
    "    tmp       = str_extract(completed_tracefile[line_num, ]$name, \"\\\\(([^)]+)\\\\)\")\n",
    "    family_id = str_remove_all(tmp, \"[()]\")\n",
    "\n",
    "    # the if to correctly handle the work to do on each process are in order of how many instance there are of that process.\n",
    "    if (process_name == \"run_phylo_ML_supertree_aln_emp\") {\n",
    "        # handle the case of Supertree - ML\n",
    "        # first get the process for the creation of the single specie paralog tree\n",
    "       \n",
    "        # add the running time to the appropriate df column.\n",
    "        # all supertree running times for the same families are summed toghether (25 values). as if there was no empbarassing parallelization happening.\n",
    "        running_time_df[family_id, ]$Supertree_ML = running_time_df[family_id, ]$Supertree_ML + completed_tracefile[line_num, ]$realtime\n",
    "        \n",
    "    } else if (process_name == \"run_phylo_ME_supertree_aln_emp\") {\n",
    "        # handle the case of Supertree - ME\n",
    "        # first get the process for the creation of the single specie paralog tree, do as above\n",
    "        running_time_df[family_id, ]$Supertree_ME = running_time_df[family_id, ]$Supertree_ME + completed_tracefile[line_num, ]$realtime\n",
    "        \n",
    "    } else if (process_name == \"superfine_supertree_emp\" ) {\n",
    "        # add the superfine step to both supertree ME and ML \n",
    "\n",
    "        # distiguish between ME and ML base on the snippet present in the family name. example: avg_branchlen_0.7_protein_family_001.sub - ME\n",
    "        method_code = sub(\".*\\\\(.* - (.*)\\\\)\", \"\\\\1\", completed_tracefile[line_num, ]$name)\n",
    "\n",
    "        # create the right column name to update the running_time_df\n",
    "        method = paste0(\"Supertree_\", method_code)\n",
    "\n",
    "        # get the family name now\n",
    "        real_family_name = str_extract(completed_tracefile[line_num, ]$name, \"PF\\\\d{5}\")\n",
    "\n",
    "        # sum this running time to the correct supertree method as well\n",
    "        running_time_df[real_family_name, ][[method]] = running_time_df[real_family_name, ][[method]] + completed_tracefile[line_num, ]$realtime\n",
    "        \n",
    "    } else if(process_name == \"run_phylo_ML_supermatrix_aln_emp\") {\n",
    "        # handle supermatrix ML\n",
    "       \n",
    "        # add the running time to create the paralog tree\n",
    "        running_time_df[family_id, ]$Supermatrix_ML = running_time_df[family_id, ]$Supermatrix_ML + completed_tracefile[line_num, ]$realtime\n",
    "        \n",
    "    } else if (process_name == \"run_phylo_ME_supermatrix_aln_emp\") {\n",
    "        # handle supermatrix ME\n",
    "\n",
    "        running_time_df[family_id, ]$Supermatrix_ME = running_time_df[family_id, ]$Supermatrix_ME + completed_tracefile[line_num, ]$realtime\n",
    "    } else if (process_name == \"only_concatenate_aln_emp\") {\n",
    "        # add the running time of the step of concatenating submsa into a single MSA creating the \"supermatrix\" to both supermatrix ME and ML\n",
    "\n",
    "        running_time_df[family_id, ]$Supermatrix_ML = running_time_df[family_id, ]$Supermatrix_ML + completed_tracefile[line_num, ]$realtime\n",
    "        running_time_df[family_id, ]$Supermatrix_ME = running_time_df[family_id, ]$Supermatrix_ME + completed_tracefile[line_num, ]$realtime\n",
    "    \n",
    "    } else if(process_name == \"run_phylo_ML_full_aln_emp\" | process_name == \"extract_species_submsa_ML_emp\") {\n",
    "        # handle BigTree ML\n",
    "\n",
    "        # add the running time of both creating the tree and extracting the species paralogs from it\n",
    "        running_time_df[family_id, ]$BigTree_ML = running_time_df[family_id, ]$BigTree_ML + completed_tracefile[line_num, ]$realtime\n",
    "        \n",
    "    } else if(process_name == \"run_phylo_ME_full_aln_emp\" | process_name == \"extract_species_submsa_ME_emp\") {\n",
    "        # handle  Bigtree ME\n",
    "\n",
    "        running_time_df[family_id, ]$BigTree_ME = running_time_df[family_id, ]$BigTree_ME + completed_tracefile[line_num, ]$realtime\n",
    "        \n",
    "    } else if(process_name == \"superfine_emp\"){\n",
    "        # add the superfine step to both bigtree ME and ML. this is used to merge toghether all paralog trees into a single one like supertrree does.\n",
    "\n",
    "        # distiguish between ME and ML base on the snippet present in the family name. example: avg_branchlen_0.7_protein_family_001.sub - ME\n",
    "        method_code = sub(\".*\\\\(.* - (.*)\\\\)\", \"\\\\1\", completed_tracefile[line_num, ]$name)\n",
    "\n",
    "        # create the right column name to update the running_time_df\n",
    "        method = paste0(\"BigTree_\", method_code)\n",
    "\n",
    "        # get the family name now\n",
    "        real_family_name = str_extract(completed_tracefile[line_num, ]$name, \"PF\\\\d{5}\")\n",
    "\n",
    "        # sum this running time to the correct supertree method as well\n",
    "        running_time_df[real_family_name, ][[method]] = running_time_df[real_family_name, ][[method]] + completed_tracefile[line_num, ]$realtime\n",
    "\n",
    "    } else if(process_name == \"run_phylo_ML_full_aln_emp_no_ref\" | process_name == \"extract_species_submsa_ML_emp_no_ref\") {\n",
    "        # handle BigTree ML without ref\n",
    "\n",
    "        # add the running time of both creating the tree and extracting the species paralogs from it\n",
    "        running_time_df[family_id, ]$BigTree_ML_no_ref = running_time_df[family_id, ]$BigTree_ML_no_ref + completed_tracefile[line_num, ]$realtime\n",
    "        \n",
    "    } else if(process_name == \"run_phylo_ME_full_aln_emp_no_ref\" | process_name == \"extract_species_submsa_ME_emp_no_ref\") {\n",
    "        # handle  Bigtree ME without ref\n",
    "\n",
    "        running_time_df[family_id, ]$BigTree_ME_no_ref = running_time_df[family_id, ]$BigTree_ME_no_ref + completed_tracefile[line_num, ]$realtime\n",
    "        \n",
    "    } else if(process_name == \"superfine_emp_no_ref\"){\n",
    "        # add the superfine step to both bigtree ME and ML. this is used to merge toghether all paralog trees into a single one like supertrree does.\n",
    "\n",
    "        # distiguish between ME and ML base on the snippet present in the family name. example: avg_branchlen_0.7_protein_family_001.sub - ME\n",
    "        method_code = sub(\".*\\\\(.* - (.*)\\\\)\", \"\\\\1\", completed_tracefile[line_num, ]$name)\n",
    "\n",
    "        # create the right column name to update the running_time_df\n",
    "        method = paste0(\"BigTree_\", method_code, \"_no_ref\")\n",
    "\n",
    "        # get the family name now\n",
    "        real_family_name = str_extract(completed_tracefile[line_num, ]$name, \"PF\\\\d{5}\")\n",
    "\n",
    "        # sum this running time to the correct supertree method as well\n",
    "        running_time_df[real_family_name, ][[method]] = running_time_df[real_family_name, ][[method]] + completed_tracefile[line_num, ]$realtime\n",
    "        \n",
    "    } else {\n",
    "        print(\"WARNING: unrecognized process name\")\n",
    "    }\n",
    "}\n",
    "\n",
    "# save the dataframe\n",
    "write.table(running_time_df, file = \"tables/RF_empirical_set_all_methods_running_time_per_family.tsv\", sep=\"\\t\", quote = FALSE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1976c35a-dfaf-4069-8702-6f7559e9c602",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>Product:</strong> 0.012950923507367"
      ],
      "text/latex": [
       "\\textbf{Product:} 0.012950923507367"
      ],
      "text/markdown": [
       "**Product:** 0.012950923507367"
      ],
      "text/plain": [
       "   Product \n",
       "0.01295092 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>Product:</strong> 0.00259445268449518"
      ],
      "text/latex": [
       "\\textbf{Product:} 0.00259445268449518"
      ],
      "text/markdown": [
       "**Product:** 0.00259445268449518"
      ],
      "text/plain": [
       "    Product \n",
       "0.002594453 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>Product:</strong> 0.00546690898051326"
      ],
      "text/latex": [
       "\\textbf{Product:} 0.00546690898051326"
      ],
      "text/markdown": [
       "**Product:** 0.00546690898051326"
      ],
      "text/plain": [
       "    Product \n",
       "0.005466909 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#\n",
    "# Fig 5c all methods ML running time lineplot. for SM and ST it is done on unit 6 with no shuffle. For Bigtree is with/without ref species.\n",
    "# x axis is average family sequence length. Since ios the closest thing that can be compared across all methods.\n",
    "\n",
    "# read the table compiled in chapter 4. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df_fam_info = read.table(\"tables/RF_empirical_set_fam_info.tsv\", sep=\"\\t\", header=T)\n",
    "\n",
    "# read the table compiled in chapter 5 one cell above.\n",
    "df_running_time_ML = read.table(\"tables/RF_empirical_set_all_methods_running_time_per_family.tsv\", sep=\"\\t\", header=T)\n",
    "\n",
    "# multiply the average length per family and nuumber of paralogs per family columns. Getting to a proxy for dimension of matrix MSA.\n",
    "df_running_time_ML[\"Product\"] = df_fam_info$num_col_ML * df_fam_info$num_paralogs * df_fam_info$num_units\n",
    "\n",
    "# filter out BigTree with ref species. it would not be a fair comparison as SM and ST are done on (6) units.\n",
    "df_running_time_ML = df_running_time_ML[, c(\"family_name\", \"BigTree_ML_no_ref\", \"Supermatrix_ML\", \"Supertree_ML\", \"Product\")]\n",
    "\n",
    "# Rename the BigTree column and divide milliseconds to seconds for running times.\n",
    "colnames(df_running_time_ML)[colnames(df_running_time_ML) == \"BigTree_ML_no_ref\"] = \"BigTree_ML\"\n",
    "df_running_time_ML$BigTree_ML     = df_running_time_ML$BigTree_ML / 1000\n",
    "df_running_time_ML$Supermatrix_ML = df_running_time_ML$Supermatrix_ML / 1000\n",
    "df_running_time_ML$Supertree_ML   = df_running_time_ML$Supertree_ML / 1000\n",
    "\n",
    "# Fit linear regression model for BigTree ML\n",
    "model_Bigtree_ML     = lm(BigTree_ML ~ Product, data = df_running_time_ML)\n",
    "slope_Bigtree_ML     = coef(model_Bigtree_ML)[2]  # Extract slope\n",
    "intercept_Bigtree_ML = coef(model_Bigtree_ML)[1]  # Extract intercept\n",
    "\n",
    "# Fit linear regression model for SM + ML\n",
    "model_SM_ML     = lm(Supermatrix_ML ~ Product, data = df_running_time_ML)\n",
    "slope_SM_ML     = coef(model_SM_ML)[2]  # Extract slope\n",
    "intercept_SM_ML = coef(model_SM_ML)[1]  # Extract intercept\n",
    "\n",
    "# Fit linear regression model for ST + ML\n",
    "model_ST_ML     = lm(Supertree_ML ~ Product, data = df_running_time_ML)\n",
    "slope_ST_ML     = coef(model_ST_ML)[2]  # Extract slope\n",
    "intercept_ST_ML = coef(model_ST_ML)[1]  # Extract intercept\n",
    "\n",
    "# print the slopes of this fitted lines. they have a meaning of running time growing factor.\n",
    "slope_Bigtree_ML\n",
    "slope_SM_ML\n",
    "slope_ST_ML\n",
    "\n",
    "# It takes a table of all families as rows (41) and it goes to a table of 246 (41*6 there are 6 methods). \n",
    "# and the column names of the methods are now a value of a new column called variable, so that is renamed to method, as well as the new column holding the running times is called value.\n",
    "df_running_time_ML_metled = melt(df_running_time_ML, id = c(\"family_name\", \"Product\"))\n",
    "colnames(df_running_time_ML_metled) = c(\"name\", \"Product\", \"Method\", \"Time\")\n",
    "\n",
    "# Determine the x-axis range. used to acctually plot the intercept on the plot since\n",
    "x_min = min(df_running_time_ML_metled$Product)\n",
    "x_max = max(df_running_time_ML_metled$Product)\n",
    "\n",
    "# plot the table \n",
    "unit_rf_5c = ggplot(data=df_running_time_ML_metled, aes(x=Product, y=Time, group=Method, color=Method)) +\n",
    "    geom_point() +\n",
    "    geom_segment(x = x_min, \n",
    "                 xend = x_max, \n",
    "                 y = intercept_Bigtree_ML + slope_Bigtree_ML * x_min, \n",
    "                 yend = intercept_Bigtree_ML + slope_Bigtree_ML * x_max, \n",
    "                 color = \"red\", \n",
    "                 linetype = \"dashed\", size = 0.8) +\n",
    "    geom_segment(x = x_min, \n",
    "                 xend = x_max, \n",
    "                 y = intercept_SM_ML + slope_SM_ML * x_min, \n",
    "                 yend = intercept_SM_ML + slope_SM_ML * x_max, \n",
    "                 color = \"dark green\", \n",
    "                 linetype = \"dashed\", size = 0.8) +\n",
    "    geom_segment(x = x_min, \n",
    "                 xend = x_max, \n",
    "                 y = intercept_ST_ML + slope_ST_ML * x_min, \n",
    "                 yend = intercept_ST_ML + slope_ST_ML * x_max, \n",
    "                 color = \"blue\", \n",
    "                 linetype = \"dashed\", size = 0.8) +\n",
    "    theme_light() + \n",
    "    ylab(\"CPU time (sec)\") +\n",
    "    xlab(\"num. columns * num. sequences\") +\n",
    "    theme(\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size=8),\n",
    "        axis.text.y = element_text(size=8),\n",
    "        axis.title = element_text(face = \"bold\"),\n",
    "        legend.position=\"right\"\n",
    "        ) \n",
    "    #scale_y_continuous(limits = c(0, 0.51)) +\n",
    "\n",
    "ggsave(filename=\"figures/CPU_time_per_family_empirical_set_lineplot_ML_SM_ST_Bigtree_noref_figure5c.png\", plot=unit_rf_5c, dpi = \"retina\", width = 5, height = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6e48c35c-7244-417d-808a-7e224445c785",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>Product:</strong> 9.01603519554755e-05"
      ],
      "text/latex": [
       "\\textbf{Product:} 9.01603519554755e-05"
      ],
      "text/markdown": [
       "**Product:** 9.01603519554755e-05"
      ],
      "text/plain": [
       "     Product \n",
       "9.016035e-05 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>Product:</strong> 4.51061490860769e-06"
      ],
      "text/latex": [
       "\\textbf{Product:} 4.51061490860769e-06"
      ],
      "text/markdown": [
       "**Product:** 4.51061490860769e-06"
      ],
      "text/plain": [
       "     Product \n",
       "4.510615e-06 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>Product:</strong> 2.14583116457355e-05"
      ],
      "text/latex": [
       "\\textbf{Product:} 2.14583116457355e-05"
      ],
      "text/markdown": [
       "**Product:** 2.14583116457355e-05"
      ],
      "text/plain": [
       "     Product \n",
       "2.145831e-05 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#\n",
    "# Fig 5d all methods ME running time lineplot. for SM and ST it is done on unit 6 with no shuffle. For Bigtree is with/without ref species.\n",
    "# x axis is average family sequence length. Since ios the closest thing that can be compared across all methods.\n",
    "\n",
    "# read the table compiled in chapter 4. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df_fam_info = read.table(\"tables/RF_empirical_set_fam_info.tsv\", sep=\"\\t\", header=T)\n",
    "\n",
    "# read the table compiled in chapter 5 one cell above.\n",
    "df_running_time_ME = read.table(\"tables/RF_empirical_set_all_methods_running_time_per_family.tsv\", sep=\"\\t\", header=T)\n",
    "\n",
    "# multiply the average length per family and nuumber of paralogs per family columns. Getting to a proxy for dimension of matrix MSA.\n",
    "df_running_time_ME[\"Product\"] = df_fam_info$num_col_ML * df_fam_info$num_paralogs * df_fam_info$num_units\n",
    "\n",
    "# filter out BigTree with ref species. it would nbot be a fair comparison as SM and ST are done on (6) units.\n",
    "df_running_time_ME = df_running_time_ME[, c(\"family_name\", \"BigTree_ME_no_ref\", \"Supermatrix_ME\", \"Supertree_ME\", \"Product\")]\n",
    "\n",
    "# Rename the BigTree column and divide milliseconds to seconds for running times.\n",
    "colnames(df_running_time_ME)[colnames(df_running_time_ME) == \"BigTree_ME_no_ref\"] = \"BigTree_ME\"\n",
    "df_running_time_ME$BigTree_ME     = df_running_time_ME$BigTree_ME / 1000\n",
    "df_running_time_ME$Supermatrix_ME = df_running_time_ME$Supermatrix_ME / 1000\n",
    "df_running_time_ME$Supertree_ME   = df_running_time_ME$Supertree_ME / 1000\n",
    "\n",
    "# Fit linear regression model for BigTree ME\n",
    "model_Bigtree_ME     = lm(BigTree_ME ~ Product, data = df_running_time_ME)\n",
    "slope_Bigtree_ME     = coef(model_Bigtree_ME)[2]  # Extract slope\n",
    "intercept_Bigtree_ME = coef(model_Bigtree_ME)[1]  # Extract intercept\n",
    "\n",
    "# Fit linear regression model for SM + ME\n",
    "model_SM_ME     = lm(Supermatrix_ME ~ Product, data = df_running_time_ME)\n",
    "slope_SM_ME     = coef(model_SM_ME)[2]  # Extract slope\n",
    "intercept_SM_ME = coef(model_SM_ME)[1]  # Extract intercept\n",
    "\n",
    "# Fit linear regression model for ST + ME\n",
    "model_ST_ME     = lm(Supertree_ME ~ Product, data = df_running_time_ME)\n",
    "slope_ST_ME     = coef(model_ST_ME)[2]  # Extract slope\n",
    "intercept_ST_ME = coef(model_ST_ME)[1]  # Extract intercept\n",
    "\n",
    "# print the slopes of this fitted lines. they have a meaning of running time growing factor.\n",
    "slope_Bigtree_ME\n",
    "slope_SM_ME\n",
    "slope_ST_ME\n",
    "\n",
    "# It takes a table of all families as rows (41) and it goes to a table of 246 (41*6 there are 6 methods). \n",
    "# and the column names of the methods are now a value of a new column called variable, so that is renamed to method, as well as the new column holding the running times is called value.\n",
    "df_running_time_ME_metled = melt(df_running_time_ME, id = c(\"family_name\", \"Product\"))\n",
    "colnames(df_running_time_ME_metled) = c(\"name\", \"Product\", \"Method\", \"Time\")\n",
    "\n",
    "# Determine the x-axis range. used to acctually plot the intercept on the plot since\n",
    "x_min = min(df_running_time_ML_metled$Product)\n",
    "x_max = max(df_running_time_ML_metled$Product)\n",
    "\n",
    "# plot the table \n",
    "unit_rf_5d = ggplot(data=df_running_time_ME_metled, aes(x=Product, y=Time, group=Method, color=Method)) +\n",
    "    geom_point() + \n",
    "    geom_segment(x = x_min, \n",
    "                 xend = x_max, \n",
    "                 y = intercept_Bigtree_ME + slope_Bigtree_ME * x_min, \n",
    "                 yend = intercept_Bigtree_ME + slope_Bigtree_ME * x_max, \n",
    "                 color = \"red\", \n",
    "                 linetype = \"dashed\", size = 0.8) +\n",
    "    geom_segment(x = x_min, \n",
    "                 xend = x_max, \n",
    "                 y = intercept_SM_ME + slope_SM_ME * x_min, \n",
    "                 yend = intercept_SM_ME + slope_SM_ME * x_max, \n",
    "                 color = \"dark green\", \n",
    "                 linetype = \"dashed\", size = 0.8) +\n",
    "    geom_segment(x = x_min, \n",
    "                 xend = x_max, \n",
    "                 y = intercept_ST_ME + slope_ST_ME * x_min, \n",
    "                 yend = intercept_ST_ME + slope_ST_ME * x_max, \n",
    "                 color = \"blue\", \n",
    "                 linetype = \"dashed\", size = 0.8) +\n",
    "    theme_light() + \n",
    "    ylab(\"CPU time (sec)\") +\n",
    "    xlab(\"num. columns * num. sequences\") +\n",
    "    theme(\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size=8),\n",
    "        axis.text.y = element_text(size=8),\n",
    "        axis.title = element_text(face = \"bold\"),\n",
    "        legend.position=\"right\"\n",
    "        ) \n",
    "    #scale_y_continuous(limits = c(0, 0.51)) +\n",
    "\n",
    "ggsave(filename=\"figures/CPU_time_per_family_empirical_set_lineplot_ME_SM_ST_Bigtree_noref_figure5d.png\", plot=unit_rf_5d, dpi = \"retina\", width = 5, height = 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3f37687-f7f3-4c01-b127-f533222b7c99",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "<br><br><br>\n",
    "# USEFULL STUFF BUT NOT INCLUDED IN PAPER\n",
    "<br><br><br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee9f9025-138b-44e7-9144-7ede52c9e660",
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Compute RF distances between supermatrix and supertree trees and ML trees from the full alignment of all the seqs\n",
    "# without Mouse MSAs \n",
    "\n",
    "# create the dataframe that will hold RF values for each tree comparison. \n",
    "# there are (7) reference trees for a family in this case. because there are 7 species and each one of them might yeld a diffrent paralog tree.\n",
    "# for this reason the unit_sample trees are compared to all (7) reference trees.  \n",
    "# all columns (except the rf_val column) will work as identifier for that comparison if put toghether.\n",
    "df_all_rf = data.frame(\n",
    "    method    = c(),\n",
    "    family    = c(),\n",
    "    species   = c(),\n",
    "    unit      = c(),\n",
    "    sample    = c(),\n",
    "    rf_val    = c()\n",
    ")\n",
    "\n",
    "# create a dataframe with miscellaneuos informations about the gene family\n",
    "df_fam_info = data.frame(\n",
    "    family_name  = c(),\n",
    "    num_seq      = c(),\n",
    "    num_paralogs = c(),\n",
    "    organism     = c()\n",
    ")\n",
    "\n",
    "\n",
    "# the directory prefix where all analazyd files are. this is created by main.nf pipeline in empirical data mode. \n",
    "seq_prefix = \"results_without_mouse_10_samples/tcoffee/\"\n",
    "\n",
    "# iterate over all family specific output directories\n",
    "for (family in pfam_dirs) {\n",
    "\n",
    "    print(basename(family))\n",
    "\n",
    "    # load the BIgtree MSA sequence names mapping scheme, aka tip 1 is pig_genename.\n",
    "    codefile_full_aln_path = file.path(family, seq_prefix, \"full_aln.code_name\")\n",
    "    codefile_full_aln      = read.table(codefile_full_aln_path)\n",
    "\n",
    "    # load only the mouse MSA sequence names mapping scheme.\n",
    "    codefile_path = file.path(family, seq_prefix, \"file.code_name\")\n",
    "    codefile      = read.table(codefile_path)\n",
    "\n",
    "    # get the codename/acronym for the organism/species in the gene family. \n",
    "    orgs_path = file.path(family, seq_prefix, paste0(basename(family), \".orthologs_org_ids_to_concatenate\"))\n",
    "    orgs      = read.table(orgs_path)\n",
    "\n",
    "    # read the paralog gene names \n",
    "    paralogs_file_path = file.path(family, paste0(basename(family), \".intersecting_genes\"))\n",
    "    paralogs_file      = read.table(paralogs_file_path)\n",
    "\n",
    "    # calculate numbers of: sequence in all family, number of species (7) and number of paralog per specie. \n",
    "    # All these three value are correlated: num_species * pralog_per_species = sequence_in_fam \n",
    "    num_seq_in_fam = length(codefile_full_aln[[1]])\n",
    "    num_of_units   = length(orgs$V1) - 1\n",
    "    num_paralogs   = length(paralogs_file$V1)\n",
    "\n",
    "    # compile the dataframe with family name and organism present\n",
    "    df_fam_info = data.frame(\n",
    "        family_name  = c(df_fam_info$family_name, c(rep(basename(family), times=7))),       # repeat to match the size of organism names\n",
    "        num_seq      = c(df_fam_info$num_seq, c(rep(num_seq_in_fam, times=7))),\n",
    "        num_paralogs = c(df_fam_info$num_paralogs, c(rep(num_paralogs, times=7))),\n",
    "        organism     = c(df_fam_info$organism, orgs[[1]])\n",
    "    )\n",
    "\n",
    "    # Get all units trees (60 = 6 organism/units (7 - maouse) * 10 samples/statistical repeats) for the combination Minimun Evolution (ME) + supermatrix method (SM).\n",
    "    # make it to a binary tree using the multi2di package\n",
    "    all_units_trees_seq_path = file.path(family, seq_prefix, \"all_units.nwk\")\n",
    "    all_units_trees_seq      = read.tree(all_units_trees_seq_path, keep.multi=TRUE)\n",
    "    all_units_trees_seq      = multi2di(all_units_trees_seq)\n",
    "\n",
    "    # do the same as above for Maximum Likelihood (ML) + SM \n",
    "    all_units_raxml_trees_seq_path = file.path(family, seq_prefix, \"all_units_RAxML_trees.nwk\")\n",
    "    all_units_raxml_trees_seq      = read.tree(all_units_raxml_trees_seq_path, keep.multi=TRUE)\n",
    "    all_units_raxml_trees_seq      = multi2di(all_units_raxml_trees_seq)\n",
    "\n",
    "    # do the same as above for ME + Super Tree (ST) method\n",
    "    all_units_superfine_trees_seq_path = file.path(family, seq_prefix, \"all_units_superfine_trees.nwk\")\n",
    "    all_units_superfine_trees_seq      = read.tree(all_units_superfine_trees_seq_path, keep.multi=TRUE)\n",
    "    all_units_superfine_trees_seq      = multi2di(all_units_superfine_trees_seq)\n",
    "\n",
    "    # do the same as above for ML + ST\n",
    "    all_units_raxml_superfine_trees_seq_path = file.path(family, seq_prefix, \"all_units_RAxML_superfine_trees.nwk\")\n",
    "    all_units_raxml_superfine_trees_seq      = read.tree(all_units_raxml_superfine_trees_seq_path, keep.multi=TRUE)\n",
    "    all_units_raxml_superfine_trees_seq      = multi2di(all_units_raxml_superfine_trees_seq)\n",
    "\n",
    "    # it will store all the species refernce paralog trees (7)\n",
    "    all_species_raxml_subtrees_from_full_aln = c()\n",
    "    \n",
    "    # read the BigTree ML \n",
    "    species_raxml_tree_from_full_aln_path = file.path(family, seq_prefix, paste0(\"RAxML_bestTree.\", basename(family), \".domain_sequences_prior_after_intersection_full_coded_raxml.nwk\"))\n",
    "    species_raxml_tree_from_full_aln = read.tree(species_raxml_tree_from_full_aln_path, keep.multi=FALSE)\n",
    "    \n",
    "    # for each tip in the above tree rename accordin to the mapping coding file for ML bigtree\n",
    "    for (j in 1:length(species_raxml_tree_from_full_aln$tip.label)) { \n",
    "        species_raxml_tree_from_full_aln$tip.label[j] = as.character(codefile_full_aln$V1[which(codefile_full_aln$V2 == species_raxml_tree_from_full_aln[[\"tip.label\"]][j])])\n",
    "    }\n",
    "    \n",
    "    # do a for loop on the species (7) \n",
    "    for (species in orgs$V1) {\n",
    "    \n",
    "        # get the list of indices referring to the organism in question\n",
    "        tips_to_keep = grep(species,species_raxml_tree_from_full_aln$tip.label)\n",
    "    \n",
    "        # extract the subtree (paralog tree) from the BigTree ML relating to the species in question\n",
    "        test_subtree = keep.tip(species_raxml_tree_from_full_aln,tips_to_keep)\n",
    "    \n",
    "        # rename each tip sustituting the specific species codename with MOUSE, effectively going from ACAP2_AILME_1 to ACAP2_MOUSE_1\n",
    "        # this makes all tips names from all species subtrees comparable altough they refer to a different sequence.\n",
    "        # this is fine since only the correctness of the topology of the tree is tested later on. \n",
    "        test_subtree$tip.label = gsub(paste0(\"_\", species, \"_\"), \"_MOUSE_\", test_subtree$tip.label)\n",
    "    \n",
    "        # the tips are further mapped to the C!, C2, C3, ecc.. nomenclature. going from  ACAP2_MOUSE_1 to C2\n",
    "        for (j in 1:length(test_subtree$tip.label)) { \n",
    "            test_subtree$tip.label[j] = as.character(codefile$V2[which(codefile$V1 == test_subtree[[\"tip.label\"]][j])])\n",
    "        }\n",
    "        \n",
    "        # trasform the subtree into the ape multiPhylo class and added it to the list with all reference paralog trees.\n",
    "        test_subtree = as.multiPhylo(test_subtree)\n",
    "        all_species_raxml_subtrees_from_full_aln = c(all_species_raxml_subtrees_from_full_aln, test_subtree)\n",
    "    }\n",
    "\n",
    "    #\n",
    "    # SM + ME rf computation block \n",
    "    # \n",
    "    \n",
    "    # iterate over all possible species reference trees\n",
    "    for (r in 1:length(all_species_raxml_subtrees_from_full_aln)) {\n",
    "    \n",
    "        # RF to work needs all trees it needs to compare into a single list (class) and then it will automaticly do all against all comparisons.\n",
    "        # here we are comparing all unit_sample (60) family SM + ME trees with the family (7) species paralog reference trees\n",
    "        trees_to_compare_ME_SM         = c(all_species_raxml_subtrees_from_full_aln[r], all_units_trees_seq)\n",
    "        class(trees_to_compare_ME_SM)  = \"multiPhylo\"\n",
    "        rf_trees_to_compare_ME_SM      = RF.dist(trees_to_compare_ME_SM, normalize=TRUE)\n",
    "    \n",
    "        # transform Rf.dist output format to a matrix and take the first row without the first element. \n",
    "        # Basically retrieve only the comparisons between reference tree and each unit tree\n",
    "        rf_ME_SM_comparisons                = data.matrix(rf_trees_to_compare_ME_SM)[1, -1]\n",
    "    \n",
    "        # for loop over the created vector to correctly update the dataframe. \n",
    "        # the (60) values in the above vector  are from different units and samples in an ordered manner, the for loop is just to \"count\" correctly.\n",
    "        # the dataframe will be updated in chuncks of 10 \n",
    "        for (u in seq(10, 60, by=10)) {\n",
    "        \n",
    "            # update the df with all relevant info. since the rf vector is chuncked in pieces of ten all other things have to be repeated accordingly.\n",
    "            df_all_rf = data.frame(\n",
    "                method    = c(df_all_rf$method, rep(\"SM_ME\", 10)),\n",
    "                family    = c(df_all_rf$family, rep(basename(family), 10)),           # use family name\n",
    "                species   = c(df_all_rf$reference, rep(orgs$V1[r] , 10)),             # for the reference tree get the species which it is from\n",
    "                unit      = c(df_all_rf$unit, rep((u/10), 10)),\n",
    "                sample    = c(df_all_rf$sample, c(1, 2, 3, 4, 5, 6 ,7 ,8, 9, 10)),\n",
    "                rf_val    = c(df_all_rf$rf_val, rf_ME_SM_comparisons[(u-9):u])        # get all sample from the same unit\n",
    "            )\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    #\n",
    "    # ST + ME rf computation block \n",
    "    # \n",
    "    \n",
    "    # repeat the previous block for super tree as well\n",
    "    for (r in 1:length(all_species_raxml_subtrees_from_full_aln)) {\n",
    "        trees_to_compare_ME_SM         = c(all_species_raxml_subtrees_from_full_aln[r], all_units_superfine_trees_seq)\n",
    "        class(trees_to_compare_ME_SM)  = \"multiPhylo\"\n",
    "        rf_trees_to_compare_ME_SM      = RF.dist(trees_to_compare_ME_SM, normalize=TRUE)\n",
    "        rf_ME_SM_comparisons                = data.matrix(rf_trees_to_compare_ME_SM)[1, -1]\n",
    "        for (u in seq(10, 60, by=10)) {\n",
    "            df_all_rf = data.frame(\n",
    "                method    = c(df_all_rf$method, rep(\"ST_ME\", 10)),\n",
    "                family    = c(df_all_rf$family, rep(basename(family), 10)),           # use family name\n",
    "                species   = c(df_all_rf$reference, rep(orgs$V1[r] , 10)),             # for the reference tree get the species which it is from\n",
    "                unit      = c(df_all_rf$unit, rep((u/10), 10)),\n",
    "                sample    = c(df_all_rf$sample, c(1, 2, 3, 4, 5, 6 ,7 ,8, 9, 10)),\n",
    "                rf_val    = c(df_all_rf$rf_val, rf_ME_SM_comparisons[(u-9):u])        # get all sample from the same unit\n",
    "            )\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "# save the dataframe\n",
    "write.table(df_all_rf, file = \"tables/RF_empirical_set_between_ref_supermatrix_and_supertree.tsv\", quote = FALSE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f14901f-e266-4a5f-a080-73c208aaadfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Table 2 Supermatrix-ME and Supertree-ME VS Big-Tree ML no ref.\n",
    "#\n",
    "\n",
    "# define how many units there are \n",
    "orgs_path      = file.path(family, seq_prefix, paste0(basename(family), \".orthologs_org_ids_to_concatenate\"))\n",
    "orgs           = read.table(orgs_path)\n",
    "num_of_species = length(orgs$V1)\n",
    "num_of_units   = length(orgs$V1) - 1\n",
    "\n",
    "# read the table compiled in chapter 4. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df_new = read.table(\"tables/RF_empirical_set_between_Bigtree_ML_no_ref_supermatrix_and_supertree.tsv\", header=T)\n",
    "\n",
    "# Group data by method and unit, then calculate mean and standard deviation for all families \n",
    "summary_table = df_new %>%\n",
    "  group_by(method, unit) %>%\n",
    "  summarise(\n",
    "    mean_rf_val = mean(rf_val, na.rm = TRUE),\n",
    "    sd_rf_val = sd(rf_val, na.rm = TRUE)\n",
    "  ) %>%\n",
    "  ungroup()\n",
    "\n",
    "# print the unit rf table and save it\n",
    "summary_table\n",
    "write.table(summary_table, file = \"tables/RF_empirical_set_per_unit_SM_ST_ME.tsv\", quote = FALSE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c285e4a1-4873-4927-bc2f-bd24472c3b0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Fig 4 Supermatrix-ME and Supertree-ME VS Big-Tree ML\n",
    "#\n",
    "\n",
    "# read the table compiled in chapter 4 cell above. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df_figure4 = read.table(\"tables/RF_empirical_set_per_unit_SM_ST_ME.tsv\", header=T)\n",
    "\n",
    "# Add prefix \"unit_\" to the \"unit\" column\n",
    "df_figure4$unit = paste0(\"unit_\", df_figure4$unit)\n",
    "\n",
    "# set limits for plot\n",
    "options(repr.plot.width = 5, repr.plot.height = 5, repr.plot.res = 300)\n",
    "\n",
    "# plot the table \n",
    "unit_rf = ggplot(data=df_figure4, aes(x=unit, y=mean_rf_val, group=method, color=method)) +\n",
    "    geom_line() + \n",
    "    geom_point() + \n",
    "    theme_light() + \n",
    "    ylab(\"Normalized RF\") + xlab(\"Topology\") +\n",
    "    theme(\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size=8),\n",
    "        axis.text.y = element_text(size=8),\n",
    "        axis.title = element_text(face = \"bold\"),\n",
    "        legend.position=\"right\"\n",
    "        ) + \n",
    "    scale_y_continuous(limits = c(0, 0.51)) +\n",
    "    # Custom legend labels while keeping default colors\n",
    "    scale_color_discrete(\n",
    "        name = \"Method\", \n",
    "        labels = c(\"ST_ME\" = \"Supertree-ME\", \"SM_ME\" = \"Supermatrix-ME\") # Custom legend labels\n",
    "    ) \n",
    "\n",
    "\n",
    "ggsave(filename=\"figures/RF_lineplot_ME_seq_vs_ML_Big_tree.png\", plot=unit_rf, dpi = \"retina\", width = 5, height = 5)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
