{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cfd49a4b-ae8f-4279-a940-d72726d29424",
   "metadata": {},
   "source": [
    "# Analysis of mouse empirical dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32bddeff-6a85-41c2-bae5-1f7160f9e179",
   "metadata": {},
   "source": [
    "TODO add here a brief description of your project or analysis. Explain the purpose, goals, or context of the notebook here.\n",
    "\n",
    "## Table of Contents\n",
    "- [1. Chapter 1 - Introduction](#chapter-1-introduction)\n",
    "- [2. Chapter 2 - R packages](#chapter-2-R-packages)\n",
    "- [3. Chapter 3 - Input specification](#chapter-3-Input-specification)\n",
    "- [4. Chapter 4 - RF distance summaries](#chapter-4-RF-distance-summaries)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8e366a2-65de-425e-9cf8-6a44c1d6fcbc",
   "metadata": {},
   "source": [
    "<a id=\"chapter-1-introduction\"></a>\n",
    "\n",
    "## Chapter 1 - Introduction\n",
    "\n",
    "TODO add description of notebook and how to use it.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c64433ad-f6bc-4891-a3cf-9165d3538d9e",
   "metadata": {},
   "source": [
    "<a id=\"chapter-2-R-packages\"></a>\n",
    "\n",
    "## Chapter 2 - R packages\n",
    "\n",
    "Here is the list of packages needed for this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a1a8e10a-6755-4ca8-8818-f267f292639b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘tidyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:reshape2’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(\"phangorn\")\n",
    "library(\"rlist\")\n",
    "library(\"reshape2\")\n",
    "library(\"ggplot2\")\n",
    "library(\"geiger\")\n",
    "library(\"adephylo\")\n",
    "library(\"phytools\")\n",
    "library(\"hrbrthemes\")\n",
    "#library(\"ggtree\")\n",
    "#library(\"pheatmap\")\n",
    "library(\"gplots\")\n",
    "library(\"ape\")\n",
    "library(\"plotrix\")\n",
    "library(\"plyr\")\n",
    "library(\"dplyr\")\n",
    "library(\"rstatix\")\n",
    "library(\"grid\")\n",
    "library(\"gridExtra\")\n",
    "library(\"ggforce\")\n",
    "library(\"viridis\")\n",
    "library(\"tidyr\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de9ad2d2-464e-4838-86ff-1f4cb300b6bf",
   "metadata": {},
   "source": [
    "<a id=\"chapter-3-Input-specification\"></a>\n",
    "\n",
    "## Chapter 3 - Input specification\n",
    "\n",
    "Change the parent_of_all_dir variable to give the location of the input. All input files are expected to be under it. Otherwise change the single input location to you liking.<br>\n",
    "All input files or directory paths are listed here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3c6c3d6d-1571-44ab-9b01-8ca5097e22b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the location of the directory that contains all inputs for analyses \n",
    "parent_of_all_dir = \"/users/cn/avignoli/1_paralogs/MOUSE_above_40_entries_clean_run/rerun_with_struct_concat/\"\n",
    "\n",
    "# The paths to all PFAM families directories\n",
    "pfam_dirs_paths = paste0(parent_of_all_dir, \"PF*\")\n",
    "\n",
    "# here we get all PFAM directories paths into an iterable used later on in the analysis\n",
    "pfam_dirs = Sys.glob(pfam_dirs_paths)\n",
    "pfam_dirs = pfam_dirs[file.info(pfam_dirs)$isdir]\n",
    "\n",
    "# Create the 'figures' directory if it doesn't exist. Created where this notebook is executed.\n",
    "if (!dir.exists(\"figures\")) {\n",
    "    dir.create(\"figures\")\n",
    "}\n",
    "\n",
    "# Create the 'tables' directory if it doesn't exist. Created where this notebook is executed.\n",
    "if (!dir.exists(\"tables\")) {\n",
    "    dir.create(\"tables\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f0a3763-0bec-4c8c-954d-4188e860d17e",
   "metadata": {},
   "source": [
    "<a id=\"chapter-4-RF-distance-summaries\"></a>\n",
    "\n",
    "## Chapter 4 - RF distance summaries\n",
    "\n",
    "TODO add description "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e36eb9b3-7da6-4b1f-818b-aaf32e99ff44",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"PF00004\"\n",
      "[1] \"PF00019\"\n",
      "[1] \"PF00022\"\n",
      "[1] \"PF00025\"\n",
      "[1] \"PF00041\"\n",
      "[1] \"PF00059\"\n",
      "[1] \"PF00061\"\n",
      "[1] \"PF00092\"\n",
      "[1] \"PF00102\"\n",
      "[1] \"PF00104\"\n",
      "[1] \"PF00106\"\n",
      "[1] \"PF00122\"\n",
      "[1] \"PF00134\"\n",
      "[1] \"PF00167\"\n",
      "[1] \"PF00168\"\n",
      "[1] \"PF00178\"\n",
      "[1] \"PF00179\"\n",
      "[1] \"PF00250\"\n",
      "[1] \"PF00307\"\n",
      "[1] \"PF00373\"\n",
      "[1] \"PF00531\"\n",
      "[1] \"PF00566\"\n",
      "[1] \"PF00595\"\n",
      "[1] \"PF00615\"\n",
      "[1] \"PF00620\"\n",
      "[1] \"PF00621\"\n",
      "[1] \"PF00622\"\n",
      "[1] \"PF00640\"\n",
      "[1] \"PF00777\"\n",
      "[1] \"PF00782\"\n",
      "[1] \"PF00787\"\n",
      "[1] \"PF00788\"\n",
      "[1] \"PF00822\"\n",
      "[1] \"PF01392\"\n",
      "[1] \"PF01412\"\n",
      "[1] \"PF01421\"\n",
      "[1] \"PF02214\"\n",
      "[1] \"PF02931\"\n",
      "[1] \"PF07679\"\n",
      "[1] \"PF08205\"\n",
      "[1] \"PF12796\"\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Compute RF distances between supermatrix and supertree trees and ML trees from the full alignment of all the seqs\n",
    "# without Mouse MSAs and structure-based MSAs\n",
    "\n",
    "# create the dataframe that will hold RF values for each tree comparison. \n",
    "# there are (7) reference trees for a family in this case. because there are 7 species and each one of them might yeld a diffrent paralog tree.\n",
    "# for this reason the unist_sample trees are compared to all (7) reference trees.  \n",
    "# all columns (except the rf_val column) will work as identifier for that comparison if put toghether.\n",
    "df_all_rf = data.frame(\n",
    "    method    = c(),\n",
    "    family    = c(),\n",
    "    species   = c(),\n",
    "    unit      = c(),\n",
    "    sample    = c(),\n",
    "    rf_val    = c()\n",
    ")\n",
    "\n",
    "# create a dataframe with miscellaneuos informations about the gene family\n",
    "df_fam_info = data.frame(\n",
    "    family_name  = c(),\n",
    "    num_seq      = c(),\n",
    "    num_paralogs = c(),\n",
    "    organism     = c()\n",
    ")\n",
    "\n",
    "\n",
    "# the directory prefix where all analazyd files are. this is created by main.nf pipeline in empirical data mode. \n",
    "seq_prefix = \"results_without_mouse_10_samples/tcoffee/\"\n",
    "\n",
    "# iterate over all family specific output directories\n",
    "for (family in pfam_dirs) {\n",
    "\n",
    "    print(basename(family))\n",
    "\n",
    "    # load the BIgtree MSA sequence names mapping scheme, aka tip 1 is pig_genename.\n",
    "    codefile_full_aln_path = file.path(family, seq_prefix, \"full_aln.code_name\")\n",
    "    codefile_full_aln      = read.table(codefile_full_aln_path)\n",
    "\n",
    "    # load only the mouse MSA sequence names mapping scheme.\n",
    "    codefile_path = file.path(family, seq_prefix, \"file.code_name\")\n",
    "    codefile      = read.table(codefile_path)\n",
    "\n",
    "    # get the codename/acronym for the organism/species in the gene family. \n",
    "    orgs_path = file.path(family, seq_prefix, paste0(basename(family), \".orthologs_org_ids_to_concatenate\"))\n",
    "    orgs      = read.table(orgs_path)\n",
    "\n",
    "    # read the paralog gene names \n",
    "    paralogs_file_path = file.path(family, paste0(basename(family), \".intersecting_genes\"))\n",
    "    paralogs_file      = read.table(paralogs_file_path)\n",
    "\n",
    "    # calculate numbers of: sequence in all family, number of species (7) and number of paralog per specie. \n",
    "    # All these three value are correlated: num_species * pralog_per_species = sequence_in_fam \n",
    "    num_seq_in_fam = length(codefile_full_aln[[1]])\n",
    "    num_of_units   = length(orgs$V1) - 1\n",
    "    num_paralogs   = length(paralogs_file$V1)\n",
    "\n",
    "    # compile the dataframe with family name and organism present\n",
    "    df_fam_info = data.frame(\n",
    "        family_name  = c(df_fam_info$family_name, c(rep(basename(family), times=7))),       # repeat to match the size of organism names\n",
    "        num_seq      = c(df_fam_info$num_seq, c(rep(num_seq_in_fam, times=7))),\n",
    "        num_paralogs = c(df_fam_info$num_paralogs, c(rep(num_paralogs, times=7))),\n",
    "        organism     = c(df_fam_info$organism, orgs[[1]])\n",
    "    )\n",
    "\n",
    "    # Get all units trees (60 = 6 organism/units (7 - maouse) * 10 samples/statistical repeats) for the combination Minimun Evolution (ME) + supermatrix method (SM).\n",
    "    # make it to a binary tree using the multi2di package\n",
    "    all_units_trees_seq_path = file.path(family, seq_prefix, \"all_units.nwk\")\n",
    "    all_units_trees_seq      = read.tree(all_units_trees_seq_path, keep.multi=TRUE)\n",
    "    all_units_trees_seq      = multi2di(all_units_trees_seq)\n",
    "\n",
    "    # do the same as above for Maximum Likelihood (ML) + SM \n",
    "    all_units_raxml_trees_seq_path = file.path(family, seq_prefix, \"all_units_RAxML_trees.nwk\")\n",
    "    all_units_raxml_trees_seq      = read.tree(all_units_raxml_trees_seq_path, keep.multi=TRUE)\n",
    "    all_units_raxml_trees_seq      = multi2di(all_units_raxml_trees_seq)\n",
    "\n",
    "    # do the same as above for ME + Super Tree (ST) method\n",
    "    all_units_superfine_trees_seq_path = file.path(family, seq_prefix, \"all_units_superfine_trees.nwk\")\n",
    "    all_units_superfine_trees_seq      = read.tree(all_units_superfine_trees_seq_path, keep.multi=TRUE)\n",
    "    all_units_superfine_trees_seq      = multi2di(all_units_superfine_trees_seq)\n",
    "\n",
    "    # do the same as above for ML + ST\n",
    "    all_units_raxml_superfine_trees_seq_path = file.path(family, seq_prefix, \"all_units_RAxML_superfine_trees.nwk\")\n",
    "    all_units_raxml_superfine_trees_seq      = read.tree(all_units_raxml_superfine_trees_seq_path, keep.multi=TRUE)\n",
    "    all_units_raxml_superfine_trees_seq      = multi2di(all_units_raxml_superfine_trees_seq)\n",
    "\n",
    "    # it will store all the species refernce paralog trees (7)\n",
    "    all_species_raxml_subtrees_from_full_aln = c()\n",
    "    \n",
    "    # read the BigTree ML \n",
    "    species_raxml_tree_from_full_aln_path = file.path(family, seq_prefix, paste0(\"RAxML_bestTree.\", basename(family), \".domain_sequences_prior_after_intersection_full_coded_raxml.nwk\"))\n",
    "    species_raxml_tree_from_full_aln = read.tree(species_raxml_tree_from_full_aln_path, keep.multi=FALSE)\n",
    "    \n",
    "    # for each tip in the above tree rename accordin to the mapping coding file for ML bigtree\n",
    "    for (j in 1:length(species_raxml_tree_from_full_aln$tip.label)) { \n",
    "        species_raxml_tree_from_full_aln$tip.label[j] = as.character(codefile_full_aln$V1[which(codefile_full_aln$V2 == species_raxml_tree_from_full_aln[[\"tip.label\"]][j])])\n",
    "    }\n",
    "    \n",
    "    # do a for loop on the species (7) \n",
    "    for (species in orgs$V1) {\n",
    "    \n",
    "        # get the list of indices referring to the organism in question\n",
    "        tips_to_keep = grep(species,species_raxml_tree_from_full_aln$tip.label)\n",
    "    \n",
    "        # extract the subtree (paralog tree) from the BigTree ML relating to the species in question\n",
    "        test_subtree = keep.tip(species_raxml_tree_from_full_aln,tips_to_keep)\n",
    "    \n",
    "        # rename each tip sustituting the specific species codename with MOUSE, effectively going from ACAP2_AILME_1 to ACAP2_MOUSE_1\n",
    "        # this makes all tips names from all species subtrees comparable altough they refer to a different sequence.\n",
    "        # this is fine since only the correctness of the topology of the tree is tested later on. \n",
    "        test_subtree$tip.label = gsub(paste0(\"_\", species, \"_\"), \"_MOUSE_\", test_subtree$tip.label)\n",
    "    \n",
    "        # the tips are further mapped to the C!, C2, C3, ecc.. nomenclature. going from  ACAP2_MOUSE_1 to C2\n",
    "        for (j in 1:length(test_subtree$tip.label)) { \n",
    "            test_subtree$tip.label[j] = as.character(codefile$V2[which(codefile$V1 == test_subtree[[\"tip.label\"]][j])])\n",
    "        }\n",
    "        \n",
    "        # trasform the subtree into the ape multiPhylo class and added it to the list with all reference paralog trees.\n",
    "        test_subtree = as.multiPhylo(test_subtree)\n",
    "        all_species_raxml_subtrees_from_full_aln = c(all_species_raxml_subtrees_from_full_aln, test_subtree)\n",
    "    }\n",
    "\n",
    "    #\n",
    "    # SM + ME rf computation block \n",
    "    # \n",
    "    \n",
    "    # iterate over all possible species reference trees\n",
    "    for (r in 1:length(all_species_raxml_subtrees_from_full_aln)) {\n",
    "    \n",
    "        # RF to work needs all trees it needs to compare into a single list (class) and then it will automaticly do all against all comparisons.\n",
    "        # here we are comparing all unit_sample (60) family SM + ME trees with the family (7) species paralog reference trees\n",
    "        trees_to_compare_ME_SM         = c(all_species_raxml_subtrees_from_full_aln[r], all_units_trees_seq)\n",
    "        class(trees_to_compare_ME_SM)  = \"multiPhylo\"\n",
    "        rf_trees_to_compare_ME_SM      = RF.dist(trees_to_compare_ME_SM, normalize=TRUE)\n",
    "    \n",
    "        # transform Rf.dist output format to a matrix and take the first row without the first element. \n",
    "        # Basically retrieve only the comparisons between reference tree and each unit tree\n",
    "        rf_ME_SM_comparisons                = data.matrix(rf_trees_to_compare_ME_SM)[1, -1]\n",
    "    \n",
    "        # for loop over the created vector to correctly update the dataframe. \n",
    "        # the (60) values in the above vector  are from different units and samples in an ordered manner, the for loop is just to \"count\" correctly.\n",
    "        # the dataframe will be updated in chuncks of 10 \n",
    "        for (u in seq(10, 60, by=10)) {\n",
    "        \n",
    "            # update the df with all relevant info. since the rf vector is chuncked in pieces of ten all other things have to be repeated accordingly.\n",
    "            df_all_rf = data.frame(\n",
    "                method    = c(df_all_rf$method, rep(\"SM_ME\", 10)),\n",
    "                family    = c(df_all_rf$family, rep(basename(family), 10)),           # use family name\n",
    "                species   = c(df_all_rf$reference, rep(orgs$V1[r] , 10)),             # for the reference tree get the species which it is from\n",
    "                unit      = c(df_all_rf$unit, rep((u/10), 10)),\n",
    "                sample    = c(df_all_rf$sample, c(1, 2, 3, 4, 5, 6 ,7 ,8, 9, 10)),\n",
    "                rf_val    = c(df_all_rf$rf_val, rf_ME_SM_comparisons[(u-9):u])        # get all sample from the same unit\n",
    "            )\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    #\n",
    "    # ST + ME rf computation block \n",
    "    # \n",
    "    \n",
    "    # repeat the previous block for super tree as well\n",
    "    for (r in 1:length(all_species_raxml_subtrees_from_full_aln)) {\n",
    "        trees_to_compare_ME_SM         = c(all_species_raxml_subtrees_from_full_aln[r], all_units_superfine_trees_seq)\n",
    "        class(trees_to_compare_ME_SM)  = \"multiPhylo\"\n",
    "        rf_trees_to_compare_ME_SM      = RF.dist(trees_to_compare_ME_SM, normalize=TRUE)\n",
    "        rf_ME_SM_comparisons                = data.matrix(rf_trees_to_compare_ME_SM)[1, -1]\n",
    "        for (u in seq(10, 60, by=10)) {\n",
    "            df_all_rf = data.frame(\n",
    "                method    = c(df_all_rf$method, rep(\"ST_ME\", 10)),\n",
    "                family    = c(df_all_rf$family, rep(basename(family), 10)),           # use family name\n",
    "                species   = c(df_all_rf$reference, rep(orgs$V1[r] , 10)),             # for the reference tree get the species which it is from\n",
    "                unit      = c(df_all_rf$unit, rep((u/10), 10)),\n",
    "                sample    = c(df_all_rf$sample, c(1, 2, 3, 4, 5, 6 ,7 ,8, 9, 10)),\n",
    "                rf_val    = c(df_all_rf$rf_val, rf_ME_SM_comparisons[(u-9):u])        # get all sample from the same unit\n",
    "            )\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "# save the dataframe\n",
    "write.table(df_all_rf, file = \"tables/RF_empirical_set_between_ref_supermatrix_and_supertree.tsv\", quote = FALSE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "09c3d96f-4e70-49e2-90b6-1d8499d92a0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'method'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 12 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>method</th><th scope=col>unit</th><th scope=col>mean_rf_val</th><th scope=col>sd_rf_val</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>SM_ME</td><td>1</td><td>0.4813773</td><td>0.2207796</td></tr>\n",
       "\t<tr><td>SM_ME</td><td>2</td><td>0.4252749</td><td>0.2183687</td></tr>\n",
       "\t<tr><td>SM_ME</td><td>3</td><td>0.3943477</td><td>0.2192552</td></tr>\n",
       "\t<tr><td>SM_ME</td><td>4</td><td>0.3688909</td><td>0.2233997</td></tr>\n",
       "\t<tr><td>SM_ME</td><td>5</td><td>0.3500012</td><td>0.2273973</td></tr>\n",
       "\t<tr><td>SM_ME</td><td>6</td><td>0.3413522</td><td>0.2263561</td></tr>\n",
       "\t<tr><td>ST_ME</td><td>1</td><td>0.4813773</td><td>0.2207796</td></tr>\n",
       "\t<tr><td>ST_ME</td><td>2</td><td>0.4558193</td><td>0.2256991</td></tr>\n",
       "\t<tr><td>ST_ME</td><td>3</td><td>0.4220801</td><td>0.2225704</td></tr>\n",
       "\t<tr><td>ST_ME</td><td>4</td><td>0.4013759</td><td>0.2250145</td></tr>\n",
       "\t<tr><td>ST_ME</td><td>5</td><td>0.3896059</td><td>0.2241575</td></tr>\n",
       "\t<tr><td>ST_ME</td><td>6</td><td>0.3771120</td><td>0.2262908</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 12 × 4\n",
       "\\begin{tabular}{llll}\n",
       " method & unit & mean\\_rf\\_val & sd\\_rf\\_val\\\\\n",
       " <chr> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t SM\\_ME & 1 & 0.4813773 & 0.2207796\\\\\n",
       "\t SM\\_ME & 2 & 0.4252749 & 0.2183687\\\\\n",
       "\t SM\\_ME & 3 & 0.3943477 & 0.2192552\\\\\n",
       "\t SM\\_ME & 4 & 0.3688909 & 0.2233997\\\\\n",
       "\t SM\\_ME & 5 & 0.3500012 & 0.2273973\\\\\n",
       "\t SM\\_ME & 6 & 0.3413522 & 0.2263561\\\\\n",
       "\t ST\\_ME & 1 & 0.4813773 & 0.2207796\\\\\n",
       "\t ST\\_ME & 2 & 0.4558193 & 0.2256991\\\\\n",
       "\t ST\\_ME & 3 & 0.4220801 & 0.2225704\\\\\n",
       "\t ST\\_ME & 4 & 0.4013759 & 0.2250145\\\\\n",
       "\t ST\\_ME & 5 & 0.3896059 & 0.2241575\\\\\n",
       "\t ST\\_ME & 6 & 0.3771120 & 0.2262908\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 12 × 4\n",
       "\n",
       "| method &lt;chr&gt; | unit &lt;int&gt; | mean_rf_val &lt;dbl&gt; | sd_rf_val &lt;dbl&gt; |\n",
       "|---|---|---|---|\n",
       "| SM_ME | 1 | 0.4813773 | 0.2207796 |\n",
       "| SM_ME | 2 | 0.4252749 | 0.2183687 |\n",
       "| SM_ME | 3 | 0.3943477 | 0.2192552 |\n",
       "| SM_ME | 4 | 0.3688909 | 0.2233997 |\n",
       "| SM_ME | 5 | 0.3500012 | 0.2273973 |\n",
       "| SM_ME | 6 | 0.3413522 | 0.2263561 |\n",
       "| ST_ME | 1 | 0.4813773 | 0.2207796 |\n",
       "| ST_ME | 2 | 0.4558193 | 0.2256991 |\n",
       "| ST_ME | 3 | 0.4220801 | 0.2225704 |\n",
       "| ST_ME | 4 | 0.4013759 | 0.2250145 |\n",
       "| ST_ME | 5 | 0.3896059 | 0.2241575 |\n",
       "| ST_ME | 6 | 0.3771120 | 0.2262908 |\n",
       "\n"
      ],
      "text/plain": [
       "   method unit mean_rf_val sd_rf_val\n",
       "1  SM_ME  1    0.4813773   0.2207796\n",
       "2  SM_ME  2    0.4252749   0.2183687\n",
       "3  SM_ME  3    0.3943477   0.2192552\n",
       "4  SM_ME  4    0.3688909   0.2233997\n",
       "5  SM_ME  5    0.3500012   0.2273973\n",
       "6  SM_ME  6    0.3413522   0.2263561\n",
       "7  ST_ME  1    0.4813773   0.2207796\n",
       "8  ST_ME  2    0.4558193   0.2256991\n",
       "9  ST_ME  3    0.4220801   0.2225704\n",
       "10 ST_ME  4    0.4013759   0.2250145\n",
       "11 ST_ME  5    0.3896059   0.2241575\n",
       "12 ST_ME  6    0.3771120   0.2262908"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#\n",
    "# Table 2 Supermatrix-ME and Supertree-ME VS Big-Tree ML\n",
    "#\n",
    "\n",
    "# define how many units there are \n",
    "orgs_path      = file.path(family, seq_prefix, paste0(basename(family), \".orthologs_org_ids_to_concatenate\"))\n",
    "orgs           = read.table(orgs_path)\n",
    "num_of_species = length(orgs$V1)\n",
    "num_of_units   = length(orgs$V1) - 1\n",
    "\n",
    "# read the table compiled in chapter 4. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df_new = read.table(\"tables/RF_empirical_set_between_ref_supermatrix_and_supertree.tsv\", header=T)\n",
    "\n",
    "# Group data by method and unit, then calculate mean and standard deviation for all families \n",
    "summary_table = df_new %>%\n",
    "  group_by(method, unit) %>%\n",
    "  summarise(\n",
    "    mean_rf_val = mean(rf_val, na.rm = TRUE),\n",
    "    sd_rf_val = sd(rf_val, na.rm = TRUE)\n",
    "  ) %>%\n",
    "  ungroup()\n",
    "\n",
    "# print the unit rf table and save it\n",
    "summary_table\n",
    "write.table(summary_table, file = \"tables/RF_empirical_set_per_unit_SM_ST_ME.tsv\", quote = FALSE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "1dcf7904-9d4d-44fb-8f7e-b86da0ade9e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Fig 4 Supermatrix-ME and Supertree-ME VS Big-Tree ML\n",
    "#\n",
    "\n",
    "# read the table compiled in chapter 4 cell above. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df_figure4 = read.table(\"tables/RF_empirical_set_per_unit_SM_ST_ME.tsv\", header=T)\n",
    "\n",
    "# Add prefix \"unit_\" to the \"unit\" column\n",
    "df_figure4$unit = paste0(\"unit_\", df_figure4$unit)\n",
    "\n",
    "# set limits for plot\n",
    "options(repr.plot.width = 5, repr.plot.height = 5, repr.plot.res = 300)\n",
    "\n",
    "# plot the table \n",
    "unit_rf = ggplot(data=df_figure4, aes(x=unit, y=mean_rf_val, group=method, color=method)) +\n",
    "    geom_line() + \n",
    "    geom_point() + \n",
    "    theme_light() + \n",
    "    ylab(\"Normalized RF\") + xlab(\"Topology\") +\n",
    "    theme(\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size=8),\n",
    "        axis.text.y = element_text(size=8),\n",
    "        axis.title = element_text(face = \"bold\"),\n",
    "        legend.position=\"right\"\n",
    "        ) + \n",
    "    scale_y_continuous(limits = c(0, 0.51)) +\n",
    "    # Custom legend labels while keeping default colors\n",
    "    scale_color_discrete(\n",
    "        name = \"Method\", \n",
    "        labels = c(\"ST_ME\" = \"Supertree-ME\", \"SM_ME\" = \"Supermatrix-ME\") # Custom legend labels\n",
    "    ) \n",
    "\n",
    "\n",
    "ggsave(filename=\"figures/RF_lineplot_ME_seq_vs_ML_Big_tree.png\", plot=unit_rf, dpi = \"retina\", width = 5, height = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa359065-c348-4b1e-9c6c-99901ea2b2a1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
