{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8d6c408c-3b82-4cbb-b94a-cb9728c78fe6",
   "metadata": {},
   "source": [
    "# Analysis of simulated data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "606b5d7a-8d48-4caf-8c72-e89dce7c1208",
   "metadata": {},
   "source": [
    "\n",
    "TODO add here a brief description of your project or analysis. Explain the purpose, goals, or context of the notebook here.\n",
    "\n",
    "## Table of Contents\n",
    "- [1. Chapter 1 - Introduction](#chapter-1-introduction)\n",
    "- [2. Chapter 2 - R packages](#chapter-2-R-packages)\n",
    "- [3. Chapter 3 - Input specification](#chapter-3-Input-specification)\n",
    "- [4. Chapter 4 - Supermatrix and Supertree trees VS true topologies](#chapter-4-Supermatrix-and-Supertree-trees-VS-true-topologies)\n",
    "- [5. Chapter 5 - BigTree VS true topologies](#chapter-5-BigTree-VS-true-topologies)\n",
    "- [6. Chapter 6 - RF distance summaries](#chapter-6-RF-distance-summaries)\n",
    "    - [6.1 Subchapter 6.1 - Table 1](#subchapter-6_1-Table-1)\n",
    "    - [6.2 Subchapter 6.2 - Figure 2](#subchapter-6_2-Figure-2)\n",
    "- [7. Chapter 7 - Running times](#chapter-7-Running-times)\n",
    "    - [7.1 Subchapter 7.1 - Figure 3](#subchapter-7_1-Figure-3)\n",
    "- [8. Chapter 8 - Bootstrap analysis](#chapter-8-Bootstrap-analysis)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d203093e-da42-45a7-8e98-853d4403630a",
   "metadata": {},
   "source": [
    "<a id=\"chapter-1-introduction\"></a>\n",
    "\n",
    "## Chapter 1 - Introduction\n",
    "\n",
    "TODO add description of notebook and how to use it.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3912a30d-42ec-4e90-99e6-8bacfeb50512",
   "metadata": {},
   "source": [
    "<a id=\"chapter-2-R-packages\"></a>\n",
    "\n",
    "## Chapter 2 - R packages\n",
    "\n",
    "Here is the list of packages needed for this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "67058cec-ea5e-4ce3-be29-faf428919086",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"phangorn\")\n",
    "library(\"ggplot2\")\n",
    "library(\"reshape2\")\n",
    "library(\"geiger\")\n",
    "library(\"adephylo\")\n",
    "library(\"phytools\")\n",
    "library(\"rlist\")\n",
    "library(\"plotrix\")\n",
    "library(\"plyr\")\n",
    "library(\"dplyr\")\n",
    "library(\"hrbrthemes\")\n",
    "library(\"rstatix\")\n",
    "library(\"gplots\")\n",
    "library(\"viridis\")\n",
    "library('gtools')\n",
    "library(gridExtra)\n",
    "library(grid)\n",
    "library(\"ggforce\")\n",
    "library(\"scales\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b609768d-d957-4b11-83ce-c27152a96fb2",
   "metadata": {},
   "source": [
    "TODO make a chapter for aqll custom functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20bb020e-4f1d-4338-ac5a-abecc8a06121",
   "metadata": {},
   "source": [
    "<a id=\"chapter-3-Input-specification\"></a>\n",
    "\n",
    "## Chapter 3 - Input specification\n",
    "\n",
    "Change the parent_of_all_dir variable to give the location of the input. All input files are expected to be under it. Otherwise change the single input location to you liking.<br>\n",
    "All input files or directory paths are listed here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "04187777-5e20-43b3-9f39-62523fc0fc0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the location of the directory that contains all inputs for analyses \n",
    "parent_of_all_dir = \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/\"\n",
    "\n",
    "# The directory (one or more) that contain all the results from the main.nf pipeline run. It can be a glob path.\n",
    "dir_containing_all_pipe_results = \"../results\" # todo make this disappear once analysis are ready\n",
    "#dir_containing_all_pipe_results = paste0(parent_of_all_dir, \"rerun_25_species_corrected_branches/rerun_sim_b_factor*\")\n",
    "\n",
    "#  TODO describe what this file does\n",
    "reference_tree_codefile = file.path(parent_of_all_dir, \"ref_tree.code\")\n",
    "\n",
    "# get the species/organism list of names used by the main.nf pipeline\n",
    "species_name_file = file.path(parent_of_all_dir, \"orthologs_org_ids_to_concatenate\")\n",
    "\n",
    "\n",
    "# each run of main.nf has one output directory. here we get that or multiple runs outputs directories.\n",
    "run_dirs = Sys.glob(dir_containing_all_pipe_results)\n",
    "run_dirs = run_dirs[file.info(run_dirs)$isdir]\n",
    "\n",
    "# Create the 'figures' directory if it doesn't exist. Created where this notebook is executed.\n",
    "if (!dir.exists(\"figures\")) {\n",
    "    dir.create(\"figures\")\n",
    "}\n",
    "\n",
    "# Create the 'tables' directory if it doesn't exist. Created where this notebook is executed.\n",
    "if (!dir.exists(\"tables\")) {\n",
    "    dir.create(\"tables\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "566da14f-4b2d-4610-9a9a-6f892dfc9964",
   "metadata": {},
   "source": [
    "<a id=\"chapter-4-Supermatrix-and-Supertree-trees-VS-true-topologies\"></a>\n",
    "\n",
    "## Chapter 4 - Supermatrix and Supertree trees VS true topologies\n",
    "\n",
    "TODO add description "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c707c5c4-51d2-4b25-a6cc-b8a9ccb9e0aa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n"
     ]
    }
   ],
   "source": [
    "# setting global lists that eill hold all family level values\n",
    "RF_avg_list = list()\n",
    "collect_branch_len_supermatrix_ml = c()\n",
    "collect_diameter_supermatrix_ml = c()\n",
    "\n",
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    # the mapping used later on to rename species trees according to a more generic sequence name\n",
    "    ref_codefile = read.table(file = reference_tree_codefile )\n",
    "    \n",
    "    # get al the family subdirectories in each mqin.nf output run dir\n",
    "    fam_dirs = grep(\"avg_branchlen_0.7_protein_family*\", list.dirs(run, recursive = FALSE), value = TRUE)\n",
    "    count_fam = 0\n",
    "\n",
    "    # this will contain the mean branch length for all true/reference paralog trees for each gene family\n",
    "    avg_branch_length = c()\n",
    "\n",
    "    # read the names of the ortholog species.\n",
    "    orgs = read.table(file = species_name_file)\n",
    "\n",
    "    # get all units trees for all combination of methods \n",
    "    for (m in fam_dirs) {\n",
    "        \n",
    "        count_fam = count_fam + 1\n",
    "\n",
    "        # print every 10 families a check on progress to screen\n",
    "        if (count_fam %% 10 == 0) {\n",
    "            print(m)\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250 = 25 organism * 10 samples/statistical repeats) for the combination Minimun Evolution (ME) + supermatrix method (SM). \n",
    "        # We unroot every tree and fully resolve it to use the RF.dist package and others. \n",
    "        all_units_file  = file.path(m, \"results/rerun\", \"all_units.nwk\")\n",
    "        all_units_trees = read.tree(all_units_file, keep.multi=TRUE)\n",
    "        all_units_trees = unroot(all_units_trees)\n",
    "        \n",
    "        # check if there any non-bynary (consensus trees) among the 250 trees present in the file above\n",
    "        if (length(which(is.binary(all_units_trees) == FALSE)) != 0) {\n",
    "            \n",
    "            # if there are non-binary trees, iterate through them and transformed into a binary one using multi2di package\n",
    "            for (tree in which(is.binary(all_units_trees) == FALSE)) {\n",
    "                all_units_trees[[tree]] = multi2di(all_units_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250) ME + supertree (ST) method and handle non-binary trees as above\n",
    "        all_superfine_file  = file.path(m, \"results/rerun\", \"all_units_superfine_trees.nwk\")\n",
    "        all_superfine_trees = read.tree(all_superfine_file, keep.multi=TRUE)\n",
    "        all_superfine_trees = unroot(all_superfine_trees)\n",
    "        if (length(which(is.binary(all_superfine_trees) == FALSE)) != 0) {\n",
    "            for (tree in which(is.binary(all_superfine_trees) == FALSE)) {\n",
    "                all_superfine_trees[[tree]] = multi2di(all_superfine_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250) Maximum Likelihood (ML) + SM \n",
    "        all_units_raxml_file  = file.path(m, \"results/rerun\", \"all_units_RAxML_trees.nwk\")\n",
    "        all_units_raxml_trees = read.tree(all_units_raxml_file, keep.multi=TRUE)\n",
    "        all_units_raxml_trees = unroot(all_units_raxml_trees)\n",
    "        if (length(which(is.binary(all_units_raxml_trees) == FALSE)) != 0) {\n",
    "            for (tree in which(is.binary(all_units_raxml_trees) == FALSE)) {\n",
    "                all_units_raxml_trees[[tree]] = multi2di(all_units_raxml_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # take only the last 10 of the 250 above trees (unit25's 10 samples/replicates)(unit25 = all 25 single units concatenated) and collect branch lengths (each branch in tree, 27 in total 2*15-3) and diameter (using patristic distance)\n",
    "        for(mytree in 241:length(all_units_raxml_trees)) {\n",
    "            collect_branch_len_supermatrix_ml = c(collect_branch_len_supermatrix_ml, all_units_raxml_trees[[mytree]]$edge.length)\n",
    "            collect_diameter_supermatrix_ml = c(collect_diameter_supermatrix_ml, max(distTips(all_units_raxml_trees[[mytree]], method = 'patristic')))\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250) ML + ST\n",
    "        all_raxml_superfine_file  = file.path(m, \"results/rerun\", \"all_units_RAxML_superfine_trees.nwk\")\n",
    "        all_raxml_superfine_trees = read.tree(all_raxml_superfine_file, keep.multi=TRUE)\n",
    "        all_raxml_superfine_trees = unroot(all_raxml_superfine_trees)\n",
    "        if (length(which(is.binary(all_raxml_superfine_trees) == FALSE)) != 0) {\n",
    "            for (tree in which(is.binary(all_raxml_superfine_trees) == FALSE)) {\n",
    "                all_raxml_superfine_trees[[tree]] = multi2di(all_raxml_superfine_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # Reference tree block, each gene family has it's own reference tree that is the one generated by the simulation step. Along this tree all sequences are generated and then alligned and passed as input for main.nf pipeline.\n",
    "        ref_tree_file = file.path(run, paste0(basename(m), \".nwk\"))\n",
    "        ref_tree      = read.tree(ref_tree_file, keep.multi=FALSE)\n",
    "\n",
    "        # Since each gene family has a predifined gene tree (ref_tree) composed of a fixed species tree pasted as leaf of the paralog tree. The tree is designed like this by choice. Each gene families varies only in the paralog tree that it has that is unique (15 tips).\n",
    "        # Here we extract the paralog tree and to do so we just need to extract the subtree with all the sequences of omne species. Since each species have the same paralog tree. So we can extract the subtree of whichever species it does not matter.\n",
    "        ref_subtree   = keep.tip(ref_tree, as.character(ref_codefile$V1))\n",
    "\n",
    "        # we then rename the tips of the paralog tree (ref_subtree) to a more generic keyname. Effectively going from rat1 to seq1. \n",
    "        for (j in 1:length(ref_subtree$tip.label)) { \n",
    "            ref_subtree$tip.label[j] = as.character(ref_codefile$V2[which(ref_codefile$V1 == ref_subtree[[\"tip.label\"]][j])])\n",
    "        }\n",
    "\n",
    "        # Unroot the tree and perform aother operation and then save the paralog tree to the specifc gene faimily directory\n",
    "        ref_subtree        = unroot(ref_subtree)\n",
    "        ref_subtree        = list(ref_subtree)\n",
    "        class(ref_subtree) = \"multiPhylo\"\n",
    "        subtree_out_file   = file.path(m, \"results/rerun\", paste0(\"ref_subtree_\", basename(m), \".nwk\"))\n",
    "        # TODO remove this comment write.tree(ref_subtree, file=subtree_out_file)\n",
    "\n",
    "        # update on the average branch length with the mean of the true/refernce paralog tree of the given gene family\n",
    "        avg_branch_length = c(avg_branch_length, mean(ref_subtree[[1]]$edge.length))\n",
    "\n",
    "        # Make a list containing all trees gathered so far and compute all possible pair comparisons among trees. There are 1001*1001-1001 unique pairewise comparisons ( a square matrix of size 1001 with 0 in diagonal ).\n",
    "        # On this matrix (plotdata) all other measuramnts are done. This is a symmetric matrix (distance matrix)\n",
    "        all_trees        = c(ref_subtree, all_units_trees, all_superfine_trees, all_units_raxml_trees, all_raxml_superfine_trees)\n",
    "        class(all_trees) = \"multiPhylo\"\n",
    "        all_vs_all_rf    = RF.dist(all_trees, normalize = TRUE)\n",
    "        plotdata         = data.matrix(all_vs_all_rf)\n",
    "\n",
    "        # all ME metrics to be computed superfine = SuperTree\n",
    "        mean_rf           = c()\n",
    "        mean_superfine_rf = c()\n",
    "\n",
    "        # all ML metrics to be computed\n",
    "        mean_raxml_rf           = c()\n",
    "        mean_raxml_superfine_rf = c()\n",
    "\n",
    "        # proceed throught the plotdata matrix taking from the second element one every ten up to 242. Basically proceed at chunkcs of 10 (samples). On those chuncks compute the mean and the Standard error. \n",
    "        # average all sample \"values\" whithin a unit. going from 250 value to 25 averages\n",
    "        # This iterate only through the values of the comparison between the reference paralofg tree (ref_subtree)(first column/row of matrix) and the 250 paraogs trees of ME + SM (all_units_trees)\n",
    "        for (i in seq(2, (length(orgs$V1)*10)+1, by=10)) {\n",
    "            mean_rf = c(mean_rf, mean(plotdata[1, i:(i+9)]))\n",
    "        }\n",
    "\n",
    "        # This instead goes from row 252 to 492 of the plotdata matrix. It still goes in chuncks of 10 (samples). \n",
    "        # So this effectively does what is done above but between the reference paralofg tree and the 250 paraogs trees of ME + ST (all_superfine_trees)\n",
    "        for (i in seq( (length(orgs$V1)*10)+2, (2*(length(orgs$V1)*10)+1), by=10 )) {\n",
    "            mean_superfine_rf = c(mean_superfine_rf, mean(plotdata[1,i:(i+9)]))\n",
    "        }\n",
    "\n",
    "        # This for loop goes from 502 to 742. It still goes in chuncks of 10 (samples).\n",
    "        # So this effectively does what is done above but between the reference paralofg tree and the 250 paraogs trees of ML + SM (all_units_raxml_trees)\n",
    "        for (i in seq( (2*(length(orgs$V1)*10))+2, (3*(length(orgs$V1)*10)+1), by=10 )) {\n",
    "            mean_raxml_rf = c(mean_raxml_rf, mean(plotdata[1,i:(i+9)]))\n",
    "        }\n",
    "\n",
    "        # This for loop goes from 752 to 992. It still goes in chuncks of 10 (samples).\n",
    "        # So this effectively does what is done above but between the reference paralofg tree and the 250 paraogs trees of ML + ST (all_raxml_superfine_trees)\n",
    "        for (i in seq( (3*(length(orgs$V1)*10))+2, (4*(length(orgs$V1)*10)+1), by=10)) {\n",
    "            mean_raxml_superfine_rf = c(mean_raxml_superfine_rf, mean(plotdata[1,i:(i+9)]))\n",
    "        }\n",
    "\n",
    "        # collect all mean RF values 25 mean_rf, 25 mean_superfine_rf ecc.. for each family. One for each unit.\n",
    "        collection_mean_rf      = c(mean_rf, mean_superfine_rf, mean_raxml_rf, mean_raxml_superfine_rf)\n",
    "  \n",
    "        # Finally update the \"global\" lists created at the beginning\n",
    "        RF_avg_list      = list.append(RF_avg_list, collection_mean_rf)\n",
    "\n",
    "    }\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9ac26094-fda0-489f-af78-a70716514196",
   "metadata": {},
   "outputs": [],
   "source": [
    "# transform the list of lists (RF_avg_list) into a dataframe (matrix) size 500 * 100 (25 units for all combinations of ME,ML and SM,ST)\n",
    "RF_avg_as_df = ldply(RF_avg_list, recursive=FALSE)\n",
    "\n",
    "# create the column names for the later dataframe, ME = minimum Evolutio, ML = Maximum Likelihood, SM = SuperMatrix , ST = SuperTree. names will be like ME_SM_unit_1\n",
    "units_names = c()\n",
    "for (i in 1:length(orgs$V1)) {\n",
    "    units_names = c(units_names, paste(\"ME_SM\", \"unit\", i, sep=\"_\"), \n",
    "                    paste(\"ME_ST\", \"unit\", i, sep=\"_\"), \n",
    "                    paste(\"ML_SM\", \"unit\", i, sep=\"_\"), \n",
    "                    paste(\"ML_ST\", \"unit\", i, sep=\"_\"))\n",
    "}\n",
    "\n",
    "# natural_sort them or human readable sort, so that the order is  ME_SM unit1...unit25, ME_ST unit1....uniy25, ML_SM and ML_ST\n",
    "units_names = mixedsort(units_names)\n",
    "\n",
    "# TODO see if standar error se was actually usefull\n",
    "RF_df = round(RF_avg_as_df, digits=3)\n",
    "\n",
    "# add the units name generated before as column names for the dataframe and save it to file\n",
    "colnames(RF_df) = units_names\n",
    "write.table(RF_df, file = \"tables/avg_RF_between_ref_supermatrix_and_supertree.tsv\", quote = FALSE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f8b0912-095a-4aef-9277-f40429e45a5e",
   "metadata": {},
   "source": [
    "<a id=\"chapter-5-BigTree-VS-true-topologies\"></a>\n",
    "\n",
    "## Chapter 5 BigTree VS true topologies\n",
    "\n",
    "TODO add description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "ce413952-5667-4c45-8d24-5a578c3527a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"results\"\n",
      "[1] \"../results/avg_branchlen_0.7_protein_family_100.sub\"\n"
     ]
    }
   ],
   "source": [
    "# Trees from BigTree (full_aln) ME/ML (paralog tree) VS true topologies \n",
    "\n",
    "paralog_ME_vs_true_tree = c()\n",
    "paralog_ML_vs_true_tree = c()\n",
    "RF_matrices = list()\n",
    "\n",
    "\n",
    "collect_branch_len_sub_from_ref = c()\n",
    "collect_diameter_sub_from_ref = c()\n",
    "\n",
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    # the mapping used later on to rename species trees according to a more generic sequence name\n",
    "    ref_codefile = read.table(file = reference_tree_codefile )\n",
    "    \n",
    "    # get al the family subdirectories in each mqin.nf output run dir\n",
    "    fam_dirs = grep(\"avg_branchlen_0.7_protein_family*\", list.dirs(run, recursive = FALSE), value = TRUE)\n",
    "    count_fam = 0\n",
    "\n",
    "    # this will contain the mean branch length for all true/reference paralog trees for each gene family\n",
    "    avg_branch_length = c()\n",
    "\n",
    "    # read the names of the ortholog species.\n",
    "    orgs = read.table(file = species_name_file)\n",
    "\n",
    "    # extract and work on all BigTree paralog files for all families. this trees are computed from the main.nf pipeline in simulation mode.\n",
    "    for (m in fam_dirs) {\n",
    "        \n",
    "        count_fam = count_fam + 1\n",
    "\n",
    "        # print every 10 families a check on progress to screen\n",
    "        if (count_fam %% 10 == 0) {\n",
    "            print(m)\n",
    "        }\n",
    "\n",
    "        # read the paralog tree obtained from the BigTree (gene family tree) created unsing ME.\n",
    "        # the paralog tree was obtained extracting each species subtree for the gene family tree, then this (25) subtrees were merged into one exactly like is done for SuperTree (using superfine).\n",
    "        fastme_paralog_tree_file          = file.path(m, \"results/rerun\", paste0(basename(m), \"_full_aln_coded.phylip_fastme_tree_paralog_tree.nwk\"))\n",
    "        fastme_paralog_tree_from_full_aln = read.tree(fastme_paralog_tree_file, keep.multi=FALSE)\n",
    "        fastme_paralog_tree_from_full_aln = unroot(fastme_paralog_tree_from_full_aln)\n",
    "        fastme_paralog_tree_from_full_aln = list(fastme_paralog_tree_from_full_aln)\n",
    "\n",
    "\n",
    "        # read the paralog tree obtained from the BigTree (gene family tree) created unsing ML. paralog tree constructed same way as above.\n",
    "        raxml_paralog_tree_file           = file.path(m, \"results/rerun\", paste0(\"RAxML_bestTree.\", basename(m), \"_full_aln_coded_raxml_paralog_tree.nwk\"))   \n",
    "        raxml_paralog_tree_from_full_aln  = read.tree(raxml_paralog_tree_file, keep.multi=FALSE)\n",
    "        raxml_paralog_tree_from_full_aln  = unroot(raxml_paralog_tree_from_full_aln)\n",
    "        raxml_paralog_tree_from_full_aln  = list(raxml_paralog_tree_from_full_aln)\n",
    "                \n",
    "        # read the reference tree (gene family tree) the one set as thruth and on which the sequences were generated along.\n",
    "        ref_tree_file = file.path(run, paste0(basename(m), \".nwk\"))\n",
    "        ref_tree      = read.tree(ref_tree_file, keep.multi=FALSE)\n",
    "\n",
    "        # each reference species paralog tree is identical in topologty (designed so) but diverges for branch lengths. here we retrieve some info for that.\n",
    "        for (species in orgs$V1) {\n",
    "            \n",
    "            # ectract tips like before and then extract subtree and unroot it. \n",
    "            # this is the refence paralog tree for each species and whithin one family all species have the same paralog tree as far as topology goes, what changes between this species paralog trees is their branch lengths.\n",
    "            tips_to_keep = grep(paste0(\"^\", species), ref_tree$tip.label, perl = TRUE)\n",
    "            test_subtree = keep.tip(ref_tree, tips_to_keep)\n",
    "            test_subtree = unroot(test_subtree)\n",
    "\n",
    "            # get branch length and diameter\n",
    "            collect_branch_len_sub_from_ref = c(collect_branch_len_sub_from_ref, test_subtree$edge.length)\n",
    "            collect_diameter_sub_from_ref   = c(collect_diameter_sub_from_ref, max(distTips(test_subtree, method = 'patristic')))\n",
    "        }\n",
    "\n",
    "        # Since each gene family has a predifined gene tree (ref_tree) composed of a fixed species tree pasted as leaf of the paralog tree. The tree is designed like this by choice. Each gene families varies only in the paralog tree that it has that is unique (15 tips).\n",
    "        # Here we extract the paralog tree and to do so we just need to extract the subtree with all the sequences of one species. Since each species have the same paralog tree. So we can extract the subtree of whichever species it does not matter.\n",
    "        ref_subtree   = keep.tip(ref_tree, as.character(ref_codefile$V1))\n",
    "\n",
    "        # we then rename the tips of the paralog tree (ref_subtree) to a more generic keyname. Effectively going from rat1 to seq1. \n",
    "        for (j in 1:length(ref_subtree$tip.label)) { \n",
    "            ref_subtree$tip.label[j] = as.character(ref_codefile$V2[which(ref_codefile$V1 == ref_subtree[[\"tip.label\"]][j])])\n",
    "        }\n",
    "\n",
    "        # Unroot the tree and put into a list\n",
    "        ref_subtree        = unroot(ref_subtree)\n",
    "        ref_subtree        = list(ref_subtree)\n",
    "\n",
    "\n",
    "        # update on the average branch length with the mean of the true/refernce paralog tree of the given gene family\n",
    "        avg_branch_length = c(avg_branch_length, mean(ref_subtree[[1]]$edge.length))\n",
    "        \n",
    "        # put all  BIgTree (ME/ML) paralog tres (2) and the reference paralog tree into one list.\n",
    "        all_trees        = c(ref_subtree, fastme_paralog_tree_from_full_aln, raxml_paralog_tree_from_full_aln)\n",
    "        class(all_trees) = \"multiPhylo\"\n",
    "        \n",
    "        # compute the RF dictance between 1 refernce paralog the other two paralog trees: ME ML. In reality all difrrences between them.\n",
    "        all_vs_all_rf = RF.dist(all_trees, normalize=TRUE)\n",
    "        plotdata      = data.matrix(all_vs_all_rf)\n",
    "        RF_matrices   = list.append(RF_matrices, plotdata)\n",
    "        \n",
    "        # do the average of all pair comparison between the species paralog trees and the reference tree. -c[1] means not consider the first column, because first row first column is RF between reference tree and itself\n",
    "        paralog_ME_vs_true_tree = c(paralog_ME_vs_true_tree, plotdata[1, 2])\n",
    "        paralog_ML_vs_true_tree = c(paralog_ML_vs_true_tree, plotdata[1, 3])\n",
    "    }\n",
    "}\n",
    "\n",
    "write.table(paralog_ME_vs_true_tree, file = \"tables/paralog_ME_vs_true_tree.tsv\", quote=FALSE, row.names = FALSE, col.names = FALSE)\n",
    "write.table(paralog_ML_vs_true_tree, file = \"tables/paralog_ML_vs_true_tree.tsv\", quote=FALSE, row.names = FALSE, col.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab7ac11a-c926-4fa6-aafd-d3c8b281e7fc",
   "metadata": {},
   "source": [
    "<a id='chapter-6-RF-distance-summaries'></a>\n",
    "\n",
    "## Chapter 6 - RF distance summaries\n",
    "\n",
    "todo add description"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0553ac22-a97a-49f6-adfe-8323e9003fba",
   "metadata": {},
   "source": [
    "<a id='subchapter-6_1-Table-1'></a>\n",
    "\n",
    "### Subchapter 6.1 - Table 1\n",
    "todo add description."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f493767f-1b5e-4251-848d-6ff64f26f4f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Method</th><th scope=col>Unit.25</th><th scope=col>Unit.1</th><th scope=col>Difference</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ME_SM     </td><td>0.24  sd  0.15</td><td>0.47  sd  0.13</td><td>0.23</td></tr>\n",
       "\t<tr><td>ME_ST     </td><td>0.23  sd  0.13</td><td>0.47  sd  0.13</td><td>0.24</td></tr>\n",
       "\t<tr><td>ML_SM     </td><td>0.19  sd  0.14</td><td>0.44  sd  0.13</td><td>0.25</td></tr>\n",
       "\t<tr><td>ML_ST     </td><td>0.2  sd  0.12 </td><td>0.44  sd  0.13</td><td>0.24</td></tr>\n",
       "\t<tr><td>ME_BigTree</td><td>0.25  sd  0.16</td><td>Na            </td><td>Na  </td></tr>\n",
       "\t<tr><td>ML_BigTree</td><td>0.18  sd  0.15</td><td>Na            </td><td>Na  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 4\n",
       "\\begin{tabular}{llll}\n",
       " Method & Unit.25 & Unit.1 & Difference\\\\\n",
       " <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t ME\\_SM      & 0.24  sd  0.15 & 0.47  sd  0.13 & 0.23\\\\\n",
       "\t ME\\_ST      & 0.23  sd  0.13 & 0.47  sd  0.13 & 0.24\\\\\n",
       "\t ML\\_SM      & 0.19  sd  0.14 & 0.44  sd  0.13 & 0.25\\\\\n",
       "\t ML\\_ST      & 0.2  sd  0.12  & 0.44  sd  0.13 & 0.24\\\\\n",
       "\t ME\\_BigTree & 0.25  sd  0.16 & Na             & Na  \\\\\n",
       "\t ML\\_BigTree & 0.18  sd  0.15 & Na             & Na  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 4\n",
       "\n",
       "| Method &lt;chr&gt; | Unit.25 &lt;chr&gt; | Unit.1 &lt;chr&gt; | Difference &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| ME_SM      | 0.24  sd  0.15 | 0.47  sd  0.13 | 0.23 |\n",
       "| ME_ST      | 0.23  sd  0.13 | 0.47  sd  0.13 | 0.24 |\n",
       "| ML_SM      | 0.19  sd  0.14 | 0.44  sd  0.13 | 0.25 |\n",
       "| ML_ST      | 0.2  sd  0.12  | 0.44  sd  0.13 | 0.24 |\n",
       "| ME_BigTree | 0.25  sd  0.16 | Na             | Na   |\n",
       "| ML_BigTree | 0.18  sd  0.15 | Na             | Na   |\n",
       "\n"
      ],
      "text/plain": [
       "  Method     Unit.25        Unit.1         Difference\n",
       "1 ME_SM      0.24  sd  0.15 0.47  sd  0.13 0.23      \n",
       "2 ME_ST      0.23  sd  0.13 0.47  sd  0.13 0.24      \n",
       "3 ML_SM      0.19  sd  0.14 0.44  sd  0.13 0.25      \n",
       "4 ML_ST      0.2  sd  0.12  0.44  sd  0.13 0.24      \n",
       "5 ME_BigTree 0.25  sd  0.16 Na             Na        \n",
       "6 ML_BigTree 0.18  sd  0.15 Na             Na        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read the table compiled in chapter 4. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df_new = read.table(\"tables/avg_RF_between_ref_supermatrix_and_supertree.tsv\", header=T)\n",
    "\n",
    "# do the mean and sd of the column. effectively getting a vector of size 100 filed with averages of averages and with ids like SM_ML_unit9\n",
    "rf_df_new_means = colMeans(df_new)\n",
    "sd_df_new_means = apply(df_new, 2, sd)\n",
    "\n",
    "# Initialize empty list to hold the extracted method names and unit names values\n",
    "methods = c(\"ME_SM\", \"ME_ST\", \"ML_SM\",  \"ML_ST\")\n",
    "units   = 25\n",
    "\n",
    "# Create an empty dataframe with row names for methods and columns for unit_1 to unit_25\n",
    "avg_rf_df_method_unit           = data.frame(Method = methods, matrix(ncol = units, nrow = 4))\n",
    "colnames(avg_rf_df_method_unit) = c(\"Method\", paste0(\"unit_\", 1:units))\n",
    "\n",
    "# Fill the dataframe using the keys of the named vector of rf_df_new_means, these names are the same for the vector sd_df_new_means too.\n",
    "for (key in names(rf_df_new_means)) {\n",
    "\n",
    "  # Split the key into parts (method and unit)\n",
    "  parts  = strsplit(key, \"_unit_\")[[1]]\n",
    "  method = parts[1]\n",
    "  unit   = as.numeric(parts[2])\n",
    "\n",
    "  # Place the value in the correct row and column of the dataframe adding the standard deviation to it.\n",
    "  avg_rf_df_method_unit[avg_rf_df_method_unit$Method == method, paste0(\"unit_\", unit)] = paste0(round(rf_df_new_means[key], digit=2), \"  sd  \", round(sd_df_new_means[key], digit=2))\n",
    "}\n",
    "\n",
    "# create the actual table with only unit1, unit25 difference values from above df\n",
    "diff   = c(as.numeric(sub(\" sd.*\", \"\", avg_rf_df_method_unit$unit_1)) - as.numeric(sub(\" sd.*\", \"\", avg_rf_df_method_unit$unit_25)))\n",
    "table1 = data.frame(Method = avg_rf_df_method_unit$Method,\n",
    "                    \"Unit 25\"    = avg_rf_df_method_unit$unit_25,\n",
    "                    \"Unit 1\"     = avg_rf_df_method_unit$unit_1,\n",
    "                    \"Difference\" = diff )\n",
    "\n",
    "# read the table compiled in chapter 5. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df_bigtree_ME = read.table(\"tables/paralog_ME_vs_true_tree.tsv\", header=F)\n",
    "df_bigtree_ML = read.table(\"tables/paralog_ML_vs_true_tree.tsv\", header=F)\n",
    "\n",
    "# make average of all family BigTree RF (ME and ML) distances to their respective paralog tree. going from 500 values to 1.\n",
    "rf_df_bigtree_ME_mean = colMeans(df_bigtree_ME)\n",
    "rf_df_bigtree_ME_sd   = apply(df_bigtree_ME, 2, sd)\n",
    "rf_df_bigtree_ML_mean = colMeans(df_bigtree_ML)\n",
    "rf_df_bigtree_ML_sd   = apply(df_bigtree_ML, 2, sd)\n",
    "\n",
    "# add the above four values to table1\n",
    "bigtree_ME_row = c(\"ME_BigTree\", paste0(round(rf_df_bigtree_ME_mean, digits = 2), \"  sd  \", round(rf_df_bigtree_ME_sd, digits = 2)), \"Na\", \"Na\")\n",
    "bigtree_ML_row = c(\"ML_BigTree\", paste0(round(rf_df_bigtree_ML_mean, digits = 2), \"  sd  \", round(rf_df_bigtree_ML_sd, digits = 2)), \"Na\", \"Na\")\n",
    "table1      = rbind(table1, bigtree_ME_row)\n",
    "table1      = rbind(table1, bigtree_ML_row)\n",
    "\n",
    "table1\n",
    "write.table(table1, file=\"tables/Table1_average_RF.tsv\", quote=FALSE, sep=\"\\t\", row.names=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "735442c0-400c-4549-9064-b9892fff8880",
   "metadata": {},
   "source": [
    "<a id=\"subchapter-6_2-Figure-2\"></a>\n",
    "\n",
    "### Subchapter 6.2 Figure 2\n",
    "\n",
    "TODO add description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0981fb0d-587e-4428-bcb5-07301de851eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "# Set plot dimensions and resolution for inline display\n",
    "options(repr.plot.width = 6, repr.plot.height = 5, repr.plot.res = 300)\n",
    "\n",
    "# Load the necessary data\n",
    "# Reading in average RF values between reference big tree ME/ML from a file computed in chapter 5.\n",
    "df_bigtree_ME = read.table(\"tables/paralog_ME_vs_true_tree.tsv\", header=F)\n",
    "df_bigtree_ML = read.table(\"tables/paralog_ML_vs_true_tree.tsv\", header=F)\n",
    "rf_df_bigtree_ME_mean = colMeans(df_bigtree_ME)\n",
    "rf_df_bigtree_ML_mean = colMeans(df_bigtree_ML)\n",
    "\n",
    "# Reading in average RF values between reference supermatrix and supertree\n",
    "df_new = read.table(\"../analysis/tables/avg_RF_between_ref_supermatrix_and_supertree.tsv\", header = TRUE)\n",
    "\n",
    "# Calculate column-wise means of the loaded data\n",
    "rf_df_new_means = colMeans(df_new)\n",
    "\n",
    "# Reshape the means data into a long format\n",
    "rf_df_new_means_melted = melt(rf_df_new_means)\n",
    "\n",
    "# Rename the first column to \"RF\" for clarity\n",
    "colnames(rf_df_new_means_melted)[1] = \"RF\"\n",
    "\n",
    "# Add a column indicating the topologies (25 levels, repeated 4 times for each method)\n",
    "rf_df_new_means_melted$Topology = c(rep(colnames(df_new)[1:25], 4))\n",
    "\n",
    "# Add a column specifying the method (Supermatrix-ME, Supertree-ME, Supermatrix-ML, Supertree-ML)\n",
    "rf_df_new_means_melted$Method = c(rep(\"Supermatrix-ME\", 25), rep(\"Supertree-ME\", 25),\n",
    "                                  rep(\"Supermatrix-ML\", 25), rep(\"Supertree-ML\", 25))\n",
    "\n",
    "# Sort the data by topology using mixedorder\n",
    "rf_df_new_means_melted = rf_df_new_means_melted[mixedorder(rf_df_new_means_melted$Topology), ]\n",
    "\n",
    "# crate the x axis label names\n",
    "x_axis_labels = gsub(\"ME_SM_\", \"\", colnames(df_new)[1:25])\n",
    "\n",
    "# Modify the Topology column to match the cleaned x-axis labels\n",
    "rf_df_new_means_melted$Topology = rep(x_axis_labels, each = 4)  # Repeat each value per method (4 times)\n",
    "\n",
    "# Define custom colors for each Method\n",
    "custom_colors = c(\"Supertree-ME\" = \"#7CB6E2\",     # Light blue \n",
    "                  \"Supermatrix-ME\" = \"#2F739B\",   # Dark blue\n",
    "                  \"Supertree-ML\" = \"#E5A23D\",     # Orange/yellow\n",
    "                  \"Supermatrix-ML\" = \"#D05159\")   # Red\n",
    "\n",
    "# Create the line plot\n",
    "p = ggplot(rf_df_new_means_melted, aes(x = factor(Topology, levels = x_axis_labels), \n",
    "                                       y = RF, group = Method, color = Method)) + \n",
    "    geom_line() +  # Add lines connecting the points by method\n",
    "    geom_point(shape = 16) +  # Use shape 16 (solid dot) for all points\n",
    "\n",
    "    # Add a horizontal dashed line representing the average RF of ME big tree\n",
    "    geom_hline(yintercept = rf_df_bigtree_ME_mean, \n",
    "               linetype = \"dashed\", color = \"#606060\") +\n",
    "    \n",
    "    # Annotate the plot with a label for the average bigtree ME RF line\n",
    "    annotate(\"text\", x = 18, y = rf_df_bigtree_ME_mean + 0.025, \n",
    "             label = \"Avg RF between BigTree-ME and true topology\", \n",
    "             size = 3, color = \"#606060\") +\n",
    "\n",
    "    # Add a horizontal dashed line representing the average RF of ML big tree\n",
    "    geom_hline(yintercept = rf_df_bigtree_ML_mean, \n",
    "               linetype = \"dashed\", color = \"#1E7005\") +\n",
    "    \n",
    "    # Annotate the plot with a label for the average bigtree ML RF line\n",
    "    annotate(\"text\", x = 12, y = rf_df_bigtree_ML_mean - 0.025, \n",
    "             label = \"Avg RF between BigTree-ML and true topology\", \n",
    "             size = 3, color = \"#1E7005\") +\n",
    "\n",
    "    # Apply a light theme for better readability\n",
    "    theme_light() + \n",
    "\n",
    "    # Adjust x-axis text to be tilted at 45 degrees and align to the right\n",
    "    theme(axis.text.x = element_text(angle = 45, hjust = 1), \n",
    "          legend.position = \"right\",\n",
    "          axis.title.x = element_text(face = \"bold\"),  \n",
    "          axis.title.y = element_text(face = \"bold\"))  + \n",
    "    \n",
    "    # Label the axes\n",
    "    ylab(\"Normalized RF\") + \n",
    "    xlab(\"Topology\") + \n",
    "\n",
    "    # Set y-axis limits between 0 and 0.5\n",
    "    scale_y_continuous(limits = c(0, 0.5)) +\n",
    "\n",
    "    # Customize color and shape scales for the different methods\n",
    "    scale_color_discrete(breaks = c(\"Supertree-ME\", \"Supermatrix-ME\", \"Supertree-ML\", \"Supermatrix-ML\")) + \n",
    "    scale_shape_discrete(breaks = c(\"Supertree-ME\", \"Supermatrix-ME\", \"Supertree-ML\", \"Supermatrix-ML\")) +\n",
    "    scale_color_manual(values = custom_colors)  # Apply custom colors\n",
    "\n",
    "# Save the plot to a file with high resolution\n",
    "ggsave(filename = \"figures/RF_lineplot_figure2.png\", plot = p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfa19d80-b20f-4a24-af34-ea65b39aa0f2",
   "metadata": {},
   "source": [
    "<a id=\"chapter-7-Running-times\"></a>\n",
    "\n",
    "## Chapter 7 - Running times\n",
    "\n",
    "TODO add description "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "062ad02a-baca-4fce-b3ee-e8ea3b3bb039",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\"\n"
     ]
    }
   ],
   "source": [
    "# read the trace files for running times\n",
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    # each main.nf output run has a trace file repoprting various information on the run itself\n",
    "    tracefile_path = file.path(run, \"trace_cpu_time.txt\")\n",
    "    tracefile      = read.table(tracefile_path, header = TRUE)\n",
    "    \n",
    "    # here we parse the trace file to retain only the processes that we want to take the running time (realtime) from\n",
    "    tracefile_mod  = tracefile[grep(\"simulated_data:run_phylo_ML_supermatrix_aln_sim|simulated_data:only_concatenate_aln_sim|simulated_data:run_phylo_ML_full_aln_sim\", tracefile$native_id), c(\"name\", \"native_id\", \"realtime\")]\n",
    "    \n",
    "    # transforming the above table to one that have the process names (simulated_data:run_phylo_ML_supermatrix_aln_sim ecc..) as column names, \n",
    "    # the rows will be still the family names (working as ID) and the values in each column will be the running times associated to the column name (process name) for that given family.\n",
    "    tracefile_mod_wide = spread(tracefile_mod, native_id, realtime)\n",
    "\n",
    "    # creating a new last column called Supermatrix_ML that has aas values the running times of concatenation plus ML tree computation (on concatenated MSA) \n",
    "    tracefile_mod_wide$Supermatrix_ML = as.numeric(tracefile_mod_wide$`simulated_data:only_concatenate_aln_sim`) + as.numeric(tracefile_mod_wide$`simulated_data:run_phylo_ML_supermatrix_aln_sim`)\n",
    "\n",
    "    # retain only the family name, BIgTree ML running time and SM + ML running time\n",
    "    tracefile_mod_wide = tracefile_mod_wide[,c(1,3,5)]\n",
    "\n",
    "    # call the second column name Big_Tree and set their value to numeric type\n",
    "    colnames(tracefile_mod_wide)[2] = \"Big_Tree\"\n",
    "    tracefile_mod_wide$Big_Tree = as.numeric(tracefile_mod_wide$Big_Tree)\n",
    "\n",
    "    # adding a first column that groups families by the run of the main.nf\n",
    "    tracefile_mod_wide$family_group = rep(basename(run), 100)\n",
    "\n",
    "    # add tables toghether raw-wise\n",
    "    if(run == run_dirs[1]) {\n",
    "        fin_tracefile_mod_wide = tracefile_mod_wide\n",
    "    } else {\n",
    "        fin_tracefile_mod_wide = rbind(fin_tracefile_mod_wide, tracefile_mod_wide)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d6a00cd4-31ce-440c-8528-94805af06047",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>Big_Tree</dt><dd>638.12</dd><dt>Supermatrix_ML</dt><dd>67.76</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[Big\\textbackslash{}\\_Tree] 638.12\n",
       "\\item[Supermatrix\\textbackslash{}\\_ML] 67.76\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "Big_Tree\n",
       ":   638.12Supermatrix_ML\n",
       ":   67.76\n",
       "\n"
      ],
      "text/plain": [
       "      Big_Tree Supermatrix_ML \n",
       "        638.12          67.76 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "588.3"
      ],
      "text/latex": [
       "588.3"
      ],
      "text/markdown": [
       "588.3"
      ],
      "text/plain": [
       "[1] 588.3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "63.85"
      ],
      "text/latex": [
       "63.85"
      ],
      "text/markdown": [
       "63.85"
      ],
      "text/plain": [
       "[1] 63.85"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# TODO make nicer and update with whatever is changed above\n",
    "\n",
    "# make the means of all (500) family running time for the BigTree and ML + SM. Bring them to seconds (/1000) and round the digits to 2\n",
    "round(colMeans(fin_tracefile_mod_wide[, c(2, 3)])/1000,digits = 2)\n",
    "\n",
    "# compute the median for the BigTree approach and the  ML + SM.\n",
    "round(median(fin_tracefile_mod_wide$Big_Tree)/1000,digits = 2)\n",
    "round(median(fin_tracefile_mod_wide$Supermatrix_ML)/1000,digits = 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b79de7a-f8b9-4dc7-8fc6-232cf86be231",
   "metadata": {},
   "source": [
    "<a id='subchapter-7_1-Figure-3'></a>\n",
    "\n",
    "### Subchapter 7.1 - Figure 3\n",
    "\n",
    "TODO add chapter link and description if necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ba993fba-7f4b-48aa-9b03-775093172898",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO make this into the section of custom functions.\n",
    "# Function to assign new values based on a mapping provided by the user\n",
    "assign_mapped_values <- function(df, column_name, mapping, new_col_name) {\n",
    "  # Check if mapping is a named vector or a list\n",
    "  if (!is.vector(mapping) || is.null(names(mapping))) {\n",
    "    stop(\"Mapping must be a named vector or list.\")\n",
    "  }\n",
    "  \n",
    "  # Get the unique values from the specified column\n",
    "  unique_values = unique(df[[column_name]])\n",
    "  \n",
    "  # Check if the mapping contains all unique values\n",
    "  if (!all(unique_values %in% names(mapping))) {\n",
    "    stop(\"Mapping must include all unique values from the specified column.\")\n",
    "  }\n",
    "  \n",
    "  # Initialize a vector to store the assigned values\n",
    "  assigned_values = vector(\"character\", nrow(df))\n",
    "  \n",
    "  # Assign values based on the mapping\n",
    "  for (i in seq_along(unique_values)) {\n",
    "    assigned_values[df[[column_name]] == unique_values[i]] <- mapping[names(mapping) == unique_values[i]]\n",
    "  }\n",
    "  \n",
    "  # Add the assigned values to the original dataframe\n",
    "  df[[new_col_name]] = assigned_values\n",
    "  \n",
    "  return(df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4568c025-fffa-45df-bfff-086d67ad2a53",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using name, family_group, length as id variables\n",
      "\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method (SM + ML and Bigtree) divided per group (by sequence length)\n",
    "#\n",
    "\n",
    "# assign to each family_group a mapping values  (as new column). basically insert the length of the sequences in each family as an additional column in the dataframe.\n",
    "# Remember each family in the same group have the same length of sequence generated, the MSA though might differ in length due to gaps.\n",
    "mapping = c(\"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\" = \"group1\\nlen = 32\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\" = \"group2\\nlen = 46\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\" = \"group3\\nlen = 50\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\" = \"group4\\nlen = 60\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\" = \"group5\\nlen = 82\")\n",
    "fin_tracefile_mod_reassigned = assign_mapped_values(fin_tracefile_mod_wide, \"family_group\", mapping, \"length\")\n",
    "\n",
    "# reorder the columns to have the numerical values at the end\n",
    "fin_tracefile_mod_reassigned = fin_tracefile_mod_reassigned[, c(\"name\", \"family_group\", \"length\", \"Big_Tree\", \"Supermatrix_ML\")]\n",
    "\n",
    "# It takes a table of all families as rows (500) and it goes to a table of 1000 (500 BigTree + 500 SM + ML) where the running times instead of being separated into two different coluns now they are in the same one.\n",
    "# name + group columns will work as unique ID.  group is the directory name of the main.nf run/s, while name is the name of the family. \n",
    "fin_tracefile_mod_melted           = melt(fin_tracefile_mod_reassigned)\n",
    "\n",
    "\n",
    "# transform running times from millisecond to seconds. And rename the column names.\n",
    "fin_tracefile_mod_melted$value     = fin_tracefile_mod_melted$value / 1000\n",
    "colnames(fin_tracefile_mod_melted) = c(\"name\", \"family_group\", \"length\", \"Method\", \"Time\")\n",
    "\n",
    "# create a table with 5 pairs of boxplot. Group1, 2.. ecc.. each group has 100 families and two running time (realtime) mesuraments for the BigTree and SM + ML approaches.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel.\n",
    "p = ggplot(fin_tracefile_mod_melted, aes(x=length, y=Time, fill=Method)) +\n",
    "    geom_boxplot(show.legend = TRUE) +\n",
    "    theme_light() +\n",
    "    labs(x = \"sequence length\",\n",
    "       y = \"CPU time (sec)\",\n",
    "       fill = \"Method\")\n",
    "ggsave(filename=\"figures/CPU_time_per_group_BigTree_SM_ML_figure3.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83903c34-3f32-431b-abc5-156cb2abeaf8",
   "metadata": {},
   "source": [
    "### Suplementary Fig todo put number of supplementary figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "81650244-8771-4d70-a7c6-57d5218f928d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method ratio (Bigtree / SM + ML ) boxplot of each family ratio by group (x axis groups y ratio Big/SM) \n",
    "# \n",
    "\n",
    "# start by taking the already present info for the families and their grouping from fig 2 cell and compute ratio Bigtree / SM + ML\n",
    "# Create a new column with the division of Big_Tree by Supermatrix_ML running times for each family (500)\n",
    "# and then remove the original columns Big_Tree Supermatrix_ML\n",
    "fin_tracefile_mod_ratio = fin_tracefile_mod_reassigned %>%\n",
    "  mutate(fold_ratio = Big_Tree / Supermatrix_ML) %>%\n",
    "  select(-Big_Tree, -Supermatrix_ML)\n",
    "\n",
    "# create a table with 5 boxplots. Group1, 2.. ecc.. each group has 100 families and  the ratio of two running time (realtime) mesuraments for the BigTree and SM + ML approaches.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel.\n",
    "p = ggplot(fin_tracefile_mod_ratio, aes(x=length, y=fold_ratio)) +\n",
    "    geom_boxplot(show.legend = FALSE) +\n",
    "    theme_light() +\n",
    "    labs(x = \"sequence length\",\n",
    "       y = \"CPU time fold ratio (BigTree / Supermatrix_ML)\")\n",
    "\n",
    "# TODO add subplot number to filename\n",
    "ggsave(filename=\"figures/CPU_time_per_group_BigTree_SM_ML_ratio.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42387886-a000-4c7b-80d3-1dc82b4f4825",
   "metadata": {},
   "source": [
    "<a id=\"chapter-8-Bootstrap-analysis\"></a>\n",
    "\n",
    "## Chapter 8 - Bootstrap analysis\n",
    "\n",
    "TODO add decriptoion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "0b0a5b14-5c84-4733-9cff-0313fc6105d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"results\"\n",
      "[1] \"../results/avg_branchlen_0.7_protein_family_100.sub\"\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# ME + SM block\n",
    "#\n",
    "\n",
    "\n",
    "# Create the dataframe that will contain all bs values and other info like presence in reference tree\n",
    "BS_df_ME_SM = data.frame(\n",
    "  group = character(), \n",
    "  family = character(), \n",
    "  unit = integer(), \n",
    "  sample = integer(),\n",
    "  node = integer(),\n",
    "  BS = numeric(), \n",
    "  ref_presence = numeric()  # Using integer type for 0 and 1 values\n",
    ")\n",
    "\n",
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "    \n",
    "    # get al the family subdirectories in each mqin.nf output run dir\n",
    "    fam_dirs = grep(\"avg_branchlen_0.7_protein_family*\", list.dirs(run, recursive = FALSE), value = TRUE)\n",
    "    count_fam = 0\n",
    "\n",
    "    # read the names of the ortholog species.\n",
    "    orgs = read.table(file = species_name_file)\n",
    "\n",
    "    # the mapping used later on to rename species trees according to a more generic sequence name\n",
    "    ref_codefile = read.table(file = reference_tree_codefile )\n",
    "    \n",
    "    # work at family level to get the BS value\n",
    "    for (m in fam_dirs) {\n",
    "        \n",
    "        count_fam = count_fam + 1\n",
    "\n",
    "        # print every 10 families a check on progress to screen\n",
    "        if (count_fam %% 10 == 0) {\n",
    "            print(m)\n",
    "        }\n",
    "\n",
    "        # read the file that contains the length of the MSA (number of columns) of the input MSA of main.nf for that given family.\n",
    "        bs_file = file.path(m, \"results/rerun\", \"BS.dat\")\n",
    "        bs      = read.table(bs_file)\n",
    "        myBS    = bs$V1\n",
    "\n",
    "        # read the reference tree (gene family tree) the one set as thruth and on which the sequences were generated along.\n",
    "        ref_tree_file = file.path(run, paste0(basename(m), \".nwk\"))\n",
    "        ref_tree      = read.tree(ref_tree_file, keep.multi=FALSE)\n",
    "\n",
    "        # Since each gene family has a predifined gene tree (ref_tree) composed of a fixed species tree pasted as leaf of the paralog tree. The tree is designed like this by choice. Each gene families varies only in the paralog tree that it has that is unique (15 tips).\n",
    "        # Here we extract the paralog tree and to do so we just need to extract the subtree with all the sequences of one species. Since each species have the same paralog tree. So we can extract the subtree of whichever species it does not matter.\n",
    "        ref_subtree   = keep.tip(ref_tree, as.character(ref_codefile$V1))\n",
    "\n",
    "        # we then rename the tips of the paralog tree (ref_subtree) to a more generic keyname. Effectively going from rat1 to seq1. \n",
    "        for (j in 1:length(ref_subtree$tip.label)) { \n",
    "            ref_subtree$tip.label[j] = as.character(ref_codefile$V2[which(ref_codefile$V1 == ref_subtree[[\"tip.label\"]][j])])\n",
    "        }\n",
    "\n",
    "        # Unroot the tree and put into a list\n",
    "        ref_paralog_tree = unroot(ref_subtree)\n",
    "      \n",
    "        # do a for loop on the number of species (25) for reading all trees of a given unit (from unit1 to unit25).\n",
    "        for (k in 1:length(orgs$V1)) {\n",
    "\n",
    "            # read the file that holds all ME SM trees for a given unit. it holds 10 trees one for each sample/reppliceate.\n",
    "            all_unit_trees_ME_SM_file  = file.path(m, \"results/rerun\", paste0(\"unit_\", k, \"_all.nwk\"))\n",
    "            all_unit_trees_ME_SM       = read.tree(file = all_unit_trees_ME_SM_file, keep.multi = TRUE)\n",
    "            all_unit_trees_ME_SM       = unroot(all_unit_trees_ME_SM)\n",
    "\n",
    "            BS_mean_unit_ME_SM = c()\n",
    "            \n",
    "            # for loop on the number of trees in all_unit_trees_ME_SM (10). \n",
    "            for (i in 1:length(all_unit_trees_ME_SM)) {\n",
    "\n",
    "                # Each unit_k_sample_i_rep.trees file contains as many trees as number of columns MSA gene family trees. This are all bootstrap replicate trees for the corresponding ME tree in all_unit_trees_ME_SM. \n",
    "                # This represents the variability in tree structure across the bootstrap replicates.\n",
    "                all_bs_rep_tree_ME_SM_file = file.path(m, \"results/rerun\", paste0(\"unit_\", k, \"_sample_\", i, \"_rep.trees\"))\n",
    "                all_bs_rep_tree_ME_SM      = read.tree(file = all_bs_rep_tree_ME_SM_file, keep.multi = TRUE)\n",
    "                all_bs_rep_tree_ME_SM      = unroot(all_bs_rep_tree_ME_SM)\n",
    "\n",
    "                # calculates the bootstrap support for each clade in the tree by counting how often each clade (node) in the original tree (all_unit_trees_ME_SM[[i]]) is found in the bootstrap replicate trees (all_bs_rep_tree_ME_SM). \n",
    "                # The [-1] removes the root node’s support value, which is usually not meaningful in unrooted trees. \n",
    "                # This is a list of 12 ( 15 tips -3, 13 internal nodes for unrooted tree and the first clade is the tree itself). because each tree has 15 tips being a paralog tree. Like this by design choiche.0\n",
    "                BS_all_unit_trees_ME_SM         = prop.clades(all_unit_trees_ME_SM[[i]], all_bs_rep_tree_ME_SM)[-1]\n",
    "                \n",
    "                # check now if a clade in the unit tree of interest is also present in the paralog tree. 1 if present NA if missing.\n",
    "                # this list and the one above are syncronyzed. meaning the first element of both point to the same clade\n",
    "                presence_of_clade_in_ref_ME_SM = prop.clades(all_unit_trees_ME_SM[[i]], ref_paralog_tree)[-1]\n",
    "\n",
    "                # Any NA values in all_trees_BS (indicating clades absent in the bootstrap replicates) are replaced with 0.             \n",
    "                BS_all_unit_trees_ME_SM[is.na(BS_all_unit_trees_ME_SM)]                 = 0\n",
    "                presence_of_clade_in_ref_ME_SM[is.na(presence_of_clade_in_ref_ME_SM)] = 0\n",
    "\n",
    "                # Then, the mean of these clade support values is calculated and stored.\n",
    "                val = mean((as.numeric(BS_all_unit_trees_ME_SM)))\n",
    "                BS_mean_unit_ME_SM = c(BS_mean_unit_ME_SM, val)\n",
    "\n",
    "                # Each clade’s support value is scaled by dividing by myBS (which is set to the total number of bootstrap replicates -> number of columns MSA gene family trees).\n",
    "                # add the clade values to the overall dataframe, teh id of each clade will be the concatenation of the first 5 columns.\n",
    "                vector3 = 1:12\n",
    "                tmp_df  = data.frame(\n",
    "                      group = rep(basename(run), 12),\n",
    "                      family = rep(m, 12),\n",
    "                      unit = rep(k, 12),\n",
    "                      sample = rep(i, 12),\n",
    "                      node = vector3,\n",
    "                      BS = c(as.numeric(BS_all_unit_trees_ME_SM) / myBS), # This scales the values to a proportion (0 to 1) for each clade.\n",
    "                      ref_presence = presence_of_clade_in_ref_ME_SM\n",
    "                    )\n",
    "\n",
    "                # add the new block to the overall dataframe\n",
    "                BS_df_ME_SM = rbind(BS_df_ME_SM, tmp_df)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "write.table(BS_df_ME_SM, file=\"tables/df_bootsrtap_clade_presence_in_ref_ME_SM.tsv\", quote=FALSE, sep=\"\\t\", row.names=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4e74a63-1ca7-4e35-99e3-46ac055741a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO make this cell into something like tha above but for ML\n",
    "\n",
    "# read the file that holds all ML SM trees for a given unit. it holds 10 trees one for each sample/reppliceate.\n",
    "            all_unit_trees_ML_SM_file = file.path(m, \"results/rerun\", paste0(\"unit_\", k, \"_all_raxml.nwk\"))\n",
    "            all_unit_trees_ML_SM      = read.tree(file = all_unit_trees_ML_SM_file, keep.multi = TRUE)\n",
    "            all_unit_trees_ML_SM      = unroot(all_unit_trees_ML_SM)\n",
    "\n",
    "            BS_mean_unit_ML_SM = c()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "7027cb71-32ab-4641-ab73-5a5a7774ddfa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 30000 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>group</th><th scope=col>family</th><th scope=col>unit</th><th scope=col>sample</th><th scope=col>node</th><th scope=col>BS</th><th scope=col>ref_presence</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>1</td><td> 1</td><td>0.06250</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>1</td><td> 2</td><td>0.71875</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>1</td><td> 3</td><td>0.15625</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>1</td><td> 4</td><td>0.31250</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>1</td><td> 5</td><td>0.21875</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>1</td><td> 6</td><td>0.28125</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>1</td><td> 7</td><td>0.40625</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>1</td><td> 8</td><td>0.28125</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>1</td><td> 9</td><td>0.37500</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>1</td><td>10</td><td>0.40625</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>1</td><td>11</td><td>0.40625</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>1</td><td>12</td><td>0.84375</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>2</td><td> 1</td><td>0.15625</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>2</td><td> 2</td><td>0.12500</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>2</td><td> 3</td><td>0.43750</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>2</td><td> 4</td><td>0.03125</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>2</td><td> 5</td><td>0.09375</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>2</td><td> 6</td><td>0.15625</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>2</td><td> 7</td><td>0.50000</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>2</td><td> 8</td><td>0.18750</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>2</td><td> 9</td><td>0.28125</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>2</td><td>10</td><td>0.09375</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>2</td><td>11</td><td>0.25000</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>2</td><td>12</td><td>0.21875</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>3</td><td> 1</td><td>0.03125</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>3</td><td> 2</td><td>0.43750</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>3</td><td> 3</td><td>0.12500</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>3</td><td> 4</td><td>0.31250</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>3</td><td> 5</td><td>0.31250</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_001.sub</td><td>1</td><td>3</td><td> 6</td><td>0.93750</td><td>1</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 8</td><td> 7</td><td>0.62500</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 8</td><td> 8</td><td>0.56250</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 8</td><td> 9</td><td>0.62500</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 8</td><td>10</td><td>0.71875</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 8</td><td>11</td><td>0.84375</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 8</td><td>12</td><td>0.65625</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 9</td><td> 1</td><td>0.65625</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 9</td><td> 2</td><td>0.31250</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 9</td><td> 3</td><td>0.12500</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 9</td><td> 4</td><td>0.46875</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 9</td><td> 5</td><td>0.56250</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 9</td><td> 6</td><td>0.31250</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 9</td><td> 7</td><td>0.62500</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 9</td><td> 8</td><td>0.65625</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 9</td><td> 9</td><td>0.56250</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 9</td><td>10</td><td>0.84375</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 9</td><td>11</td><td>0.81250</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td> 9</td><td>12</td><td>0.62500</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td>10</td><td> 1</td><td>0.71875</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td>10</td><td> 2</td><td>0.43750</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td>10</td><td> 3</td><td>0.00000</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td>10</td><td> 4</td><td>0.31250</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td>10</td><td> 5</td><td>0.65625</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td>10</td><td> 6</td><td>0.28125</td><td>0</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td>10</td><td> 7</td><td>0.68750</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td>10</td><td> 8</td><td>0.50000</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td>10</td><td> 9</td><td>0.75000</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td>10</td><td>10</td><td>0.87500</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td>10</td><td>11</td><td>0.75000</td><td>1</td></tr>\n",
       "\t<tr><td>results</td><td>../results/avg_branchlen_0.7_protein_family_100.sub</td><td>25</td><td>10</td><td>12</td><td>0.59375</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 30000 × 7\n",
       "\\begin{tabular}{lllllll}\n",
       " group & family & unit & sample & node & BS & ref\\_presence\\\\\n",
       " <chr> & <chr> & <int> & <int> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 1 &  1 & 0.06250 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 1 &  2 & 0.71875 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 1 &  3 & 0.15625 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 1 &  4 & 0.31250 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 1 &  5 & 0.21875 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 1 &  6 & 0.28125 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 1 &  7 & 0.40625 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 1 &  8 & 0.28125 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 1 &  9 & 0.37500 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 1 & 10 & 0.40625 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 1 & 11 & 0.40625 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 1 & 12 & 0.84375 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 2 &  1 & 0.15625 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 2 &  2 & 0.12500 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 2 &  3 & 0.43750 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 2 &  4 & 0.03125 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 2 &  5 & 0.09375 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 2 &  6 & 0.15625 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 2 &  7 & 0.50000 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 2 &  8 & 0.18750 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 2 &  9 & 0.28125 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 2 & 10 & 0.09375 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 2 & 11 & 0.25000 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 2 & 12 & 0.21875 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 3 &  1 & 0.03125 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 3 &  2 & 0.43750 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 3 &  3 & 0.12500 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 3 &  4 & 0.31250 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 3 &  5 & 0.31250 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub & 1 & 3 &  6 & 0.93750 & 1\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  8 &  7 & 0.62500 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  8 &  8 & 0.56250 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  8 &  9 & 0.62500 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  8 & 10 & 0.71875 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  8 & 11 & 0.84375 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  8 & 12 & 0.65625 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  9 &  1 & 0.65625 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  9 &  2 & 0.31250 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  9 &  3 & 0.12500 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  9 &  4 & 0.46875 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  9 &  5 & 0.56250 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  9 &  6 & 0.31250 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  9 &  7 & 0.62500 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  9 &  8 & 0.65625 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  9 &  9 & 0.56250 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  9 & 10 & 0.84375 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  9 & 11 & 0.81250 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 &  9 & 12 & 0.62500 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 & 10 &  1 & 0.71875 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 & 10 &  2 & 0.43750 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 & 10 &  3 & 0.00000 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 & 10 &  4 & 0.31250 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 & 10 &  5 & 0.65625 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 & 10 &  6 & 0.28125 & 0\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 & 10 &  7 & 0.68750 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 & 10 &  8 & 0.50000 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 & 10 &  9 & 0.75000 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 & 10 & 10 & 0.87500 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 & 10 & 11 & 0.75000 & 1\\\\\n",
       "\t results & ../results/avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub & 25 & 10 & 12 & 0.59375 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 30000 × 7\n",
       "\n",
       "| group &lt;chr&gt; | family &lt;chr&gt; | unit &lt;int&gt; | sample &lt;int&gt; | node &lt;int&gt; | BS &lt;dbl&gt; | ref_presence &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 1 |  1 | 0.06250 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 1 |  2 | 0.71875 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 1 |  3 | 0.15625 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 1 |  4 | 0.31250 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 1 |  5 | 0.21875 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 1 |  6 | 0.28125 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 1 |  7 | 0.40625 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 1 |  8 | 0.28125 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 1 |  9 | 0.37500 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 1 | 10 | 0.40625 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 1 | 11 | 0.40625 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 1 | 12 | 0.84375 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 2 |  1 | 0.15625 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 2 |  2 | 0.12500 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 2 |  3 | 0.43750 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 2 |  4 | 0.03125 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 2 |  5 | 0.09375 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 2 |  6 | 0.15625 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 2 |  7 | 0.50000 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 2 |  8 | 0.18750 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 2 |  9 | 0.28125 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 2 | 10 | 0.09375 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 2 | 11 | 0.25000 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 2 | 12 | 0.21875 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 3 |  1 | 0.03125 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 3 |  2 | 0.43750 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 3 |  3 | 0.12500 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 3 |  4 | 0.31250 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 3 |  5 | 0.31250 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_001.sub | 1 | 3 |  6 | 0.93750 | 1 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  8 |  7 | 0.62500 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  8 |  8 | 0.56250 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  8 |  9 | 0.62500 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  8 | 10 | 0.71875 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  8 | 11 | 0.84375 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  8 | 12 | 0.65625 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  9 |  1 | 0.65625 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  9 |  2 | 0.31250 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  9 |  3 | 0.12500 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  9 |  4 | 0.46875 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  9 |  5 | 0.56250 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  9 |  6 | 0.31250 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  9 |  7 | 0.62500 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  9 |  8 | 0.65625 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  9 |  9 | 0.56250 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  9 | 10 | 0.84375 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  9 | 11 | 0.81250 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 |  9 | 12 | 0.62500 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 | 10 |  1 | 0.71875 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 | 10 |  2 | 0.43750 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 | 10 |  3 | 0.00000 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 | 10 |  4 | 0.31250 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 | 10 |  5 | 0.65625 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 | 10 |  6 | 0.28125 | 0 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 | 10 |  7 | 0.68750 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 | 10 |  8 | 0.50000 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 | 10 |  9 | 0.75000 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 | 10 | 10 | 0.87500 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 | 10 | 11 | 0.75000 | 1 |\n",
       "| results | ../results/avg_branchlen_0.7_protein_family_100.sub | 25 | 10 | 12 | 0.59375 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "      group   family                                              unit sample\n",
       "1     results ../results/avg_branchlen_0.7_protein_family_001.sub 1    1     \n",
       "2     results ../results/avg_branchlen_0.7_protein_family_001.sub 1    1     \n",
       "3     results ../results/avg_branchlen_0.7_protein_family_001.sub 1    1     \n",
       "4     results ../results/avg_branchlen_0.7_protein_family_001.sub 1    1     \n",
       "5     results ../results/avg_branchlen_0.7_protein_family_001.sub 1    1     \n",
       "6     results ../results/avg_branchlen_0.7_protein_family_001.sub 1    1     \n",
       "7     results ../results/avg_branchlen_0.7_protein_family_001.sub 1    1     \n",
       "8     results ../results/avg_branchlen_0.7_protein_family_001.sub 1    1     \n",
       "9     results ../results/avg_branchlen_0.7_protein_family_001.sub 1    1     \n",
       "10    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    1     \n",
       "11    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    1     \n",
       "12    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    1     \n",
       "13    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    2     \n",
       "14    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    2     \n",
       "15    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    2     \n",
       "16    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    2     \n",
       "17    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    2     \n",
       "18    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    2     \n",
       "19    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    2     \n",
       "20    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    2     \n",
       "21    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    2     \n",
       "22    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    2     \n",
       "23    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    2     \n",
       "24    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    2     \n",
       "25    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    3     \n",
       "26    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    3     \n",
       "27    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    3     \n",
       "28    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    3     \n",
       "29    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    3     \n",
       "30    results ../results/avg_branchlen_0.7_protein_family_001.sub 1    3     \n",
       "⋮     ⋮       ⋮                                                   ⋮    ⋮     \n",
       "29971 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    8    \n",
       "29972 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    8    \n",
       "29973 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    8    \n",
       "29974 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    8    \n",
       "29975 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    8    \n",
       "29976 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    8    \n",
       "29977 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    9    \n",
       "29978 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    9    \n",
       "29979 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    9    \n",
       "29980 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    9    \n",
       "29981 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    9    \n",
       "29982 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    9    \n",
       "29983 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    9    \n",
       "29984 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    9    \n",
       "29985 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    9    \n",
       "29986 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    9    \n",
       "29987 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    9    \n",
       "29988 results ../results/avg_branchlen_0.7_protein_family_100.sub 25    9    \n",
       "29989 results ../results/avg_branchlen_0.7_protein_family_100.sub 25   10    \n",
       "29990 results ../results/avg_branchlen_0.7_protein_family_100.sub 25   10    \n",
       "29991 results ../results/avg_branchlen_0.7_protein_family_100.sub 25   10    \n",
       "29992 results ../results/avg_branchlen_0.7_protein_family_100.sub 25   10    \n",
       "29993 results ../results/avg_branchlen_0.7_protein_family_100.sub 25   10    \n",
       "29994 results ../results/avg_branchlen_0.7_protein_family_100.sub 25   10    \n",
       "29995 results ../results/avg_branchlen_0.7_protein_family_100.sub 25   10    \n",
       "29996 results ../results/avg_branchlen_0.7_protein_family_100.sub 25   10    \n",
       "29997 results ../results/avg_branchlen_0.7_protein_family_100.sub 25   10    \n",
       "29998 results ../results/avg_branchlen_0.7_protein_family_100.sub 25   10    \n",
       "29999 results ../results/avg_branchlen_0.7_protein_family_100.sub 25   10    \n",
       "30000 results ../results/avg_branchlen_0.7_protein_family_100.sub 25   10    \n",
       "      node BS      ref_presence\n",
       "1      1   0.06250 1           \n",
       "2      2   0.71875 1           \n",
       "3      3   0.15625 1           \n",
       "4      4   0.31250 0           \n",
       "5      5   0.21875 1           \n",
       "6      6   0.28125 0           \n",
       "7      7   0.40625 1           \n",
       "8      8   0.28125 1           \n",
       "9      9   0.37500 1           \n",
       "10    10   0.40625 0           \n",
       "11    11   0.40625 0           \n",
       "12    12   0.84375 1           \n",
       "13     1   0.15625 1           \n",
       "14     2   0.12500 0           \n",
       "15     3   0.43750 1           \n",
       "16     4   0.03125 1           \n",
       "17     5   0.09375 1           \n",
       "18     6   0.15625 1           \n",
       "19     7   0.50000 1           \n",
       "20     8   0.18750 1           \n",
       "21     9   0.28125 1           \n",
       "22    10   0.09375 0           \n",
       "23    11   0.25000 0           \n",
       "24    12   0.21875 0           \n",
       "25     1   0.03125 0           \n",
       "26     2   0.43750 0           \n",
       "27     3   0.12500 0           \n",
       "28     4   0.31250 0           \n",
       "29     5   0.31250 1           \n",
       "30     6   0.93750 1           \n",
       "⋮     ⋮    ⋮       ⋮           \n",
       "29971  7   0.62500 1           \n",
       "29972  8   0.56250 1           \n",
       "29973  9   0.62500 1           \n",
       "29974 10   0.71875 1           \n",
       "29975 11   0.84375 1           \n",
       "29976 12   0.65625 0           \n",
       "29977  1   0.65625 1           \n",
       "29978  2   0.31250 0           \n",
       "29979  3   0.12500 0           \n",
       "29980  4   0.46875 1           \n",
       "29981  5   0.56250 1           \n",
       "29982  6   0.31250 0           \n",
       "29983  7   0.62500 1           \n",
       "29984  8   0.65625 1           \n",
       "29985  9   0.56250 1           \n",
       "29986 10   0.84375 1           \n",
       "29987 11   0.81250 1           \n",
       "29988 12   0.62500 0           \n",
       "29989  1   0.71875 1           \n",
       "29990  2   0.43750 0           \n",
       "29991  3   0.00000 0           \n",
       "29992  4   0.31250 1           \n",
       "29993  5   0.65625 1           \n",
       "29994  6   0.28125 0           \n",
       "29995  7   0.68750 1           \n",
       "29996  8   0.50000 1           \n",
       "29997  9   0.75000 1           \n",
       "29998 10   0.87500 1           \n",
       "29999 11   0.75000 1           \n",
       "30000 12   0.59375 0           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "BS_df_ME_SM"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92bfb310-9e7c-43cb-8a9d-dce71d69920d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "<br><br><br>\n",
    "# USEFULL STUFF BUT NOT INCLUDED IN PAPER\n",
    "<br><br><br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "e1ff4f0e-ec70-4989-89c7-f007d09969c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using name, family_group as id variables\n",
      "\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method (SM + ML and Bigtree)   OLD VERSION\n",
    "#\n",
    "\n",
    "# It takes a table of all families as rows (500) and it goes to a table of 1000 (500 BigTree + 500 SM + ML) where the running times instead of being separated into two different coluns now they are in the same one.\n",
    "# name + group columns will work as unique ID.  group is the directory name of the main.nf run/s, while name is the name of the family. \n",
    "fin_tracefile_mod_wide_melted           = melt(fin_tracefile_mod_wide)\n",
    "\n",
    "# transform running times from millisecond to seconds. And rename the column names.\n",
    "fin_tracefile_mod_wide_melted$value     = fin_tracefile_mod_wide_melted$value / 1000\n",
    "colnames(fin_tracefile_mod_wide_melted) = c(\"name\", \"family_group\", \"Method\", \"Time\")\n",
    "\n",
    "# crate a table with two boxplots (x axis) with Method columns as values and on the y the running times.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel. \n",
    "p = ggplot(fin_tracefile_mod_wide_melted,aes(x=Method, y=Time, fill=Method)) +\n",
    "    geom_boxplot(show.legend = FALSE) +\n",
    "    theme_light() +\n",
    "    ylab(\"CPU time (sec)\") + xlab(\"Method\")\n",
    "ggsave(filename=\"figures/CPU_time_per_method.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6df071e6-9d31-4158-ae52-3384e9e35a6b",
   "metadata": {},
   "source": [
    "### OLD version all families toghther"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53a77a3f-22e3-4136-b883-32db83aaa230",
   "metadata": {},
   "source": [
    "### TODO make the following better or remove it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "a73a53a9-f1a1-40c9-84e9-09965fe17f26",
   "metadata": {},
   "outputs": [],
   "source": [
    "# update the trace matrix with the length of the input msa\n",
    "library(\"Biostrings\") \n",
    "fin_tracefile_mod_with_len = fin_tracefile_mod_wide\n",
    "\n",
    "# for each family go and read the input MSA (input of main.nf) to get length of MSA (number of columns).\n",
    "# Iterate over each row of the dataframe\n",
    "for (i in 1:nrow(fin_tracefile_mod_with_len)) {\n",
    "    # Construct the full file path using group (directory) and name (filename) to read the input msa\n",
    "    # TODO make it not depend on the rerun_bla_bla direcrtory\n",
    "    input_msa_path = file.path(paste0(parent_of_all_dir, \"rerun_25_species_corrected_branches\"), fin_tracefile_mod_with_len$family_group[i], paste0(gsub(\"[()]\", \"\", fin_tracefile_mod_with_len$name[i]), \".ma\"))\n",
    "\n",
    "    # read the msa in fasta format and get the length of the first sequence (aka length of the alignment)\n",
    "    msa_file = readAAStringSet(input_msa_path)\n",
    "    msa_len  = length(msa_file[[1]])\n",
    "    \n",
    "    # add the length to the table \n",
    "    fin_tracefile_mod_with_len$length[i] = msa_len\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1b9d82dd-7856-44ec-99f5-60086ee8b8d1",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error: object 'fin_tracefile_mod_with_len' not found\n",
     "output_type": "error",
     "traceback": [
      "Error: object 'fin_tracefile_mod_with_len' not found\nTraceback:\n",
      "1. select(., -Big_Tree, -Supermatrix_ML)",
      "2. mutate(., fold_ratio = Big_Tree/Supermatrix_ML)",
      "3. .handleSimpleError(function (cnd) \n . {\n .     watcher$capture_plot_and_output()\n .     cnd <- sanitize_call(cnd)\n .     watcher$push(cnd)\n .     switch(on_error, continue = invokeRestart(\"eval_continue\"), \n .         stop = invokeRestart(\"eval_stop\"), error = invokeRestart(\"eval_error\", \n .             cnd))\n . }, \"object 'fin_tracefile_mod_with_len' not found\", base::quote(eval(expr, \n .     envir)))"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method (SM + ML and Bigtree) ratio plot of each family (x axis number columns BigMSA) \n",
    "# \n",
    "\n",
    "# Create a new column with the division of Big_Tree by Supermatrix_ML running times for each family (500)\n",
    "# and then remove the original columns Big_Tree Supermatrix_ML\n",
    "fin_tracefile_mod_ratio = fin_tracefile_mod_with_len %>%\n",
    "  mutate(fold_ratio = Big_Tree / Supermatrix_ML) %>%\n",
    "  select(-Big_Tree, -Supermatrix_ML)\n",
    "\n",
    "# create a table with on x axis the length of the MSA (n columns)  and on y the fold ratio (BigTree / SM + ML) of running times. The plot will have now 500 points.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel.\n",
    "p = ggplot(fin_tracefile_mod_ratio, aes(x=length, y=fold_ratio)) +\n",
    "    geom_point(color = 'blue', size = 2, alpha = 0.6) +  # Add points\n",
    "    theme_light() +\n",
    "    labs(x = \"MSA length (n# columns)\",\n",
    "       y = \"fold ratio (BigTree / Supermatrix_ML)\")\n",
    "ggsave(filename=\"figures/CPU_time_scatterplot_msa_len_BigTree_SM_ML.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1a03b1b-cfef-4abf-a0f0-d0fd195c15bf",
   "metadata": {},
   "source": [
    "### TODO the below was just a 10 familoy trial make it go away "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "873d9c1c-ab54-40b6-aed9-90caa7d43601",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO make this into the section of custom functions.\n",
    "# Function to convert to milliseconds\n",
    "convert_to_ms <- function(time_str) {\n",
    "    # Extract components using regular expressions\n",
    "    millis = \"0\"\n",
    "    seconds = \"0\"\n",
    "    minutes = \"0\"\n",
    "    if (str_detect(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=ms)\")) {\n",
    "        millis <- str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=ms)\")\n",
    "    } else {\n",
    "        seconds <- str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=s)\")\n",
    "        minutes <- str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=m)\")\n",
    "    }\n",
    "    \n",
    "    # Convert each component to milliseconds (coercing to numeric where needed)\n",
    "    seconds_ms <- as.numeric(seconds) * 1000\n",
    "    minutes_ms <- as.numeric(minutes) * 60 * 1000\n",
    "    millis_ms <- as.numeric(millis)\n",
    "    \n",
    "    # Sum all components and handle NAs by treating them as 0\n",
    "    total_ms <- sum(c(seconds_ms, minutes_ms, millis_ms), na.rm = TRUE)\n",
    "    return(total_ms)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "9a84bc61-7c7b-4805-a7dc-47071e51fc39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"results\"\n"
     ]
    }
   ],
   "source": [
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    # each main.nf output run has a trace file repoprting various information on the run itself\n",
    "    tracefile_path = file.path(run, \"pipeline_info/execution_trace_2024-09-27_17-39-02.txt\")\n",
    "    tracefile      = read.table(tracefile_path, sep = \"\\t\", header = TRUE)\n",
    "\n",
    "    # here we parse the trace file to retain only the processes that we want to take the running time (realtime) from\n",
    "    tracefile_mod = tracefile[grep(\"simulated_data:run_phylo_ML_supermatrix_aln_sim|simulated_data:only_concatenate_aln_sim|simulated_data:run_phylo_ML_full_aln_sim\", tracefile$name), c(\"name\", \"realtime\")]\n",
    "\n",
    "    # Use strsplit to split the 'name' column by whitespace\n",
    "    split_name    = strsplit(as.character(tracefile_mod$name), \" \")\n",
    "\n",
    "    # Create new columns\n",
    "    tracefile_mod$name       = sapply(split_name, `[`, 2) # First part \n",
    "    tracefile_mod$native_id  = sapply(split_name, `[`, 1) # Second part \n",
    "\n",
    "    # Reorder columns so 'native_id' becomes the second column\n",
    "    tracefile_mod = tracefile_mod[, c(\"name\", \"native_id\", \"realtime\")]\n",
    "\n",
    "    # trasform from human readable time to millisecond time\n",
    "    tracefile_mod$realtime = sapply(tracefile_mod$realtime, convert_to_ms)\n",
    "\n",
    "    # transforming the above table to one that have the process names (simulated_data:run_phylo_ML_supermatrix_aln_sim ecc..) as column names, \n",
    "    # the rows will be still the family names (working as ID) and the values in each column will be the running times associated to the column name (process name) for that given family.\n",
    "    tracefile_mod_wide = spread(tracefile_mod, native_id, realtime)\n",
    "    \n",
    "    # creating a new last column called Supermatrix_ML that has aas values the running times of concatenation plus ML tree computation (on concatenated MSA) \n",
    "    tracefile_mod_wide$Supermatrix_ML = as.numeric(tracefile_mod_wide$`simulated_data:only_concatenate_aln_sim`) + as.numeric(tracefile_mod_wide$`simulated_data:run_phylo_ML_supermatrix_aln_sim`)\n",
    "\n",
    "    # retain only the family name, BIgTree ML running time and SM + ML running time\n",
    "    tracefile_mod_wide = tracefile_mod_wide[,c(1,3,5)]\n",
    "\n",
    "    # call the second column name Big_Tree and set their value to numeric type\n",
    "    colnames(tracefile_mod_wide)[2] = \"Big_Tree\"\n",
    "    tracefile_mod_wide$Big_Tree = as.numeric(tracefile_mod_wide$Big_Tree)\n",
    "\n",
    "    # adding a first column that groups families by the run of the main.nf\n",
    "    tracefile_mod_wide$group = rep(basename(run), 10)\n",
    "\n",
    "    # add tables toghether raw-wise\n",
    "    if(run == run_dirs[1]) {\n",
    "        fin_tracefile_mod_wide = tracefile_mod_wide\n",
    "    } else {\n",
    "        fin_tracefile_mod_wide = rbind(fin_tracefile_mod_wide, tracefile_mod_wide)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "ee66b8ad-a4df-4ad9-bdf2-7f8d7a13c8c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>Big_Tree</dt><dd>195.4</dd><dt>Supermatrix_ML</dt><dd>23.47</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[Big\\textbackslash{}\\_Tree] 195.4\n",
       "\\item[Supermatrix\\textbackslash{}\\_ML] 23.47\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "Big_Tree\n",
       ":   195.4Supermatrix_ML\n",
       ":   23.47\n",
       "\n"
      ],
      "text/plain": [
       "      Big_Tree Supermatrix_ML \n",
       "        195.40          23.47 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "196"
      ],
      "text/latex": [
       "196"
      ],
      "text/markdown": [
       "196"
      ],
      "text/plain": [
       "[1] 196"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "21.87"
      ],
      "text/latex": [
       "21.87"
      ],
      "text/markdown": [
       "21.87"
      ],
      "text/plain": [
       "[1] 21.87"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# make the means of all (500) family running time for the BigTree and ML + SM. Bring them to seconds (/1000) and round the digits to 2\n",
    "round(colMeans(fin_tracefile_mod_wide[, c(2, 3)])/1000,digits = 2)\n",
    "\n",
    "# compute the median for the BigTree approach and the  ML + SM.\n",
    "round(median(fin_tracefile_mod_wide$Big_Tree)/1000,digits = 2)\n",
    "round(median(fin_tracefile_mod_wide$Supermatrix_ML)/1000,digits = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "ed1275aa-e0bd-4628-bab2-5eb5a54271ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using name, group as id variables\n",
      "\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "\n",
    "fin_tracefile_mod_wide_melted           = melt(fin_tracefile_mod_wide)\n",
    "\n",
    "fin_tracefile_mod_wide_melted$value     = fin_tracefile_mod_wide_melted$value / 1000\n",
    "colnames(fin_tracefile_mod_wide_melted) = c(\"name\",\"group\",\"Method\",\"Time\")\n",
    "\n",
    "p = ggplot(fin_tracefile_mod_wide_melted,aes(x=Method,y=Time,fill=Method)) +\n",
    "    geom_boxplot(show.legend = FALSE) +\n",
    "    theme_light() +\n",
    "    ylab(\"CPU time (sec)\") + xlab(\"Method\")\n",
    "ggsave(filename=\"CPU_time_per_method_10_fam.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31c3465f-5bf7-465b-bf62-e004f9441080",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
