{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8d6c408c-3b82-4cbb-b94a-cb9728c78fe6",
   "metadata": {},
   "source": [
    "# Analysis of simulated data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "606b5d7a-8d48-4caf-8c72-e89dce7c1208",
   "metadata": {},
   "source": [
    "\n",
    "TODO add here a brief description of your project or analysis. Explain the purpose, goals, or context of the notebook here.\n",
    "\n",
    "## Table of Contents\n",
    "- [1. Chapter 1 - Introduction](#chapter-1-introduction)\n",
    "- [2. Chapter 2 - R packages](#chapter-2-R-packages)\n",
    "- [3. Chapter 3 - Input specification](#chapter-3-Input-specification)\n",
    "- [4. Chapter 4 - Supermatrix and Supertree trees VS true topologies](#chapter-4-Supermatrix-and-Supertree-trees-VS-true-topologies)\n",
    "- [5. Chapter 5 - BigTree VS true topologies](#chapter-5-BigTree-VS-true-topologies)\n",
    "- [6. Chapter 6 - RF distance summaries](#chapter-6-RF-distance-summaries)\n",
    "    - [6.1 Subchapter 6.1 - Table 1](#subchapter-6_1-Table-1)\n",
    "    - [6.2 Subchapter 6.2 - Figure 2](#subchapter-6_2-Figure-2)\n",
    "- [7. Chapter 7 - Running times](#chapter-7-Running-times)\n",
    "    - [7.1 Subchapter 7.1 - Figure 3](#subchapter-7_1-Figure-3)\n",
    "- [8. Chapter 8 - Bootstrap analysis](#chapter-8-Bootstrap-analysis)\n",
    "    - [8.1 Subchapter 8.1 - AUC computation](#subchapter-8_1-AUC-computation)\n",
    "    - [8.2 Subchapter 8.2 - MCC computation](#subchapter-8_2-MCC-computation)\n",
    "    - [8.3 Subchapter 8.3 - Reference presence vs BS ](#subchapter-8_3-Reference-presence-vs-BS)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d203093e-da42-45a7-8e98-853d4403630a",
   "metadata": {},
   "source": [
    "<a id=\"chapter-1-introduction\"></a>\n",
    "\n",
    "## Chapter 1 - Introduction\n",
    "\n",
    "TODO add description of notebook and how to use it.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3912a30d-42ec-4e90-99e6-8bacfeb50512",
   "metadata": {},
   "source": [
    "<a id=\"chapter-2-R-packages\"></a>\n",
    "\n",
    "## Chapter 2 - R packages\n",
    "\n",
    "Here is the list of packages needed for this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "67058cec-ea5e-4ce3-be29-faf428919086",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: ape\n",
      "\n",
      "Loading required package: phytools\n",
      "\n",
      "Loading required package: maps\n",
      "\n",
      "Loading required package: ade4\n",
      "\n",
      "\n",
      "Attaching package: ‘plotrix’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:phytools’:\n",
      "\n",
      "    rescale\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘plyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:maps’:\n",
      "\n",
      "    ozone\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ape’:\n",
      "\n",
      "    where\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘rstatix’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    desc, mutate\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘gplots’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plotrix’:\n",
      "\n",
      "    plotCI\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    lowess\n",
      "\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "\n",
      "Attaching package: ‘viridis’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:maps’:\n",
      "\n",
      "    unemp\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘gridExtra’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘scales’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:viridis’:\n",
      "\n",
      "    viridis_pal\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plotrix’:\n",
      "\n",
      "    rescale\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:phytools’:\n",
      "\n",
      "    rescale\n",
      "\n",
      "\n",
      "Type 'citation(\"pROC\")' for a citation.\n",
      "\n",
      "\n",
      "Attaching package: ‘pROC’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:phangorn’:\n",
      "\n",
      "    coords\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    cov, smooth, var\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘tidyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:reshape2’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(\"phangorn\")\n",
    "library(\"ggplot2\")\n",
    "library(\"reshape2\")\n",
    "library(\"geiger\")\n",
    "library(\"adephylo\")\n",
    "library(\"phytools\")\n",
    "library(\"rlist\")\n",
    "library(\"plotrix\")\n",
    "library(\"plyr\")\n",
    "library(\"dplyr\")\n",
    "library(\"hrbrthemes\")\n",
    "library(\"rstatix\")\n",
    "library(\"gplots\")\n",
    "library(\"viridis\")\n",
    "library('gtools')\n",
    "library(gridExtra)\n",
    "library(grid)\n",
    "library(\"ggforce\")\n",
    "library(\"scales\")\n",
    "library(\"pROC\")\n",
    "library(\"tidyr\")\n",
    "library(\"stringr\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b609768d-d957-4b11-83ce-c27152a96fb2",
   "metadata": {},
   "source": [
    "TODO make a chapter for aqll custom functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20bb020e-4f1d-4338-ac5a-abecc8a06121",
   "metadata": {},
   "source": [
    "<a id=\"chapter-3-Input-specification\"></a>\n",
    "\n",
    "## Chapter 3 - Input specification\n",
    "\n",
    "Change the parent_of_all_dir variable to give the location of the input. All input files are expected to be under it. Otherwise change the single input location to you liking.<br>\n",
    "All input files or directory paths are listed here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "04187777-5e20-43b3-9f39-62523fc0fc0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the location of the directory that contains all inputs for analyses \n",
    "parent_of_all_dir = \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/\"\n",
    "\n",
    "# The directory (one or more) that contain all the results from the main.nf pipeline run. It can be a glob path.\n",
    "dir_containing_all_pipe_results = paste0(parent_of_all_dir, \"rerun_25_species_corrected_branches/rerun_sim_b_factor*\")\n",
    "\n",
    "#  TODO describe what this file does\n",
    "reference_tree_codefile = file.path(parent_of_all_dir, \"ref_tree.code\")\n",
    "\n",
    "# get the species/organism list of names used by the main.nf pipeline\n",
    "species_name_file = file.path(parent_of_all_dir, \"orthologs_org_ids_to_concatenate\")\n",
    "\n",
    "\n",
    "# each run of main.nf has one output directory. here we get that or multiple runs outputs directories.\n",
    "run_dirs = Sys.glob(dir_containing_all_pipe_results)\n",
    "run_dirs = run_dirs[file.info(run_dirs)$isdir]\n",
    "\n",
    "# Create the 'figures' directory if it doesn't exist. Created where this notebook is executed.\n",
    "if (!dir.exists(\"figures\")) {\n",
    "    dir.create(\"figures\")\n",
    "}\n",
    "\n",
    "# Create the 'tables' directory if it doesn't exist. Created where this notebook is executed.\n",
    "if (!dir.exists(\"tables\")) {\n",
    "    dir.create(\"tables\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "566da14f-4b2d-4610-9a9a-6f892dfc9964",
   "metadata": {},
   "source": [
    "<a id=\"chapter-4-Supermatrix-and-Supertree-trees-VS-true-topologies\"></a>\n",
    "\n",
    "## Chapter 4 - Supermatrix and Supertree trees VS true topologies\n",
    "\n",
    "TODO add description "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c707c5c4-51d2-4b25-a6cc-b8a9ccb9e0aa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n"
     ]
    }
   ],
   "source": [
    "# setting global lists that eill hold all family level values\n",
    "RF_avg_list = list()\n",
    "collect_branch_len_supermatrix_ml = c()\n",
    "collect_diameter_supermatrix_ml = c()\n",
    "\n",
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    # the mapping used later on to rename species trees according to a more generic sequence name\n",
    "    ref_codefile = read.table(file = reference_tree_codefile )\n",
    "    \n",
    "    # get al the family subdirectories in each mqin.nf output run dir\n",
    "    fam_dirs = grep(\"avg_branchlen_0.7_protein_family*\", list.dirs(run, recursive = FALSE), value = TRUE)\n",
    "    count_fam = 0\n",
    "\n",
    "    # this will contain the mean branch length for all true/reference paralog trees for each gene family\n",
    "    avg_branch_length = c()\n",
    "\n",
    "    # read the names of the ortholog species.\n",
    "    orgs = read.table(file = species_name_file)\n",
    "\n",
    "    # get all units trees for all combination of methods \n",
    "    for (m in fam_dirs) {\n",
    "        \n",
    "        count_fam = count_fam + 1\n",
    "\n",
    "        # print every 10 families a check on progress to screen\n",
    "        if (count_fam %% 10 == 0) {\n",
    "            print(m)\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250 = 25 organism * 10 samples/statistical repeats) for the combination Minimun Evolution (ME) + supermatrix method (SM). \n",
    "        # We unroot every tree and fully resolve it to use the RF.dist package and others. \n",
    "        all_units_file  = file.path(m, \"results/rerun\", \"all_units.nwk\")\n",
    "        all_units_trees = read.tree(all_units_file, keep.multi=TRUE)\n",
    "        all_units_trees = unroot(all_units_trees)\n",
    "        \n",
    "        # check if there any non-bynary (consensus trees) among the 250 trees present in the file above\n",
    "        if (length(which(is.binary(all_units_trees) == FALSE)) != 0) {\n",
    "            \n",
    "            # if there are non-binary trees, iterate through them and transformed into a binary one using multi2di package\n",
    "            for (tree in which(is.binary(all_units_trees) == FALSE)) {\n",
    "                all_units_trees[[tree]] = multi2di(all_units_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250) ME + supertree (ST) method and handle non-binary trees as above\n",
    "        all_superfine_file  = file.path(m, \"results/rerun\", \"all_units_superfine_trees.nwk\")\n",
    "        all_superfine_trees = read.tree(all_superfine_file, keep.multi=TRUE)\n",
    "        all_superfine_trees = unroot(all_superfine_trees)\n",
    "        if (length(which(is.binary(all_superfine_trees) == FALSE)) != 0) {\n",
    "            for (tree in which(is.binary(all_superfine_trees) == FALSE)) {\n",
    "                all_superfine_trees[[tree]] = multi2di(all_superfine_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250) Maximum Likelihood (ML) + SM \n",
    "        all_units_raxml_file  = file.path(m, \"results/rerun\", \"all_units_RAxML_trees.nwk\")\n",
    "        all_units_raxml_trees = read.tree(all_units_raxml_file, keep.multi=TRUE)\n",
    "        all_units_raxml_trees = unroot(all_units_raxml_trees)\n",
    "        if (length(which(is.binary(all_units_raxml_trees) == FALSE)) != 0) {\n",
    "            for (tree in which(is.binary(all_units_raxml_trees) == FALSE)) {\n",
    "                all_units_raxml_trees[[tree]] = multi2di(all_units_raxml_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # take only the last 10 of the 250 above trees (unit25's 10 samples/replicates)(unit25 = all 25 single units concatenated) and collect branch lengths (each branch in tree, 27 in total 2*15-3) and diameter (using patristic distance)\n",
    "        for(mytree in 241:length(all_units_raxml_trees)) {\n",
    "            collect_branch_len_supermatrix_ml = c(collect_branch_len_supermatrix_ml, all_units_raxml_trees[[mytree]]$edge.length)\n",
    "            collect_diameter_supermatrix_ml = c(collect_diameter_supermatrix_ml, max(distTips(all_units_raxml_trees[[mytree]], method = 'patristic')))\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250) ML + ST\n",
    "        all_raxml_superfine_file  = file.path(m, \"results/rerun\", \"all_units_RAxML_superfine_trees.nwk\")\n",
    "        all_raxml_superfine_trees = read.tree(all_raxml_superfine_file, keep.multi=TRUE)\n",
    "        all_raxml_superfine_trees = unroot(all_raxml_superfine_trees)\n",
    "        if (length(which(is.binary(all_raxml_superfine_trees) == FALSE)) != 0) {\n",
    "            for (tree in which(is.binary(all_raxml_superfine_trees) == FALSE)) {\n",
    "                all_raxml_superfine_trees[[tree]] = multi2di(all_raxml_superfine_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # Reference tree block, each gene family has it's own reference tree that is the one generated by the simulation step. Along this tree all sequences are generated and then alligned and passed as input for main.nf pipeline.\n",
    "        ref_tree_file = file.path(run, paste0(basename(m), \".nwk\"))\n",
    "        ref_tree      = read.tree(ref_tree_file, keep.multi=FALSE)\n",
    "\n",
    "        # Since each gene family has a predifined gene tree (ref_tree) composed of a fixed species tree pasted as leaf of the paralog tree. The tree is designed like this by choice. Each gene families varies only in the paralog tree that it has that is unique (15 tips).\n",
    "        # Here we extract the paralog tree and to do so we just need to extract the subtree with all the sequences of omne species. Since each species have the same paralog tree. So we can extract the subtree of whichever species it does not matter.\n",
    "        ref_subtree   = keep.tip(ref_tree, as.character(ref_codefile$V1))\n",
    "\n",
    "        # we then rename the tips of the paralog tree (ref_subtree) to a more generic keyname. Effectively going from rat1 to seq1. \n",
    "        for (j in 1:length(ref_subtree$tip.label)) { \n",
    "            ref_subtree$tip.label[j] = as.character(ref_codefile$V2[which(ref_codefile$V1 == ref_subtree[[\"tip.label\"]][j])])\n",
    "        }\n",
    "\n",
    "        # Unroot the tree and perform aother operation and then save the paralog tree to the specifc gene faimily directory\n",
    "        ref_subtree        = unroot(ref_subtree)\n",
    "        ref_subtree        = list(ref_subtree)\n",
    "        class(ref_subtree) = \"multiPhylo\"\n",
    "        subtree_out_file   = file.path(m, \"results/rerun\", paste0(\"ref_subtree_\", basename(m), \".nwk\"))\n",
    "        # TODO remove this comment write.tree(ref_subtree, file=subtree_out_file)\n",
    "\n",
    "        # update on the average branch length with the mean of the true/refernce paralog tree of the given gene family\n",
    "        avg_branch_length = c(avg_branch_length, mean(ref_subtree[[1]]$edge.length))\n",
    "\n",
    "        # Make a list containing all trees gathered so far and compute all possible pair comparisons among trees. There are 1001*1001-1001 unique pairewise comparisons ( a square matrix of size 1001 with 0 in diagonal ).\n",
    "        # On this matrix (plotdata) all other measuramnts are done. This is a symmetric matrix (distance matrix)\n",
    "        all_trees        = c(ref_subtree, all_units_trees, all_superfine_trees, all_units_raxml_trees, all_raxml_superfine_trees)\n",
    "        class(all_trees) = \"multiPhylo\"\n",
    "        all_vs_all_rf    = RF.dist(all_trees, normalize = TRUE)\n",
    "        plotdata         = data.matrix(all_vs_all_rf)\n",
    "\n",
    "        # all ME metrics to be computed superfine = SuperTree\n",
    "        mean_rf           = c()\n",
    "        mean_superfine_rf = c()\n",
    "\n",
    "        # all ML metrics to be computed\n",
    "        mean_raxml_rf           = c()\n",
    "        mean_raxml_superfine_rf = c()\n",
    "\n",
    "        # proceed throught the plotdata matrix taking from the second element one every ten up to 242. Basically proceed at chunkcs of 10 (samples). On those chuncks compute the mean and the Standard error. \n",
    "        # average all sample \"values\" whithin a unit. going from 250 value to 25 averages\n",
    "        # This iterate only through the values of the comparison between the reference paralofg tree (ref_subtree)(first column/row of matrix) and the 250 paraogs trees of ME + SM (all_units_trees)\n",
    "        for (i in seq(2, (length(orgs$V1)*10)+1, by=10)) {\n",
    "            mean_rf = c(mean_rf, mean(plotdata[1, i:(i+9)]))\n",
    "        }\n",
    "\n",
    "        # This instead goes from row 252 to 492 of the plotdata matrix. It still goes in chuncks of 10 (samples). \n",
    "        # So this effectively does what is done above but between the reference paralofg tree and the 250 paraogs trees of ME + ST (all_superfine_trees)\n",
    "        for (i in seq( (length(orgs$V1)*10)+2, (2*(length(orgs$V1)*10)+1), by=10 )) {\n",
    "            mean_superfine_rf = c(mean_superfine_rf, mean(plotdata[1,i:(i+9)]))\n",
    "        }\n",
    "\n",
    "        # This for loop goes from 502 to 742. It still goes in chuncks of 10 (samples).\n",
    "        # So this effectively does what is done above but between the reference paralofg tree and the 250 paraogs trees of ML + SM (all_units_raxml_trees)\n",
    "        for (i in seq( (2*(length(orgs$V1)*10))+2, (3*(length(orgs$V1)*10)+1), by=10 )) {\n",
    "            mean_raxml_rf = c(mean_raxml_rf, mean(plotdata[1,i:(i+9)]))\n",
    "        }\n",
    "\n",
    "        # This for loop goes from 752 to 992. It still goes in chuncks of 10 (samples).\n",
    "        # So this effectively does what is done above but between the reference paralofg tree and the 250 paraogs trees of ML + ST (all_raxml_superfine_trees)\n",
    "        for (i in seq( (3*(length(orgs$V1)*10))+2, (4*(length(orgs$V1)*10)+1), by=10)) {\n",
    "            mean_raxml_superfine_rf = c(mean_raxml_superfine_rf, mean(plotdata[1,i:(i+9)]))\n",
    "        }\n",
    "\n",
    "        # collect all mean RF values 25 mean_rf, 25 mean_superfine_rf ecc.. for each family. One for each unit.\n",
    "        collection_mean_rf      = c(mean_rf, mean_superfine_rf, mean_raxml_rf, mean_raxml_superfine_rf)\n",
    "  \n",
    "        # Finally update the \"global\" lists created at the beginning\n",
    "        RF_avg_list      = list.append(RF_avg_list, collection_mean_rf)\n",
    "\n",
    "    }\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9ac26094-fda0-489f-af78-a70716514196",
   "metadata": {},
   "outputs": [],
   "source": [
    "# transform the list of lists (RF_avg_list) into a dataframe (matrix) size 500 * 100 (25 units for all combinations of ME,ML and SM,ST)\n",
    "RF_avg_as_df = ldply(RF_avg_list, recursive=FALSE)\n",
    "\n",
    "# create the column names for the later dataframe, ME = minimum Evolutio, ML = Maximum Likelihood, SM = SuperMatrix , ST = SuperTree. names will be like ME_SM_unit_1\n",
    "units_names = c()\n",
    "for (i in 1:length(orgs$V1)) {\n",
    "    units_names = c(units_names, paste(\"ME_SM\", \"unit\", i, sep=\"_\"), \n",
    "                    paste(\"ME_ST\", \"unit\", i, sep=\"_\"), \n",
    "                    paste(\"ML_SM\", \"unit\", i, sep=\"_\"), \n",
    "                    paste(\"ML_ST\", \"unit\", i, sep=\"_\"))\n",
    "}\n",
    "\n",
    "# natural_sort them or human readable sort, so that the order is  ME_SM unit1...unit25, ME_ST unit1....uniy25, ML_SM and ML_ST\n",
    "units_names = mixedsort(units_names)\n",
    "\n",
    "# TODO see if standar error se was actually usefull\n",
    "RF_df = round(RF_avg_as_df, digits=3)\n",
    "\n",
    "# add the units name generated before as column names for the dataframe and save it to file\n",
    "colnames(RF_df) = units_names\n",
    "write.table(RF_df, file = \"tables/avg_RF_between_ref_supermatrix_and_supertree.tsv\", quote = FALSE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f8b0912-095a-4aef-9277-f40429e45a5e",
   "metadata": {},
   "source": [
    "<a id=\"chapter-5-BigTree-VS-true-topologies\"></a>\n",
    "\n",
    "## Chapter 5 BigTree VS true topologies\n",
    "\n",
    "TODO add description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ce413952-5667-4c45-8d24-5a578c3527a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n"
     ]
    }
   ],
   "source": [
    "# Trees from BigTree (full_aln) ME/ML (paralog tree) VS true topologies \n",
    "\n",
    "paralog_ME_vs_true_tree = c()\n",
    "paralog_ML_vs_true_tree = c()\n",
    "RF_matrices = list()\n",
    "\n",
    "\n",
    "collect_branch_len_sub_from_ref = c()\n",
    "collect_diameter_sub_from_ref = c()\n",
    "\n",
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    # the mapping used later on to rename species trees according to a more generic sequence name\n",
    "    ref_codefile = read.table(file = reference_tree_codefile )\n",
    "    \n",
    "    # get al the family subdirectories in each mqin.nf output run dir\n",
    "    fam_dirs = grep(\"avg_branchlen_0.7_protein_family*\", list.dirs(run, recursive = FALSE), value = TRUE)\n",
    "    count_fam = 0\n",
    "\n",
    "    # this will contain the mean branch length for all true/reference paralog trees for each gene family\n",
    "    avg_branch_length = c()\n",
    "\n",
    "    # read the names of the ortholog species.\n",
    "    orgs = read.table(file = species_name_file)\n",
    "\n",
    "    # extract and work on all BigTree paralog files for all families. this trees are computed from the main.nf pipeline in simulation mode.\n",
    "    for (m in fam_dirs) {\n",
    "        \n",
    "        count_fam = count_fam + 1\n",
    "\n",
    "        # print every 10 families a check on progress to screen\n",
    "        if (count_fam %% 10 == 0) {\n",
    "            print(m)\n",
    "        }\n",
    "\n",
    "        # read the paralog tree obtained from the BigTree (gene family tree) created unsing ME.\n",
    "        # the paralog tree was obtained extracting each species subtree for the gene family tree, then this (25) subtrees were merged into one exactly like is done for SuperTree (using superfine).\n",
    "        fastme_paralog_tree_file          = file.path(m, \"results/rerun\", paste0(basename(m), \"_full_aln_coded.phylip_fastme_tree_paralog_tree.nwk\"))\n",
    "        fastme_paralog_tree_from_full_aln = read.tree(fastme_paralog_tree_file, keep.multi=FALSE)\n",
    "        fastme_paralog_tree_from_full_aln = unroot(fastme_paralog_tree_from_full_aln)\n",
    "        fastme_paralog_tree_from_full_aln = list(fastme_paralog_tree_from_full_aln)\n",
    "\n",
    "\n",
    "        # read the paralog tree obtained from the BigTree (gene family tree) created unsing ML. paralog tree constructed same way as above.\n",
    "        raxml_paralog_tree_file           = file.path(m, \"results/rerun\", paste0(\"RAxML_bestTree.\", basename(m), \"_full_aln_coded_raxml_paralog_tree.nwk\"))   \n",
    "        raxml_paralog_tree_from_full_aln  = read.tree(raxml_paralog_tree_file, keep.multi=FALSE)\n",
    "        raxml_paralog_tree_from_full_aln  = unroot(raxml_paralog_tree_from_full_aln)\n",
    "        raxml_paralog_tree_from_full_aln  = list(raxml_paralog_tree_from_full_aln)\n",
    "                \n",
    "        # read the reference tree (gene family tree) the one set as thruth and on which the sequences were generated along.\n",
    "        ref_tree_file = file.path(run, paste0(basename(m), \".nwk\"))\n",
    "        ref_tree      = read.tree(ref_tree_file, keep.multi=FALSE)\n",
    "\n",
    "        # each reference species paralog tree is identical in topologty (designed so) but diverges for branch lengths. here we retrieve some info for that.\n",
    "        for (species in orgs$V1) {\n",
    "            \n",
    "            # ectract tips like before and then extract subtree and unroot it. \n",
    "            # this is the refence paralog tree for each species and whithin one family all species have the same paralog tree as far as topology goes, what changes between this species paralog trees is their branch lengths.\n",
    "            tips_to_keep = grep(paste0(\"^\", species), ref_tree$tip.label, perl = TRUE)\n",
    "            test_subtree = keep.tip(ref_tree, tips_to_keep)\n",
    "            test_subtree = unroot(test_subtree)\n",
    "\n",
    "            # get branch length and diameter\n",
    "            collect_branch_len_sub_from_ref = c(collect_branch_len_sub_from_ref, test_subtree$edge.length)\n",
    "            collect_diameter_sub_from_ref   = c(collect_diameter_sub_from_ref, max(distTips(test_subtree, method = 'patristic')))\n",
    "        }\n",
    "\n",
    "        # Since each gene family has a predifined gene tree (ref_tree) composed of a fixed species tree pasted as leaf of the paralog tree. The tree is designed like this by choice. Each gene families varies only in the paralog tree that it has that is unique (15 tips).\n",
    "        # Here we extract the paralog tree and to do so we just need to extract the subtree with all the sequences of one species. Since each species have the same paralog tree. So we can extract the subtree of whichever species it does not matter.\n",
    "        ref_subtree   = keep.tip(ref_tree, as.character(ref_codefile$V1))\n",
    "\n",
    "        # we then rename the tips of the paralog tree (ref_subtree) to a more generic keyname. Effectively going from rat1 to seq1. \n",
    "        for (j in 1:length(ref_subtree$tip.label)) { \n",
    "            ref_subtree$tip.label[j] = as.character(ref_codefile$V2[which(ref_codefile$V1 == ref_subtree[[\"tip.label\"]][j])])\n",
    "        }\n",
    "\n",
    "        # Unroot the tree and put into a list\n",
    "        ref_subtree        = unroot(ref_subtree)\n",
    "        ref_subtree        = list(ref_subtree)\n",
    "\n",
    "\n",
    "        # update on the average branch length with the mean of the true/refernce paralog tree of the given gene family\n",
    "        avg_branch_length = c(avg_branch_length, mean(ref_subtree[[1]]$edge.length))\n",
    "        \n",
    "        # put all  BIgTree (ME/ML) paralog tres (2) and the reference paralog tree into one list.\n",
    "        all_trees        = c(ref_subtree, fastme_paralog_tree_from_full_aln, raxml_paralog_tree_from_full_aln)\n",
    "        class(all_trees) = \"multiPhylo\"\n",
    "        \n",
    "        # compute the RF dictance between 1 refernce paralog the other two paralog trees: ME ML. In reality all difrrences between them.\n",
    "        all_vs_all_rf = RF.dist(all_trees, normalize=TRUE)\n",
    "        plotdata      = data.matrix(all_vs_all_rf)\n",
    "        RF_matrices   = list.append(RF_matrices, plotdata)\n",
    "        \n",
    "        # do the average of all pair comparison between the species paralog trees and the reference tree. -c[1] means not consider the first column, because first row first column is RF between reference tree and itself\n",
    "        paralog_ME_vs_true_tree = c(paralog_ME_vs_true_tree, plotdata[1, 2])\n",
    "        paralog_ML_vs_true_tree = c(paralog_ML_vs_true_tree, plotdata[1, 3])\n",
    "    }\n",
    "}\n",
    "\n",
    "write.table(paralog_ME_vs_true_tree, file = \"tables/paralog_ME_vs_true_tree.tsv\", quote=FALSE, row.names = FALSE, col.names = FALSE)\n",
    "write.table(paralog_ML_vs_true_tree, file = \"tables/paralog_ML_vs_true_tree.tsv\", quote=FALSE, row.names = FALSE, col.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab7ac11a-c926-4fa6-aafd-d3c8b281e7fc",
   "metadata": {},
   "source": [
    "<a id='chapter-6-RF-distance-summaries'></a>\n",
    "\n",
    "## Chapter 6 - RF distance summaries\n",
    "\n",
    "todo add description"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0553ac22-a97a-49f6-adfe-8323e9003fba",
   "metadata": {},
   "source": [
    "<a id='subchapter-6_1-Table-1'></a>\n",
    "\n",
    "### Subchapter 6.1 - Table 1\n",
    "todo add description."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f493767f-1b5e-4251-848d-6ff64f26f4f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Method</th><th scope=col>Unit.25</th><th scope=col>Unit.1</th><th scope=col>Difference</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ME_SM     </td><td>0.24  sd  0.15</td><td>0.47  sd  0.13</td><td>0.23</td></tr>\n",
       "\t<tr><td>ME_ST     </td><td>0.23  sd  0.13</td><td>0.47  sd  0.13</td><td>0.24</td></tr>\n",
       "\t<tr><td>ML_SM     </td><td>0.19  sd  0.14</td><td>0.44  sd  0.13</td><td>0.25</td></tr>\n",
       "\t<tr><td>ML_ST     </td><td>0.2  sd  0.12 </td><td>0.44  sd  0.13</td><td>0.24</td></tr>\n",
       "\t<tr><td>ME_BigTree</td><td>0.25  sd  0.16</td><td>Na            </td><td>Na  </td></tr>\n",
       "\t<tr><td>ML_BigTree</td><td>0.18  sd  0.15</td><td>Na            </td><td>Na  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 4\n",
       "\\begin{tabular}{llll}\n",
       " Method & Unit.25 & Unit.1 & Difference\\\\\n",
       " <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t ME\\_SM      & 0.24  sd  0.15 & 0.47  sd  0.13 & 0.23\\\\\n",
       "\t ME\\_ST      & 0.23  sd  0.13 & 0.47  sd  0.13 & 0.24\\\\\n",
       "\t ML\\_SM      & 0.19  sd  0.14 & 0.44  sd  0.13 & 0.25\\\\\n",
       "\t ML\\_ST      & 0.2  sd  0.12  & 0.44  sd  0.13 & 0.24\\\\\n",
       "\t ME\\_BigTree & 0.25  sd  0.16 & Na             & Na  \\\\\n",
       "\t ML\\_BigTree & 0.18  sd  0.15 & Na             & Na  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 4\n",
       "\n",
       "| Method &lt;chr&gt; | Unit.25 &lt;chr&gt; | Unit.1 &lt;chr&gt; | Difference &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| ME_SM      | 0.24  sd  0.15 | 0.47  sd  0.13 | 0.23 |\n",
       "| ME_ST      | 0.23  sd  0.13 | 0.47  sd  0.13 | 0.24 |\n",
       "| ML_SM      | 0.19  sd  0.14 | 0.44  sd  0.13 | 0.25 |\n",
       "| ML_ST      | 0.2  sd  0.12  | 0.44  sd  0.13 | 0.24 |\n",
       "| ME_BigTree | 0.25  sd  0.16 | Na             | Na   |\n",
       "| ML_BigTree | 0.18  sd  0.15 | Na             | Na   |\n",
       "\n"
      ],
      "text/plain": [
       "  Method     Unit.25        Unit.1         Difference\n",
       "1 ME_SM      0.24  sd  0.15 0.47  sd  0.13 0.23      \n",
       "2 ME_ST      0.23  sd  0.13 0.47  sd  0.13 0.24      \n",
       "3 ML_SM      0.19  sd  0.14 0.44  sd  0.13 0.25      \n",
       "4 ML_ST      0.2  sd  0.12  0.44  sd  0.13 0.24      \n",
       "5 ME_BigTree 0.25  sd  0.16 Na             Na        \n",
       "6 ML_BigTree 0.18  sd  0.15 Na             Na        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read the table compiled in chapter 4. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df_new = read.table(\"tables/avg_RF_between_ref_supermatrix_and_supertree.tsv\", header=T)\n",
    "\n",
    "# do the mean and sd of the column. effectively getting a vector of size 100 filed with averages of averages and with ids like SM_ML_unit9\n",
    "rf_df_new_means = colMeans(df_new)\n",
    "sd_df_new_means = apply(df_new, 2, sd)\n",
    "\n",
    "# Initialize empty list to hold the extracted method names and unit names values\n",
    "methods = c(\"ME_SM\", \"ME_ST\", \"ML_SM\",  \"ML_ST\")\n",
    "units   = 25\n",
    "\n",
    "# Create an empty dataframe with row names for methods and columns for unit_1 to unit_25\n",
    "avg_rf_df_method_unit           = data.frame(Method = methods, matrix(ncol = units, nrow = 4))\n",
    "colnames(avg_rf_df_method_unit) = c(\"Method\", paste0(\"unit_\", 1:units))\n",
    "\n",
    "# Fill the dataframe using the keys of the named vector of rf_df_new_means, these names are the same for the vector sd_df_new_means too.\n",
    "for (key in names(rf_df_new_means)) {\n",
    "\n",
    "  # Split the key into parts (method and unit)\n",
    "  parts  = strsplit(key, \"_unit_\")[[1]]\n",
    "  method = parts[1]\n",
    "  unit   = as.numeric(parts[2])\n",
    "\n",
    "  # Place the value in the correct row and column of the dataframe adding the standard deviation to it.\n",
    "  avg_rf_df_method_unit[avg_rf_df_method_unit$Method == method, paste0(\"unit_\", unit)] = paste0(round(rf_df_new_means[key], digit=2), \"  sd  \", round(sd_df_new_means[key], digit=2))\n",
    "}\n",
    "\n",
    "# create the actual table with only unit1, unit25 difference values from above df\n",
    "diff   = c(as.numeric(sub(\" sd.*\", \"\", avg_rf_df_method_unit$unit_1)) - as.numeric(sub(\" sd.*\", \"\", avg_rf_df_method_unit$unit_25)))\n",
    "table1 = data.frame(Method = avg_rf_df_method_unit$Method,\n",
    "                    \"Unit 25\"    = avg_rf_df_method_unit$unit_25,\n",
    "                    \"Unit 1\"     = avg_rf_df_method_unit$unit_1,\n",
    "                    \"Difference\" = diff )\n",
    "\n",
    "# read the table compiled in chapter 5. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df_bigtree_ME = read.table(\"tables/paralog_ME_vs_true_tree.tsv\", header=F)\n",
    "df_bigtree_ML = read.table(\"tables/paralog_ML_vs_true_tree.tsv\", header=F)\n",
    "\n",
    "# make average of all family BigTree RF (ME and ML) distances to their respective paralog tree. going from 500 values to 1.\n",
    "rf_df_bigtree_ME_mean = colMeans(df_bigtree_ME)\n",
    "rf_df_bigtree_ME_sd   = apply(df_bigtree_ME, 2, sd)\n",
    "rf_df_bigtree_ML_mean = colMeans(df_bigtree_ML)\n",
    "rf_df_bigtree_ML_sd   = apply(df_bigtree_ML, 2, sd)\n",
    "\n",
    "# add the above four values to table1\n",
    "bigtree_ME_row = c(\"ME_BigTree\", paste0(round(rf_df_bigtree_ME_mean, digits = 2), \"  sd  \", round(rf_df_bigtree_ME_sd, digits = 2)), \"Na\", \"Na\")\n",
    "bigtree_ML_row = c(\"ML_BigTree\", paste0(round(rf_df_bigtree_ML_mean, digits = 2), \"  sd  \", round(rf_df_bigtree_ML_sd, digits = 2)), \"Na\", \"Na\")\n",
    "table1      = rbind(table1, bigtree_ME_row)\n",
    "table1      = rbind(table1, bigtree_ML_row)\n",
    "\n",
    "table1\n",
    "write.table(table1, file=\"tables/Table1_average_RF.tsv\", quote=FALSE, sep=\"\\t\", row.names=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "735442c0-400c-4549-9064-b9892fff8880",
   "metadata": {},
   "source": [
    "<a id=\"subchapter-6_2-Figure-2\"></a>\n",
    "\n",
    "### Subchapter 6.2 Figure 2\n",
    "\n",
    "TODO add description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0981fb0d-587e-4428-bcb5-07301de851eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "# Set plot dimensions and resolution for inline display\n",
    "options(repr.plot.width = 6, repr.plot.height = 5, repr.plot.res = 300)\n",
    "\n",
    "# Load the necessary data\n",
    "# Reading in average RF values between reference big tree ME/ML from a file computed in chapter 5.\n",
    "df_bigtree_ME = read.table(\"tables/paralog_ME_vs_true_tree.tsv\", header=F)\n",
    "df_bigtree_ML = read.table(\"tables/paralog_ML_vs_true_tree.tsv\", header=F)\n",
    "rf_df_bigtree_ME_mean = colMeans(df_bigtree_ME)\n",
    "rf_df_bigtree_ML_mean = colMeans(df_bigtree_ML)\n",
    "\n",
    "# Reading in average RF values between reference supermatrix and supertree\n",
    "df_new = read.table(\"../analysis/tables/avg_RF_between_ref_supermatrix_and_supertree.tsv\", header = TRUE)\n",
    "\n",
    "# Calculate column-wise means of the loaded data\n",
    "rf_df_new_means = colMeans(df_new)\n",
    "\n",
    "# Reshape the means data into a long format\n",
    "rf_df_new_means_melted = melt(rf_df_new_means)\n",
    "\n",
    "# Rename the first column to \"RF\" for clarity\n",
    "colnames(rf_df_new_means_melted)[1] = \"RF\"\n",
    "\n",
    "# Add a column indicating the topologies (25 levels, repeated 4 times for each method)\n",
    "rf_df_new_means_melted$Topology = c(rep(colnames(df_new)[1:25], 4))\n",
    "\n",
    "# Add a column specifying the method (Supermatrix-ME, Supertree-ME, Supermatrix-ML, Supertree-ML)\n",
    "rf_df_new_means_melted$Method = c(rep(\"Supermatrix-ME\", 25), rep(\"Supertree-ME\", 25),\n",
    "                                  rep(\"Supermatrix-ML\", 25), rep(\"Supertree-ML\", 25))\n",
    "\n",
    "# Sort the data by topology using mixedorder\n",
    "rf_df_new_means_melted = rf_df_new_means_melted[mixedorder(rf_df_new_means_melted$Topology), ]\n",
    "\n",
    "# crate the x axis label names\n",
    "x_axis_labels = gsub(\"ME_SM_\", \"\", colnames(df_new)[1:25])\n",
    "\n",
    "# Modify the Topology column to match the cleaned x-axis labels\n",
    "rf_df_new_means_melted$Topology = rep(x_axis_labels, each = 4)  # Repeat each value per method (4 times)\n",
    "\n",
    "# Define custom colors for each Method\n",
    "custom_colors = c(\"Supertree-ME\" = \"#7CB6E2\",     # Light blue \n",
    "                  \"Supermatrix-ME\" = \"#2F739B\",   # Dark blue\n",
    "                  \"Supertree-ML\" = \"#E5A23D\",     # Orange/yellow\n",
    "                  \"Supermatrix-ML\" = \"#D05159\")   # Red\n",
    "\n",
    "# Create the line plot\n",
    "p = ggplot(rf_df_new_means_melted, aes(x = factor(Topology, levels = x_axis_labels), \n",
    "                                       y = RF, group = Method, color = Method)) + \n",
    "    geom_line() +  # Add lines connecting the points by method\n",
    "    geom_point(shape = 16) +  # Use shape 16 (solid dot) for all points\n",
    "\n",
    "    # Add a horizontal dashed line representing the average RF of ME big tree\n",
    "    geom_hline(yintercept = rf_df_bigtree_ME_mean, \n",
    "               linetype = \"dashed\", color = \"#606060\") +\n",
    "    \n",
    "    # Annotate the plot with a label for the average bigtree ME RF line\n",
    "    annotate(\"text\", x = 18, y = rf_df_bigtree_ME_mean + 0.025, \n",
    "             label = \"Avg RF between BigTree-ME and true topology\", \n",
    "             size = 3, color = \"#606060\") +\n",
    "\n",
    "    # Add a horizontal dashed line representing the average RF of ML big tree\n",
    "    geom_hline(yintercept = rf_df_bigtree_ML_mean, \n",
    "               linetype = \"dashed\", color = \"#1E7005\") +\n",
    "    \n",
    "    # Annotate the plot with a label for the average bigtree ML RF line\n",
    "    annotate(\"text\", x = 12, y = rf_df_bigtree_ML_mean - 0.025, \n",
    "             label = \"Avg RF between BigTree-ML and true topology\", \n",
    "             size = 3, color = \"#1E7005\") +\n",
    "\n",
    "    # Apply a light theme for better readability\n",
    "    theme_light() + \n",
    "\n",
    "    # Adjust x-axis text to be tilted at 45 degrees and align to the right\n",
    "    theme(axis.text.x = element_text(angle = 45, hjust = 1), \n",
    "          legend.position = \"right\",\n",
    "          axis.title.x = element_text(face = \"bold\"),  \n",
    "          axis.title.y = element_text(face = \"bold\"))  + \n",
    "    \n",
    "    # Label the axes\n",
    "    ylab(\"Normalized RF\") + \n",
    "    xlab(\"Topology\") + \n",
    "\n",
    "    # Set y-axis limits between 0 and 0.5\n",
    "    scale_y_continuous(limits = c(0, 0.5)) +\n",
    "\n",
    "    # Customize color and shape scales for the different methods\n",
    "    scale_color_discrete(breaks = c(\"Supertree-ME\", \"Supermatrix-ME\", \"Supertree-ML\", \"Supermatrix-ML\")) + \n",
    "    scale_shape_discrete(breaks = c(\"Supertree-ME\", \"Supermatrix-ME\", \"Supertree-ML\", \"Supermatrix-ML\")) +\n",
    "    scale_color_manual(values = custom_colors)  # Apply custom colors\n",
    "\n",
    "# Save the plot to a file with high resolution\n",
    "ggsave(filename = \"figures/RF_lineplot_figure2.png\", plot = p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfa19d80-b20f-4a24-af34-ea65b39aa0f2",
   "metadata": {},
   "source": [
    "<a id=\"chapter-7-Running-times\"></a>\n",
    "\n",
    "## Chapter 7 - Running times\n",
    "\n",
    "TODO add description "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "062ad02a-baca-4fce-b3ee-e8ea3b3bb039",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\"\n"
     ]
    }
   ],
   "source": [
    "# read the trace files for running times\n",
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    # each main.nf output run has a trace file repoprting various information on the run itself\n",
    "    tracefile_path = file.path(run, \"trace_cpu_time.txt\")\n",
    "    tracefile      = read.table(tracefile_path, header = TRUE)\n",
    "    \n",
    "    # here we parse the trace file to retain only the processes that we want to take the running time (realtime) from\n",
    "    tracefile_mod  = tracefile[grep(\"simulated_data:run_phylo_ML_supermatrix_aln_sim|simulated_data:only_concatenate_aln_sim|simulated_data:run_phylo_ML_full_aln_sim\", tracefile$native_id), c(\"name\", \"native_id\", \"realtime\")]\n",
    "    \n",
    "    # transforming the above table to one that have the process names (simulated_data:run_phylo_ML_supermatrix_aln_sim ecc..) as column names, \n",
    "    # the rows will be still the family names (working as ID) and the values in each column will be the running times associated to the column name (process name) for that given family.\n",
    "    tracefile_mod_wide = spread(tracefile_mod, native_id, realtime)\n",
    "\n",
    "    # creating a new last column called Supermatrix_ML that has aas values the running times of concatenation plus ML tree computation (on concatenated MSA) \n",
    "    tracefile_mod_wide$Supermatrix_ML = as.numeric(tracefile_mod_wide$`simulated_data:only_concatenate_aln_sim`) + as.numeric(tracefile_mod_wide$`simulated_data:run_phylo_ML_supermatrix_aln_sim`)\n",
    "\n",
    "    # retain only the family name, BIgTree ML running time and SM + ML running time\n",
    "    tracefile_mod_wide = tracefile_mod_wide[,c(1,3,5)]\n",
    "\n",
    "    # call the second column name Big_Tree and set their value to numeric type\n",
    "    colnames(tracefile_mod_wide)[2] = \"Big_Tree\"\n",
    "    tracefile_mod_wide$Big_Tree = as.numeric(tracefile_mod_wide$Big_Tree)\n",
    "\n",
    "    # adding a first column that groups families by the run of the main.nf\n",
    "    tracefile_mod_wide$family_group = rep(basename(run), 100)\n",
    "\n",
    "    # add tables toghether raw-wise\n",
    "    if(run == run_dirs[1]) {\n",
    "        fin_tracefile_mod_wide = tracefile_mod_wide\n",
    "    } else {\n",
    "        fin_tracefile_mod_wide = rbind(fin_tracefile_mod_wide, tracefile_mod_wide)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d6a00cd4-31ce-440c-8528-94805af06047",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>Big_Tree</dt><dd>638.12</dd><dt>Supermatrix_ML</dt><dd>67.76</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[Big\\textbackslash{}\\_Tree] 638.12\n",
       "\\item[Supermatrix\\textbackslash{}\\_ML] 67.76\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "Big_Tree\n",
       ":   638.12Supermatrix_ML\n",
       ":   67.76\n",
       "\n"
      ],
      "text/plain": [
       "      Big_Tree Supermatrix_ML \n",
       "        638.12          67.76 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "588.3"
      ],
      "text/latex": [
       "588.3"
      ],
      "text/markdown": [
       "588.3"
      ],
      "text/plain": [
       "[1] 588.3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "63.85"
      ],
      "text/latex": [
       "63.85"
      ],
      "text/markdown": [
       "63.85"
      ],
      "text/plain": [
       "[1] 63.85"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# TODO make nicer and update with whatever is changed above\n",
    "\n",
    "# make the means of all (500) family running time for the BigTree and ML + SM. Bring them to seconds (/1000) and round the digits to 2\n",
    "round(colMeans(fin_tracefile_mod_wide[, c(2, 3)])/1000,digits = 2)\n",
    "\n",
    "# compute the median for the BigTree approach and the  ML + SM.\n",
    "round(median(fin_tracefile_mod_wide$Big_Tree)/1000,digits = 2)\n",
    "round(median(fin_tracefile_mod_wide$Supermatrix_ML)/1000,digits = 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b79de7a-f8b9-4dc7-8fc6-232cf86be231",
   "metadata": {},
   "source": [
    "<a id='subchapter-7_1-Figure-3'></a>\n",
    "\n",
    "### Subchapter 7.1 - Figure 3\n",
    "\n",
    "TODO add description if necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ba993fba-7f4b-48aa-9b03-775093172898",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO make this into the section of custom functions.\n",
    "# Function to assign new values based on a mapping provided by the user\n",
    "assign_mapped_values <- function(df, column_name, mapping, new_col_name) {\n",
    "  # Check if mapping is a named vector or a list\n",
    "  if (!is.vector(mapping) || is.null(names(mapping))) {\n",
    "    stop(\"Mapping must be a named vector or list.\")\n",
    "  }\n",
    "  \n",
    "  # Get the unique values from the specified column\n",
    "  unique_values = unique(df[[column_name]])\n",
    "  \n",
    "  # Check if the mapping contains all unique values\n",
    "  if (!all(unique_values %in% names(mapping))) {\n",
    "    stop(\"Mapping must include all unique values from the specified column.\")\n",
    "  }\n",
    "  \n",
    "  # Initialize a vector to store the assigned values\n",
    "  assigned_values = vector(\"character\", nrow(df))\n",
    "  \n",
    "  # Assign values based on the mapping\n",
    "  for (i in seq_along(unique_values)) {\n",
    "    assigned_values[df[[column_name]] == unique_values[i]] <- mapping[names(mapping) == unique_values[i]]\n",
    "  }\n",
    "  \n",
    "  # Add the assigned values to the original dataframe\n",
    "  df[[new_col_name]] = assigned_values\n",
    "  \n",
    "  return(df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4568c025-fffa-45df-bfff-086d67ad2a53",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using name, family_group, length as id variables\n",
      "\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method (SM + ML and Bigtree) divided per group (by sequence length)\n",
    "#\n",
    "\n",
    "# assign to each family_group a mapping values  (as new column). basically insert the length of the sequences in each family as an additional column in the dataframe.\n",
    "# Remember each family in the same group have the same length of sequence generated, the MSA though might differ in length due to gaps.\n",
    "mapping = c(\"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\" = \"group1\\nlen = 32\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\" = \"group2\\nlen = 46\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\" = \"group3\\nlen = 50\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\" = \"group4\\nlen = 60\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\" = \"group5\\nlen = 82\")\n",
    "fin_tracefile_mod_reassigned = assign_mapped_values(fin_tracefile_mod_wide, \"family_group\", mapping, \"length\")\n",
    "\n",
    "# reorder the columns to have the numerical values at the end\n",
    "fin_tracefile_mod_reassigned = fin_tracefile_mod_reassigned[, c(\"name\", \"family_group\", \"length\", \"Big_Tree\", \"Supermatrix_ML\")]\n",
    "\n",
    "# It takes a table of all families as rows (500) and it goes to a table of 1000 (500 BigTree + 500 SM + ML) where the running times instead of being separated into two different coluns now they are in the same one.\n",
    "# name + group columns will work as unique ID.  group is the directory name of the main.nf run/s, while name is the name of the family. \n",
    "fin_tracefile_mod_melted           = melt(fin_tracefile_mod_reassigned)\n",
    "\n",
    "\n",
    "# transform running times from millisecond to seconds. And rename the column names.\n",
    "fin_tracefile_mod_melted$value     = fin_tracefile_mod_melted$value / 1000\n",
    "colnames(fin_tracefile_mod_melted) = c(\"name\", \"family_group\", \"length\", \"Method\", \"Time\")\n",
    "\n",
    "# create a table with 5 pairs of boxplot. Group1, 2.. ecc.. each group has 100 families and two running time (realtime) mesuraments for the BigTree and SM + ML approaches.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel.\n",
    "p = ggplot(fin_tracefile_mod_melted, aes(x=length, y=Time, fill=Method)) +\n",
    "    geom_boxplot(show.legend = TRUE) +\n",
    "    theme_light() +\n",
    "    labs(x = \"sequence length\",\n",
    "       y = \"CPU time (sec)\",\n",
    "       fill = \"Method\") +\n",
    "\n",
    "    # Adjust x-axis text to be tilted at 45 degrees and align to the right\n",
    "    theme(\n",
    "        axis.title.x = element_text(face = \"bold\"),  \n",
    "        axis.title.y = element_text(face = \"bold\")\n",
    "    ) +\n",
    "\n",
    "    # Manually set colors for the two methods \"#1E7005\" bright green and \"#D05159\" redish\n",
    "    scale_fill_manual(\n",
    "        values = c(\"Big_Tree\" = \"#1E7005\", \"Supermatrix_ML\" = \"#D05159\"),\n",
    "        labels = c(\"BigTree-ML\", \"Supermatrix-ML\")\n",
    "    )\n",
    "\n",
    "ggsave(filename=\"figures/CPU_time_per_group_BigTree_SM_ML_figure3.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83903c34-3f32-431b-abc5-156cb2abeaf8",
   "metadata": {},
   "source": [
    "### Suplementary Fig todo put number of supplementary figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "81650244-8771-4d70-a7c6-57d5218f928d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method ratio (Bigtree / SM + ML ) boxplot of each family ratio by group (x axis groups y ratio Big/SM) \n",
    "# \n",
    "\n",
    "# start by taking the already present info for the families and their grouping from fig 2 cell and compute ratio Bigtree / SM + ML\n",
    "# Create a new column with the division of Big_Tree by Supermatrix_ML running times for each family (500)\n",
    "# and then remove the original columns Big_Tree Supermatrix_ML\n",
    "fin_tracefile_mod_ratio = fin_tracefile_mod_reassigned %>%\n",
    "  mutate(fold_ratio = Big_Tree / Supermatrix_ML) %>%\n",
    "  select(-Big_Tree, -Supermatrix_ML)\n",
    "\n",
    "# create a table with 5 boxplots. Group1, 2.. ecc.. each group has 100 families and  the ratio of two running time (realtime) mesuraments for the BigTree and SM + ML approaches.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel.\n",
    "p = ggplot(fin_tracefile_mod_ratio, aes(x=length, y=fold_ratio)) +\n",
    "    geom_boxplot(show.legend = FALSE) +\n",
    "    theme_light() +\n",
    "    labs(x = \"sequence length\",\n",
    "       y = \"CPU time fold ratio (BigTree / Supermatrix_ML)\")\n",
    "\n",
    "# TODO add subplot number to filename\n",
    "ggsave(filename=\"figures/CPU_time_per_group_BigTree_SM_ML_ratio.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42387886-a000-4c7b-80d3-1dc82b4f4825",
   "metadata": {},
   "source": [
    "<a id=\"chapter-8-Bootstrap-analysis\"></a>\n",
    "\n",
    "## Chapter 8 - Bootstrap analysis\n",
    "\n",
    "TODO add decriptoion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0b0a5b14-5c84-4733-9cff-0313fc6105d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# ME + SM block\n",
    "#\n",
    "\n",
    "\n",
    "# Create the dataframe that will contain all bs values and other info like presence in reference tree\n",
    "BS_df_ME_SM = data.frame(\n",
    "  group = character(), \n",
    "  family = character(), \n",
    "  unit = integer(), \n",
    "  sample = integer(),\n",
    "  node = integer(),\n",
    "  BS = numeric(), \n",
    "  ref_presence = numeric()  # Using integer type for 0 and 1 values\n",
    ")\n",
    "\n",
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "    \n",
    "    # get al the family subdirectories in each mqin.nf output run dir\n",
    "    fam_dirs = grep(\"avg_branchlen_0.7_protein_family*\", list.dirs(run, recursive = FALSE), value = TRUE)\n",
    "    count_fam = 0\n",
    "\n",
    "    # read the names of the ortholog species.\n",
    "    orgs = read.table(file = species_name_file)\n",
    "\n",
    "    # the mapping used later on to rename species trees according to a more generic sequence name\n",
    "    ref_codefile = read.table(file = reference_tree_codefile )\n",
    "    \n",
    "    # work at family level to get the BS value\n",
    "    for (m in fam_dirs) {\n",
    "        \n",
    "        count_fam = count_fam + 1\n",
    "\n",
    "        # print every 10 families a check on progress to screen\n",
    "        if (count_fam %% 10 == 0) {\n",
    "            print(m)\n",
    "        }\n",
    "\n",
    "        # read the file that contains the length of the MSA (number of columns) of the input MSA of main.nf for that given family.\n",
    "        bs_file = file.path(m, \"results/rerun\", \"BS.dat\")\n",
    "        bs      = read.table(bs_file)\n",
    "        myBS    = 32                       # set manually because that is the scaling factor for the smallest of the groups. not dynamical anymore\n",
    "        #myBS    = bs$V1\n",
    "\n",
    "        # read the reference tree (gene family tree) the one set as thruth and on which the sequences were generated along.\n",
    "        ref_tree_file = file.path(run, paste0(basename(m), \".nwk\"))\n",
    "        ref_tree      = read.tree(ref_tree_file, keep.multi=FALSE)\n",
    "\n",
    "        # Since each gene family has a predifined gene tree (ref_tree) composed of a fixed species tree pasted as leaf of the paralog tree. The tree is designed like this by choice. Each gene families varies only in the paralog tree that it has that is unique (15 tips).\n",
    "        # Here we extract the paralog tree and to do so we just need to extract the subtree with all the sequences of one species. Since each species have the same paralog tree. So we can extract the subtree of whichever species it does not matter.\n",
    "        ref_subtree   = keep.tip(ref_tree, as.character(ref_codefile$V1))\n",
    "\n",
    "        # we then rename the tips of the paralog tree (ref_subtree) to a more generic keyname. Effectively going from rat1 to seq1. \n",
    "        for (j in 1:length(ref_subtree$tip.label)) { \n",
    "            ref_subtree$tip.label[j] = as.character(ref_codefile$V2[which(ref_codefile$V1 == ref_subtree[[\"tip.label\"]][j])])\n",
    "        }\n",
    "\n",
    "        # Unroot the tree and put into a list\n",
    "        ref_paralog_tree = unroot(ref_subtree)\n",
    "      \n",
    "        # do a for loop on the number of species (25) for reading all trees of a given unit (from unit1 to unit25).\n",
    "        for (k in 1:length(orgs$V1)) {\n",
    "\n",
    "            # read the file that holds all ME SM trees for a given unit. it holds 10 trees one for each sample/reppliceate.\n",
    "            all_unit_trees_ME_SM_file  = file.path(m, \"results/rerun\", paste0(\"unit_\", k, \"_all.nwk\"))\n",
    "            all_unit_trees_ME_SM       = read.tree(file = all_unit_trees_ME_SM_file, keep.multi = TRUE)\n",
    "            all_unit_trees_ME_SM       = unroot(all_unit_trees_ME_SM)\n",
    "\n",
    "            BS_mean_unit_ME_SM = c()\n",
    "            \n",
    "            # for loop on the number of trees in all_unit_trees_ME_SM (10). \n",
    "            for (i in 1:length(all_unit_trees_ME_SM)) {\n",
    "\n",
    "                # Each unit_k_sample_i_rep.trees file contains as many trees as number of columns MSA gene family trees. This are all bootstrap replicate trees for the corresponding ME tree in all_unit_trees_ME_SM. \n",
    "                # This represents the variability in tree structure across the bootstrap replicates.\n",
    "                all_bs_rep_tree_ME_SM_file = file.path(m, \"results/rerun\", paste0(\"unit_\", k, \"_sample_\", i, \"_rep.trees\"))\n",
    "                all_bs_rep_tree_ME_SM      = read.tree(file = all_bs_rep_tree_ME_SM_file, keep.multi = TRUE)\n",
    "                all_bs_rep_tree_ME_SM      = unroot(all_bs_rep_tree_ME_SM)\n",
    "\n",
    "                # get only the first 32 of them. So that each group has the same number of replicate trees in it. \n",
    "                # TODO remove this workaround variable number of trees\n",
    "                all_bs_rep_tree_ME_SM = all_bs_rep_tree_ME_SM[1:32]\n",
    "\n",
    "                # calculates the bootstrap support for each clade in the tree by counting how often each clade (node) in the original tree (all_unit_trees_ME_SM[[i]]) is found in the bootstrap replicate trees (all_bs_rep_tree_ME_SM). \n",
    "                # The [-1] removes the root node’s support value, which is usually not meaningful in unrooted trees. \n",
    "                # This is a list of 12 ( 15 tips -3, 13 internal nodes for unrooted tree and the first clade is the tree itself). because each tree has 15 tips being a paralog tree. Like this by design choiche.0\n",
    "                BS_all_unit_trees_ME_SM         = prop.clades(all_unit_trees_ME_SM[[i]], all_bs_rep_tree_ME_SM)[-1]\n",
    "                \n",
    "                # check now if a clade in the unit tree of interest is also present in the paralog tree. 1 if present NA if missing.\n",
    "                # this list and the one above are syncronyzed. meaning the first element of both point to the same clade\n",
    "                presence_of_clade_in_ref_ME_SM = prop.clades(all_unit_trees_ME_SM[[i]], ref_paralog_tree)[-1]\n",
    "\n",
    "                # Any NA values in all_trees_BS (indicating clades absent in the bootstrap replicates) are replaced with 0.             \n",
    "                BS_all_unit_trees_ME_SM[is.na(BS_all_unit_trees_ME_SM)]               = 0\n",
    "                presence_of_clade_in_ref_ME_SM[is.na(presence_of_clade_in_ref_ME_SM)] = 0\n",
    "\n",
    "                # Then, the mean of these clade support values is calculated and stored.\n",
    "                val = mean((as.numeric(BS_all_unit_trees_ME_SM)))\n",
    "                BS_mean_unit_ME_SM = c(BS_mean_unit_ME_SM, val)\n",
    "\n",
    "                # Each clade’s support value is scaled by dividing by myBS (which is set to the total number of bootstrap replicates -> number of columns MSA gene family trees).\n",
    "                # add the clade values to the overall dataframe, teh id of each clade will be the concatenation of the first 5 columns.\n",
    "                vector3 = 1:12\n",
    "                tmp_df  = data.frame(\n",
    "                      group = rep(basename(run), 12),\n",
    "                      family = rep(m, 12),\n",
    "                      unit = rep(k, 12),\n",
    "                      sample = rep(i, 12),\n",
    "                      node = vector3,\n",
    "                      BS = c(as.numeric(BS_all_unit_trees_ME_SM) / myBS), # This scales the values to a proportion (0 to 1) for each clade.\n",
    "                      ref_presence = presence_of_clade_in_ref_ME_SM\n",
    "                    )\n",
    "\n",
    "                # add the new block to the overall dataframe\n",
    "                BS_df_ME_SM = rbind(BS_df_ME_SM, tmp_df)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "# save as gizepp so it can be pushed to github otherwise is too big.\n",
    "write.table(BS_df_ME_SM, gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ME_SM.tsv.gz\"), quote=FALSE, sep=\"\\t\", row.names=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4286e44c-b2f5-443e-8d2d-011ce20a5ccc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/avignoli/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# ML + SM block (like cell above)\n",
    "#\n",
    "\n",
    "\n",
    "# Create the dataframe that will contain all bs values and other info like presence in reference tree\n",
    "BS_df_ML_SM = data.frame(\n",
    "  group = character(), \n",
    "  family = character(), \n",
    "  unit = integer(), \n",
    "  sample = integer(),\n",
    "  node = integer(),\n",
    "  BS = numeric(), \n",
    "  ref_presence = numeric()  # Using integer type for 0 and 1 values\n",
    ")\n",
    "\n",
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "    \n",
    "    # get al the family subdirectories in each mqin.nf output run dir\n",
    "    fam_dirs = grep(\"avg_branchlen_0.7_protein_family*\", list.dirs(run, recursive = FALSE), value = TRUE)\n",
    "    count_fam = 0\n",
    "\n",
    "    # read the names of the ortholog species.\n",
    "    orgs = read.table(file = species_name_file)\n",
    "\n",
    "    # the mapping used later on to rename species trees according to a more generic sequence name\n",
    "    ref_codefile = read.table(file = reference_tree_codefile )\n",
    "    \n",
    "    # work at family level to get the BS value\n",
    "    for (m in fam_dirs) {\n",
    "        \n",
    "        count_fam = count_fam + 1\n",
    "\n",
    "        # print every 10 families a check on progress to screen\n",
    "        if (count_fam %% 10 == 0) {\n",
    "            print(m)\n",
    "        }\n",
    "\n",
    "        # read the file that contains the length of the MSA (number of columns) of the input MSA of main.nf for that given family.\n",
    "        bs_file = file.path(m, \"results/rerun\", \"BS.dat\")\n",
    "        bs      = read.table(bs_file)\n",
    "        myBS    = 32                       # set manually because that is the scaling factor for the smallest of the groups. not dynamical anymore\n",
    "        #myBS    = bs$V1\n",
    "\n",
    "        # read the reference tree (gene family tree) the one set as thruth and on which the sequences were generated along.\n",
    "        ref_tree_file = file.path(run, paste0(basename(m), \".nwk\"))\n",
    "        ref_tree      = read.tree(ref_tree_file, keep.multi=FALSE)\n",
    "\n",
    "        # Since each gene family has a predifined gene tree (ref_tree) composed of a fixed species tree pasted as leaf of the paralog tree. The tree is designed like this by choice. Each gene families varies only in the paralog tree that it has that is unique (15 tips).\n",
    "        # Here we extract the paralog tree and to do so we just need to extract the subtree with all the sequences of one species. Since each species have the same paralog tree. So we can extract the subtree of whichever species it does not matter.\n",
    "        ref_subtree   = keep.tip(ref_tree, as.character(ref_codefile$V1))\n",
    "\n",
    "        # we then rename the tips of the paralog tree (ref_subtree) to a more generic keyname. Effectively going from rat1 to seq1. \n",
    "        for (j in 1:length(ref_subtree$tip.label)) { \n",
    "            ref_subtree$tip.label[j] = as.character(ref_codefile$V2[which(ref_codefile$V1 == ref_subtree[[\"tip.label\"]][j])])\n",
    "        }\n",
    "\n",
    "        # Unroot the tree and put into a list\n",
    "        ref_paralog_tree = unroot(ref_subtree)\n",
    "      \n",
    "        # do a for loop on the number of species (25) for reading all trees of a given unit (from unit1 to unit25).\n",
    "        for (k in 1:length(orgs$V1)) {\n",
    "\n",
    "            # read the file that holds all ML SM trees for a given unit. it holds 10 trees one for each sample/reppliceate.\n",
    "            all_unit_trees_ML_SM_file  = file.path(m, \"results/rerun\", paste0(\"unit_\", k, \"_all_raxml.nwk\"))\n",
    "            all_unit_trees_ML_SM       = read.tree(file = all_unit_trees_ML_SM_file, keep.multi = TRUE)\n",
    "            all_unit_trees_ML_SM       = unroot(all_unit_trees_ML_SM)\n",
    "\n",
    "            BS_mean_unit_ML_SM = c()\n",
    "            \n",
    "            # for loop on the number of trees in all_unit_trees_ML_SM (10). \n",
    "            for (i in 1:length(all_unit_trees_ML_SM)) {\n",
    "\n",
    "                # Each unit_k_sample_i_rep.trees file contains as many trees as number of columns MSA gene family trees. This are all bootstrap replicate trees for the corresponding ML tree in all_unit_trees_ML_SM. \n",
    "                # This represents the variability in tree structure across the bootstrap replicates.\n",
    "                all_bs_rep_tree_ML_SM_file = file.path(m, \"results/rerun\", paste0(\"unit_\", k, \"_sample_\", i, \"_raxml_rep.trees\"))\n",
    "                all_bs_rep_tree_ML_SM      = read.tree(file = all_bs_rep_tree_ML_SM_file, keep.multi = TRUE)\n",
    "                all_bs_rep_tree_ML_SM      = unroot(all_bs_rep_tree_ML_SM)\n",
    "\n",
    "                # get only the first 32 of them. So that each group has the same number of replicate trees in it. \n",
    "                # TODO remove this workaround variable number of trees\n",
    "                all_bs_rep_tree_ML_SM = all_bs_rep_tree_ML_SM[1:32]\n",
    "\n",
    "                # calculates the bootstrap support for each clade in the tree by counting how often each clade (node) in the original tree (all_unit_trees_ML_SM[[i]]) is found in the bootstrap replicate trees (all_bs_rep_tree_ML_SM). \n",
    "                # The [-1] removes the root node’s support value, which is usually not meaningful in unrooted trees. \n",
    "                # This is a list of 12 ( 15 tips -3, 13 internal nodes for unrooted tree and the first clade is the tree itself). because each tree has 15 tips being a paralog tree. Like this by design choiche.0\n",
    "                BS_all_unit_trees_ML_SM         = prop.clades(all_unit_trees_ML_SM[[i]], all_bs_rep_tree_ML_SM)[-1]\n",
    "                \n",
    "                # check now if a clade in the unit tree of interest is also present in the paralog tree. 1 if present NA if missing.\n",
    "                # this list and the one above are syncronyzed. meaning the first element of both point to the same clade\n",
    "                presence_of_clade_in_ref_ML_SM = prop.clades(all_unit_trees_ML_SM[[i]], ref_paralog_tree)[-1]\n",
    "\n",
    "                # Any NA values in all_trees_BS (indicating clades absent in the bootstrap replicates) are replaced with 0.             \n",
    "                BS_all_unit_trees_ML_SM[is.na(BS_all_unit_trees_ML_SM)]               = 0\n",
    "                presence_of_clade_in_ref_ML_SM[is.na(presence_of_clade_in_ref_ML_SM)] = 0\n",
    "\n",
    "                # Then, the mean of these clade support values is calculated and stored.\n",
    "                val = mean((as.numeric(BS_all_unit_trees_ML_SM)))\n",
    "                BS_mean_unit_ML_SM = c(BS_mean_unit_ML_SM, val)\n",
    "\n",
    "                # Each clade’s support value is scaled by dividing by myBS (which is set to the total number of bootstrap replicates -> number of columns MSA gene family trees).\n",
    "                # add the clade values to the overall dataframe, teh id of each clade will be the concatenation of the first 5 columns.\n",
    "                vector3 = 1:12\n",
    "                tmp_df1  = data.frame(\n",
    "                      group = rep(basename(run), 12),\n",
    "                      family = rep(m, 12),\n",
    "                      unit = rep(k, 12),\n",
    "                      sample = rep(i, 12),\n",
    "                      node = vector3,\n",
    "                      BS = c(as.numeric(BS_all_unit_trees_ML_SM) / myBS), # This scales the values to a proportion (0 to 1) for each clade.\n",
    "                      ref_presence = presence_of_clade_in_ref_ML_SM\n",
    "                    )\n",
    "\n",
    "                # add the new block to the overall dataframe\n",
    "                BS_df_ML_SM = rbind(BS_df_ML_SM, tmp_df1)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "write.table(BS_df_ML_SM, gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ML_SM.tsv.gz\"), quote=FALSE, sep=\"\\t\", row.names=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ec13acc-3ca6-4416-a749-e1bad293098b",
   "metadata": {},
   "source": [
    "<a id='subchapter-8_1-AUC-computation'></a>\n",
    "\n",
    "### Subchapter 8.1 - AUC computation\n",
    "\n",
    "TODO add description if necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4ae0ece3-08e4-4d90-b6a7-28784736ac35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# figure 4.a\n",
    "\n",
    "# Reading Bootstrap values for each node and presence of node in family paralog reference tree. \n",
    "# This is done for ME and ML + SM, this tables are computed in chapter 8 so that this block does not dipend directly onit once the table is created.\n",
    "df_bs_ME_SM = read.table(gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ME_SM.tsv.gz\"), header = TRUE)\n",
    "df_bs_ML_SM = read.table(gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ML_SM.tsv.gz\"), header = TRUE)\n",
    "\n",
    "# Get unique unit values to loop over. assuming two dataframes have same number of units.\n",
    "unique_units = unique(df_bs_ME_SM$unit)\n",
    "\n",
    "# Initialize a dataframe to store AUC values for each unit, with names as \"unit_1\", \"unit_2\", etc.\n",
    "auc_df = data.frame(\n",
    "    unit      = paste0(\"unit_\", unique_units),\n",
    "    auc_ME_SM = numeric(length(unique_units)),\n",
    "    auc_ML_SM = numeric(length(unique_units))\n",
    ")\n",
    "\n",
    "\n",
    "# Calculate AUC for each unique unit, using all nodes that belong to it\n",
    "for (unit_num in unique_units) {\n",
    "    \n",
    "    # Subset the data for the current unit ME + SM\n",
    "    unit_data_ME_SM = subset(df_bs_ME_SM, unit == unit_num)\n",
    "\n",
    "    # Ensure there are enough positive and negative samples to calculate AUC ME + SM \n",
    "    if (length(unique(unit_data_ME_SM$ref_presence)) > 1) {\n",
    "    \n",
    "        # Calculate AUC using pROC\n",
    "        roc_obj_ME_SM              = roc(unit_data_ME_SM$ref_presence, unit_data_ME_SM$BS, quiet = TRUE)\n",
    "        auc_df$auc_ME_SM[unit_num] = auc(roc_obj_ME_SM)\n",
    "        \n",
    "    } else {\n",
    "    \n",
    "        # If there's only one class, AUC is undefined, assign NA\n",
    "        auc_df$auc_ME_SM[unit_num] = NA\n",
    "    }\n",
    "\n",
    "    # Subset the data for the current unit ML + SM and repeat above procedure\n",
    "    unit_data_ML_SM = subset(df_bs_ML_SM, unit == unit_num)\n",
    "\n",
    "    if (length(unique(unit_data_ML_SM$ref_presence)) > 1) {\n",
    "        roc_obj_ML_SM              = roc(unit_data_ML_SM$ref_presence, unit_data_ML_SM$BS, quiet = TRUE)\n",
    "        auc_df$auc_ML_SM[unit_num] = auc(roc_obj_ML_SM)\n",
    "    } else {\n",
    "        auc_df$auc_ML_SM[unit_num] = NA\n",
    "    }\n",
    "   \n",
    "}\n",
    "\n",
    "# Reshape data from wide to long format\n",
    "auc_df_long = auc_df %>%\n",
    "  pivot_longer(cols = starts_with(\"auc\"), \n",
    "               names_to = \"metric\", \n",
    "               values_to = \"auc_value\")\n",
    "\n",
    "# Set the unit column as a factor ordered by mixedorder\n",
    "auc_df_long$unit = factor(auc_df_long$unit, levels=auc_df$unit[mixedorder(auc_df$unit)])\n",
    "\n",
    "auc_p = ggplot(auc_df_long, aes(x = unit, y = auc_value, color = metric, shape = metric, group = metric)) +\n",
    "\n",
    "    # Adjust line thickness if needed\n",
    "    geom_line(size = 1) + \n",
    "\n",
    "    # Adjust point size if needed\n",
    "    geom_point(size = 3) +  \n",
    "\n",
    "    # 15: square, 17: triangle\n",
    "    scale_shape_manual(\n",
    "        values = c(auc_ME_SM = 15, auc_ML_SM = 17),\n",
    "        labels = c(\"Supermatrix-ME\", \"Supermatrix-ML\") \n",
    "    ) +\n",
    "\n",
    "    # Set custom colors \"#2F739B\" darck blue and \"#D05159\" redish \n",
    "    scale_color_manual(\n",
    "        values = c(auc_ME_SM = \"#2F739B\", auc_ML_SM = \"#D05159\"),\n",
    "        labels = c(\"Supermatrix-ME\", \"Supermatrix-ML\")  \n",
    "    ) +\n",
    "\n",
    "    # Apply a light theme for better readability\n",
    "    theme_light() + \n",
    "\n",
    "    # Adjust x-axis text to be tilted at 45 degrees and align to the right\n",
    "    theme(axis.text.x = element_text(angle = 45, hjust = 1), \n",
    "        legend.position = c(0.80, 0.20),  # Position legend in the bottom-right corner (x = right, y = bottom)\n",
    "        legend.margin = margin(10, 10, 10, 10),  # Add a bit of margin around the legend\n",
    "        axis.title.x = element_text(face = \"bold\"),  \n",
    "        axis.title.y = element_text(face = \"bold\")\n",
    "    )  + \n",
    "    \n",
    "    # Label the axes\n",
    "    ylab(\"ROC AUC of Bootstrap Support\") + \n",
    "    xlab(\"Topology\") + \n",
    "\n",
    "    # Set y-axis limits between 0.5 and 1\n",
    "    scale_y_continuous(limits = c(0.5, 1)) \n",
    "\n",
    "ggsave(filename=\"figures/BS_lineplot_auc_per_unit_ME_ML_SM_figure4a.png\", plot=auc_p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "2cd49d4e-034d-486d-a5df-3474db88ad2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "# PLot 4.b\n",
    "\n",
    "# Reading Bootstrap values for each node and presence of node in family paralog reference tree\n",
    "df_bs_ME_SM = read.table(gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ME_SM.tsv.gz\"), header = TRUE)\n",
    "df_bs_ML_SM = read.table(gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ML_SM.tsv.gz\"), header = TRUE)\n",
    "\n",
    "# Get unique unit assuming two dataframes have the same number of units.\n",
    "unique_units = unique(df_bs_ME_SM$unit)\n",
    "\n",
    "# Subset the data for the last (25) unit, ME + SM and ML + SM \n",
    "unit_data_ME_SM = subset(df_bs_ME_SM, unit == length(unique_units))\n",
    "unit_data_ML_SM = subset(df_bs_ML_SM, unit == length(unique_units))\n",
    "\n",
    "# Calculate AUC and ROC curves for ME + SM\n",
    "if (length(unique(unit_data_ME_SM$ref_presence)) > 1) {\n",
    "    unit_25_roc_obj_ME_SM = roc(unit_data_ME_SM$ref_presence, unit_data_ME_SM$BS, quiet = TRUE)\n",
    "    auc_unit_25_ME_SM = auc(unit_25_roc_obj_ME_SM)\n",
    "    roc_data_ME_SM = data.frame(\n",
    "        fpr = 1 - unit_25_roc_obj_ME_SM$specificities,\n",
    "        tpr = unit_25_roc_obj_ME_SM$sensitivities,\n",
    "        curve = \"ME + SM\"\n",
    "    )\n",
    "} else {\n",
    "    roc_data_ME_SM = NULL\n",
    "    auc_unit_25_ME_SM = NA\n",
    "}\n",
    "\n",
    "# Calculate AUC and ROC curves for ML + SM\n",
    "if (length(unique(unit_data_ML_SM$ref_presence)) > 1) {\n",
    "    unit_25_roc_obj_ML_SM = roc(unit_data_ML_SM$ref_presence, unit_data_ML_SM$BS, quiet = TRUE)\n",
    "    auc_unit_25_ML_SM = auc(unit_25_roc_obj_ML_SM)\n",
    "    roc_data_ML_SM = data.frame(\n",
    "        fpr = 1 - unit_25_roc_obj_ML_SM$specificities,\n",
    "        tpr = unit_25_roc_obj_ML_SM$sensitivities,\n",
    "        curve = \"ML + SM\"\n",
    "    )\n",
    "} else {\n",
    "    roc_data_ML_SM = NULL\n",
    "    auc_unit_25_ML_SM = NA\n",
    "}\n",
    "\n",
    "# Combine both ROC datasets\n",
    "roc_data_combined = rbind(roc_data_ME_SM, roc_data_ML_SM)\n",
    "\n",
    "# Plot both ROC curves in a single plot\n",
    "p_combined = ggplot(roc_data_combined, aes(x = fpr, y = tpr, color = curve)) +\n",
    "    geom_line(size = 2) +\n",
    "    geom_abline(slope = 1, intercept = 0, linetype = \"dashed\", color = \"grey\", size = 1.5) +\n",
    "    labs(x = \"False Positive Rate\", y = \"True Positive Rate\") +\n",
    "    scale_color_manual(\n",
    "        values = c(\"ME + SM\" = \"#2F739B\", \"ML + SM\" = \"#D05159\"),\n",
    "        name = \"Unit_25\",\n",
    "        labels = c(\n",
    "            paste(\"\\nSupermatrix-ME\\nAUC =\", round(auc_unit_25_ME_SM, 3)),\n",
    "            paste(\"\\nSupermatrix-ML\\nAUC =\", round(auc_unit_25_ML_SM, 3))\n",
    "        )\n",
    "    ) +\n",
    "    theme_light(base_size = 18) +\n",
    "    theme(\n",
    "        legend.position = c(0.70, 0.20),\n",
    "        legend.margin = margin(20, 20, 20, 20),\n",
    "        legend.text = element_text(size = 16),\n",
    "        axis.title.x = element_text(face = \"bold\", size = 18),\n",
    "        axis.title.y = element_text(face = \"bold\", size = 18)\n",
    "    )\n",
    "\n",
    "# Save the combined plot\n",
    "ggsave(filename=\"figures/BS_lineplot_auc_unit_25_ME_ML_SM_combined_figure4b.png\", plot=p_combined, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e56a6edd-2161-4d25-95e6-ea9c8324f62f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 5 rows containing non-finite outside the scale range\n",
      "(`stat_boxplot()`).”\n"
     ]
    }
   ],
   "source": [
    "# PLot 4.c\n",
    "\n",
    "# Reading Bootstrap values for each node and presence of node in family paralog reference tree\n",
    "df_bs_ME_SM = read.table(gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ME_SM.tsv.gz\"), header = TRUE)\n",
    "df_bs_ML_SM = read.table(gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ML_SM.tsv.gz\"), header = TRUE)\n",
    "\n",
    "# get the ammount of unique families (500), the two dataframes are assumed to have same number\n",
    "unique_family_names = unique(df_bs_ME_SM$family)\n",
    "\n",
    "# prpare the dataframe for the auc\n",
    "df_family_auc = data.frame(\n",
    "    family  = c(),\n",
    "    group   = c(),\n",
    "    method  = c(),\n",
    "    auc_val = c()\n",
    ")\n",
    "\n",
    "\n",
    "# for each family compute the auc\n",
    "for (fam_num in 1:length(unique_family_names)) {\n",
    "\n",
    "    # get only family node datas for ME + SM\n",
    "    family_data_ME_SM = subset(df_bs_ME_SM, family == unique_family_names[fam_num])\n",
    "    \n",
    "    # calculate the ROC curve and then the AUC\n",
    "    fam_roc_obj_ME_SM = roc(family_data_ME_SM$ref_presence, family_data_ME_SM$BS, quiet = TRUE)\n",
    "    auc_fam_ME_SM     = auc(fam_roc_obj_ME_SM)\n",
    "\n",
    "    # updatet the dataframe with relevant information, changing some fields for readability\n",
    "    df_family_auc = data.frame(\n",
    "        family  = c(df_family_auc$family, paste0(\"family_\", fam_num)),\n",
    "        group   = c(df_family_auc$group, family_data_ME_SM$group[1]),    # a family can only be part of one group \n",
    "        method  = c(df_family_auc$method, \"Supermatrix-ME\"),\n",
    "        auc_val = c(df_family_auc$auc_val, auc_fam_ME_SM)\n",
    "    )\n",
    "\n",
    "    # repeat all of the above for ML + SM\n",
    "    family_data_ML_SM = subset(df_bs_ML_SM, family == unique_family_names[fam_num])\n",
    "    \n",
    "    fam_roc_obj_ML_SM = roc(family_data_ML_SM$ref_presence, family_data_ML_SM$BS, quiet = TRUE)\n",
    "    auc_fam_ML_SM     = auc(fam_roc_obj_ML_SM)\n",
    "\n",
    "    df_family_auc = data.frame(\n",
    "        family  = c(df_family_auc$family, paste0(\"family_\", fam_num)),\n",
    "        group   = c(df_family_auc$group, family_data_ML_SM$group[1]),    # a family can only be part of one group \n",
    "        method  = c(df_family_auc$method, \"Supermatrix-ML\"),\n",
    "        auc_val = c(df_family_auc$auc_val, auc_fam_ML_SM)\n",
    "    )    \n",
    "}\n",
    "\n",
    "# Map group names based on provided mapping\n",
    "groupnames_mapping = list(\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\" = \"group1\\nlen = 32\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\" = \"group2\\nlen = 46\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\" = \"group3\\nlen = 50\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\" = \"group4\\nlen = 60\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\" = \"group5\\nlen = 82\"\n",
    ")\n",
    "\n",
    "# Apply the group name mapping\n",
    "df_family_auc = df_family_auc %>%\n",
    "  mutate(\n",
    "    group = unlist(groupnames_mapping[group])\n",
    "  )\n",
    "\n",
    "# Define colors for methods\n",
    "method_colors = c(\"Supermatrix-ME\" = \"#2F739B\", \"Supermatrix-ML\" = \"#D05159\")\n",
    "\n",
    "# Plot using ggplot2\n",
    "p_family = ggplot(df_family_auc, aes(x = group, y = auc_val, color = method)) +\n",
    "    geom_boxplot(alpha = 0.6) +  # Adding boxplot layer without outliers\n",
    "    scale_color_manual(values = method_colors) +  # Set colors based on method\n",
    "    labs(x = \"Sequence length\", y = \"ROC AUC of Bootstrap Support\", color = \"Method\") +  # Set axis and legend labels\n",
    "    theme_light() +\n",
    "    theme(\n",
    "    axis.title = element_text(face = \"bold\"),  # Make axis labels bold\n",
    "    legend.position = \"top\"\n",
    "    ) + \n",
    "    scale_y_continuous(limits = c(0.5, 1))\n",
    "\n",
    "ggsave(filename=\"figures/BS_boxplot_auc_ME_ML_SM_combined_figure4c.png\", plot=p_family, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e7f73b68-b87d-4083-8eed-e8011719c78e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "# TODO add fiigure number\n",
    "\n",
    "# Reading Bootstrap values for each node and presence of node in family paralog reference tree\n",
    "df_bs_ME_SM = read.table(gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ME_SM.tsv.gz\"), header = TRUE)\n",
    "df_bs_ML_SM = read.table(gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ML_SM.tsv.gz\"), header = TRUE)\n",
    "\n",
    "# get the ammount of unique families (500), the two dataframes are assumed to have same number\n",
    "unique_family_names = unique(df_bs_ME_SM$family)\n",
    "\n",
    "# prpare the dataframe for the auc for both ME and ML\n",
    "df_family_auc_ME_SM = data.frame(\n",
    "    family       = c(),\n",
    "    group        = c(),\n",
    "    method       = c(),\n",
    "    auc_delta    = c(),\n",
    "    avg_BS_delta = c()\n",
    ")\n",
    "df_family_auc_ML_SM = df_family_auc_ME_SM\n",
    "\n",
    "# for each family compute the auc\n",
    "for (fam_num in 1:length(unique_family_names)) {\n",
    "\n",
    "    # get only family node datas for ME + SM for unit1 and unit25\n",
    "    family_data_ME_SM = subset(df_bs_ME_SM, family == unique_family_names[fam_num])\n",
    "    unit1_data_ME_SM  = subset(family_data_ME_SM, unit == 1)\n",
    "    unit25_data_ME_SM = subset(family_data_ME_SM, unit == 25)\n",
    "    \n",
    "    # compute mean BS value\n",
    "    delta_fam_avg_BS_ME_SM =  mean(unit25_data_ME_SM$BS) - mean(unit1_data_ME_SM$BS)\n",
    "    \n",
    "    # calculate the ROC curve and then the AUC for unit1. if there are only 1 or 0 in the reference column AUC can not be computed.\n",
    "    if (length(unique(unit1_data_ME_SM$ref_presence)) > 1) {\n",
    "        unit1_roc_obj_ME_SM = roc(unit1_data_ME_SM$ref_presence, unit1_data_ME_SM$BS, quiet = TRUE)\n",
    "        auc_unit1_ME_SM     = auc(unit1_roc_obj_ME_SM)\n",
    "    # if there are only 1 it means THIS chiunk of nodes was good so it should be rewarded.\n",
    "    } else if (unique(unit1_data_ME_SM$ref_presence == 1)) {        \n",
    "        auc_unit1_ME_SM     = 1.0\n",
    "    # if it where only zeroes the chunk should be punished.\n",
    "    } else {\n",
    "        auc_unit1_ME_SM     = 0.5\n",
    "    }\n",
    "\n",
    "    # repeat as above for unit25\n",
    "    if (length(unique(unit25_data_ME_SM$ref_presence)) > 1) {\n",
    "        unit255_roc_obj_ME_SM = roc(unit25_data_ME_SM$ref_presence, unit25_data_ME_SM$BS, quiet = TRUE)\n",
    "        auc_unit1_ME_SM     = auc(unit1_roc_obj_ME_SM)\n",
    "    } else if (unique(unit25_data_ME_SM$ref_presence == 1)) {        \n",
    "        auc_unit25_ME_SM     = 1.0\n",
    "    } else {\n",
    "        auc_unit25_ME_SM     = 0.5\n",
    "    }\n",
    "\n",
    "    # make the delta AUC\n",
    "    delta_fam_auc_ME_SM  = auc_unit25_ME_SM - auc_unit1_ME_SM\n",
    "\n",
    "    # updatet the dataframe with relevant information, changing some fields for readability\n",
    "    df_family_auc_ME_SM = data.frame(\n",
    "        family       = c(df_family_auc_ME_SM$family, paste0(\"family_\", fam_num)),\n",
    "        group        = c(df_family_auc_ME_SM$group, family_data_ME_SM$group[1]),    # a family can only be part of one group \n",
    "        method       = c(df_family_auc_ME_SM$method, \"Supermatrix-ME\"),\n",
    "        auc_delta    = c(df_family_auc_ME_SM$auc_delta, delta_fam_avg_BS_ME_SM),\n",
    "        avg_BS_delta = c(df_family_auc_ME_SM$avg_BS_delta, delta_fam_auc_ME_SM)\n",
    "    )\n",
    "\n",
    "    # repeat all of the above for ML + SM\n",
    "    family_data_ML_SM = subset(df_bs_ML_SM, family == unique_family_names[fam_num])\n",
    "    unit1_data_ML_SM  = subset(family_data_ML_SM, unit == 1)\n",
    "    unit25_data_ML_SM = subset(family_data_ML_SM, unit == 25)\n",
    "\n",
    "    delta_fam_avg_BS_ML_SM =  mean(unit25_data_ML_SM$BS) - mean(unit1_data_ML_SM$BS)\n",
    "\n",
    "    if (length(unique(unit1_data_ML_SM$ref_presence)) > 1) {\n",
    "        unit1_roc_obj_ML_SM = roc(unit1_data_ML_SM$ref_presence, unit1_data_ML_SM$BS, quiet = TRUE)\n",
    "        auc_unit1_ML_SM     = auc(unit1_roc_obj_ML_SM)\n",
    "    } else if (unique(unit1_data_ML_SM$ref_presence == 1)) {        \n",
    "        auc_unit1_ML_SM     = 1.0\n",
    "    } else {\n",
    "        auc_unit1_ML_SM     = 0.5\n",
    "    }\n",
    "\n",
    "    if (length(unique(unit25_data_ML_SM$ref_presence)) > 1) {\n",
    "        unit255_roc_obj_ML_SM = roc(unit25_data_ML_SM$ref_presence, unit25_data_ML_SM$BS, quiet = TRUE)\n",
    "        auc_unit1_ML_SM     = auc(unit1_roc_obj_ML_SM)\n",
    "    } else if (unique(unit25_data_ML_SM$ref_presence == 1)) {        \n",
    "        auc_unit25_ML_SM     = 1.0\n",
    "    } else {\n",
    "        auc_unit25_ML_SM     = 0.5\n",
    "    }\n",
    "\n",
    "    delta_fam_auc_ML_SM  = auc_unit25_ML_SM - auc_unit1_ML_SM\n",
    "\n",
    "    df_family_auc_ML_SM = data.frame(\n",
    "        family       = c(df_family_auc_ML_SM$family, paste0(\"family_\", fam_num)),\n",
    "        group        = c(df_family_auc_ML_SM$group, family_data_ML_SM$group[1]),    \n",
    "        method       = c(df_family_auc_ML_SM$method, \"Supermatrix-ML\"),\n",
    "        auc_delta    = c(df_family_auc_ML_SM$auc_delta, delta_fam_avg_BS_ML_SM),\n",
    "        avg_BS_delta = c(df_family_auc_ML_SM$avg_BS_delta, delta_fam_auc_ML_SM)\n",
    "    )    \n",
    "}\n",
    "\n",
    "# Map group names based on provided mapping\n",
    "groupnames_mapping = list(\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\" = \"group1 len = 32\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\" = \"group2 len = 46\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\" = \"group3 len = 50\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\" = \"group4 len = 60\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\" = \"group5 len = 82\"\n",
    ")\n",
    "\n",
    "# Apply the group name mapping\n",
    "df_family_auc_ME_SM = df_family_auc_ME_SM %>%\n",
    "  mutate(\n",
    "    group = unlist(groupnames_mapping[group])\n",
    "  )\n",
    "df_family_auc_ML_SM = df_family_auc_ML_SM %>%\n",
    "  mutate(\n",
    "    group = unlist(groupnames_mapping[group])\n",
    "  )\n",
    "\n",
    "# Fit linear regression model for ME + SM \n",
    "model_ME_SM     = lm(auc_delta ~ avg_BS_delta, data = df_family_auc_ME_SM)\n",
    "slope_ME_SM     = coef(model_ME_SM)[2]  # Extract slope\n",
    "intercept_ME_SM = coef(model_ME_SM)[1]  # Extract intercept\n",
    "r_squared_ME_SM = summary(model_ME_SM)$r.squared  # Extract R-squared value\n",
    "\n",
    "# Fit linear regression model for ML + SM \n",
    "model_ML_SM     = lm(auc_delta ~ avg_BS_delta, data = df_family_auc_ML_SM)\n",
    "slope_ML_SM     = coef(model_ML_SM)[2]  # Extract slope\n",
    "intercept_ML_SM = coef(model_ML_SM)[1]  # Extract intercept\n",
    "r_squared_ML_SM = summary(model_ML_SM)$r.squared  # Extract R-squared value\n",
    "\n",
    "# Create regression equation string for both ME and ML SM\n",
    "regression_eq_ME_SM = paste0(\"R² = \", round(r_squared_ME_SM, 2))\n",
    "regression_eq_ML_SM = paste0(\"R² = \", round(r_squared_ML_SM, 2))\n",
    "\n",
    "# Custom colors for each group, change these as needed\n",
    "custom_colors = c(\n",
    "  \"group2 len = 46\" = \"#33a02c\",   # Change to your preferred color\n",
    "  \"group1 len = 32\" = \"#1f78b4\",   # Change to your preferred color\n",
    "  \"group3 len = 50\" = \"#e31a1c\",   # Change to your preferred color\n",
    "  \"group4 len = 60\" = \"#ff7f00\",   # Change to your preferred color\n",
    "  \"group5 len = 82\" = \"#6a3d9a\"    # Change to your preferred color\n",
    ")\n",
    "\n",
    "# Plot ME\n",
    "p_ME_auc_bs = ggplot(df_family_auc_ME_SM, aes(x = avg_BS_delta, y = auc_delta, color = group)) +\n",
    "    geom_point(size = 3, alpha = 0.6) +  # Size of points\n",
    "    geom_abline(intercept = intercept_ME_SM, slope = slope_ME_SM, color = \"black\", linetype = \"dashed\") +  # Add regression line\n",
    "    annotate(\"text\", x = 0.03, y = 0.65, label = regression_eq_ME_SM, hjust = 0, size = 5) +  # Add regression equation text\n",
    "    labs(x = \"unit25 - unit1 Average node Bootstrap Support\", y = \"unit25 - unit1 ROC AUC of Bootstrap Support\") +  # Custom axis labels\n",
    "    theme_light() +  # Clean theme\n",
    "    theme(\n",
    "    axis.title = element_text(face = \"bold\", size = 14),  # Bold axis labels, adjustable font size\n",
    "    axis.text = element_text(size = 12)  # Axis text size\n",
    "    ) +\n",
    "    #scale_color_manual(values = custom_colors) +  # Apply custom colors\n",
    "    guides(color = guide_legend(title = \"Supermatrix-ME\"))   # Legend with title\n",
    "    #scale_x_continuous(limits = c(0, 1)) +\n",
    "    #scale_y_continuous(limits = c(0.5, 1))\n",
    "\n",
    "# Plot ML\n",
    "p_ML_auc_bs = ggplot(df_family_auc_ML_SM, aes(x = avg_BS_delta, y = auc_delta, color = group)) +\n",
    "    geom_point(size = 3, alpha = 0.6) +  # Size of points\n",
    "    geom_abline(intercept = intercept_ML_SM, slope = slope_ML_SM, color = \"black\", linetype = \"dashed\") +  # Add regression line\n",
    "    annotate(\"text\", x = 0.05, y = 0.57, label = regression_eq_ML_SM, hjust = 0, size = 5) +  # Add regression equation text\n",
    "    labs(x = \"unit25 - unit1 Average node Bootstrap Support\", y = \"unit25 - unit1 ROC AUC of Bootstrap Support\") +  # Custom axis labels\n",
    "    theme_light() +  # Clean theme\n",
    "    theme(\n",
    "    axis.title = element_text(face = \"bold\", size = 14),  # Bold axis labels, adjustable font size\n",
    "    axis.text = element_text(size = 12)  # Axis text size\n",
    "    ) +\n",
    "    #scale_color_manual(values = custom_colors) +  # Apply custom colors\n",
    "    guides(color = guide_legend(title = \"Supermatrix-ML\"))   # Legend with title\n",
    "    #scale_x_continuous(limits = c(0, 1)) +\n",
    "    #scale_y_continuous(limits = c(0.5, 1))\n",
    "\n",
    "ggsave(filename=\"figures/BS_delta_vs_auc_delta_scatterplot_ME.png\", plot=p_ME_auc_bs, dpi = \"retina\")\n",
    "ggsave(filename=\"figures/BS_delta_vs_auc_delta_scatterplot_ML.png\", plot=p_ML_auc_bs, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d80c45ec-c2f3-4983-804e-3482d4f90a31",
   "metadata": {},
   "source": [
    "<a id='subchapter-8_2-MCC-computation'></a>\n",
    "\n",
    "### Subchapter 8.2 - MCC computation\n",
    "\n",
    "TODO add description if necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "8f78ca97-1cec-40ba-8351-f92e778c768f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to find the optimal threshold that maximizes the Matthews Correlation Coefficient (MCC)\n",
    "# TODO add to custom function block\n",
    "\n",
    "get_best_mcc <- function(predicted_values, proven_positives, thresholds){\n",
    "    \n",
    "    # Initialize variables to store the best threshold and best MCC found so far\n",
    "    best_threshold = 0\n",
    "    best_mcc = -1  # Start with -1 as initial MCC to ensure any positive MCC will replace it\n",
    "    \n",
    "    # Loop through each threshold in the provided list to evaluate MCC\n",
    "    for(threshold in thresholds){\n",
    "        \n",
    "        # Apply the threshold to classify predictions as positive or negative\n",
    "        predicted_labels <- ifelse(predicted_values > threshold, 1, 0)\n",
    "        predicted_labels = factor(predicted_labels, levels = c(0,1))  # Ensure binary factor levels for confusion matrix\n",
    "        \n",
    "        # Generate confusion matrix to calculate TP, TN, FP, and FN\n",
    "        confusion_matrix = table(predicted_labels, proven_positives)\n",
    "\n",
    "        # converting them to numeric to handle very large numbers later on in the multiplication step\n",
    "        tp = as.numeric(confusion_matrix[2,2])  # True Positives\n",
    "        tn = as.numeric(confusion_matrix[1,1])  # True Negatives\n",
    "        fp = as.numeric(confusion_matrix[1,2]) # False Positives\n",
    "        fn = as.numeric(confusion_matrix[2,1])  # False Negatives\n",
    "        \n",
    "        # Calculate True Positive Rate (TPR) and False Positive Rate (FPR) for ROC analysis (optional)\n",
    "        tpr = tp / sum(proven_positives)              # True Positive Rate (Sensitivity)\n",
    "        fpr = fp / sum(proven_positives == 0)         # False Positive Rate\n",
    "\n",
    "        # Handle potential NaN values in TPR and FPR by setting them to 0\n",
    "        if(is.nan(tpr)){\n",
    "            tpr = 0\n",
    "        }\n",
    "        if(is.nan(fpr)){\n",
    "            fpr = 0\n",
    "        }\n",
    "\n",
    "        # Calculate Matthews Correlation Coefficient (MCC) for current threshold\n",
    "        mcc = (tp*tn - fp*fn) / sqrt((tp + fp)*(tp + fn)*(tn + fp)*(tn + fn))\n",
    "        \n",
    "        # Handle NaN values in MCC by setting MCC to -1\n",
    "        if(is.nan(mcc)){\n",
    "            mcc = -1\n",
    "        }\n",
    "        \n",
    "        # Update best MCC, threshold, and metrics if current MCC is higher than previous best\n",
    "        if(mcc > best_mcc){\n",
    "            best_mcc = mcc\n",
    "            best_threshold = threshold\n",
    "            best_sensitivity = confusion_matrix[2,2] / sum(proven_positives)     # Sensitivity\n",
    "            best_specificity = confusion_matrix[1,1] / sum(proven_positives == 0) # Specificity\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    # Return a list with the best MCC, corresponding threshold, sensitivity, and specificity\n",
    "    return(list(best_mcc = best_mcc, best_threshold = best_threshold, \n",
    "                best_sensitivity = best_sensitivity, best_specificity = best_specificity))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "efbedafe-0159-4022-b88a-e9ebe3c7fbe0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "# TODO insert plot number\n",
    "\n",
    "# Reading Bootstrap values for each node and presence of node in family paralog reference tree\n",
    "df_bs_ME_SM = read.table(gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ME_SM.tsv.gz\"), header = TRUE)\n",
    "df_bs_ML_SM = read.table(gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ML_SM.tsv.gz\"), header = TRUE)\n",
    "\n",
    "# get the ammount of unique families (500), the two dataframes are assumed to have same number\n",
    "unique_family_names = unique(df_bs_ME_SM$family)\n",
    "\n",
    "# prepare the dataframe for the best MCC values\n",
    "df_family_mcc = data.frame(\n",
    "    family         = c(),\n",
    "    group          = c(),\n",
    "    method         = c(),\n",
    "    mcc_val        = c(),\n",
    "    best_threshold = c()\n",
    ")\n",
    "\n",
    "# initialize threshold range and step\n",
    "thresholds = seq(0.2, 1, 0.01)\n",
    "\n",
    "# for each family compute the best MCC value\n",
    "for (fam_num in 1:length(unique_family_names)) {\n",
    "\n",
    "    # get only family node datas for ME + SM\n",
    "    family_data_ME_SM = subset(df_bs_ME_SM, family == unique_family_names[fam_num])\n",
    "    \n",
    "    # calculate best threshold according to MCC and best MCC value\n",
    "    fam_mcc_infos_ME_SM      = get_best_mcc(family_data_ME_SM$BS, family_data_ME_SM$ref_presence, thresholds)\n",
    "\n",
    "    # updatet the dataframe with relevant information, changing some fields for readability\n",
    "    df_family_mcc = data.frame(\n",
    "        family         = c(df_family_mcc$family, paste0(\"family_\", fam_num)),\n",
    "        group          = c(df_family_mcc$group, family_data_ME_SM$group[1]),    # a family can only be part of one group \n",
    "        method         = c(df_family_mcc$method, \"Supermatrix-ME\"),\n",
    "        mcc_val        = c(df_family_mcc$mcc_val, fam_mcc_infos_ME_SM$best_mcc),\n",
    "        best_threshold = c(df_family_mcc$best_threshold, fam_mcc_infos_ME_SM$best_threshold)\n",
    "    )\n",
    "\n",
    "    # repeat all of the above for ML + SM\n",
    "    family_data_ML_SM = subset(df_bs_ML_SM, family == unique_family_names[fam_num])\n",
    "    \n",
    "    fam_mcc_infos_ML_SM      = get_best_mcc(family_data_ML_SM$BS, family_data_ML_SM$ref_presence, thresholds)\n",
    "\n",
    "    df_family_mcc = data.frame(\n",
    "        family         = c(df_family_mcc$family, paste0(\"family_\", fam_num)),\n",
    "        group          = c(df_family_mcc$group, family_data_ML_SM$group[1]),    # a family can only be part of one group \n",
    "        method         = c(df_family_mcc$method, \"Supermatrix-ML\"),\n",
    "        mcc_val        = c(df_family_mcc$mcc_val, fam_mcc_infos_ML_SM$best_mcc),\n",
    "        best_threshold = c(df_family_mcc$best_threshold, fam_mcc_infos_ML_SM$best_threshold)\n",
    "    )    \n",
    "}\n",
    "\n",
    "# Map group names based on provided mapping\n",
    "groupnames_mapping = list(\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\" = \"group1\\nlen = 32\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\" = \"group2\\nlen = 46\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\" = \"group3\\nlen = 50\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\" = \"group4\\nlen = 60\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\" = \"group5\\nlen = 82\"\n",
    ")\n",
    "\n",
    "# Apply the group name mapping\n",
    "df_family_mcc = df_family_mcc %>%\n",
    "  mutate(\n",
    "    group = unlist(groupnames_mapping[group])\n",
    "  )\n",
    "\n",
    "# Define colors for methods\n",
    "method_colors = c(\"Supermatrix-ME\" = \"#2F739B\", \"Supermatrix-ML\" = \"#D05159\")\n",
    "\n",
    "# Plot using ggplot2\n",
    "p_family_mcc = ggplot(df_family_mcc, aes(x = group, y = best_threshold, color = method)) +\n",
    "    geom_boxplot(alpha = 0.6) +  # Adding boxplot layer without outliers\n",
    "    scale_color_manual(values = method_colors) +  # Set colors based on method\n",
    "    labs(x = \"Sequence length\", y = \"Best BS threshold cut-off according to MCC\", color = \"Method\") +  # Set axis and legend labels\n",
    "    theme_light() +\n",
    "    theme(\n",
    "    axis.title = element_text(face = \"bold\"),  # Make axis labels bold\n",
    "    legend.position = \"top\"\n",
    "    ) + \n",
    "    scale_y_continuous(limits = c(0, 1))\n",
    "\n",
    "ggsave(filename=\"figures/BS_boxplot_mcc_best_threshold_ME_ML_SM_combined.png\", plot=p_family_mcc, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "560fd3e4-3356-4360-bfe6-064370d47d03",
   "metadata": {},
   "source": [
    "<a id='subchapter-8_3-Reference-presence-vs-BS'></a>\n",
    "\n",
    "### Subchapter 8.3 - Reference presence vs BS \n",
    "\n",
    "TODO add description if necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a1ed6e5c-f229-4dfd-889c-390dfc9add9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "# todo add figure number\n",
    "\n",
    "# Reading Bootstrap values for each node and presence of node in family paralog reference tree\n",
    "df_bs_ME_SM = read.table(gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ME_SM.tsv.gz\"), header = TRUE)\n",
    "df_bs_ML_SM = read.table(gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ML_SM.tsv.gz\"), header = TRUE)\n",
    "\n",
    "# get the ammount of unique families (500), the two dataframes are assumed to have same number\n",
    "unique_family_names = unique(df_bs_ME_SM$family)\n",
    "\n",
    "# prpare the dataframe for the auc for both ME and ML\n",
    "df_family_bs_ref_ME_SM = data.frame(\n",
    "    family      = c(),\n",
    "    group       = c(),\n",
    "    method      = c(),\n",
    "    percent_ref = c(),\n",
    "    avg_BS      = c()\n",
    ")\n",
    "df_family_bs_ref_ML_SM = df_family_bs_ref_ME_SM\n",
    "\n",
    "# for each family compute the percentage ref presence and abg bootstrap val\n",
    "for (fam_num in 1:length(unique_family_names)) {\n",
    "\n",
    "    # get only family node datas for ME + SM \n",
    "    family_data_ME_SM = subset(df_bs_ME_SM, family == unique_family_names[fam_num])\n",
    "    \n",
    "    # updatet the dataframe with relevant information, changing some fields for readability\n",
    "    # computing percentage node presence in reference and average BS \n",
    "    df_family_bs_ref_ME_SM = data.frame(\n",
    "        family       = c(df_family_bs_ref_ME_SM$family, paste0(\"family_\", fam_num)),\n",
    "        group        = c(df_family_bs_ref_ME_SM$group, family_data_ME_SM$group[1]),    # a family can only be part of one group \n",
    "        method       = c(df_family_bs_ref_ME_SM$method, \"Supermatrix-ME\"),\n",
    "        percent_ref  = c(df_family_bs_ref_ME_SM$percent_ref, sum(family_data_ME_SM$ref_presence==1)/length(family_data_ME_SM$ref_presence)),\n",
    "        avg_BS       = c(df_family_bs_ref_ME_SM$avg_BS, mean(family_data_ME_SM$BS))\n",
    "    )\n",
    "    \n",
    "    # repeat for ML\n",
    "    family_data_ML_SM = subset(df_bs_ML_SM, family == unique_family_names[fam_num])\n",
    "    df_family_bs_ref_ML_SM = data.frame(\n",
    "        family       = c(df_family_bs_ref_ML_SM$family, paste0(\"family_\", fam_num)),\n",
    "        group        = c(df_family_bs_ref_ML_SM$group, family_data_ML_SM$group[1]),    # a family can only be part of one group \n",
    "        method       = c(df_family_bs_ref_ML_SM$method, \"Supermatrix-ML\"),\n",
    "        percent_ref  = c(df_family_bs_ref_ML_SM$percent_ref, sum(family_data_ML_SM$ref_presence==1)/length(family_data_ML_SM$ref_presence)),\n",
    "        avg_BS       = c(df_family_bs_ref_ML_SM$avg_BS, mean(family_data_ML_SM$BS))\n",
    "    )\n",
    "}\n",
    "\n",
    "# Map group names based on provided mapping\n",
    "groupnames_mapping = list(\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\" = \"group1 len = 32\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\" = \"group2 len = 46\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\" = \"group3 len = 50\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\" = \"group4 len = 60\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\" = \"group5 len = 82\"\n",
    ")\n",
    "\n",
    "# Apply the group name mapping\n",
    "df_family_bs_ref_ME_SM = df_family_bs_ref_ME_SM %>%\n",
    "  mutate(\n",
    "    group = unlist(groupnames_mapping[group])\n",
    "  )\n",
    "df_family_bs_ref_ML_SM = df_family_bs_ref_ML_SM %>%\n",
    "  mutate(\n",
    "    group = unlist(groupnames_mapping[group])\n",
    "  )\n",
    "\n",
    "# Fit linear regression model for ME + SM \n",
    "model_ME_SM     = lm(percent_ref ~ avg_BS, data = df_family_bs_ref_ME_SM)\n",
    "slope_ME_SM     = coef(model_ME_SM)[2]  # Extract slope\n",
    "intercept_ME_SM = coef(model_ME_SM)[1]  # Extract intercept\n",
    "r_squared_ME_SM = summary(model_ME_SM)$r.squared  # Extract R-squared value\n",
    "\n",
    "# Fit linear regression model for ML + SM \n",
    "model_ML_SM     = lm(percent_ref ~ avg_BS, data = df_family_bs_ref_ML_SM)\n",
    "slope_ML_SM     = coef(model_ML_SM)[2]  # Extract slope\n",
    "intercept_ML_SM = coef(model_ML_SM)[1]  # Extract intercept\n",
    "r_squared_ML_SM = summary(model_ML_SM)$r.squared  # Extract R-squared value\n",
    "\n",
    "# Create regression equation string for both ME and ML SM\n",
    "regression_eq_ME_SM = paste0(\"R² = \", round(r_squared_ME_SM, 2))\n",
    "regression_eq_ML_SM = paste0(\"R² = \", round(r_squared_ML_SM, 2))\n",
    "\n",
    "# Custom colors for each group, change these as needed\n",
    "custom_colors = c(\n",
    "  \"group2 len = 46\" = \"#33a02c\",   # Change to your preferred color\n",
    "  \"group1 len = 32\" = \"#1f78b4\",   # Change to your preferred color\n",
    "  \"group3 len = 50\" = \"#e31a1c\",   # Change to your preferred color\n",
    "  \"group4 len = 60\" = \"#ff7f00\",   # Change to your preferred color\n",
    "  \"group5 len = 82\" = \"#6a3d9a\"    # Change to your preferred color\n",
    ")\n",
    "\n",
    "# Plot ME\n",
    "p_ME_auc_bs = ggplot(df_family_bs_ref_ME_SM, aes(x = avg_BS, y = percent_ref, color = group)) +\n",
    "    geom_point(size = 3, alpha = 0.6) +  # Size of points\n",
    "    geom_abline(intercept = intercept_ME_SM, slope = slope_ME_SM, color = \"black\", linetype = \"dashed\") +  # Add regression line\n",
    "    annotate(\"text\", x = 0.03, y = 0.65, label = regression_eq_ME_SM, hjust = 0, size = 5) +  # Add regression equation text\n",
    "    labs(x = \"average Bootstrap Support\", y = \"% nodes in reference tree\") +  # Custom axis labels\n",
    "    theme_light() +  # Clean theme\n",
    "    theme(\n",
    "    axis.title = element_text(face = \"bold\", size = 14),  # Bold axis labels, adjustable font size\n",
    "    axis.text = element_text(size = 12)  # Axis text size\n",
    "    ) +\n",
    "    #scale_color_manual(values = custom_colors) +  # Apply custom colors\n",
    "    guides(color = guide_legend(title = \"Supermatrix-ME\")) +  # Legend with title\n",
    "    scale_x_continuous(limits = c(0, 1)) +\n",
    "    scale_y_continuous(limits = c(0, 1))\n",
    "\n",
    "# Plot ML\n",
    "p_ML_auc_bs = ggplot(df_family_bs_ref_ML_SM, aes(x = avg_BS, y = percent_ref, color = group)) +\n",
    "    geom_point(size = 3, alpha = 0.6) +  # Size of points\n",
    "    geom_abline(intercept = intercept_ML_SM, slope = slope_ML_SM, color = \"black\", linetype = \"dashed\") +  # Add regression line\n",
    "    annotate(\"text\", x = 0.03, y = 0.65, label = regression_eq_ML_SM, hjust = 0, size = 5) +  # Add regression equation text\n",
    "    labs(x = \"average Bootstrap Support\", y = \"% nodes in reference tree\") +  # Custom axis labels\n",
    "    theme_light() +  # Clean theme\n",
    "    theme(\n",
    "    axis.title = element_text(face = \"bold\", size = 14),  # Bold axis labels, adjustable font size\n",
    "    axis.text = element_text(size = 12)  # Axis text size\n",
    "    ) +\n",
    "    #scale_color_manual(values = custom_colors) +  # Apply custom colors\n",
    "    guides(color = guide_legend(title = \"Supermatrix-ML\")) +  # Legend with title\n",
    "    scale_x_continuous(limits = c(0, 1)) +\n",
    "    scale_y_continuous(limits = c(0, 1))\n",
    "\n",
    "ggsave(filename=\"figures/BS_vs_ref_presence_scatterplot_ME.png\", plot=p_ME_auc_bs, dpi = \"retina\")\n",
    "ggsave(filename=\"figures/BS_vs_ref_presence_scatterplot_ML.png\", plot=p_ML_auc_bs, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "befc9c5d-7b53-44be-936f-4c81f504237c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 10 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Row_Name</th><th scope=col>group1_BS</th><th scope=col>group1_ref</th><th scope=col>group2_BS</th><th scope=col>group2_ref</th><th scope=col>group3_BS</th><th scope=col>group3_ref</th><th scope=col>group4_BS</th><th scope=col>group4_ref</th><th scope=col>group5_BS</th><th scope=col>group5_ref</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td><span style=white-space:pre-wrap>0&lt;BS&lt;=0.1  </span></td><td>1034</td><td>10.93 %</td><td> 802</td><td>12.34 %</td><td> 664</td><td>10.24 %</td><td> 624</td><td>13.62 %</td><td> 386</td><td>14.51 %</td></tr>\n",
       "\t<tr><td>0.1&lt;BS&lt;=0.2</td><td>1527</td><td>16.83 %</td><td>1365</td><td>18.61 %</td><td>1267</td><td>19.02 %</td><td>1012</td><td>20.45 %</td><td> 714</td><td>21.71 %</td></tr>\n",
       "\t<tr><td>0.2&lt;BS&lt;=0.3</td><td>1711</td><td>24.84 %</td><td>1592</td><td>27.95 %</td><td>1549</td><td>28.92 %</td><td>1336</td><td>31.81 %</td><td>1038</td><td>33.62 %</td></tr>\n",
       "\t<tr><td>0.3&lt;BS&lt;=0.4</td><td>1672</td><td>34.33 %</td><td>1627</td><td>38.11 %</td><td>1589</td><td>36.69 %</td><td>1478</td><td>39.78 %</td><td>1296</td><td>43.75 %</td></tr>\n",
       "\t<tr><td>0.4&lt;BS&lt;=0.5</td><td>2004</td><td>45.66 %</td><td>1911</td><td>48.87 %</td><td>1998</td><td>52.45 %</td><td>2012</td><td>52.44 %</td><td>1886</td><td>55.62 %</td></tr>\n",
       "\t<tr><td>0.5&lt;BS&lt;=0.6</td><td>1179</td><td>57.17 %</td><td>1290</td><td>58.14 %</td><td>1359</td><td>62.77 %</td><td>1321</td><td>65.33 %</td><td>1313</td><td>69.46 %</td></tr>\n",
       "\t<tr><td>0.6&lt;BS&lt;=0.7</td><td> 956</td><td>65.79 %</td><td>1006</td><td>71.77 %</td><td>1052</td><td>70.06 %</td><td>1203</td><td>75.23 %</td><td>1249</td><td>77.66 %</td></tr>\n",
       "\t<tr><td>0.7&lt;BS&lt;=0.8</td><td> 727</td><td>74.14 %</td><td> 842</td><td>76.72 %</td><td> 873</td><td>79.73 %</td><td>1044</td><td>83.72 %</td><td>1241</td><td>84.93 %</td></tr>\n",
       "\t<tr><td>0.8&lt;BS&lt;=0.9</td><td> 573</td><td>81.68 %</td><td> 711</td><td>84.53 %</td><td> 816</td><td>89.71 %</td><td> 904</td><td>90.71 %</td><td>1180</td><td>91.86 %</td></tr>\n",
       "\t<tr><td><span style=white-space:pre-wrap>0.9&lt;BS&lt;=1  </span></td><td> 475</td><td>92.42 %</td><td> 748</td><td>93.72 %</td><td> 747</td><td>94.78 %</td><td>1007</td><td>96.52 %</td><td>1668</td><td>97.18 %</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 11\n",
       "\\begin{tabular}{lllllllllll}\n",
       " Row\\_Name & group1\\_BS & group1\\_ref & group2\\_BS & group2\\_ref & group3\\_BS & group3\\_ref & group4\\_BS & group4\\_ref & group5\\_BS & group5\\_ref\\\\\n",
       " <chr> & <int> & <chr> & <int> & <chr> & <int> & <chr> & <int> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t 0<BS<=0.1   & 1034 & 10.93 \\% &  802 & 12.34 \\% &  664 & 10.24 \\% &  624 & 13.62 \\% &  386 & 14.51 \\%\\\\\n",
       "\t 0.1<BS<=0.2 & 1527 & 16.83 \\% & 1365 & 18.61 \\% & 1267 & 19.02 \\% & 1012 & 20.45 \\% &  714 & 21.71 \\%\\\\\n",
       "\t 0.2<BS<=0.3 & 1711 & 24.84 \\% & 1592 & 27.95 \\% & 1549 & 28.92 \\% & 1336 & 31.81 \\% & 1038 & 33.62 \\%\\\\\n",
       "\t 0.3<BS<=0.4 & 1672 & 34.33 \\% & 1627 & 38.11 \\% & 1589 & 36.69 \\% & 1478 & 39.78 \\% & 1296 & 43.75 \\%\\\\\n",
       "\t 0.4<BS<=0.5 & 2004 & 45.66 \\% & 1911 & 48.87 \\% & 1998 & 52.45 \\% & 2012 & 52.44 \\% & 1886 & 55.62 \\%\\\\\n",
       "\t 0.5<BS<=0.6 & 1179 & 57.17 \\% & 1290 & 58.14 \\% & 1359 & 62.77 \\% & 1321 & 65.33 \\% & 1313 & 69.46 \\%\\\\\n",
       "\t 0.6<BS<=0.7 &  956 & 65.79 \\% & 1006 & 71.77 \\% & 1052 & 70.06 \\% & 1203 & 75.23 \\% & 1249 & 77.66 \\%\\\\\n",
       "\t 0.7<BS<=0.8 &  727 & 74.14 \\% &  842 & 76.72 \\% &  873 & 79.73 \\% & 1044 & 83.72 \\% & 1241 & 84.93 \\%\\\\\n",
       "\t 0.8<BS<=0.9 &  573 & 81.68 \\% &  711 & 84.53 \\% &  816 & 89.71 \\% &  904 & 90.71 \\% & 1180 & 91.86 \\%\\\\\n",
       "\t 0.9<BS<=1   &  475 & 92.42 \\% &  748 & 93.72 \\% &  747 & 94.78 \\% & 1007 & 96.52 \\% & 1668 & 97.18 \\%\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 11\n",
       "\n",
       "| Row_Name &lt;chr&gt; | group1_BS &lt;int&gt; | group1_ref &lt;chr&gt; | group2_BS &lt;int&gt; | group2_ref &lt;chr&gt; | group3_BS &lt;int&gt; | group3_ref &lt;chr&gt; | group4_BS &lt;int&gt; | group4_ref &lt;chr&gt; | group5_BS &lt;int&gt; | group5_ref &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 0&lt;BS&lt;=0.1   | 1034 | 10.93 % |  802 | 12.34 % |  664 | 10.24 % |  624 | 13.62 % |  386 | 14.51 % |\n",
       "| 0.1&lt;BS&lt;=0.2 | 1527 | 16.83 % | 1365 | 18.61 % | 1267 | 19.02 % | 1012 | 20.45 % |  714 | 21.71 % |\n",
       "| 0.2&lt;BS&lt;=0.3 | 1711 | 24.84 % | 1592 | 27.95 % | 1549 | 28.92 % | 1336 | 31.81 % | 1038 | 33.62 % |\n",
       "| 0.3&lt;BS&lt;=0.4 | 1672 | 34.33 % | 1627 | 38.11 % | 1589 | 36.69 % | 1478 | 39.78 % | 1296 | 43.75 % |\n",
       "| 0.4&lt;BS&lt;=0.5 | 2004 | 45.66 % | 1911 | 48.87 % | 1998 | 52.45 % | 2012 | 52.44 % | 1886 | 55.62 % |\n",
       "| 0.5&lt;BS&lt;=0.6 | 1179 | 57.17 % | 1290 | 58.14 % | 1359 | 62.77 % | 1321 | 65.33 % | 1313 | 69.46 % |\n",
       "| 0.6&lt;BS&lt;=0.7 |  956 | 65.79 % | 1006 | 71.77 % | 1052 | 70.06 % | 1203 | 75.23 % | 1249 | 77.66 % |\n",
       "| 0.7&lt;BS&lt;=0.8 |  727 | 74.14 % |  842 | 76.72 % |  873 | 79.73 % | 1044 | 83.72 % | 1241 | 84.93 % |\n",
       "| 0.8&lt;BS&lt;=0.9 |  573 | 81.68 % |  711 | 84.53 % |  816 | 89.71 % |  904 | 90.71 % | 1180 | 91.86 % |\n",
       "| 0.9&lt;BS&lt;=1   |  475 | 92.42 % |  748 | 93.72 % |  747 | 94.78 % | 1007 | 96.52 % | 1668 | 97.18 % |\n",
       "\n"
      ],
      "text/plain": [
       "   Row_Name    group1_BS group1_ref group2_BS group2_ref group3_BS group3_ref\n",
       "1  0<BS<=0.1   1034      10.93 %     802      12.34 %     664      10.24 %   \n",
       "2  0.1<BS<=0.2 1527      16.83 %    1365      18.61 %    1267      19.02 %   \n",
       "3  0.2<BS<=0.3 1711      24.84 %    1592      27.95 %    1549      28.92 %   \n",
       "4  0.3<BS<=0.4 1672      34.33 %    1627      38.11 %    1589      36.69 %   \n",
       "5  0.4<BS<=0.5 2004      45.66 %    1911      48.87 %    1998      52.45 %   \n",
       "6  0.5<BS<=0.6 1179      57.17 %    1290      58.14 %    1359      62.77 %   \n",
       "7  0.6<BS<=0.7  956      65.79 %    1006      71.77 %    1052      70.06 %   \n",
       "8  0.7<BS<=0.8  727      74.14 %     842      76.72 %     873      79.73 %   \n",
       "9  0.8<BS<=0.9  573      81.68 %     711      84.53 %     816      89.71 %   \n",
       "10 0.9<BS<=1    475      92.42 %     748      93.72 %     747      94.78 %   \n",
       "   group4_BS group4_ref group5_BS group5_ref\n",
       "1   624      13.62 %     386      14.51 %   \n",
       "2  1012      20.45 %     714      21.71 %   \n",
       "3  1336      31.81 %    1038      33.62 %   \n",
       "4  1478      39.78 %    1296      43.75 %   \n",
       "5  2012      52.44 %    1886      55.62 %   \n",
       "6  1321      65.33 %    1313      69.46 %   \n",
       "7  1203      75.23 %    1249      77.66 %   \n",
       "8  1044      83.72 %    1241      84.93 %   \n",
       "9   904      90.71 %    1180      91.86 %   \n",
       "10 1007      96.52 %    1668      97.18 %   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 10 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Row_Name</th><th scope=col>group1_BS</th><th scope=col>group1_ref</th><th scope=col>group2_BS</th><th scope=col>group2_ref</th><th scope=col>group3_BS</th><th scope=col>group3_ref</th><th scope=col>group4_BS</th><th scope=col>group4_ref</th><th scope=col>group5_BS</th><th scope=col>group5_ref</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td><span style=white-space:pre-wrap>0&lt;BS&lt;=0.1  </span></td><td> 567</td><td>18.52 %</td><td> 339</td><td>15.63 %</td><td> 279</td><td>25.81 %</td><td> 146</td><td>16.44 %</td><td><span style=white-space:pre-wrap>  59</span></td><td>28.81 %</td></tr>\n",
       "\t<tr><td>0.1&lt;BS&lt;=0.2</td><td>1160</td><td>28.36 %</td><td> 895</td><td>31.28 %</td><td> 787</td><td>37.99 %</td><td> 498</td><td>26.71 %</td><td> 315</td><td>46.67 %</td></tr>\n",
       "\t<tr><td>0.2&lt;BS&lt;=0.3</td><td>1607</td><td><span style=white-space:pre-wrap>44 %   </span></td><td>1294</td><td>43.28 %</td><td>1223</td><td>50.86 %</td><td> 859</td><td>41.44 %</td><td> 643</td><td>48.99 %</td></tr>\n",
       "\t<tr><td>0.3&lt;BS&lt;=0.4</td><td>1771</td><td>57.88 %</td><td>1578</td><td>58.68 %</td><td>1558</td><td>58.92 %</td><td>1205</td><td>55.1 % </td><td> 883</td><td>60.48 %</td></tr>\n",
       "\t<tr><td>0.4&lt;BS&lt;=0.5</td><td>2236</td><td>70.26 %</td><td>2093</td><td>73.34 %</td><td>2084</td><td>73.03 %</td><td>2042</td><td>71.35 %</td><td>1527</td><td>73.74 %</td></tr>\n",
       "\t<tr><td>0.5&lt;BS&lt;=0.6</td><td>1483</td><td>81.86 %</td><td>1497</td><td>85.24 %</td><td>1477</td><td>85.04 %</td><td>1530</td><td>85.49 %</td><td>1428</td><td>82.21 %</td></tr>\n",
       "\t<tr><td>0.6&lt;BS&lt;=0.7</td><td>1141</td><td>88.52 %</td><td>1313</td><td>90.56 %</td><td>1324</td><td>92.67 %</td><td>1661</td><td>93.5 % </td><td>1527</td><td>92.4 % </td></tr>\n",
       "\t<tr><td>0.7&lt;BS&lt;=0.8</td><td> 886</td><td>94.02 %</td><td>1166</td><td>93.65 %</td><td>1183</td><td>96.7 % </td><td>1573</td><td>96.5 % </td><td>1619</td><td>96.85 %</td></tr>\n",
       "\t<tr><td>0.8&lt;BS&lt;=0.9</td><td> 652</td><td>97.85 %</td><td> 943</td><td>97.67 %</td><td>1062</td><td>98.96 %</td><td>1246</td><td>99.52 %</td><td>1719</td><td>99.3 % </td></tr>\n",
       "\t<tr><td><span style=white-space:pre-wrap>0.9&lt;BS&lt;=1  </span></td><td> 466</td><td><span style=white-space:pre-wrap>100 %  </span></td><td> 862</td><td>99.07 %</td><td>1004</td><td>99.9 % </td><td>1236</td><td><span style=white-space:pre-wrap>100 %  </span></td><td>2272</td><td>99.96 %</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 11\n",
       "\\begin{tabular}{lllllllllll}\n",
       " Row\\_Name & group1\\_BS & group1\\_ref & group2\\_BS & group2\\_ref & group3\\_BS & group3\\_ref & group4\\_BS & group4\\_ref & group5\\_BS & group5\\_ref\\\\\n",
       " <chr> & <int> & <chr> & <int> & <chr> & <int> & <chr> & <int> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t 0<BS<=0.1   &  567 & 18.52 \\% &  339 & 15.63 \\% &  279 & 25.81 \\% &  146 & 16.44 \\% &   59 & 28.81 \\%\\\\\n",
       "\t 0.1<BS<=0.2 & 1160 & 28.36 \\% &  895 & 31.28 \\% &  787 & 37.99 \\% &  498 & 26.71 \\% &  315 & 46.67 \\%\\\\\n",
       "\t 0.2<BS<=0.3 & 1607 & 44 \\%    & 1294 & 43.28 \\% & 1223 & 50.86 \\% &  859 & 41.44 \\% &  643 & 48.99 \\%\\\\\n",
       "\t 0.3<BS<=0.4 & 1771 & 57.88 \\% & 1578 & 58.68 \\% & 1558 & 58.92 \\% & 1205 & 55.1 \\%  &  883 & 60.48 \\%\\\\\n",
       "\t 0.4<BS<=0.5 & 2236 & 70.26 \\% & 2093 & 73.34 \\% & 2084 & 73.03 \\% & 2042 & 71.35 \\% & 1527 & 73.74 \\%\\\\\n",
       "\t 0.5<BS<=0.6 & 1483 & 81.86 \\% & 1497 & 85.24 \\% & 1477 & 85.04 \\% & 1530 & 85.49 \\% & 1428 & 82.21 \\%\\\\\n",
       "\t 0.6<BS<=0.7 & 1141 & 88.52 \\% & 1313 & 90.56 \\% & 1324 & 92.67 \\% & 1661 & 93.5 \\%  & 1527 & 92.4 \\% \\\\\n",
       "\t 0.7<BS<=0.8 &  886 & 94.02 \\% & 1166 & 93.65 \\% & 1183 & 96.7 \\%  & 1573 & 96.5 \\%  & 1619 & 96.85 \\%\\\\\n",
       "\t 0.8<BS<=0.9 &  652 & 97.85 \\% &  943 & 97.67 \\% & 1062 & 98.96 \\% & 1246 & 99.52 \\% & 1719 & 99.3 \\% \\\\\n",
       "\t 0.9<BS<=1   &  466 & 100 \\%   &  862 & 99.07 \\% & 1004 & 99.9 \\%  & 1236 & 100 \\%   & 2272 & 99.96 \\%\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 11\n",
       "\n",
       "| Row_Name &lt;chr&gt; | group1_BS &lt;int&gt; | group1_ref &lt;chr&gt; | group2_BS &lt;int&gt; | group2_ref &lt;chr&gt; | group3_BS &lt;int&gt; | group3_ref &lt;chr&gt; | group4_BS &lt;int&gt; | group4_ref &lt;chr&gt; | group5_BS &lt;int&gt; | group5_ref &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 0&lt;BS&lt;=0.1   |  567 | 18.52 % |  339 | 15.63 % |  279 | 25.81 % |  146 | 16.44 % |   59 | 28.81 % |\n",
       "| 0.1&lt;BS&lt;=0.2 | 1160 | 28.36 % |  895 | 31.28 % |  787 | 37.99 % |  498 | 26.71 % |  315 | 46.67 % |\n",
       "| 0.2&lt;BS&lt;=0.3 | 1607 | 44 %    | 1294 | 43.28 % | 1223 | 50.86 % |  859 | 41.44 % |  643 | 48.99 % |\n",
       "| 0.3&lt;BS&lt;=0.4 | 1771 | 57.88 % | 1578 | 58.68 % | 1558 | 58.92 % | 1205 | 55.1 %  |  883 | 60.48 % |\n",
       "| 0.4&lt;BS&lt;=0.5 | 2236 | 70.26 % | 2093 | 73.34 % | 2084 | 73.03 % | 2042 | 71.35 % | 1527 | 73.74 % |\n",
       "| 0.5&lt;BS&lt;=0.6 | 1483 | 81.86 % | 1497 | 85.24 % | 1477 | 85.04 % | 1530 | 85.49 % | 1428 | 82.21 % |\n",
       "| 0.6&lt;BS&lt;=0.7 | 1141 | 88.52 % | 1313 | 90.56 % | 1324 | 92.67 % | 1661 | 93.5 %  | 1527 | 92.4 %  |\n",
       "| 0.7&lt;BS&lt;=0.8 |  886 | 94.02 % | 1166 | 93.65 % | 1183 | 96.7 %  | 1573 | 96.5 %  | 1619 | 96.85 % |\n",
       "| 0.8&lt;BS&lt;=0.9 |  652 | 97.85 % |  943 | 97.67 % | 1062 | 98.96 % | 1246 | 99.52 % | 1719 | 99.3 %  |\n",
       "| 0.9&lt;BS&lt;=1   |  466 | 100 %   |  862 | 99.07 % | 1004 | 99.9 %  | 1236 | 100 %   | 2272 | 99.96 % |\n",
       "\n"
      ],
      "text/plain": [
       "   Row_Name    group1_BS group1_ref group2_BS group2_ref group3_BS group3_ref\n",
       "1  0<BS<=0.1    567      18.52 %     339      15.63 %     279      25.81 %   \n",
       "2  0.1<BS<=0.2 1160      28.36 %     895      31.28 %     787      37.99 %   \n",
       "3  0.2<BS<=0.3 1607      44 %       1294      43.28 %    1223      50.86 %   \n",
       "4  0.3<BS<=0.4 1771      57.88 %    1578      58.68 %    1558      58.92 %   \n",
       "5  0.4<BS<=0.5 2236      70.26 %    2093      73.34 %    2084      73.03 %   \n",
       "6  0.5<BS<=0.6 1483      81.86 %    1497      85.24 %    1477      85.04 %   \n",
       "7  0.6<BS<=0.7 1141      88.52 %    1313      90.56 %    1324      92.67 %   \n",
       "8  0.7<BS<=0.8  886      94.02 %    1166      93.65 %    1183      96.7 %    \n",
       "9  0.8<BS<=0.9  652      97.85 %     943      97.67 %    1062      98.96 %   \n",
       "10 0.9<BS<=1    466      100 %       862      99.07 %    1004      99.9 %    \n",
       "   group4_BS group4_ref group5_BS group5_ref\n",
       "1   146      16.44 %      59      28.81 %   \n",
       "2   498      26.71 %     315      46.67 %   \n",
       "3   859      41.44 %     643      48.99 %   \n",
       "4  1205      55.1 %      883      60.48 %   \n",
       "5  2042      71.35 %    1527      73.74 %   \n",
       "6  1530      85.49 %    1428      82.21 %   \n",
       "7  1661      93.5 %     1527      92.4 %    \n",
       "8  1573      96.5 %     1619      96.85 %   \n",
       "9  1246      99.52 %    1719      99.3 %    \n",
       "10 1236      100 %      2272      99.96 %   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# TODO put table number\n",
    "# table binning nodes based on their BS value and computing percentages for ref presence.\n",
    "# ME + SM \n",
    "\n",
    "# Reading Bootstrap values for each node and presence of node in family paralog reference tree\n",
    "df_bs_ME_SM = read.table(gzfile(\"../analysis/tables/df_bootsrtap_clade_presence_in_ref_ME_SM.tsv.gz\"), header = TRUE)\n",
    "\n",
    "# get how many units (25) \n",
    "num_units = length(unique(df_bs_ME_SM$unit))\n",
    "\n",
    "# get how many groups there are (5) each with (100) families. groups define sequence length\n",
    "num_groups = length(unique(df_bs_ME_SM$group))\n",
    "\n",
    "# Map group names based on provided mapping\n",
    "groupnames_mapping = list(\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\" = \"group1\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\" = \"group2\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\" = \"group3\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\" = \"group4\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\" = \"group5\"\n",
    ")\n",
    "\n",
    "# Apply the group name mapping\n",
    "df_bs_ME_SM = df_bs_ME_SM %>%\n",
    "  mutate(\n",
    "    group = unlist(groupnames_mapping[group])\n",
    "  )\n",
    "\n",
    "# Create empty data frames with 10 rows (one for each BS bin) to match the output vectors.\n",
    "df_unit1  = data.frame(matrix(ncol = 0, nrow = 10))\n",
    "df_unit25 = data.frame(matrix(ncol = 0, nrow = 10))\n",
    "\n",
    "# Loop through five iterations (one for each group).\n",
    "for (n in 1:num_groups) {\n",
    "  \n",
    "  # Filter data for the specific group and unit for the current iteration.\n",
    "  unit1_data  = subset(df_bs_ME_SM, group == paste0(\"group\", n) & unit == 1)\n",
    "  unit25_data = subset(df_bs_ME_SM, group == paste0(\"group\", n) & unit == num_units)\n",
    "  \n",
    "  # Initialize empty vectors to hold the binned values for the current iteration.\n",
    "  unit1_bin_percent_nodes   = c()\n",
    "  unit1_bin_percent_in_ref  = c()\n",
    "  unit25_bin_percent_nodes  = c()\n",
    "  unit25_bin_percent_in_ref = c()\n",
    "  \n",
    "  # Create bins from 0.1 to 1.0 in steps of 0.1 and calculate percentages for each bin.\n",
    "  for (i in seq(0.1, 1.0, by = 0.1)) {\n",
    "    \n",
    "    # Filter rows within the BS value range for both unit1 and unit25.\n",
    "    filtered_rows_unit1  = subset(unit1_data, BS > (i - 0.1) & BS <= i)\n",
    "    filtered_rows_unit25 = subset(unit25_data, BS > (i - 0.1) & BS <= i)\n",
    "    \n",
    "    # Calculate the number of nodes in the bin.\n",
    "    unit1_bin_percent_nodes  = c(unit1_bin_percent_nodes, length(filtered_rows_unit1$node))\n",
    "    unit25_bin_percent_nodes = c(unit25_bin_percent_nodes, length(filtered_rows_unit25$node))\n",
    "    #unit25_bin_percent_nodes = c(unit25_bin_percent_nodes, length(filtered_rows_unit25$node) / length(unit25_data$node))\n",
    "    \n",
    "    # Calculate the percentage of nodes in reference out of those in the bin.\n",
    "    unit1_bin_percent_in_ref  = c(unit1_bin_percent_in_ref, paste0(round(sum(filtered_rows_unit1$ref_presence == 1) / length(filtered_rows_unit1$node) * 100, 2), \" %\"))\n",
    "    unit25_bin_percent_in_ref = c(unit25_bin_percent_in_ref, paste0(round(sum(filtered_rows_unit25$ref_presence == 1) / length(filtered_rows_unit25$node)* 100, 2), \" %\"))\n",
    "  }\n",
    "  \n",
    "  # Add the vectors as new columns in the data frames, naming the columns based on the current group.\n",
    "  df_unit1[[paste0(\"group\", n, \"_BS\")]]   = unit1_bin_percent_nodes\n",
    "  df_unit1[[paste0(\"group\", n, \"_ref\")]]  = unit1_bin_percent_in_ref\n",
    "  df_unit25[[paste0(\"group\", n, \"_BS\")]]  = unit25_bin_percent_nodes\n",
    "  df_unit25[[paste0(\"group\", n, \"_ref\")]] = unit25_bin_percent_in_ref\n",
    "}\n",
    "\n",
    "# Define custom row labels representing BIN values\n",
    "row_names = c(\"0<BS<=0.1\", \"0.1<BS<=0.2\", \"0.2<BS<=0.3\", \"0.3<BS<=0.4\", \n",
    "               \"0.4<BS<=0.5\", \"0.5<BS<=0.6\", \"0.6<BS<=0.7\", \"0.7<BS<=0.8\", \n",
    "               \"0.8<BS<=0.9\", \"0.9<BS<=1\")\n",
    "\n",
    "# Add row names as the first column of the data frame\n",
    "df_unit1  = data.frame(\"Row_Name\" = row_names, df_unit1)\n",
    "df_unit25 = data.frame(\"Row_Name\" = row_names, df_unit25)\n",
    "                              \n",
    "# Output the data frames.\n",
    "df_unit1\n",
    "df_unit25\n",
    "\n",
    "# save tables in files\n",
    "write.table(df_unit1, file=\"tables/unit_1_per_group_nodes_BS_ref_ME_SM.tsv\", quote=FALSE, sep=\"\\t\", row.names=FALSE)\n",
    "write.table(df_unit25, file=\"tables/unit_25_per_group_nodes_BS_ref_ME_SM.tsv\", quote=FALSE, sep=\"\\t\", row.names=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1ff5c32e-2a7c-47e5-9b59-77ad400d8610",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 10 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Row_Name</th><th scope=col>group1_BS</th><th scope=col>group1_ref</th><th scope=col>group2_BS</th><th scope=col>group2_ref</th><th scope=col>group3_BS</th><th scope=col>group3_ref</th><th scope=col>group4_BS</th><th scope=col>group4_ref</th><th scope=col>group5_BS</th><th scope=col>group5_ref</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td><span style=white-space:pre-wrap>0&lt;BS&lt;=0.1  </span></td><td>1031</td><td><span style=white-space:pre-wrap>13 %   </span></td><td> 813</td><td>14.15 %</td><td> 703</td><td>13.09 %</td><td> 519</td><td>13.49 %</td><td> 252</td><td>18.65 %</td></tr>\n",
       "\t<tr><td>0.1&lt;BS&lt;=0.2</td><td>1470</td><td>21.29 %</td><td>1283</td><td>23.69 %</td><td>1109</td><td>22.81 %</td><td> 907</td><td>27.78 %</td><td> 574</td><td>26.48 %</td></tr>\n",
       "\t<tr><td>0.2&lt;BS&lt;=0.3</td><td>1616</td><td>30.57 %</td><td>1544</td><td>32.32 %</td><td>1404</td><td>33.69 %</td><td>1280</td><td>33.12 %</td><td> 920</td><td>35.87 %</td></tr>\n",
       "\t<tr><td>0.3&lt;BS&lt;=0.4</td><td>1636</td><td>39.73 %</td><td>1631</td><td>39.36 %</td><td>1625</td><td>42.71 %</td><td>1490</td><td>43.15 %</td><td>1249</td><td>47.48 %</td></tr>\n",
       "\t<tr><td>0.4&lt;BS&lt;=0.5</td><td>2022</td><td>49.7 % </td><td>2016</td><td>50.89 %</td><td>1975</td><td>51.44 %</td><td>2035</td><td>52.63 %</td><td>1853</td><td>58.12 %</td></tr>\n",
       "\t<tr><td>0.5&lt;BS&lt;=0.6</td><td>1248</td><td>58.57 %</td><td>1237</td><td>61.76 %</td><td>1317</td><td>61.88 %</td><td>1449</td><td>66.46 %</td><td>1376</td><td>67.51 %</td></tr>\n",
       "\t<tr><td>0.6&lt;BS&lt;=0.7</td><td>1002</td><td>65.87 %</td><td>1073</td><td>71.11 %</td><td>1117</td><td>72.78 %</td><td>1263</td><td>76.17 %</td><td>1376</td><td>78.85 %</td></tr>\n",
       "\t<tr><td>0.7&lt;BS&lt;=0.8</td><td> 783</td><td>77.65 %</td><td> 875</td><td>78.63 %</td><td> 993</td><td>80.06 %</td><td>1064</td><td>84.4 % </td><td>1266</td><td>84.99 %</td></tr>\n",
       "\t<tr><td>0.8&lt;BS&lt;=0.9</td><td> 594</td><td>84.34 %</td><td> 716</td><td>87.01 %</td><td> 800</td><td>89.25 %</td><td> 961</td><td>92.4 % </td><td>1288</td><td>92.08 %</td></tr>\n",
       "\t<tr><td><span style=white-space:pre-wrap>0.9&lt;BS&lt;=1  </span></td><td> 428</td><td>90.42 %</td><td> 694</td><td>93.95 %</td><td> 859</td><td>93.83 %</td><td> 965</td><td>96.79 %</td><td>1816</td><td>97.74 %</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 11\n",
       "\\begin{tabular}{lllllllllll}\n",
       " Row\\_Name & group1\\_BS & group1\\_ref & group2\\_BS & group2\\_ref & group3\\_BS & group3\\_ref & group4\\_BS & group4\\_ref & group5\\_BS & group5\\_ref\\\\\n",
       " <chr> & <int> & <chr> & <int> & <chr> & <int> & <chr> & <int> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t 0<BS<=0.1   & 1031 & 13 \\%    &  813 & 14.15 \\% &  703 & 13.09 \\% &  519 & 13.49 \\% &  252 & 18.65 \\%\\\\\n",
       "\t 0.1<BS<=0.2 & 1470 & 21.29 \\% & 1283 & 23.69 \\% & 1109 & 22.81 \\% &  907 & 27.78 \\% &  574 & 26.48 \\%\\\\\n",
       "\t 0.2<BS<=0.3 & 1616 & 30.57 \\% & 1544 & 32.32 \\% & 1404 & 33.69 \\% & 1280 & 33.12 \\% &  920 & 35.87 \\%\\\\\n",
       "\t 0.3<BS<=0.4 & 1636 & 39.73 \\% & 1631 & 39.36 \\% & 1625 & 42.71 \\% & 1490 & 43.15 \\% & 1249 & 47.48 \\%\\\\\n",
       "\t 0.4<BS<=0.5 & 2022 & 49.7 \\%  & 2016 & 50.89 \\% & 1975 & 51.44 \\% & 2035 & 52.63 \\% & 1853 & 58.12 \\%\\\\\n",
       "\t 0.5<BS<=0.6 & 1248 & 58.57 \\% & 1237 & 61.76 \\% & 1317 & 61.88 \\% & 1449 & 66.46 \\% & 1376 & 67.51 \\%\\\\\n",
       "\t 0.6<BS<=0.7 & 1002 & 65.87 \\% & 1073 & 71.11 \\% & 1117 & 72.78 \\% & 1263 & 76.17 \\% & 1376 & 78.85 \\%\\\\\n",
       "\t 0.7<BS<=0.8 &  783 & 77.65 \\% &  875 & 78.63 \\% &  993 & 80.06 \\% & 1064 & 84.4 \\%  & 1266 & 84.99 \\%\\\\\n",
       "\t 0.8<BS<=0.9 &  594 & 84.34 \\% &  716 & 87.01 \\% &  800 & 89.25 \\% &  961 & 92.4 \\%  & 1288 & 92.08 \\%\\\\\n",
       "\t 0.9<BS<=1   &  428 & 90.42 \\% &  694 & 93.95 \\% &  859 & 93.83 \\% &  965 & 96.79 \\% & 1816 & 97.74 \\%\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 11\n",
       "\n",
       "| Row_Name &lt;chr&gt; | group1_BS &lt;int&gt; | group1_ref &lt;chr&gt; | group2_BS &lt;int&gt; | group2_ref &lt;chr&gt; | group3_BS &lt;int&gt; | group3_ref &lt;chr&gt; | group4_BS &lt;int&gt; | group4_ref &lt;chr&gt; | group5_BS &lt;int&gt; | group5_ref &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 0&lt;BS&lt;=0.1   | 1031 | 13 %    |  813 | 14.15 % |  703 | 13.09 % |  519 | 13.49 % |  252 | 18.65 % |\n",
       "| 0.1&lt;BS&lt;=0.2 | 1470 | 21.29 % | 1283 | 23.69 % | 1109 | 22.81 % |  907 | 27.78 % |  574 | 26.48 % |\n",
       "| 0.2&lt;BS&lt;=0.3 | 1616 | 30.57 % | 1544 | 32.32 % | 1404 | 33.69 % | 1280 | 33.12 % |  920 | 35.87 % |\n",
       "| 0.3&lt;BS&lt;=0.4 | 1636 | 39.73 % | 1631 | 39.36 % | 1625 | 42.71 % | 1490 | 43.15 % | 1249 | 47.48 % |\n",
       "| 0.4&lt;BS&lt;=0.5 | 2022 | 49.7 %  | 2016 | 50.89 % | 1975 | 51.44 % | 2035 | 52.63 % | 1853 | 58.12 % |\n",
       "| 0.5&lt;BS&lt;=0.6 | 1248 | 58.57 % | 1237 | 61.76 % | 1317 | 61.88 % | 1449 | 66.46 % | 1376 | 67.51 % |\n",
       "| 0.6&lt;BS&lt;=0.7 | 1002 | 65.87 % | 1073 | 71.11 % | 1117 | 72.78 % | 1263 | 76.17 % | 1376 | 78.85 % |\n",
       "| 0.7&lt;BS&lt;=0.8 |  783 | 77.65 % |  875 | 78.63 % |  993 | 80.06 % | 1064 | 84.4 %  | 1266 | 84.99 % |\n",
       "| 0.8&lt;BS&lt;=0.9 |  594 | 84.34 % |  716 | 87.01 % |  800 | 89.25 % |  961 | 92.4 %  | 1288 | 92.08 % |\n",
       "| 0.9&lt;BS&lt;=1   |  428 | 90.42 % |  694 | 93.95 % |  859 | 93.83 % |  965 | 96.79 % | 1816 | 97.74 % |\n",
       "\n"
      ],
      "text/plain": [
       "   Row_Name    group1_BS group1_ref group2_BS group2_ref group3_BS group3_ref\n",
       "1  0<BS<=0.1   1031      13 %        813      14.15 %     703      13.09 %   \n",
       "2  0.1<BS<=0.2 1470      21.29 %    1283      23.69 %    1109      22.81 %   \n",
       "3  0.2<BS<=0.3 1616      30.57 %    1544      32.32 %    1404      33.69 %   \n",
       "4  0.3<BS<=0.4 1636      39.73 %    1631      39.36 %    1625      42.71 %   \n",
       "5  0.4<BS<=0.5 2022      49.7 %     2016      50.89 %    1975      51.44 %   \n",
       "6  0.5<BS<=0.6 1248      58.57 %    1237      61.76 %    1317      61.88 %   \n",
       "7  0.6<BS<=0.7 1002      65.87 %    1073      71.11 %    1117      72.78 %   \n",
       "8  0.7<BS<=0.8  783      77.65 %     875      78.63 %     993      80.06 %   \n",
       "9  0.8<BS<=0.9  594      84.34 %     716      87.01 %     800      89.25 %   \n",
       "10 0.9<BS<=1    428      90.42 %     694      93.95 %     859      93.83 %   \n",
       "   group4_BS group4_ref group5_BS group5_ref\n",
       "1   519      13.49 %     252      18.65 %   \n",
       "2   907      27.78 %     574      26.48 %   \n",
       "3  1280      33.12 %     920      35.87 %   \n",
       "4  1490      43.15 %    1249      47.48 %   \n",
       "5  2035      52.63 %    1853      58.12 %   \n",
       "6  1449      66.46 %    1376      67.51 %   \n",
       "7  1263      76.17 %    1376      78.85 %   \n",
       "8  1064      84.4 %     1266      84.99 %   \n",
       "9   961      92.4 %     1288      92.08 %   \n",
       "10  965      96.79 %    1816      97.74 %   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 10 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Row_Name</th><th scope=col>group1_BS</th><th scope=col>group1_ref</th><th scope=col>group2_BS</th><th scope=col>group2_ref</th><th scope=col>group3_BS</th><th scope=col>group3_ref</th><th scope=col>group4_BS</th><th scope=col>group4_ref</th><th scope=col>group5_BS</th><th scope=col>group5_ref</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td><span style=white-space:pre-wrap>0&lt;BS&lt;=0.1  </span></td><td> 438</td><td>27.63 %</td><td> 223</td><td>25.11 %</td><td> 237</td><td>25.32 %</td><td><span style=white-space:pre-wrap>  92</span></td><td><span style=white-space:pre-wrap>50 %   </span></td><td><span style=white-space:pre-wrap>  44</span></td><td>52.27 %</td></tr>\n",
       "\t<tr><td>0.1&lt;BS&lt;=0.2</td><td>1043</td><td>40.46 %</td><td> 677</td><td>41.36 %</td><td> 522</td><td>37.16 %</td><td> 359</td><td>47.63 %</td><td> 156</td><td>53.85 %</td></tr>\n",
       "\t<tr><td>0.2&lt;BS&lt;=0.3</td><td>1456</td><td>54.33 %</td><td>1172</td><td>55.8 % </td><td> 912</td><td>53.29 %</td><td> 723</td><td>56.02 %</td><td> 382</td><td>59.42 %</td></tr>\n",
       "\t<tr><td>0.3&lt;BS&lt;=0.4</td><td>1638</td><td>63.49 %</td><td>1648</td><td>63.17 %</td><td>1378</td><td>64.08 %</td><td>1133</td><td>65.75 %</td><td> 726</td><td>63.77 %</td></tr>\n",
       "\t<tr><td>0.4&lt;BS&lt;=0.5</td><td>2208</td><td>74.68 %</td><td>2278</td><td>74.98 %</td><td>2060</td><td>75.44 %</td><td>1971</td><td>75.39 %</td><td>1525</td><td>72.26 %</td></tr>\n",
       "\t<tr><td>0.5&lt;BS&lt;=0.6</td><td>1532</td><td>83.62 %</td><td>1463</td><td>85.71 %</td><td>1527</td><td>83.5 % </td><td>1556</td><td>86.5 % </td><td>1265</td><td>81.9 % </td></tr>\n",
       "\t<tr><td>0.6&lt;BS&lt;=0.7</td><td>1302</td><td>90.78 %</td><td>1319</td><td>92.72 %</td><td>1493</td><td>91.16 %</td><td>1571</td><td>93.06 %</td><td>1435</td><td>92.13 %</td></tr>\n",
       "\t<tr><td>0.7&lt;BS&lt;=0.8</td><td>1061</td><td>96.32 %</td><td>1217</td><td>96.38 %</td><td>1377</td><td>94.26 %</td><td>1591</td><td>97.55 %</td><td>1645</td><td>97.14 %</td></tr>\n",
       "\t<tr><td>0.8&lt;BS&lt;=0.9</td><td> 812</td><td>98.77 %</td><td>1019</td><td>98.92 %</td><td>1230</td><td>98.37 %</td><td>1454</td><td>99.24 %</td><td>1935</td><td>99.74 %</td></tr>\n",
       "\t<tr><td><span style=white-space:pre-wrap>0.9&lt;BS&lt;=1  </span></td><td> 469</td><td><span style=white-space:pre-wrap>100 %  </span></td><td> 962</td><td>99.9 % </td><td>1235</td><td>99.84 %</td><td>1549</td><td>99.61 %</td><td>2886</td><td>99.97 %</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 11\n",
       "\\begin{tabular}{lllllllllll}\n",
       " Row\\_Name & group1\\_BS & group1\\_ref & group2\\_BS & group2\\_ref & group3\\_BS & group3\\_ref & group4\\_BS & group4\\_ref & group5\\_BS & group5\\_ref\\\\\n",
       " <chr> & <int> & <chr> & <int> & <chr> & <int> & <chr> & <int> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t 0<BS<=0.1   &  438 & 27.63 \\% &  223 & 25.11 \\% &  237 & 25.32 \\% &   92 & 50 \\%    &   44 & 52.27 \\%\\\\\n",
       "\t 0.1<BS<=0.2 & 1043 & 40.46 \\% &  677 & 41.36 \\% &  522 & 37.16 \\% &  359 & 47.63 \\% &  156 & 53.85 \\%\\\\\n",
       "\t 0.2<BS<=0.3 & 1456 & 54.33 \\% & 1172 & 55.8 \\%  &  912 & 53.29 \\% &  723 & 56.02 \\% &  382 & 59.42 \\%\\\\\n",
       "\t 0.3<BS<=0.4 & 1638 & 63.49 \\% & 1648 & 63.17 \\% & 1378 & 64.08 \\% & 1133 & 65.75 \\% &  726 & 63.77 \\%\\\\\n",
       "\t 0.4<BS<=0.5 & 2208 & 74.68 \\% & 2278 & 74.98 \\% & 2060 & 75.44 \\% & 1971 & 75.39 \\% & 1525 & 72.26 \\%\\\\\n",
       "\t 0.5<BS<=0.6 & 1532 & 83.62 \\% & 1463 & 85.71 \\% & 1527 & 83.5 \\%  & 1556 & 86.5 \\%  & 1265 & 81.9 \\% \\\\\n",
       "\t 0.6<BS<=0.7 & 1302 & 90.78 \\% & 1319 & 92.72 \\% & 1493 & 91.16 \\% & 1571 & 93.06 \\% & 1435 & 92.13 \\%\\\\\n",
       "\t 0.7<BS<=0.8 & 1061 & 96.32 \\% & 1217 & 96.38 \\% & 1377 & 94.26 \\% & 1591 & 97.55 \\% & 1645 & 97.14 \\%\\\\\n",
       "\t 0.8<BS<=0.9 &  812 & 98.77 \\% & 1019 & 98.92 \\% & 1230 & 98.37 \\% & 1454 & 99.24 \\% & 1935 & 99.74 \\%\\\\\n",
       "\t 0.9<BS<=1   &  469 & 100 \\%   &  962 & 99.9 \\%  & 1235 & 99.84 \\% & 1549 & 99.61 \\% & 2886 & 99.97 \\%\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 11\n",
       "\n",
       "| Row_Name &lt;chr&gt; | group1_BS &lt;int&gt; | group1_ref &lt;chr&gt; | group2_BS &lt;int&gt; | group2_ref &lt;chr&gt; | group3_BS &lt;int&gt; | group3_ref &lt;chr&gt; | group4_BS &lt;int&gt; | group4_ref &lt;chr&gt; | group5_BS &lt;int&gt; | group5_ref &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 0&lt;BS&lt;=0.1   |  438 | 27.63 % |  223 | 25.11 % |  237 | 25.32 % |   92 | 50 %    |   44 | 52.27 % |\n",
       "| 0.1&lt;BS&lt;=0.2 | 1043 | 40.46 % |  677 | 41.36 % |  522 | 37.16 % |  359 | 47.63 % |  156 | 53.85 % |\n",
       "| 0.2&lt;BS&lt;=0.3 | 1456 | 54.33 % | 1172 | 55.8 %  |  912 | 53.29 % |  723 | 56.02 % |  382 | 59.42 % |\n",
       "| 0.3&lt;BS&lt;=0.4 | 1638 | 63.49 % | 1648 | 63.17 % | 1378 | 64.08 % | 1133 | 65.75 % |  726 | 63.77 % |\n",
       "| 0.4&lt;BS&lt;=0.5 | 2208 | 74.68 % | 2278 | 74.98 % | 2060 | 75.44 % | 1971 | 75.39 % | 1525 | 72.26 % |\n",
       "| 0.5&lt;BS&lt;=0.6 | 1532 | 83.62 % | 1463 | 85.71 % | 1527 | 83.5 %  | 1556 | 86.5 %  | 1265 | 81.9 %  |\n",
       "| 0.6&lt;BS&lt;=0.7 | 1302 | 90.78 % | 1319 | 92.72 % | 1493 | 91.16 % | 1571 | 93.06 % | 1435 | 92.13 % |\n",
       "| 0.7&lt;BS&lt;=0.8 | 1061 | 96.32 % | 1217 | 96.38 % | 1377 | 94.26 % | 1591 | 97.55 % | 1645 | 97.14 % |\n",
       "| 0.8&lt;BS&lt;=0.9 |  812 | 98.77 % | 1019 | 98.92 % | 1230 | 98.37 % | 1454 | 99.24 % | 1935 | 99.74 % |\n",
       "| 0.9&lt;BS&lt;=1   |  469 | 100 %   |  962 | 99.9 %  | 1235 | 99.84 % | 1549 | 99.61 % | 2886 | 99.97 % |\n",
       "\n"
      ],
      "text/plain": [
       "   Row_Name    group1_BS group1_ref group2_BS group2_ref group3_BS group3_ref\n",
       "1  0<BS<=0.1    438      27.63 %     223      25.11 %     237      25.32 %   \n",
       "2  0.1<BS<=0.2 1043      40.46 %     677      41.36 %     522      37.16 %   \n",
       "3  0.2<BS<=0.3 1456      54.33 %    1172      55.8 %      912      53.29 %   \n",
       "4  0.3<BS<=0.4 1638      63.49 %    1648      63.17 %    1378      64.08 %   \n",
       "5  0.4<BS<=0.5 2208      74.68 %    2278      74.98 %    2060      75.44 %   \n",
       "6  0.5<BS<=0.6 1532      83.62 %    1463      85.71 %    1527      83.5 %    \n",
       "7  0.6<BS<=0.7 1302      90.78 %    1319      92.72 %    1493      91.16 %   \n",
       "8  0.7<BS<=0.8 1061      96.32 %    1217      96.38 %    1377      94.26 %   \n",
       "9  0.8<BS<=0.9  812      98.77 %    1019      98.92 %    1230      98.37 %   \n",
       "10 0.9<BS<=1    469      100 %       962      99.9 %     1235      99.84 %   \n",
       "   group4_BS group4_ref group5_BS group5_ref\n",
       "1    92      50 %         44      52.27 %   \n",
       "2   359      47.63 %     156      53.85 %   \n",
       "3   723      56.02 %     382      59.42 %   \n",
       "4  1133      65.75 %     726      63.77 %   \n",
       "5  1971      75.39 %    1525      72.26 %   \n",
       "6  1556      86.5 %     1265      81.9 %    \n",
       "7  1571      93.06 %    1435      92.13 %   \n",
       "8  1591      97.55 %    1645      97.14 %   \n",
       "9  1454      99.24 %    1935      99.74 %   \n",
       "10 1549      99.61 %    2886      99.97 %   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# TODO put table number\n",
    "# table binning nodes based on their BS value and computing percentages for ref presence.\n",
    "# ML + SM \n",
    "\n",
    "# Reading Bootstrap values for each node and presence of node in family paralog reference tree\n",
    "df_bs_ML_SM = read.table(gzfile(\"tables/df_bootsrtap_clade_presence_in_ref_ML_SM.tsv.gz\"), header = TRUE)\n",
    "\n",
    "# get how many units (25) \n",
    "num_units = length(unique(df_bs_ML_SM$unit))\n",
    "\n",
    "# get how many groups there are (5) each with (100) families. groups define sequence length\n",
    "num_groups = length(unique(df_bs_ML_SM$group))\n",
    "\n",
    "# Map group names based on provided mapping\n",
    "groupnames_mapping = list(\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\" = \"group1\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\" = \"group2\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\" = \"group3\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\" = \"group4\",\n",
    "  \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\" = \"group5\"\n",
    ")\n",
    "\n",
    "# Apply the group name mapping\n",
    "df_bs_ML_SM = df_bs_ML_SM %>%\n",
    "  mutate(\n",
    "    group = unlist(groupnames_mapping[group])\n",
    "  )\n",
    "\n",
    "# Create empty data frames with 10 rows (one for each BS bin) to match the output vectors.\n",
    "df_unit1  = data.frame(matrix(ncol = 0, nrow = 10))\n",
    "df_unit25 = data.frame(matrix(ncol = 0, nrow = 10))\n",
    "\n",
    "# Loop through five iterations (one for each group).\n",
    "for (n in 1:num_groups) {\n",
    "  \n",
    "  # Filter data for the specific group and unit for the current iteration.\n",
    "  unit1_data  = subset(df_bs_ML_SM, group == paste0(\"group\", n) & unit == 1)\n",
    "  unit25_data = subset(df_bs_ML_SM, group == paste0(\"group\", n) & unit == num_units)\n",
    "  \n",
    "  # Initialize empty vectors to hold the binned values for the current iteration.\n",
    "  unit1_bin_percent_nodes   = c()\n",
    "  unit1_bin_percent_in_ref  = c()\n",
    "  unit25_bin_percent_nodes  = c()\n",
    "  unit25_bin_percent_in_ref = c()\n",
    "  \n",
    "  # Create bins from 0.1 to 1.0 in steps of 0.1 and calculate percentages for each bin.\n",
    "  for (i in seq(0.1, 1.0, by = 0.1)) {\n",
    "    \n",
    "    # Filter rows within the BS value range for both unit1 and unit25.\n",
    "    filtered_rows_unit1  = subset(unit1_data, BS > (i - 0.1) & BS <= i)\n",
    "    filtered_rows_unit25 = subset(unit25_data, BS > (i - 0.1) & BS <= i)\n",
    "    \n",
    "    # Calculate the percentage of nodes in the bin.\n",
    "    unit1_bin_percent_nodes  = c(unit1_bin_percent_nodes, length(filtered_rows_unit1$node))\n",
    "    unit25_bin_percent_nodes = c(unit25_bin_percent_nodes, length(filtered_rows_unit25$node))\n",
    "    \n",
    "    # Calculate the percentage of nodes in reference out of those in the bin.\n",
    "    unit1_bin_percent_in_ref  = c(unit1_bin_percent_in_ref, paste0(round(sum(filtered_rows_unit1$ref_presence == 1) / length(filtered_rows_unit1$node) * 100, 2), \" %\"))\n",
    "    unit25_bin_percent_in_ref = c(unit25_bin_percent_in_ref, paste0(round(sum(filtered_rows_unit25$ref_presence == 1) / length(filtered_rows_unit25$node)* 100, 2), \" %\"))\n",
    "  }\n",
    "  \n",
    "  # Add the vectors as new columns in the data frames, naming the columns based on the current group.\n",
    "  df_unit1[[paste0(\"group\", n, \"_BS\")]]   = unit1_bin_percent_nodes\n",
    "  df_unit1[[paste0(\"group\", n, \"_ref\")]]  = unit1_bin_percent_in_ref\n",
    "  df_unit25[[paste0(\"group\", n, \"_BS\")]]  = unit25_bin_percent_nodes\n",
    "  df_unit25[[paste0(\"group\", n, \"_ref\")]] = unit25_bin_percent_in_ref\n",
    "}\n",
    "\n",
    "# Define custom row labels representing BIN values\n",
    "row_names = c(\"0<BS<=0.1\", \"0.1<BS<=0.2\", \"0.2<BS<=0.3\", \"0.3<BS<=0.4\", \n",
    "               \"0.4<BS<=0.5\", \"0.5<BS<=0.6\", \"0.6<BS<=0.7\", \"0.7<BS<=0.8\", \n",
    "               \"0.8<BS<=0.9\", \"0.9<BS<=1\")\n",
    "\n",
    "# Add row names as the first column of the data frame\n",
    "df_unit1  = data.frame(\"Row_Name\" = row_names, df_unit1)\n",
    "df_unit25 = data.frame(\"Row_Name\" = row_names, df_unit25)\n",
    "                              \n",
    "# Output the data frames.\n",
    "df_unit1\n",
    "df_unit25\n",
    "\n",
    "# save tables in files\n",
    "write.table(df_unit1, file=\"tables/unit_1_per_group_nodes_BS_ref_ML_SM.tsv\", quote=FALSE, sep=\"\\t\", row.names=FALSE)\n",
    "write.table(df_unit25, file=\"tables/unit_25_per_group_nodes_BS_ref_ML_SM.tsv\", quote=FALSE, sep=\"\\t\", row.names=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f7936601-ed72-41d4-93fb-50b688070bee",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mUsing `size` aesthetic for lines was deprecated in ggplot2 3.4.0.\n",
      "\u001b[36mℹ\u001b[39m Please use `linewidth` instead.”\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n",
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n"
     ]
    }
   ],
   "source": [
    "# TODO add plot number\n",
    "# plot distribution of nodes in BS bins\n",
    "# ME + SM\n",
    "\n",
    "# Reading Bootstrap values for each node and presence of node in family paralog reference tree binned bi BS values\n",
    "df_unit1_bs_ME_SM  = read.table(\"tables/unit_1_per_group_nodes_BS_ref_ME_SM.tsv\", sep=\"\\t\", header = TRUE)\n",
    "df_unit25_bs_ME_SM = read.table(\"tables/unit_25_per_group_nodes_BS_ref_ME_SM.tsv\", sep=\"\\t\", header = TRUE)\n",
    "\n",
    "# retain only columns needed for plotting\n",
    "df_unit1_bs_ME_SM  = df_unit1_bs_ME_SM %>% select(\"Row_Name\", \"group1_BS\", \"group5_BS\")\n",
    "df_unit25_bs_ME_SM = df_unit25_bs_ME_SM %>% select(\"Row_Name\", \"group1_BS\", \"group5_BS\")\n",
    "\n",
    "# Add a source column to identify the data frame\n",
    "df_unit1_bs_ME_SM$source  = \"unit1\"\n",
    "df_unit25_bs_ME_SM$source = \"unit25\"\n",
    "\n",
    "# Combine the 2 data frames\n",
    "df_combined = bind_rows(df_unit1_bs_ME_SM, df_unit25_bs_ME_SM)\n",
    "\n",
    "# Convert Row_Name to numeric bins (midpoints of the bins for plotting)\n",
    "df_combined = df_combined %>%\n",
    "  mutate(\n",
    "    bin = case_when(\n",
    "      Row_Name == \"0<BS<=0.1\" ~ 0.05,\n",
    "      Row_Name == \"0.1<BS<=0.2\" ~ 0.15,\n",
    "      Row_Name == \"0.2<BS<=0.3\" ~ 0.25,\n",
    "      Row_Name == \"0.3<BS<=0.4\" ~ 0.35,\n",
    "      Row_Name == \"0.4<BS<=0.5\" ~ 0.45,\n",
    "      Row_Name == \"0.5<BS<=0.6\" ~ 0.55,\n",
    "      Row_Name == \"0.6<BS<=0.7\" ~ 0.65,\n",
    "      Row_Name == \"0.7<BS<=0.8\" ~ 0.75,\n",
    "      Row_Name == \"0.8<BS<=0.9\" ~ 0.85,\n",
    "      Row_Name == \"0.9<BS<=1\" ~ 0.95\n",
    "    )\n",
    "  )\n",
    "\n",
    "# Reshape the data for easier plotting with ggplot2\n",
    "df_long = df_combined %>%\n",
    "  pivot_longer(cols = starts_with(\"group\"), names_to = \"group\", values_to = \"num_nodes\") %>%\n",
    "  filter(group %in% c(\"group1_BS\", \"group5_BS\"))\n",
    "\n",
    "# Define bin midpoints\n",
    "bin_midpoints = c(0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0)\n",
    "\n",
    "# define y ticks\n",
    "y_ticks = c(0, 500, 1000, 1500, 2000)\n",
    "\n",
    "# Plotting\n",
    "num_nodes_distribution_p = ggplot(df_long, aes(x = bin, y = num_nodes, color = interaction(source, group), group = interaction(source, group))) +\n",
    "  geom_point(size = 2.2, alpha = 0.5) +\n",
    "  geom_smooth(method = \"loess\", se = FALSE, size = 1.5) +\n",
    "  # Customize colors\n",
    "  scale_color_manual(values = c(\"blue\", \"red\", \"#1AE6F5\", \"orange\"),\n",
    "                    labels = c(\"unit1 group1 len = 32\", \"unit1 group5 len = 82\", \"unit25 group1 len = 32\", \"unit25 group5 len = 82\")) +\n",
    "  # Set x-axis breaks to bin midpoints and y ticks\n",
    "  scale_x_continuous(breaks = bin_midpoints,\n",
    "                    limits = c(-0.001, 1.001)) +\n",
    "  scale_y_continuous(breaks = y_ticks) +\n",
    "  # Add vertical dashed lines\n",
    "  geom_vline(xintercept = 0.5, size = 1.5, linetype = \"dashed\", color = \"#808080\") +\n",
    "  geom_vline(xintercept = 0.7, size = 1.5, linetype = \"dashed\", color = \"#00CC66\") +\n",
    "  # Labels and theme customization\n",
    "  labs(\n",
    "    x = \"bins of Bootstrap value\",\n",
    "    y = \"number of nodes\",\n",
    "    color = \"Supermatrix-ME\"\n",
    "  ) +\n",
    "  theme_light() +\n",
    "  theme(\n",
    "    axis.text.x = element_text(angle = 45, hjust = 1), \n",
    "    axis.title = element_text(face = \"bold\"),\n",
    "    legend.text = element_text(size = 10),  # Adjust this size as desired\n",
    "    # Control major x-axis grid lines to show only where ticks are\n",
    "    panel.grid.major.x = element_line(color = \"lightgrey\"),\n",
    "    panel.grid.minor.x = element_blank()  # Remove minor grid lines on x-axis\n",
    "  )\n",
    "  \n",
    "ggsave(filename=\"figures/BS_lineplot_num_nodes_distribution_bin_unit1_25_ME_SM.png\", plot=num_nodes_distribution_p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7742a2d-821a-48ed-b4eb-befa942c28c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n",
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n"
     ]
    }
   ],
   "source": [
    "# TODO add plot number\n",
    "# plot distribution of nodes in BS bins\n",
    "# ML + SM\n",
    "\n",
    "# Reading Bootstrap values for each node and presence of node in family paralog reference tree binned bi BS values\n",
    "df_unit1_bs_ML_SM  = read.table(\"tables/unit_1_per_group_nodes_BS_ref_ML_SM.tsv\", sep=\"\\t\", header = TRUE)\n",
    "df_unit25_bs_ML_SM = read.table(\"tables/unit_25_per_group_nodes_BS_ref_ML_SM.tsv\", sep=\"\\t\", header = TRUE)\n",
    "\n",
    "# retain only columns needed for plotting\n",
    "df_unit1_bs_ML_SM  = df_unit1_bs_ML_SM %>% select(\"Row_Name\", \"group1_BS\", \"group5_BS\")\n",
    "df_unit25_bs_ML_SM = df_unit25_bs_ML_SM %>% select(\"Row_Name\", \"group1_BS\", \"group5_BS\")\n",
    "\n",
    "# Add a source column to identify the data frame\n",
    "df_unit1_bs_ML_SM$source  = \"unit1\"\n",
    "df_unit25_bs_ML_SM$source = \"unit25\"\n",
    "\n",
    "# Combine the 2 data frames\n",
    "df_combined = bind_rows(df_unit1_bs_ML_SM, df_unit25_bs_ML_SM)\n",
    "\n",
    "# Convert Row_Name to numeric bins (midpoints of the bins for plotting)\n",
    "df_combined = df_combined %>%\n",
    "  mutate(\n",
    "    bin = case_when(\n",
    "      Row_Name == \"0<BS<=0.1\" ~ 0.05,\n",
    "      Row_Name == \"0.1<BS<=0.2\" ~ 0.15,\n",
    "      Row_Name == \"0.2<BS<=0.3\" ~ 0.25,\n",
    "      Row_Name == \"0.3<BS<=0.4\" ~ 0.35,\n",
    "      Row_Name == \"0.4<BS<=0.5\" ~ 0.45,\n",
    "      Row_Name == \"0.5<BS<=0.6\" ~ 0.55,\n",
    "      Row_Name == \"0.6<BS<=0.7\" ~ 0.65,\n",
    "      Row_Name == \"0.7<BS<=0.8\" ~ 0.75,\n",
    "      Row_Name == \"0.8<BS<=0.9\" ~ 0.85,\n",
    "      Row_Name == \"0.9<BS<=1\" ~ 0.95\n",
    "    )\n",
    "  )\n",
    "\n",
    "# Reshape the data for easier plotting with ggplot2\n",
    "df_long = df_combined %>%\n",
    "  pivot_longer(cols = starts_with(\"group\"), names_to = \"group\", values_to = \"num_nodes\") %>%\n",
    "  filter(group %in% c(\"group1_BS\", \"group5_BS\"))\n",
    "\n",
    "# Define bin midpoints\n",
    "bin_midpoints <- c(0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0)\n",
    "\n",
    "# define y ticks\n",
    "y_ticks = c(0, 500, 1000, 1500, 2000, 2500)\n",
    "\n",
    "# Plotting\n",
    "num_nodes_distribution_p = ggplot(df_long, aes(x = bin, y = num_nodes, color = interaction(source, group), group = interaction(source, group))) +\n",
    "  geom_point(size = 2.2, alpha = 0.5) +\n",
    "  geom_smooth(method = \"loess\", se = FALSE, size = 1.5) +\n",
    "  # Customize colors\n",
    "  scale_color_manual(values = c(\"blue\", \"red\", \"#1AE6F5\", \"orange\"),\n",
    "                    labels = c(\"unit1 group1 len = 32\", \"unit1 group5 len = 82\", \"unit25 group1 len = 32\", \"unit25 group5 len = 82\")) +\n",
    "  # Set x-axis breaks to bin midpoints and y ticks\n",
    "  scale_x_continuous(breaks = bin_midpoints,\n",
    "                    limits = c(-0.001, 1.001)) +\n",
    "  scale_y_continuous(breaks = y_ticks) +\n",
    "  # Add vertical dashed lines\n",
    "  geom_vline(xintercept = 0.5, size = 1.5, linetype = \"dashed\", color = \"#808080\") +\n",
    "  geom_vline(xintercept = 0.7, size = 1.5, linetype = \"dashed\", color = \"#00CC66\") +\n",
    "  # Labels and theme customization\n",
    "  labs(\n",
    "    x = \"bins of Bootstrap value\",\n",
    "    y = \"number of nodes\",\n",
    "    color = \"Supermatrix-ML\"\n",
    "  ) +\n",
    "  theme_light() +\n",
    "  theme(\n",
    "    axis.text.x = element_text(angle = 45, hjust = 1), \n",
    "    axis.title = element_text(face = \"bold\"),\n",
    "    legend.text = element_text(size = 10),  # Adjust this size as desired\n",
    "    # Control major x-axis grid lines to show only where ticks are\n",
    "    panel.grid.major.x = element_line(color = \"lightgrey\"),\n",
    "    panel.grid.minor.x = element_blank()  # Remove minor grid lines on x-axis\n",
    "  )\n",
    "\n",
    "ggsave(filename=\"figures/BS_lineplot_num_nodes_distribution_bin_unit1_25_ML_SM.png\", plot=num_nodes_distribution_p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5460857b-dcd5-423d-b6dd-b855ca4880e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n",
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n"
     ]
    }
   ],
   "source": [
    "# TODO add plot number\n",
    "# plot percentage of nodes that are in ref for out of those in the bin\n",
    "# ME + SM\n",
    "\n",
    "# Reading Bootstrap values for each node and presence of node in family paralog reference tree binned bi BS values\n",
    "df_unit1_bs_ME_SM  = read.table(\"tables/unit_1_per_group_nodes_BS_ref_ME_SM.tsv\", sep=\"\\t\", header = TRUE)\n",
    "df_unit25_bs_ME_SM = read.table(\"tables/unit_25_per_group_nodes_BS_ref_ME_SM.tsv\", sep=\"\\t\", header = TRUE)\n",
    "\n",
    "# retain only columns needed for plotting\n",
    "df_unit1_ref_ME_SM  = df_unit1_bs_ME_SM %>% select(\"Row_Name\", \"group1_ref\", \"group5_ref\")\n",
    "df_unit25_ref_ME_SM = df_unit25_bs_ME_SM %>% select(\"Row_Name\", \"group1_ref\", \"group5_ref\")\n",
    "\n",
    "# Add a source column to identify the data frame\n",
    "df_unit1_ref_ME_SM$source  = \"unit1\"\n",
    "df_unit25_ref_ME_SM$source = \"unit25\"\n",
    "\n",
    "# Combine the 2 data frames\n",
    "df_combined = bind_rows(df_unit1_ref_ME_SM, df_unit25_ref_ME_SM)\n",
    "\n",
    "# Convert Row_Name to numeric bins (midpoints of the bins for plotting)\n",
    "df_combined = df_combined %>%\n",
    "  mutate(\n",
    "    bin = case_when(\n",
    "      Row_Name == \"0<BS<=0.1\" ~ 0.05,\n",
    "      Row_Name == \"0.1<BS<=0.2\" ~ 0.15,\n",
    "      Row_Name == \"0.2<BS<=0.3\" ~ 0.25,\n",
    "      Row_Name == \"0.3<BS<=0.4\" ~ 0.35,\n",
    "      Row_Name == \"0.4<BS<=0.5\" ~ 0.45,\n",
    "      Row_Name == \"0.5<BS<=0.6\" ~ 0.55,\n",
    "      Row_Name == \"0.6<BS<=0.7\" ~ 0.65,\n",
    "      Row_Name == \"0.7<BS<=0.8\" ~ 0.75,\n",
    "      Row_Name == \"0.8<BS<=0.9\" ~ 0.85,\n",
    "      Row_Name == \"0.9<BS<=1\" ~ 0.95\n",
    "    )\n",
    "  )\n",
    "\n",
    "# Convert the columns to numeric and remove the percentage\n",
    "df_combined = df_combined %>%\n",
    "  mutate(\n",
    "    group1_ref = as.numeric(str_remove(group1_ref, \" %\")),\n",
    "    group5_ref = as.numeric(str_remove(group5_ref, \" %\"))\n",
    "  )\n",
    "\n",
    "# Reshape the data for easier plotting with ggplot2\n",
    "df_long = df_combined %>%\n",
    "  pivot_longer(cols = starts_with(\"group\"), names_to = \"group\", values_to = \"percentage_ref\") %>%\n",
    "  filter(group %in% c(\"group1_ref\", \"group5_ref\"))\n",
    "\n",
    "# Define bin midpoints\n",
    "bin_midpoints <- c(0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0)\n",
    "\n",
    "# define y ticks\n",
    "y_ticks = c(0, 20, 40, 60, 80, 90, 100)\n",
    "minor_y_breaks = setdiff(seq(0, 100, by = 10), c(90))  # Exclude minor ticks at y=30 and y=70\n",
    "\n",
    "\n",
    "# Plotting\n",
    "percentage_nodes_p = ggplot(df_long, aes(x = bin, y = percentage_ref, color = interaction(source, group), group = interaction(source, group))) +\n",
    "  geom_point(size = 2.2, alpha = 0.5) +\n",
    "  geom_smooth(method = \"loess\", se = FALSE, size = 1.5) +\n",
    "  # Customize colors\n",
    "  scale_color_manual(values = c(\"blue\", \"red\", \"#1AE6F5\", \"orange\"),\n",
    "                    labels = c(\"unit1 group1 len = 32\", \"unit1 group5 len = 82\", \"unit25 group1 len = 32\", \"unit25 group5 len = 82\")) +\n",
    "  # Set x-axis breaks to bin midpoints and y ticks\n",
    "  scale_x_continuous(breaks = bin_midpoints,\n",
    "                    limits = c(-0.001, 1.001)) +\n",
    "  scale_y_continuous(breaks = y_ticks,\n",
    "                    minor_breaks = minor_y_breaks,\n",
    "                    limits = c(-0.01, 100.1)) +\n",
    "  # Add vertical dashed lines\n",
    "  geom_vline(xintercept = 0.5, size = 1.5, linetype = \"dashed\", color = \"#808080\") +\n",
    "  geom_vline(xintercept = 0.7, size = 1.5, linetype = \"dashed\", color = \"#00CC66\") +\n",
    "  # add orizontal line\n",
    "  geom_hline(yintercept = 90, size = 0.8, linetype = \"dashed\", color = \"lightgrey\") +\n",
    "  # Labels and theme customization\n",
    "  labs(\n",
    "    x = \"bins of Bootstrap value\",\n",
    "    y = \"% of nodes in bin that are present in reference tree\",\n",
    "    color = \"Supermatrix-ME\"\n",
    "  ) +\n",
    "  theme_light() +\n",
    "  theme(\n",
    "    axis.text.x = element_text(angle = 45, hjust = 1), \n",
    "    axis.title = element_text(face = \"bold\"),\n",
    "    legend.text = element_text(size = 10),  # Adjust this size as desired\n",
    "    # Control major x-axis grid lines to show only where ticks are\n",
    "    panel.grid.major.x = element_line(color = \"lightgrey\"),\n",
    "    panel.grid.minor.x = element_blank()  # Remove minor grid lines on x-axis\n",
    "  )\n",
    "\n",
    "ggsave(filename=\"figures/BS_lineplot_percent_nodes_in_bin_in_ref_unit1_25_ME_SM.png\", plot=percentage_nodes_p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "246148c7-3483-4b32-b6aa-7924d55a1fb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n",
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n"
     ]
    }
   ],
   "source": [
    "# TODO add plot number\n",
    "# plot percentage of nodes that are in ref for out of those in the bin\n",
    "# ML + SM\n",
    "\n",
    "# Reading Bootstrap values for each node and presence of node in family paralog reference tree binned bi BS values\n",
    "df_unit1_bs_ML_SM  = read.table(\"tables/unit_1_per_group_nodes_BS_ref_ML_SM.tsv\", sep=\"\\t\", header = TRUE)\n",
    "df_unit25_bs_ML_SM = read.table(\"tables/unit_25_per_group_nodes_BS_ref_ML_SM.tsv\", sep=\"\\t\", header = TRUE)\n",
    "\n",
    "# retain only columns needed for plotting\n",
    "df_unit1_ref_ML_SM  = df_unit1_bs_ML_SM %>% select(\"Row_Name\", \"group1_ref\", \"group5_ref\")\n",
    "df_unit25_ref_ML_SM = df_unit25_bs_ML_SM %>% select(\"Row_Name\", \"group1_ref\", \"group5_ref\")\n",
    "\n",
    "# Add a source column to identify the data frame\n",
    "df_unit1_ref_ML_SM$source  = \"unit1\"\n",
    "df_unit25_ref_ML_SM$source = \"unit25\"\n",
    "\n",
    "# Combine the 2 data frames\n",
    "df_combined = bind_rows(df_unit1_ref_ML_SM, df_unit25_ref_ML_SM)\n",
    "\n",
    "# Convert Row_Name to numeric bins (midpoints of the bins for plotting)\n",
    "df_combined = df_combined %>%\n",
    "  mutate(\n",
    "    bin = case_when(\n",
    "      Row_Name == \"0<BS<=0.1\" ~ 0.05,\n",
    "      Row_Name == \"0.1<BS<=0.2\" ~ 0.15,\n",
    "      Row_Name == \"0.2<BS<=0.3\" ~ 0.25,\n",
    "      Row_Name == \"0.3<BS<=0.4\" ~ 0.35,\n",
    "      Row_Name == \"0.4<BS<=0.5\" ~ 0.45,\n",
    "      Row_Name == \"0.5<BS<=0.6\" ~ 0.55,\n",
    "      Row_Name == \"0.6<BS<=0.7\" ~ 0.65,\n",
    "      Row_Name == \"0.7<BS<=0.8\" ~ 0.75,\n",
    "      Row_Name == \"0.8<BS<=0.9\" ~ 0.85,\n",
    "      Row_Name == \"0.9<BS<=1\" ~ 0.95\n",
    "    )\n",
    "  )\n",
    "\n",
    "# Convert the columns to numeric and remove the percentage\n",
    "df_combined = df_combined %>%\n",
    "  mutate(\n",
    "    group1_ref = as.numeric(str_remove(group1_ref, \" %\")),\n",
    "    group5_ref = as.numeric(str_remove(group5_ref, \" %\"))\n",
    "  )\n",
    "\n",
    "# Reshape the data for easier plotting with ggplot2\n",
    "df_long = df_combined %>%\n",
    "  pivot_longer(cols = starts_with(\"group\"), names_to = \"group\", values_to = \"percentage_ref\") %>%\n",
    "  filter(group %in% c(\"group1_ref\", \"group5_ref\"))\n",
    "\n",
    "# Define bin midpoints\n",
    "bin_midpoints <- c(0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0)\n",
    "\n",
    "# define y ticks\n",
    "y_ticks = c(0, 20, 40, 60, 80, 90, 100)\n",
    "minor_y_breaks = setdiff(seq(0, 100, by = 10), c(90))  # Exclude minor ticks at y=30 and y=70\n",
    "\n",
    "\n",
    "# Plotting\n",
    "percentage_nodes_p = ggplot(df_long, aes(x = bin, y = percentage_ref, color = interaction(source, group), group = interaction(source, group))) +\n",
    "  geom_point(size = 2.2, alpha = 0.5) +\n",
    "  geom_smooth(method = \"loess\", se = FALSE, size = 1.5) +\n",
    "  # Customize colors\n",
    "  scale_color_manual(values = c(\"blue\", \"red\", \"#1AE6F5\", \"orange\"),\n",
    "                    labels = c(\"unit1 group1 len = 32\", \"unit1 group5 len = 82\", \"unit25 group1 len = 32\", \"unit25 group5 len = 82\")) +\n",
    "  # Set x-axis breaks to bin midpoints and y ticks\n",
    "  scale_x_continuous(breaks = bin_midpoints,\n",
    "                    limits = c(-0.001, 1.001)) +\n",
    "  scale_y_continuous(breaks = y_ticks,\n",
    "                    minor_breaks = minor_y_breaks,\n",
    "                    limits = c(-0.01, 101)) +\n",
    "  # Add vertical dashed lines\n",
    "  geom_vline(xintercept = 0.5, size = 1.5, linetype = \"dashed\", color = \"#808080\") +\n",
    "  geom_vline(xintercept = 0.7, size = 1.5, linetype = \"dashed\", color = \"#00CC66\") +\n",
    "  # add orizontal line\n",
    "  geom_hline(yintercept = 90, size = 0.8, linetype = \"dashed\", color = \"lightgrey\") +\n",
    "  # Labels and theme customization\n",
    "  labs(\n",
    "    x = \"bins of Bootstrap value\",\n",
    "    y = \"% of nodes in bin that are present in reference tree\",\n",
    "    color = \"Supermatrix-ML\"\n",
    "  ) +\n",
    "  theme_light() +\n",
    "  theme(\n",
    "    axis.text.x = element_text(angle = 45, hjust = 1), \n",
    "    axis.title = element_text(face = \"bold\"),\n",
    "    legend.text = element_text(size = 10),  # Adjust this size as desired\n",
    "    # Control major x-axis grid lines to show only where ticks are\n",
    "    panel.grid.major.x = element_line(color = \"lightgrey\"),\n",
    "    panel.grid.minor.x = element_blank()  # Remove minor grid lines on x-axis\n",
    "  )\n",
    "\n",
    "ggsave(filename=\"figures/BS_lineplot_percent_nodes_in_bin_in_ref_unit1_25_ML_SM.png\", plot=percentage_nodes_p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92bfb310-9e7c-43cb-8a9d-dce71d69920d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "<br><br><br>\n",
    "# USEFULL STUFF BUT NOT INCLUDED IN PAPER\n",
    "<br><br><br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "6eee8eed-a86f-45e6-a38b-5ee931ae8f4f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mUsing `size` aesthetic for lines was deprecated in ggplot2 3.4.0.\n",
      "\u001b[36mℹ\u001b[39m Please use `linewidth` instead.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mA numeric `legend.position` argument in `theme()` was deprecated in ggplot2\n",
      "3.5.0.\n",
      "\u001b[36mℹ\u001b[39m Please use the `legend.position.inside` argument of `theme()` instead.”\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "# related to plot 4\n",
    "\n",
    "# Load data for ME + SM and ML + SM\n",
    "df_bs_ME_SM = read.table(gzfile(\"../analysis/tables/df_bootsrtap_clade_presence_in_ref_ME_SM.tsv.gz\"), header = TRUE)\n",
    "df_bs_ML_SM = read.table(gzfile(\"../analysis/tables/df_bootsrtap_clade_presence_in_ref_ML_SM.tsv.gz\"), header = TRUE)\n",
    "\n",
    "# define the mapping for numbering of groups\n",
    "groupnames_mapping = list(\"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\"=\"group1\", \n",
    " \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\"=\"group2\",\n",
    " \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\"=\"group3\",\n",
    " \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\"=\"group4\",\n",
    " \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\"=\"group5\"\n",
    " )\n",
    "\n",
    "# Rename group values to \"group1\", \"group2\", etc., for simplicity\n",
    "df_bs_ME_SM = df_bs_ME_SM %>%\n",
    "  mutate( group = unlist(groupnames_mapping[group])\n",
    "         )\n",
    "df_bs_ML_SM = df_bs_ML_SM %>%\n",
    "  mutate( group = unlist(groupnames_mapping[group])\n",
    "         )\n",
    "\n",
    "# Get unique units and groups\n",
    "unique_units  = unique(df_bs_ME_SM$unit)\n",
    "unique_groups = unique(df_bs_ME_SM$group)\n",
    "\n",
    "# Initialize a dataframe to store AUC values for each unit-group combination\n",
    "auc_df = expand.grid(\n",
    "  unit  = unique_units,\n",
    "  group = unique_groups,\n",
    "  stringsAsFactors = FALSE\n",
    ") %>%\n",
    "  mutate(\n",
    "    auc_ME_SM = NA_real_,\n",
    "    auc_ML_SM = NA_real_\n",
    "  )\n",
    "\n",
    "# Calculate AUC for each unit-group combination\n",
    "for (unit_num in unique_units) {\n",
    "  for (grp in unique_groups) {\n",
    "    \n",
    "    # Subset data for ME + SM\n",
    "    unit_group_data_ME_SM = subset(df_bs_ME_SM, unit == unit_num & group == grp)\n",
    "    if (length(unique(unit_group_data_ME_SM$ref_presence)) > 1) {\n",
    "      roc_obj_ME_SM                                                   = roc(unit_group_data_ME_SM$ref_presence, unit_group_data_ME_SM$BS, quiet = TRUE)\n",
    "      auc_df$auc_ME_SM[auc_df$unit == unit_num & auc_df$group == grp] = auc(roc_obj_ME_SM)\n",
    "    } else {\n",
    "      auc_df$auc_ME_SM[auc_df$unit == unit_num & auc_df$group == grp] = NA\n",
    "    }\n",
    "    \n",
    "    # Subset data for ML + SM\n",
    "    unit_group_data_ML_SM = subset(df_bs_ML_SM, unit == unit_num & group == grp)\n",
    "    if (length(unique(unit_group_data_ML_SM$ref_presence)) > 1) {\n",
    "      roc_obj_ML_SM                                                   = roc(unit_group_data_ML_SM$ref_presence, unit_group_data_ML_SM$BS, quiet = TRUE)\n",
    "      auc_df$auc_ML_SM[auc_df$unit == unit_num & auc_df$group == grp] = auc(roc_obj_ML_SM)\n",
    "    } else {\n",
    "      auc_df$auc_ML_SM[auc_df$unit == unit_num & auc_df$group == grp] = NA\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "# trsform to unit_1 ecc.. unit column\n",
    "auc_df = auc_df %>%\n",
    "  mutate(unit = paste0(\"unit_\", unit))\n",
    "\n",
    "# Reshape data from wide to long format for plotting\n",
    "auc_df_long = auc_df %>%\n",
    "  pivot_longer(cols      = starts_with(\"auc\"), \n",
    "               names_to  = \"metric\", \n",
    "               values_to = \"auc_value\") %>%\n",
    "  mutate(\n",
    "    method = case_when(\n",
    "      metric == \"auc_ME_SM\" ~ \"Supermatrix-ME\",\n",
    "      metric == \"auc_ML_SM\" ~ \"Supermatrix-ML\"\n",
    "    )\n",
    "  )\n",
    "\n",
    "# Order the 'unit' column based on mixed order for better x-axis presentation\n",
    "auc_df_long$unit = factor(auc_df_long$unit, levels = unique(auc_df_long$unit)[mixedorder(unique(auc_df_long$unit))])\n",
    "\n",
    "# Generate the custom labels for the legend plot\n",
    "custom_label = c(\n",
    "    \"Supermatrix-ME  group1  len = 32\",\n",
    "    \"Supermatrix-ML  group1  len = 32\",\n",
    "    \"Supermatrix-ME  group2  len = 46\",\n",
    "    \"Supermatrix-ML  group2  len = 46\",\n",
    "    \"Supermatrix-ME  group3  len = 50\",\n",
    "    \"Supermatrix-ML  group3  len = 50\",\n",
    "    \"Supermatrix-ME  group4  len = 60\",\n",
    "    \"Supermatrix-ML  group4  len = 60\",\n",
    "    \"Supermatrix-ME  group5  len = 82\",\n",
    "    \"Supermatrix-ML  group5  len = 82\"   \n",
    ")\n",
    "\n",
    "# designing the color palette of the lineplots, odd numbers are reds and refer to ME, even numbers are blues and refer to ML \n",
    "colors_palette = c('#FFB3B3', '#B3CDE3', '#F39A9C', '#92B6D1', '#E78285', '#71A0BF', '#DB696F', '#5089AC', '#D05159', '#2F739B')\n",
    "\n",
    "# Plotting\n",
    "auc_p = ggplot(auc_df_long, aes(x = unit, y = auc_value, color = interaction(method, group), shape = interaction(method, group), group = interaction(method, group))) +\n",
    "  geom_line(size = 0.7) +\n",
    "  geom_point(size = 2) +\n",
    "  \n",
    "  # Set custom shapes and colors for each method-group combination\n",
    "  scale_shape_manual(\n",
    "    values = rep(c(15, 17), each = 5),  # Use different shapes for ME (15: square) and ML (17: triangle)\n",
    "    labels = custom_label\n",
    "  ) +\n",
    "  scale_color_manual(\n",
    "    values = colors_palette,  # Use different colors for ME (dark blue) and ML (reddish)\n",
    "    labels = custom_label\n",
    "  ) +\n",
    "  \n",
    "  theme_light() +\n",
    "  theme(\n",
    "    axis.text.x = element_text(angle = 45, hjust = 1),\n",
    "    legend.position = c(0.70, 0.30),\n",
    "    legend.margin = margin(10, 10, 10, 10),\n",
    "    axis.title.x = element_text(face = \"bold\"),\n",
    "    axis.title.y = element_text(face = \"bold\")\n",
    "  ) +\n",
    "  \n",
    "  ylab(\"ROC AUC of Bootstrap Support\") +\n",
    "  xlab(\"Topology\") + \n",
    "  scale_y_continuous(limits = c(0.6, 0.9))\n",
    "\n",
    "# Save the plot\n",
    "ggsave(filename = \"figures/BS_lineplot_auc_per_unit_group_ME_ML_SM.png\", plot = auc_p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "e1ff4f0e-ec70-4989-89c7-f007d09969c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using name, family_group as id variables\n",
      "\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method (SM + ML and Bigtree)   OLD VERSION\n",
    "#\n",
    "\n",
    "# It takes a table of all families as rows (500) and it goes to a table of 1000 (500 BigTree + 500 SM + ML) where the running times instead of being separated into two different coluns now they are in the same one.\n",
    "# name + group columns will work as unique ID.  group is the directory name of the main.nf run/s, while name is the name of the family. \n",
    "fin_tracefile_mod_wide_melted           = melt(fin_tracefile_mod_wide)\n",
    "\n",
    "# transform running times from millisecond to seconds. And rename the column names.\n",
    "fin_tracefile_mod_wide_melted$value     = fin_tracefile_mod_wide_melted$value / 1000\n",
    "colnames(fin_tracefile_mod_wide_melted) = c(\"name\", \"family_group\", \"Method\", \"Time\")\n",
    "\n",
    "# crate a table with two boxplots (x axis) with Method columns as values and on the y the running times.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel. \n",
    "p = ggplot(fin_tracefile_mod_wide_melted,aes(x=Method, y=Time, fill=Method)) +\n",
    "    geom_boxplot(show.legend = FALSE) +\n",
    "    theme_light() +\n",
    "    ylab(\"CPU time (sec)\") + xlab(\"Method\")\n",
    "ggsave(filename=\"figures/CPU_time_per_method.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6df071e6-9d31-4158-ae52-3384e9e35a6b",
   "metadata": {},
   "source": [
    "### OLD version all families toghther"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53a77a3f-22e3-4136-b883-32db83aaa230",
   "metadata": {},
   "source": [
    "### TODO make the following better or remove it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "a73a53a9-f1a1-40c9-84e9-09965fe17f26",
   "metadata": {},
   "outputs": [],
   "source": [
    "# update the trace matrix with the length of the input msa\n",
    "library(\"Biostrings\") \n",
    "fin_tracefile_mod_with_len = fin_tracefile_mod_wide\n",
    "\n",
    "# for each family go and read the input MSA (input of main.nf) to get length of MSA (number of columns).\n",
    "# Iterate over each row of the dataframe\n",
    "for (i in 1:nrow(fin_tracefile_mod_with_len)) {\n",
    "    # Construct the full file path using group (directory) and name (filename) to read the input msa\n",
    "    # TODO make it not depend on the rerun_bla_bla direcrtory\n",
    "    input_msa_path = file.path(paste0(parent_of_all_dir, \"rerun_25_species_corrected_branches\"), fin_tracefile_mod_with_len$family_group[i], paste0(gsub(\"[()]\", \"\", fin_tracefile_mod_with_len$name[i]), \".ma\"))\n",
    "\n",
    "    # read the msa in fasta format and get the length of the first sequence (aka length of the alignment)\n",
    "    msa_file = readAAStringSet(input_msa_path)\n",
    "    msa_len  = length(msa_file[[1]])\n",
    "    \n",
    "    # add the length to the table \n",
    "    fin_tracefile_mod_with_len$length[i] = msa_len\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1b9d82dd-7856-44ec-99f5-60086ee8b8d1",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error: object 'fin_tracefile_mod_with_len' not found\n",
     "output_type": "error",
     "traceback": [
      "Error: object 'fin_tracefile_mod_with_len' not found\nTraceback:\n",
      "1. select(., -Big_Tree, -Supermatrix_ML)",
      "2. mutate(., fold_ratio = Big_Tree/Supermatrix_ML)",
      "3. .handleSimpleError(function (cnd) \n . {\n .     watcher$capture_plot_and_output()\n .     cnd <- sanitize_call(cnd)\n .     watcher$push(cnd)\n .     switch(on_error, continue = invokeRestart(\"eval_continue\"), \n .         stop = invokeRestart(\"eval_stop\"), error = invokeRestart(\"eval_error\", \n .             cnd))\n . }, \"object 'fin_tracefile_mod_with_len' not found\", base::quote(eval(expr, \n .     envir)))"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method (SM + ML and Bigtree) ratio plot of each family (x axis number columns BigMSA) \n",
    "# \n",
    "\n",
    "# Create a new column with the division of Big_Tree by Supermatrix_ML running times for each family (500)\n",
    "# and then remove the original columns Big_Tree Supermatrix_ML\n",
    "fin_tracefile_mod_ratio = fin_tracefile_mod_with_len %>%\n",
    "  mutate(fold_ratio = Big_Tree / Supermatrix_ML) %>%\n",
    "  select(-Big_Tree, -Supermatrix_ML)\n",
    "\n",
    "# create a table with on x axis the length of the MSA (n columns)  and on y the fold ratio (BigTree / SM + ML) of running times. The plot will have now 500 points.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel.\n",
    "p = ggplot(fin_tracefile_mod_ratio, aes(x=length, y=fold_ratio)) +\n",
    "    geom_point(color = 'blue', size = 2, alpha = 0.6) +  # Add points\n",
    "    theme_light() +\n",
    "    labs(x = \"MSA length (n# columns)\",\n",
    "       y = \"fold ratio (BigTree / Supermatrix_ML)\")\n",
    "ggsave(filename=\"figures/CPU_time_scatterplot_msa_len_BigTree_SM_ML.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1a03b1b-cfef-4abf-a0f0-d0fd195c15bf",
   "metadata": {},
   "source": [
    "### TODO the below was just a 10 familoy trial make it go away "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "873d9c1c-ab54-40b6-aed9-90caa7d43601",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO make this into the section of custom functions.\n",
    "# Function to convert to milliseconds\n",
    "convert_to_ms <- function(time_str) {\n",
    "    # Extract components using regular expressions\n",
    "    millis = \"0\"\n",
    "    seconds = \"0\"\n",
    "    minutes = \"0\"\n",
    "    if (str_detect(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=ms)\")) {\n",
    "        millis <- str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=ms)\")\n",
    "    } else {\n",
    "        seconds <- str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=s)\")\n",
    "        minutes <- str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=m)\")\n",
    "    }\n",
    "    \n",
    "    # Convert each component to milliseconds (coercing to numeric where needed)\n",
    "    seconds_ms <- as.numeric(seconds) * 1000\n",
    "    minutes_ms <- as.numeric(minutes) * 60 * 1000\n",
    "    millis_ms <- as.numeric(millis)\n",
    "    \n",
    "    # Sum all components and handle NAs by treating them as 0\n",
    "    total_ms <- sum(c(seconds_ms, minutes_ms, millis_ms), na.rm = TRUE)\n",
    "    return(total_ms)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "9a84bc61-7c7b-4805-a7dc-47071e51fc39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"results\"\n"
     ]
    }
   ],
   "source": [
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    # each main.nf output run has a trace file repoprting various information on the run itself\n",
    "    tracefile_path = file.path(run, \"pipeline_info/execution_trace_2024-09-27_17-39-02.txt\")\n",
    "    tracefile      = read.table(tracefile_path, sep = \"\\t\", header = TRUE)\n",
    "\n",
    "    # here we parse the trace file to retain only the processes that we want to take the running time (realtime) from\n",
    "    tracefile_mod = tracefile[grep(\"simulated_data:run_phylo_ML_supermatrix_aln_sim|simulated_data:only_concatenate_aln_sim|simulated_data:run_phylo_ML_full_aln_sim\", tracefile$name), c(\"name\", \"realtime\")]\n",
    "\n",
    "    # Use strsplit to split the 'name' column by whitespace\n",
    "    split_name    = strsplit(as.character(tracefile_mod$name), \" \")\n",
    "\n",
    "    # Create new columns\n",
    "    tracefile_mod$name       = sapply(split_name, `[`, 2) # First part \n",
    "    tracefile_mod$native_id  = sapply(split_name, `[`, 1) # Second part \n",
    "\n",
    "    # Reorder columns so 'native_id' becomes the second column\n",
    "    tracefile_mod = tracefile_mod[, c(\"name\", \"native_id\", \"realtime\")]\n",
    "\n",
    "    # trasform from human readable time to millisecond time\n",
    "    tracefile_mod$realtime = sapply(tracefile_mod$realtime, convert_to_ms)\n",
    "\n",
    "    # transforming the above table to one that have the process names (simulated_data:run_phylo_ML_supermatrix_aln_sim ecc..) as column names, \n",
    "    # the rows will be still the family names (working as ID) and the values in each column will be the running times associated to the column name (process name) for that given family.\n",
    "    tracefile_mod_wide = spread(tracefile_mod, native_id, realtime)\n",
    "    \n",
    "    # creating a new last column called Supermatrix_ML that has aas values the running times of concatenation plus ML tree computation (on concatenated MSA) \n",
    "    tracefile_mod_wide$Supermatrix_ML = as.numeric(tracefile_mod_wide$`simulated_data:only_concatenate_aln_sim`) + as.numeric(tracefile_mod_wide$`simulated_data:run_phylo_ML_supermatrix_aln_sim`)\n",
    "\n",
    "    # retain only the family name, BIgTree ML running time and SM + ML running time\n",
    "    tracefile_mod_wide = tracefile_mod_wide[,c(1,3,5)]\n",
    "\n",
    "    # call the second column name Big_Tree and set their value to numeric type\n",
    "    colnames(tracefile_mod_wide)[2] = \"Big_Tree\"\n",
    "    tracefile_mod_wide$Big_Tree = as.numeric(tracefile_mod_wide$Big_Tree)\n",
    "\n",
    "    # adding a first column that groups families by the run of the main.nf\n",
    "    tracefile_mod_wide$group = rep(basename(run), 10)\n",
    "\n",
    "    # add tables toghether raw-wise\n",
    "    if(run == run_dirs[1]) {\n",
    "        fin_tracefile_mod_wide = tracefile_mod_wide\n",
    "    } else {\n",
    "        fin_tracefile_mod_wide = rbind(fin_tracefile_mod_wide, tracefile_mod_wide)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "ee66b8ad-a4df-4ad9-bdf2-7f8d7a13c8c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>Big_Tree</dt><dd>195.4</dd><dt>Supermatrix_ML</dt><dd>23.47</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[Big\\textbackslash{}\\_Tree] 195.4\n",
       "\\item[Supermatrix\\textbackslash{}\\_ML] 23.47\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "Big_Tree\n",
       ":   195.4Supermatrix_ML\n",
       ":   23.47\n",
       "\n"
      ],
      "text/plain": [
       "      Big_Tree Supermatrix_ML \n",
       "        195.40          23.47 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "196"
      ],
      "text/latex": [
       "196"
      ],
      "text/markdown": [
       "196"
      ],
      "text/plain": [
       "[1] 196"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "21.87"
      ],
      "text/latex": [
       "21.87"
      ],
      "text/markdown": [
       "21.87"
      ],
      "text/plain": [
       "[1] 21.87"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# make the means of all (500) family running time for the BigTree and ML + SM. Bring them to seconds (/1000) and round the digits to 2\n",
    "round(colMeans(fin_tracefile_mod_wide[, c(2, 3)])/1000,digits = 2)\n",
    "\n",
    "# compute the median for the BigTree approach and the  ML + SM.\n",
    "round(median(fin_tracefile_mod_wide$Big_Tree)/1000,digits = 2)\n",
    "round(median(fin_tracefile_mod_wide$Supermatrix_ML)/1000,digits = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "ed1275aa-e0bd-4628-bab2-5eb5a54271ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using name, group as id variables\n",
      "\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "\n",
    "fin_tracefile_mod_wide_melted           = melt(fin_tracefile_mod_wide)\n",
    "\n",
    "fin_tracefile_mod_wide_melted$value     = fin_tracefile_mod_wide_melted$value / 1000\n",
    "colnames(fin_tracefile_mod_wide_melted) = c(\"name\",\"group\",\"Method\",\"Time\")\n",
    "\n",
    "p = ggplot(fin_tracefile_mod_wide_melted,aes(x=Method,y=Time,fill=Method)) +\n",
    "    geom_boxplot(show.legend = FALSE) +\n",
    "    theme_light() +\n",
    "    ylab(\"CPU time (sec)\") + xlab(\"Method\")\n",
    "ggsave(filename=\"CPU_time_per_method_10_fam.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31c3465f-5bf7-465b-bf62-e004f9441080",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
