{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8d6c408c-3b82-4cbb-b94a-cb9728c78fe6",
   "metadata": {},
   "source": [
    "# Analysis of simulated data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "606b5d7a-8d48-4caf-8c72-e89dce7c1208",
   "metadata": {},
   "source": [
    "\n",
    "TODO add here a brief description of your project or analysis. Explain the purpose, goals, or context of the notebook here.\n",
    "\n",
    "## Table of Contents\n",
    "- [1. Chapter 1 - Introduction](#chapter-1-introduction)\n",
    "- [2. Chapter 2 - R packages](#chapter-2-R-packages)\n",
    "- [3. Chapter 3 - Input specification](#chapter-3-Input-specification)\n",
    "- [4. Chapter 4 - Supermatrix and Supertree trees VS true topologies](#chapter-4-Supermatrix-and-Supertree-trees-VS-true-topologies)\n",
    "- [5. Chapter 5 - BigTree VS true topologies](#chapter-5-BigTree-VS-true-topologies)\n",
    "- [6. Chapter 6 - RF distance summaries](#chapter-6-RF-distance-summaries)\n",
    "    - [6.1 Subchapter 6.1 - Table 1](#subchapter-6_1-Table-1)\n",
    "- [7. Chapter 7 - Running times](#chapter-7-Running-times)\n",
    "    - [7.1 Subchapter 7.1 - Figure 2](#subchapter-7_1-Figure-2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d203093e-da42-45a7-8e98-853d4403630a",
   "metadata": {},
   "source": [
    "<a id=\"chapter-1-introduction\"></a>\n",
    "\n",
    "## Chapter 1 - Introduction\n",
    "\n",
    "TODO add description of notebook and how to use it.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3912a30d-42ec-4e90-99e6-8bacfeb50512",
   "metadata": {},
   "source": [
    "<a id=\"chapter-2-R-packages\"></a>\n",
    "\n",
    "## Chapter 2 - R packages\n",
    "\n",
    "Here is the list of packages needed for this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "67058cec-ea5e-4ce3-be29-faf428919086",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: ape\n",
      "\n",
      "Loading required package: phytools\n",
      "\n",
      "Loading required package: maps\n",
      "\n",
      "Loading required package: ade4\n",
      "\n",
      "\n",
      "Attaching package: ‘plotrix’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:phytools’:\n",
      "\n",
      "    rescale\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘plyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:maps’:\n",
      "\n",
      "    ozone\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ape’:\n",
      "\n",
      "    where\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘rstatix’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    desc, mutate\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘gplots’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plotrix’:\n",
      "\n",
      "    plotCI\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    lowess\n",
      "\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "\n",
      "Attaching package: ‘viridis’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:maps’:\n",
      "\n",
      "    unemp\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘gridExtra’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘scales’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:viridis’:\n",
      "\n",
      "    viridis_pal\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plotrix’:\n",
      "\n",
      "    rescale\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:phytools’:\n",
      "\n",
      "    rescale\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(\"phangorn\")\n",
    "library(\"ggplot2\")\n",
    "library(\"reshape2\")\n",
    "library(\"geiger\")\n",
    "library(\"adephylo\")\n",
    "library(\"phytools\")\n",
    "library(\"rlist\")\n",
    "library(\"plotrix\")\n",
    "library(\"plyr\")\n",
    "library(\"dplyr\")\n",
    "library(\"hrbrthemes\")\n",
    "library(\"rstatix\")\n",
    "library(\"gplots\")\n",
    "library(\"viridis\")\n",
    "library('gtools')\n",
    "library(gridExtra)\n",
    "library(grid)\n",
    "library(\"ggforce\")\n",
    "library(\"scales\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b609768d-d957-4b11-83ce-c27152a96fb2",
   "metadata": {},
   "source": [
    "TODO make a chapter for aqll custom functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20bb020e-4f1d-4338-ac5a-abecc8a06121",
   "metadata": {},
   "source": [
    "<a id=\"chapter-3-Input-specification\"></a>\n",
    "\n",
    "## Chapter 3 - Input specification\n",
    "\n",
    "Change the parent_of_all_dir variable to give the location of the input. All input files are expected to be under it. Otherwise change the single input location to you liking.<br>\n",
    "All input files or directory paths are listed here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "04187777-5e20-43b3-9f39-62523fc0fc0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the location of the directory that contains all inputs for analyses \n",
    "parent_of_all_dir = \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/\"\n",
    "\n",
    "# The directory (one or more) that contain all the results from the main.nf pipeline run. It can be a glob path.\n",
    "#dir_containing_all_pipe_results = \"../results\" # todo make this disappear once analysis are ready\n",
    "dir_containing_all_pipe_results = paste0(parent_of_all_dir, \"rerun_25_species_corrected_branches/rerun_sim_b_factor*\")\n",
    "\n",
    "#  TODO describe what this file does\n",
    "reference_tree_codefile = file.path(parent_of_all_dir, \"ref_tree.code\")\n",
    "\n",
    "# get the species/organism list of names used by the main.nf pipeline\n",
    "species_name_file = file.path(parent_of_all_dir, \"orthologs_org_ids_to_concatenate\")\n",
    "\n",
    "\n",
    "# each run of main.nf has one output directory. here we get that or multiple runs outputs directories.\n",
    "run_dirs = Sys.glob(dir_containing_all_pipe_results)\n",
    "run_dirs = run_dirs[file.info(run_dirs)$isdir]\n",
    "\n",
    "# Create the 'figures' directory if it doesn't exist. Created where this notebook is executed.\n",
    "if (!dir.exists(\"figures\")) {\n",
    "    dir.create(\"figures\")\n",
    "}\n",
    "\n",
    "# Create the 'tables' directory if it doesn't exist. Created where this notebook is executed.\n",
    "if (!dir.exists(\"tables\")) {\n",
    "    dir.create(\"tables\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "566da14f-4b2d-4610-9a9a-6f892dfc9964",
   "metadata": {},
   "source": [
    "<a id=\"chapter-4-Supermatrix-and-Supertree-trees-VS-true-topologies\"></a>\n",
    "\n",
    "## Chapter 4 - Supermatrix and Supertree trees VS true topologies\n",
    "\n",
    "TODO add description "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c707c5c4-51d2-4b25-a6cc-b8a9ccb9e0aa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n"
     ]
    }
   ],
   "source": [
    "# setting global lists that eill hold all family level values\n",
    "RF_avg_list = list()\n",
    "collect_branch_len_supermatrix_ml = c()\n",
    "collect_diameter_supermatrix_ml = c()\n",
    "\n",
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    # the mapping used later on to rename species trees according to a more generic sequence name\n",
    "    ref_codefile = read.table(file = reference_tree_codefile )\n",
    "    \n",
    "    # get al the family subdirectories in each mqin.nf output run dir\n",
    "    fam_dirs = grep(\"avg_branchlen_0.7_protein_family*\", list.dirs(run, recursive = FALSE), value = TRUE)\n",
    "    count_fam = 0\n",
    "\n",
    "    # this will contain the mean branch length for all true/reference paralog trees for each gene family\n",
    "    avg_branch_length = c()\n",
    "\n",
    "    # read the names of the ortholog species.\n",
    "    orgs = read.table(file = species_name_file)\n",
    "\n",
    "    # get all units trees for all combination of methods \n",
    "    for (m in fam_dirs) {\n",
    "        \n",
    "        count_fam = count_fam + 1\n",
    "\n",
    "        # print every 10 families a check on progress to screen\n",
    "        if (count_fam %% 10 == 0) {\n",
    "            print(m)\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250 = 25 organism * 10 samples/statistical repeats) for the combination Minimun Evolution (ME) + supermatrix method (SM). \n",
    "        # We unroot every tree and fully resolve it to use the RF.dist package and others. \n",
    "        all_units_file  = file.path(m, \"results/rerun\", \"all_units.nwk\")\n",
    "        all_units_trees = read.tree(all_units_file, keep.multi=TRUE)\n",
    "        all_units_trees = unroot(all_units_trees)\n",
    "        \n",
    "        # check if there any non-bynary (consensus trees) among the 250 trees present in the file above\n",
    "        if (length(which(is.binary(all_units_trees) == FALSE)) != 0) {\n",
    "            \n",
    "            # if there are non-binary trees, iterate through them and transformed into a binary one using multi2di package\n",
    "            for (tree in which(is.binary(all_units_trees) == FALSE)) {\n",
    "                all_units_trees[[tree]] = multi2di(all_units_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250) ME + supertree (ST) method and handle non-binary trees as above\n",
    "        all_superfine_file  = file.path(m, \"results/rerun\", \"all_units_superfine_trees.nwk\")\n",
    "        all_superfine_trees = read.tree(all_superfine_file, keep.multi=TRUE)\n",
    "        all_superfine_trees = unroot(all_superfine_trees)\n",
    "        if (length(which(is.binary(all_superfine_trees) == FALSE)) != 0) {\n",
    "            for (tree in which(is.binary(all_superfine_trees) == FALSE)) {\n",
    "                all_superfine_trees[[tree]] = multi2di(all_superfine_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250) Maximum Likelihood (ML) + SM \n",
    "        all_units_raxml_file  = file.path(m, \"results/rerun\", \"all_units_RAxML_trees.nwk\")\n",
    "        all_units_raxml_trees = read.tree(all_units_raxml_file, keep.multi=TRUE)\n",
    "        all_units_raxml_trees = unroot(all_units_raxml_trees)\n",
    "        if (length(which(is.binary(all_units_raxml_trees) == FALSE)) != 0) {\n",
    "            for (tree in which(is.binary(all_units_raxml_trees) == FALSE)) {\n",
    "                all_units_raxml_trees[[tree]] = multi2di(all_units_raxml_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # take only the last 10 of the 250 above trees (unit25's 10 samples/replicates)(unit25 = all 25 single units concatenated) and collect branch lengths (each branch in tree, 27 in total 2*15-3) and diameter (using patristic distance)\n",
    "        for(mytree in 241:length(all_units_raxml_trees)) {\n",
    "            collect_branch_len_supermatrix_ml = c(collect_branch_len_supermatrix_ml, all_units_raxml_trees[[mytree]]$edge.length)\n",
    "            collect_diameter_supermatrix_ml = c(collect_diameter_supermatrix_ml, max(distTips(all_units_raxml_trees[[mytree]], method = 'patristic')))\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250) ML + ST\n",
    "        all_raxml_superfine_file  = file.path(m, \"results/rerun\", \"all_units_RAxML_superfine_trees.nwk\")\n",
    "        all_raxml_superfine_trees = read.tree(all_raxml_superfine_file, keep.multi=TRUE)\n",
    "        all_raxml_superfine_trees = unroot(all_raxml_superfine_trees)\n",
    "        if (length(which(is.binary(all_raxml_superfine_trees) == FALSE)) != 0) {\n",
    "            for (tree in which(is.binary(all_raxml_superfine_trees) == FALSE)) {\n",
    "                all_raxml_superfine_trees[[tree]] = multi2di(all_raxml_superfine_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # Reference tree block, each gene family has it's own reference tree that is the one generated by the simulation step. Along this tree all sequences are generated and then alligned and passed as input for main.nf pipeline.\n",
    "        ref_tree_file = file.path(run, paste0(basename(m), \".nwk\"))\n",
    "        ref_tree      = read.tree(ref_tree_file, keep.multi=FALSE)\n",
    "\n",
    "        # Since each gene family has a predifined gene tree (ref_tree) composed of a fixed species tree pasted as leaf of the paralog tree. The tree is designed like this by choice. Each gene families varies only in the paralog tree that it has that is unique (15 tips).\n",
    "        # Here we extract the paralog tree and to do so we just need to extract the subtree with all the sequences of omne species. Since each species have the same paralog tree. So we can extract the subtree of whichever species it does not matter.\n",
    "        ref_subtree   = keep.tip(ref_tree,as.character(ref_codefile$V1))\n",
    "\n",
    "        # we then rename the tips of the paralog tree (ref_subtree) to a more generic keyname. Effectively going from rat1 to seq1. \n",
    "        for (j in 1:length(ref_subtree$tip.label)) { \n",
    "            ref_subtree$tip.label[j] = as.character(ref_codefile$V2[which(ref_codefile$V1 == ref_subtree[[\"tip.label\"]][j])])\n",
    "        }\n",
    "\n",
    "        # Unroot the tree and perform aother operation and then save the paralog tree to the specifc gene faimily directory\n",
    "        ref_subtree        = unroot(ref_subtree)\n",
    "        ref_subtree        = list(ref_subtree)\n",
    "        class(ref_subtree) = \"multiPhylo\"\n",
    "        subtree_out_file   = file.path(m, \"results/rerun\", paste0(\"ref_subtree_\", basename(m), \".nwk\"))\n",
    "        # TODO remove this comment write.tree(ref_subtree, file=subtree_out_file)\n",
    "\n",
    "        # update on the average branch length with the mean of the true/refernce paralog tree of the given gene family\n",
    "        avg_branch_length = c(avg_branch_length, mean(ref_subtree[[1]]$edge.length))\n",
    "\n",
    "        # Make a list containing all trees gathered so far and compute all possible pair comparisons among trees. There are 1001*1001-1001 unique pairewise comparisons ( a square matrix of size 1001 with 0 in diagonal ).\n",
    "        # On this matrix (plotdata) all other measuramnts are done. This is a symmetric matrix (distance matrix)\n",
    "        all_trees        = c(ref_subtree, all_units_trees, all_superfine_trees, all_units_raxml_trees, all_raxml_superfine_trees)\n",
    "        class(all_trees) = \"multiPhylo\"\n",
    "        all_vs_all_rf    = RF.dist(all_trees, normalize = TRUE)\n",
    "        plotdata         = data.matrix(all_vs_all_rf)\n",
    "\n",
    "        # all ME metrics to be computed superfine = SuperTree\n",
    "        mean_rf           = c()\n",
    "        mean_superfine_rf = c()\n",
    "\n",
    "        # all ML metrics to be computed\n",
    "        mean_raxml_rf           = c()\n",
    "        mean_raxml_superfine_rf = c()\n",
    "\n",
    "        # proceed throught the plotdata matrix taking from the second element one every ten up to 242. Basically proceed at chunkcs of 10 (samples). On those chuncks compute the mean and the Standard error. \n",
    "        # average all sample \"values\" whithin a unit. going from 250 value to 25 averages\n",
    "        # This iterate only through the values of the comparison between the reference paralofg tree (ref_subtree)(first column/row of matrix) and the 250 paraogs trees of ME + SM (all_units_trees)\n",
    "        for (i in seq(2, (length(orgs$V1)*10)+1, by=10)) {\n",
    "            mean_rf = c(mean_rf, mean(plotdata[1, i:(i+9)]))\n",
    "        }\n",
    "\n",
    "        # This instead goes from row 252 to 492 of the plotdata matrix. It still goes in chuncks of 10 (samples). \n",
    "        # So this effectively does what is done above but between the reference paralofg tree and the 250 paraogs trees of ME + ST (all_superfine_trees)\n",
    "        for (i in seq( (length(orgs$V1)*10)+2, (2*(length(orgs$V1)*10)+1), by=10 )) {\n",
    "            mean_superfine_rf = c(mean_superfine_rf, mean(plotdata[1,i:(i+9)]))\n",
    "        }\n",
    "\n",
    "        # This for loop goes from 502 to 742. It still goes in chuncks of 10 (samples).\n",
    "        # So this effectively does what is done above but between the reference paralofg tree and the 250 paraogs trees of ML + SM (all_units_raxml_trees)\n",
    "        for (i in seq( (2*(length(orgs$V1)*10))+2, (3*(length(orgs$V1)*10)+1), by=10 )) {\n",
    "            mean_raxml_rf = c(mean_raxml_rf, mean(plotdata[1,i:(i+9)]))\n",
    "        }\n",
    "\n",
    "        # This for loop goes from 752 to 992. It still goes in chuncks of 10 (samples).\n",
    "        # So this effectively does what is done above but between the reference paralofg tree and the 250 paraogs trees of ML + ST (all_raxml_superfine_trees)\n",
    "        for (i in seq( (3*(length(orgs$V1)*10))+2, (4*(length(orgs$V1)*10)+1), by=10)) {\n",
    "            mean_raxml_superfine_rf = c(mean_raxml_superfine_rf, mean(plotdata[1,i:(i+9)]))\n",
    "        }\n",
    "\n",
    "        # collect all mean RF values 25 mean_rf, 25 mean_superfine_rf ecc.. for each family. One for each unit.\n",
    "        collection_mean_rf      = c(mean_rf, mean_superfine_rf, mean_raxml_rf, mean_raxml_superfine_rf)\n",
    "\n",
    "        # create the column names for the later dataframe, ME = minimum Evolutio, ML = Maximum Likelihood, SM = SuperMatrix , ST = SuperTree. names will be like ME_SM_unit_1\n",
    "        units_names = c()\n",
    "        for (i in 1:length(orgs$V1)) {\n",
    "            units_names = c(units_names, paste(\"ME_SM\", \"unit\", i, sep=\"_\"), \n",
    "                            paste(\"ME_ST\", \"unit\", i, sep=\"_\"), \n",
    "                            paste(\"ML_SM\", \"unit\", i, sep=\"_\"), \n",
    "                            paste(\"ML_ST\", \"unit\", i, sep=\"_\"))\n",
    "        }\n",
    "\n",
    "        # natural_sort them or human readable sort, so that the order is  ME_SM, ME_ST, ML_SM and ML_ST\n",
    "        mixedsort(units_names)\n",
    "\n",
    "        # Finally update the \"global\" lists created at the beginning\n",
    "        RF_avg_list      = list.append(RF_avg_list, collection_mean_rf)\n",
    "\n",
    "    }\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9ac26094-fda0-489f-af78-a70716514196",
   "metadata": {},
   "outputs": [],
   "source": [
    "# transform the list of lists (RF_avg_list) into a dataframe (matrix) size 500 * 100 (25 units for all combinations of ME,ML and SM,ST)\n",
    "RF_avg_as_df = ldply(RF_avg_list, recursive=FALSE)\n",
    "\n",
    "# TODO see if standar error se was actually usefull\n",
    "RF_df = round(RF_avg_as_df, digits=3)\n",
    "\n",
    "# add the units name generated before as column names for the dataframe and save it to file\n",
    "colnames(RF_df) = units_names\n",
    "write.table(RF_df, file = \"tables/avg_RF_between_ref_supermatrix_and_supertree.tsv\", quote = FALSE, row.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f8b0912-095a-4aef-9277-f40429e45a5e",
   "metadata": {},
   "source": [
    "<a id=\"chapter-5-BigTree-VS-true-topologies\"></a>\n",
    "\n",
    "## Chapter 5 BigTree VS true topologies\n",
    "\n",
    "TODO add description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ce413952-5667-4c45-8d24-5a578c3527a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_010.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_020.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_030.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_040.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_050.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_060.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_070.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_080.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_090.sub\"\n",
      "[1] \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species/avg_branchlen_0.7_protein_family_100.sub\"\n"
     ]
    }
   ],
   "source": [
    "# Trees from BigTree (Gene family tree) VS true topologies\n",
    "\n",
    "mean_rf_subtrees_from_full_aln_vs_true_tree = c()\n",
    "sd_rf_subtrees_from_full_aln_vs_true_tree = c()\n",
    "RF_matrices = list()\n",
    "\n",
    "collect_branch_len = c()\n",
    "collect_diameter = c()\n",
    "collect_branch_len_sub_from_ref = c()\n",
    "collect_diameter_sub_from_ref = c()\n",
    "collect_branch_len_big_trees = c()\n",
    "collect_diameter_big_trees = c()\n",
    "\n",
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    # the mapping used later on to rename species trees according to a more generic sequence name\n",
    "    ref_codefile = read.table(file = reference_tree_codefile )\n",
    "    \n",
    "    # get al the family subdirectories in each mqin.nf output run dir\n",
    "    fam_dirs = grep(\"avg_branchlen_0.7_protein_family*\", list.dirs(run, recursive = FALSE), value = TRUE)\n",
    "    count_fam = 0\n",
    "\n",
    "    # this will contain the mean branch length for all true/reference paralog trees for each gene family\n",
    "    avg_branch_length = c()\n",
    "\n",
    "    # read the names of the ortholog species.\n",
    "    orgs = read.table(file = species_name_file)\n",
    "\n",
    "    # extract and work on all BigTree msa for all families. this MSA (500) is the input of the main.nf for simulated data workflow, whose outputs are analyzed in this notebook.\n",
    "    for (m in fam_dirs) {\n",
    "        \n",
    "        count_fam = count_fam + 1\n",
    "\n",
    "        # print every 10 families a check on progress to screen\n",
    "        if (count_fam %% 10 == 0) {\n",
    "            print(m)\n",
    "        }\n",
    "\n",
    "        # each sequence in the MSA id mapping. Each sequence header is mapped to a new generic one, this file holds such mapping.\n",
    "        full_aln_codepath = file.path(m, \"results/rerun\", \"full_aln.code_name\")\n",
    "        full_aln_codefile = read.table(file = full_aln_codepath)\n",
    "\n",
    "        # variable to store all species subtrees extracted from the Gene family MSA aka BigTree MSA\n",
    "        all_species_raxml_subtrees_from_full_aln = c()\n",
    "\n",
    "        # get the BigTree tree. full_aln=BigTree and unroot it \n",
    "        bigtree_file  = file.path(m, \"results/rerun\", paste0(\"RAxML_bestTree.\", basename(m), \"_full_aln_coded_raxml.nwk\"))\n",
    "        full_aln_tree = read.tree(bigtree_file)\n",
    "        full_aln_tree = unroot(full_aln_tree)\n",
    "\n",
    "        # update the branch len of the BigTrees vector. each tree has 375 tips (25 species * 15 paralog), so there are 747 (2*375-3)  branch length for each tree.\n",
    "        collect_branch_len_big_trees = c(collect_branch_len_big_trees, full_aln_tree$edge.length)\n",
    "\n",
    "        # update the diameter of the BigTrees. diameter=max distance between two tips in the tree. The distance is computed using the Patristic method.\n",
    "        collect_diameter_big_trees   = c(collect_diameter_big_trees, max(distTips(full_aln_tree, method = 'patristic')))\n",
    "        \n",
    "        # for each tip in the tree (375) rename it's label according to the mapping of full_aln.code_name. essentially going from C181 -> Human_15\n",
    "        for (j in 1:length(full_aln_tree$tip.label)) { \n",
    "            full_aln_tree$tip.label[j] = as.character(full_aln_codefile$V1[which(full_aln_codefile$V2 == full_aln_tree[[\"tip.label\"]][j])])\n",
    "        }\n",
    "\n",
    "        # Now actually extract and work of subTree level. The subtrees are the paralog trees for a given species. That's why itarate over all species names and the change tps label naming.\n",
    "        for (species in orgs$V1) {\n",
    "\n",
    "            # For each species, the code searches the tip labels in the tree for those that start with the species name (\"^\" and perl = TRUE) and stores the corresponding indices in the tips_to_keep variable.\n",
    "            tips_to_keep = grep(paste0(\"^\", species), full_aln_tree$tip.label, perl = TRUE)\n",
    "\n",
    "            # now that we have the inidces actually extrat the species paralog subtree. Unroot it.\n",
    "            test_subtree = keep.tip(full_aln_tree, tips_to_keep)\n",
    "            test_subtree = unroot(test_subtree)\n",
    "\n",
    "            # rename the tips label again. going from Chicken_15 -> Seq_15. this for all species.\n",
    "            test_subtree$tip.label = gsub(paste0(species, \"_\"), \"Seq_\", test_subtree$tip.label)\n",
    "\n",
    "            # collect the subtrees branch length and the diameter as above.\n",
    "            collect_branch_len = c(collect_branch_len, test_subtree$edge.length)\n",
    "            collect_diameter = c(collect_diameter, max(distTips(test_subtree, method = 'patristic')))\n",
    "\n",
    "            # change class of tree and add it to a vector\n",
    "            test_subtree = as.multiPhylo(test_subtree)\n",
    "            all_species_raxml_subtrees_from_full_aln = c(all_species_raxml_subtrees_from_full_aln, test_subtree)\n",
    "        }\n",
    "        class(all_species_raxml_subtrees_from_full_aln) = \"multiPhylo\"\n",
    "\n",
    "        # read the reference tree (gene family tree) the one set as throuth and on which the sequences were generated along.\n",
    "        ref_tree_file = file.path(run, paste0(basename(m), \".nwk\"))\n",
    "        ref_tree      = read.tree(ref_tree_file, keep.multi=FALSE)\n",
    "\n",
    "        # refernce tree block similar to the for loop above\n",
    "        for (species in orgs$V1) {\n",
    "            \n",
    "            # ectract tips like before and then extract subtree and unroot it. \n",
    "            # this is the refence paralog tree for each species and whithin one family all species have the same paralog tree as far as topology goes, what changes between this species paralog trees is their branch lengths.\n",
    "            tips_to_keep = grep(paste0(\"^\", species), ref_tree$tip.label, perl = TRUE)\n",
    "            test_subtree = keep.tip(ref_tree, tips_to_keep)\n",
    "            test_subtree = unroot(test_subtree)\n",
    "\n",
    "            # get branch length and diameter\n",
    "            collect_branch_len_sub_from_ref = c(collect_branch_len_sub_from_ref, test_subtree$edge.length)\n",
    "            collect_diameter_sub_from_ref   = c(collect_diameter_sub_from_ref, max(distTips(test_subtree, method = 'patristic')))\n",
    "        }\n",
    "\n",
    "        # Since each gene family has a predifined gene tree (ref_tree) composed of a fixed species tree pasted as leaf of the paralog tree. The tree is designed like this by choice. Each gene families varies only in the paralog tree that it has that is unique (15 tips).\n",
    "        # Here we extract the paralog tree and to do so we just need to extract the subtree with all the sequences of omne species. Since each species have the same paralog tree. So we can extract the subtree of whichever species it does not matter.\n",
    "        ref_subtree   = keep.tip(ref_tree,as.character(ref_codefile$V1))\n",
    "\n",
    "        # we then rename the tips of the paralog tree (ref_subtree) to a more generic keyname. Effectively going from rat1 to seq1. \n",
    "        for (j in 1:length(ref_subtree$tip.label)) { \n",
    "            ref_subtree$tip.label[j] = as.character(ref_codefile$V2[which(ref_codefile$V1 == ref_subtree[[\"tip.label\"]][j])])\n",
    "        }\n",
    "\n",
    "        # Unroot the tree and perform aother operation and then save the paralog tree to the specifc gene faimily directory\n",
    "        ref_subtree        = unroot(ref_subtree)\n",
    "        ref_subtree        = list(ref_subtree)\n",
    "        class(ref_subtree) = \"multiPhylo\"\n",
    "\n",
    "        # update on the average branch length with the mean of the true/refernce paralog tree of the given gene family\n",
    "        avg_branch_length = c(avg_branch_length, mean(ref_subtree[[1]]$edge.length))\n",
    "\n",
    "        # put all  BIgTree (big MSA + ML) species subtrees and the reference paralog tree into one single vector so Roubinson-Fould dictance can be computed.\n",
    "        all_trees        = c(ref_subtree, all_species_raxml_subtrees_from_full_aln)\n",
    "        class(all_trees) = \"multiPhylo\"\n",
    "\n",
    "        # compute the RF dictance between 1 refernce paralog tree and all (25) species paralog tress extracted from BigTree.\n",
    "        all_vs_all_rf = RF.dist(all_trees, normalize=TRUE)\n",
    "        plotdata      = data.matrix(all_vs_all_rf)\n",
    "        RF_matrices   = list.append(RF_matrices, plotdata)\n",
    "\n",
    "        # do the average of all pair comparison between the species paralog trees and the reference tree. -c[1] means not consider the first column, because first row first column is RF between reference tree and itself\n",
    "        mean_rf_subtrees_from_full_aln_vs_true_tree = c(mean_rf_subtrees_from_full_aln_vs_true_tree, mean(plotdata[1, -c(1)]))\n",
    "        sd_rf_subtrees_from_full_aln_vs_true_tree   = c(sd_rf_subtrees_from_full_aln_vs_true_tree, sd(plotdata[1, -c(1)]))\n",
    "    }\n",
    "}\n",
    "\n",
    "write.table(mean_rf_subtrees_from_full_aln_vs_true_tree, file = \"tables/avg_RF_between_ref_big_tree_subtrees.tsv\", quote=FALSE, row.names = FALSE, col.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab7ac11a-c926-4fa6-aafd-d3c8b281e7fc",
   "metadata": {},
   "source": [
    "<a id='chapter-6-RF-distance-summaries'></a>\n",
    "\n",
    "## Chapter 6 - RF distance summaries\n",
    "\n",
    "todo add description"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0553ac22-a97a-49f6-adfe-8323e9003fba",
   "metadata": {},
   "source": [
    "<a id='subchapter-6_1-Table-1'></a>\n",
    "\n",
    "### Subchapter 6.1 - Table 1\n",
    "todo add description."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f493767f-1b5e-4251-848d-6ff64f26f4f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 4 × 26</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Method</th><th scope=col>unit_1</th><th scope=col>unit_2</th><th scope=col>unit_3</th><th scope=col>unit_4</th><th scope=col>unit_5</th><th scope=col>unit_6</th><th scope=col>unit_7</th><th scope=col>unit_8</th><th scope=col>unit_9</th><th scope=col>⋯</th><th scope=col>unit_16</th><th scope=col>unit_17</th><th scope=col>unit_18</th><th scope=col>unit_19</th><th scope=col>unit_20</th><th scope=col>unit_21</th><th scope=col>unit_22</th><th scope=col>unit_23</th><th scope=col>unit_24</th><th scope=col>unit_25</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ME_SM</td><td>0.47  sd  0.13</td><td>0.29  sd  0.13</td><td>0.26  sd  0.14</td><td>0.25  sd  0.14</td><td>0.25  sd  0.14</td><td>0.24  sd  0.14</td><td>0.24  sd  0.15</td><td>0.34  sd  0.13</td><td>0.28  sd  0.13</td><td>⋯</td><td>0.21  sd  0.13</td><td>0.2  sd  0.13</td><td>0.19  sd  0.13</td><td>0.19  sd  0.14</td><td>0.42  sd  0.13</td><td>0.27  sd  0.13</td><td>0.23  sd  0.12</td><td>0.22  sd  0.12</td><td>0.21  sd  0.12</td><td>0.2  sd  0.12</td></tr>\n",
       "\t<tr><td>ML_SM</td><td>0.33  sd  0.13</td><td>0.27  sd  0.13</td><td>0.26  sd  0.14</td><td>0.25  sd  0.14</td><td>0.24  sd  0.14</td><td>0.24  sd  0.14</td><td>0.44  sd  0.13</td><td>0.3  sd  0.13 </td><td>0.26  sd  0.13</td><td>⋯</td><td>0.2  sd  0.13 </td><td>0.2  sd  0.13</td><td>0.19  sd  0.14</td><td>0.19  sd  0.14</td><td>0.31  sd  0.13</td><td>0.25  sd  0.12</td><td>0.23  sd  0.12</td><td>0.21  sd  0.12</td><td>0.21  sd  0.12</td><td>0.2  sd  0.13</td></tr>\n",
       "\t<tr><td>ME_ST</td><td>0.37  sd  0.13</td><td>0.28  sd  0.14</td><td>0.26  sd  0.14</td><td>0.25  sd  0.14</td><td>0.24  sd  0.14</td><td>0.24  sd  0.14</td><td>0.47  sd  0.13</td><td>0.32  sd  0.13</td><td>0.27  sd  0.13</td><td>⋯</td><td>0.2  sd  0.13 </td><td>0.2  sd  0.13</td><td>0.19  sd  0.14</td><td>0.19  sd  0.14</td><td>0.34  sd  0.13</td><td>0.26  sd  0.12</td><td>0.23  sd  0.12</td><td>0.21  sd  0.12</td><td>0.21  sd  0.12</td><td>0.2  sd  0.12</td></tr>\n",
       "\t<tr><td>ML_ST</td><td>0.31  sd  0.14</td><td>0.27  sd  0.14</td><td>0.25  sd  0.14</td><td>0.25  sd  0.14</td><td>0.24  sd  0.14</td><td>0.24  sd  0.14</td><td>0.37  sd  0.13</td><td>0.29  sd  0.13</td><td>0.26  sd  0.13</td><td>⋯</td><td>0.2  sd  0.13 </td><td>0.2  sd  0.14</td><td>0.19  sd  0.14</td><td>0.44  sd  0.13</td><td>0.29  sd  0.13</td><td>0.24  sd  0.12</td><td>0.22  sd  0.12</td><td>0.21  sd  0.12</td><td>0.2  sd  0.12 </td><td>0.2  sd  0.12</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 × 26\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " Method & unit\\_1 & unit\\_2 & unit\\_3 & unit\\_4 & unit\\_5 & unit\\_6 & unit\\_7 & unit\\_8 & unit\\_9 & ⋯ & unit\\_16 & unit\\_17 & unit\\_18 & unit\\_19 & unit\\_20 & unit\\_21 & unit\\_22 & unit\\_23 & unit\\_24 & unit\\_25\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t ME\\_SM & 0.47  sd  0.13 & 0.29  sd  0.13 & 0.26  sd  0.14 & 0.25  sd  0.14 & 0.25  sd  0.14 & 0.24  sd  0.14 & 0.24  sd  0.15 & 0.34  sd  0.13 & 0.28  sd  0.13 & ⋯ & 0.21  sd  0.13 & 0.2  sd  0.13 & 0.19  sd  0.13 & 0.19  sd  0.14 & 0.42  sd  0.13 & 0.27  sd  0.13 & 0.23  sd  0.12 & 0.22  sd  0.12 & 0.21  sd  0.12 & 0.2  sd  0.12\\\\\n",
       "\t ML\\_SM & 0.33  sd  0.13 & 0.27  sd  0.13 & 0.26  sd  0.14 & 0.25  sd  0.14 & 0.24  sd  0.14 & 0.24  sd  0.14 & 0.44  sd  0.13 & 0.3  sd  0.13  & 0.26  sd  0.13 & ⋯ & 0.2  sd  0.13  & 0.2  sd  0.13 & 0.19  sd  0.14 & 0.19  sd  0.14 & 0.31  sd  0.13 & 0.25  sd  0.12 & 0.23  sd  0.12 & 0.21  sd  0.12 & 0.21  sd  0.12 & 0.2  sd  0.13\\\\\n",
       "\t ME\\_ST & 0.37  sd  0.13 & 0.28  sd  0.14 & 0.26  sd  0.14 & 0.25  sd  0.14 & 0.24  sd  0.14 & 0.24  sd  0.14 & 0.47  sd  0.13 & 0.32  sd  0.13 & 0.27  sd  0.13 & ⋯ & 0.2  sd  0.13  & 0.2  sd  0.13 & 0.19  sd  0.14 & 0.19  sd  0.14 & 0.34  sd  0.13 & 0.26  sd  0.12 & 0.23  sd  0.12 & 0.21  sd  0.12 & 0.21  sd  0.12 & 0.2  sd  0.12\\\\\n",
       "\t ML\\_ST & 0.31  sd  0.14 & 0.27  sd  0.14 & 0.25  sd  0.14 & 0.25  sd  0.14 & 0.24  sd  0.14 & 0.24  sd  0.14 & 0.37  sd  0.13 & 0.29  sd  0.13 & 0.26  sd  0.13 & ⋯ & 0.2  sd  0.13  & 0.2  sd  0.14 & 0.19  sd  0.14 & 0.44  sd  0.13 & 0.29  sd  0.13 & 0.24  sd  0.12 & 0.22  sd  0.12 & 0.21  sd  0.12 & 0.2  sd  0.12  & 0.2  sd  0.12\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 × 26\n",
       "\n",
       "| Method &lt;chr&gt; | unit_1 &lt;chr&gt; | unit_2 &lt;chr&gt; | unit_3 &lt;chr&gt; | unit_4 &lt;chr&gt; | unit_5 &lt;chr&gt; | unit_6 &lt;chr&gt; | unit_7 &lt;chr&gt; | unit_8 &lt;chr&gt; | unit_9 &lt;chr&gt; | ⋯ ⋯ | unit_16 &lt;chr&gt; | unit_17 &lt;chr&gt; | unit_18 &lt;chr&gt; | unit_19 &lt;chr&gt; | unit_20 &lt;chr&gt; | unit_21 &lt;chr&gt; | unit_22 &lt;chr&gt; | unit_23 &lt;chr&gt; | unit_24 &lt;chr&gt; | unit_25 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| ME_SM | 0.47  sd  0.13 | 0.29  sd  0.13 | 0.26  sd  0.14 | 0.25  sd  0.14 | 0.25  sd  0.14 | 0.24  sd  0.14 | 0.24  sd  0.15 | 0.34  sd  0.13 | 0.28  sd  0.13 | ⋯ | 0.21  sd  0.13 | 0.2  sd  0.13 | 0.19  sd  0.13 | 0.19  sd  0.14 | 0.42  sd  0.13 | 0.27  sd  0.13 | 0.23  sd  0.12 | 0.22  sd  0.12 | 0.21  sd  0.12 | 0.2  sd  0.12 |\n",
       "| ML_SM | 0.33  sd  0.13 | 0.27  sd  0.13 | 0.26  sd  0.14 | 0.25  sd  0.14 | 0.24  sd  0.14 | 0.24  sd  0.14 | 0.44  sd  0.13 | 0.3  sd  0.13  | 0.26  sd  0.13 | ⋯ | 0.2  sd  0.13  | 0.2  sd  0.13 | 0.19  sd  0.14 | 0.19  sd  0.14 | 0.31  sd  0.13 | 0.25  sd  0.12 | 0.23  sd  0.12 | 0.21  sd  0.12 | 0.21  sd  0.12 | 0.2  sd  0.13 |\n",
       "| ME_ST | 0.37  sd  0.13 | 0.28  sd  0.14 | 0.26  sd  0.14 | 0.25  sd  0.14 | 0.24  sd  0.14 | 0.24  sd  0.14 | 0.47  sd  0.13 | 0.32  sd  0.13 | 0.27  sd  0.13 | ⋯ | 0.2  sd  0.13  | 0.2  sd  0.13 | 0.19  sd  0.14 | 0.19  sd  0.14 | 0.34  sd  0.13 | 0.26  sd  0.12 | 0.23  sd  0.12 | 0.21  sd  0.12 | 0.21  sd  0.12 | 0.2  sd  0.12 |\n",
       "| ML_ST | 0.31  sd  0.14 | 0.27  sd  0.14 | 0.25  sd  0.14 | 0.25  sd  0.14 | 0.24  sd  0.14 | 0.24  sd  0.14 | 0.37  sd  0.13 | 0.29  sd  0.13 | 0.26  sd  0.13 | ⋯ | 0.2  sd  0.13  | 0.2  sd  0.14 | 0.19  sd  0.14 | 0.44  sd  0.13 | 0.29  sd  0.13 | 0.24  sd  0.12 | 0.22  sd  0.12 | 0.21  sd  0.12 | 0.2  sd  0.12  | 0.2  sd  0.12 |\n",
       "\n"
      ],
      "text/plain": [
       "  Method unit_1         unit_2         unit_3         unit_4        \n",
       "1 ME_SM  0.47  sd  0.13 0.29  sd  0.13 0.26  sd  0.14 0.25  sd  0.14\n",
       "2 ML_SM  0.33  sd  0.13 0.27  sd  0.13 0.26  sd  0.14 0.25  sd  0.14\n",
       "3 ME_ST  0.37  sd  0.13 0.28  sd  0.14 0.26  sd  0.14 0.25  sd  0.14\n",
       "4 ML_ST  0.31  sd  0.14 0.27  sd  0.14 0.25  sd  0.14 0.25  sd  0.14\n",
       "  unit_5         unit_6         unit_7         unit_8         unit_9         ⋯\n",
       "1 0.25  sd  0.14 0.24  sd  0.14 0.24  sd  0.15 0.34  sd  0.13 0.28  sd  0.13 ⋯\n",
       "2 0.24  sd  0.14 0.24  sd  0.14 0.44  sd  0.13 0.3  sd  0.13  0.26  sd  0.13 ⋯\n",
       "3 0.24  sd  0.14 0.24  sd  0.14 0.47  sd  0.13 0.32  sd  0.13 0.27  sd  0.13 ⋯\n",
       "4 0.24  sd  0.14 0.24  sd  0.14 0.37  sd  0.13 0.29  sd  0.13 0.26  sd  0.13 ⋯\n",
       "  unit_16        unit_17       unit_18        unit_19        unit_20       \n",
       "1 0.21  sd  0.13 0.2  sd  0.13 0.19  sd  0.13 0.19  sd  0.14 0.42  sd  0.13\n",
       "2 0.2  sd  0.13  0.2  sd  0.13 0.19  sd  0.14 0.19  sd  0.14 0.31  sd  0.13\n",
       "3 0.2  sd  0.13  0.2  sd  0.13 0.19  sd  0.14 0.19  sd  0.14 0.34  sd  0.13\n",
       "4 0.2  sd  0.13  0.2  sd  0.14 0.19  sd  0.14 0.44  sd  0.13 0.29  sd  0.13\n",
       "  unit_21        unit_22        unit_23        unit_24        unit_25      \n",
       "1 0.27  sd  0.13 0.23  sd  0.12 0.22  sd  0.12 0.21  sd  0.12 0.2  sd  0.12\n",
       "2 0.25  sd  0.12 0.23  sd  0.12 0.21  sd  0.12 0.21  sd  0.12 0.2  sd  0.13\n",
       "3 0.26  sd  0.12 0.23  sd  0.12 0.21  sd  0.12 0.21  sd  0.12 0.2  sd  0.12\n",
       "4 0.24  sd  0.12 0.22  sd  0.12 0.21  sd  0.12 0.2  sd  0.12  0.2  sd  0.12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Method</th><th scope=col>Unit.25</th><th scope=col>Unit.1</th><th scope=col>Difference</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ME_SM  </td><td>0.2  sd  0.12 </td><td>0.47  sd  0.13</td><td>0.27</td></tr>\n",
       "\t<tr><td>ML_SM  </td><td>0.2  sd  0.13 </td><td>0.33  sd  0.13</td><td>0.13</td></tr>\n",
       "\t<tr><td>ME_ST  </td><td>0.2  sd  0.12 </td><td>0.37  sd  0.13</td><td>0.17</td></tr>\n",
       "\t<tr><td>ML_ST  </td><td>0.2  sd  0.12 </td><td>0.31  sd  0.14</td><td>0.11</td></tr>\n",
       "\t<tr><td>BigTree</td><td>0.18  sd  0.15</td><td>Na            </td><td>Na  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 × 4\n",
       "\\begin{tabular}{llll}\n",
       " Method & Unit.25 & Unit.1 & Difference\\\\\n",
       " <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t ME\\_SM   & 0.2  sd  0.12  & 0.47  sd  0.13 & 0.27\\\\\n",
       "\t ML\\_SM   & 0.2  sd  0.13  & 0.33  sd  0.13 & 0.13\\\\\n",
       "\t ME\\_ST   & 0.2  sd  0.12  & 0.37  sd  0.13 & 0.17\\\\\n",
       "\t ML\\_ST   & 0.2  sd  0.12  & 0.31  sd  0.14 & 0.11\\\\\n",
       "\t BigTree & 0.18  sd  0.15 & Na             & Na  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 × 4\n",
       "\n",
       "| Method &lt;chr&gt; | Unit.25 &lt;chr&gt; | Unit.1 &lt;chr&gt; | Difference &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| ME_SM   | 0.2  sd  0.12  | 0.47  sd  0.13 | 0.27 |\n",
       "| ML_SM   | 0.2  sd  0.13  | 0.33  sd  0.13 | 0.13 |\n",
       "| ME_ST   | 0.2  sd  0.12  | 0.37  sd  0.13 | 0.17 |\n",
       "| ML_ST   | 0.2  sd  0.12  | 0.31  sd  0.14 | 0.11 |\n",
       "| BigTree | 0.18  sd  0.15 | Na             | Na   |\n",
       "\n"
      ],
      "text/plain": [
       "  Method  Unit.25        Unit.1         Difference\n",
       "1 ME_SM   0.2  sd  0.12  0.47  sd  0.13 0.27      \n",
       "2 ML_SM   0.2  sd  0.13  0.33  sd  0.13 0.13      \n",
       "3 ME_ST   0.2  sd  0.12  0.37  sd  0.13 0.17      \n",
       "4 ML_ST   0.2  sd  0.12  0.31  sd  0.14 0.11      \n",
       "5 BigTree 0.18  sd  0.15 Na             Na        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read the table compiled in chapter 4. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df_new = read.table(\"tables/avg_RF_between_ref_supermatrix_and_supertree.tsv\", header=T)\n",
    "\n",
    "# do the mean and sd of the column. effectively getting a vector of size 100 filed with averages of averages and with ids like SM_ML_unit9\n",
    "rf_df_new_means = colMeans(df_new)\n",
    "sd_df_new_means = apply(df_new, 2, sd)\n",
    "\n",
    "# Initialize empty list to hold the extracted method names and unit names values\n",
    "methods = c(\"ME_SM\", \"ML_SM\", \"ME_ST\", \"ML_ST\")\n",
    "units   = 25\n",
    "\n",
    "# Create an empty dataframe with row names for methods and columns for unit_1 to unit_25\n",
    "avg_rf_df_method_unit           = data.frame(Method = methods, matrix(ncol = units, nrow = 4))\n",
    "colnames(avg_rf_df_method_unit) = c(\"Method\", paste0(\"unit_\", 1:units))\n",
    "\n",
    "# Fill the dataframe using the keys of the named vector of rf_df_new_means, these names are the same for the vector sd_df_new_means too.\n",
    "for (key in names(rf_df_new_means)) {\n",
    "\n",
    "  # Split the key into parts (method and unit)\n",
    "  parts  = strsplit(key, \"_unit_\")[[1]]\n",
    "  method = parts[1]\n",
    "  unit   = as.numeric(parts[2])\n",
    "\n",
    "  # Place the value in the correct row and column of the dataframe adding the standard deviation to it.\n",
    "  avg_rf_df_method_unit[avg_rf_df_method_unit$Method == method, paste0(\"unit_\", unit)] = paste0(round(rf_df_new_means[key], digit=2), \"  sd  \", round(sd_df_new_means[key], digit=2))\n",
    "}\n",
    "\n",
    "# create the actual table with only unit1, unit25 difference values from above df\n",
    "diff   = c(as.numeric(sub(\" sd.*\", \"\", avg_rf_df_method_unit$unit_1)) - as.numeric(sub(\" sd.*\", \"\", avg_rf_df_method_unit$unit_25)))\n",
    "table1 = data.frame(Method = avg_rf_df_method_unit$Method,\n",
    "                    \"Unit 25\"    = avg_rf_df_method_unit$unit_25,\n",
    "                    \"Unit 1\"     = avg_rf_df_method_unit$unit_1,\n",
    "                    \"Difference\" = diff )\n",
    "\n",
    "# read the table compiled in chapter 5. So this cell does not depend on that one to run, it just needs the table file.\n",
    "df_bigtree_new = read.table(\"tables/avg_RF_between_ref_big_tree_subtrees.tsv\", header=F)\n",
    "\n",
    "# make average of all family BigTree RF distances to their respective paralog tree. going from 500 values to 1.\n",
    "rf_df_bigtree_new_mean = colMeans(df_bigtree_new)\n",
    "rf_df_bigtree_new_sd   = apply(df_bigtree_new, 2, sd)\n",
    "\n",
    "# add the above two values to table1\n",
    "bigtree_row = c(\"BigTree\", paste0(round(rf_df_bigtree_new_mean, digits = 2), \"  sd  \", round(rf_df_bigtree_new_sd, digits = 2)), \"Na\", \"Na\")\n",
    "table1      = rbind(table1, bigtree_row) \n",
    "\n",
    "avg_rf_df_method_unit\n",
    "\n",
    "table1\n",
    "#write.table(table1, file=\"Table1_average_RF.tsv\",quote=FALSE, sep=\"\\t\",row.names=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfa19d80-b20f-4a24-af34-ea65b39aa0f2",
   "metadata": {},
   "source": [
    "<a id=\"chapter-7-Running-times\"></a>\n",
    "\n",
    "## Chapter 7 - Running times\n",
    "\n",
    "TODO add description "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "062ad02a-baca-4fce-b3ee-e8ea3b3bb039",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\"\n"
     ]
    }
   ],
   "source": [
    "# read the trace files for running times\n",
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    # each main.nf output run has a trace file repoprting various information on the run itself\n",
    "    tracefile_path = file.path(run, \"trace_cpu_time.txt\")\n",
    "    tracefile      = read.table(tracefile_path, header = TRUE)\n",
    "    \n",
    "    # here we parse the trace file to retain only the processes that we want to take the running time (realtime) from\n",
    "    tracefile_mod  = tracefile[grep(\"simulated_data:run_phylo_ML_supermatrix_aln_sim|simulated_data:only_concatenate_aln_sim|simulated_data:run_phylo_ML_full_aln_sim\", tracefile$native_id), c(\"name\", \"native_id\", \"realtime\")]\n",
    "    \n",
    "    # transforming the above table to one that have the process names (simulated_data:run_phylo_ML_supermatrix_aln_sim ecc..) as column names, \n",
    "    # the rows will be still the family names (working as ID) and the values in each column will be the running times associated to the column name (process name) for that given family.\n",
    "    tracefile_mod_wide = spread(tracefile_mod, native_id, realtime)\n",
    "\n",
    "    # creating a new last column called Supermatrix_ML that has aas values the running times of concatenation plus ML tree computation (on concatenated MSA) \n",
    "    tracefile_mod_wide$Supermatrix_ML = as.numeric(tracefile_mod_wide$`simulated_data:only_concatenate_aln_sim`) + as.numeric(tracefile_mod_wide$`simulated_data:run_phylo_ML_supermatrix_aln_sim`)\n",
    "\n",
    "    # retain only the family name, BIgTree ML running time and SM + ML running time\n",
    "    tracefile_mod_wide = tracefile_mod_wide[,c(1,3,5)]\n",
    "\n",
    "    # call the second column name Big_Tree and set their value to numeric type\n",
    "    colnames(tracefile_mod_wide)[2] = \"Big_Tree\"\n",
    "    tracefile_mod_wide$Big_Tree = as.numeric(tracefile_mod_wide$Big_Tree)\n",
    "\n",
    "    # adding a first column that groups families by the run of the main.nf\n",
    "    tracefile_mod_wide$family_group = rep(basename(run), 100)\n",
    "\n",
    "    # add tables toghether raw-wise\n",
    "    if(run == run_dirs[1]) {\n",
    "        fin_tracefile_mod_wide = tracefile_mod_wide\n",
    "    } else {\n",
    "        fin_tracefile_mod_wide = rbind(fin_tracefile_mod_wide, tracefile_mod_wide)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d6a00cd4-31ce-440c-8528-94805af06047",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>Big_Tree</dt><dd>638.12</dd><dt>Supermatrix_ML</dt><dd>67.76</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[Big\\textbackslash{}\\_Tree] 638.12\n",
       "\\item[Supermatrix\\textbackslash{}\\_ML] 67.76\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "Big_Tree\n",
       ":   638.12Supermatrix_ML\n",
       ":   67.76\n",
       "\n"
      ],
      "text/plain": [
       "      Big_Tree Supermatrix_ML \n",
       "        638.12          67.76 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "588.3"
      ],
      "text/latex": [
       "588.3"
      ],
      "text/markdown": [
       "588.3"
      ],
      "text/plain": [
       "[1] 588.3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "63.85"
      ],
      "text/latex": [
       "63.85"
      ],
      "text/markdown": [
       "63.85"
      ],
      "text/plain": [
       "[1] 63.85"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# TODO make nicer and update with whatever is changed above\n",
    "\n",
    "# make the means of all (500) family running time for the BigTree and ML + SM. Bring them to seconds (/1000) and round the digits to 2\n",
    "round(colMeans(fin_tracefile_mod_wide[, c(2, 3)])/1000,digits = 2)\n",
    "\n",
    "# compute the median for the BigTree approach and the  ML + SM.\n",
    "round(median(fin_tracefile_mod_wide$Big_Tree)/1000,digits = 2)\n",
    "round(median(fin_tracefile_mod_wide$Supermatrix_ML)/1000,digits = 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b79de7a-f8b9-4dc7-8fc6-232cf86be231",
   "metadata": {},
   "source": [
    "<a id='subchapter-7_1-Figure-2'></a>\n",
    "\n",
    "### Subchapter 7.1 - Figure 2\n",
    "\n",
    "TODO add chapter link and description if necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ba993fba-7f4b-48aa-9b03-775093172898",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO make this into the section of custom functions.\n",
    "# Function to assign new values based on a mapping provided by the user\n",
    "assign_mapped_values <- function(df, column_name, mapping, new_col_name) {\n",
    "  # Check if mapping is a named vector or a list\n",
    "  if (!is.vector(mapping) || is.null(names(mapping))) {\n",
    "    stop(\"Mapping must be a named vector or list.\")\n",
    "  }\n",
    "  \n",
    "  # Get the unique values from the specified column\n",
    "  unique_values = unique(df[[column_name]])\n",
    "  \n",
    "  # Check if the mapping contains all unique values\n",
    "  if (!all(unique_values %in% names(mapping))) {\n",
    "    stop(\"Mapping must include all unique values from the specified column.\")\n",
    "  }\n",
    "  \n",
    "  # Initialize a vector to store the assigned values\n",
    "  assigned_values = vector(\"character\", nrow(df))\n",
    "  \n",
    "  # Assign values based on the mapping\n",
    "  for (i in seq_along(unique_values)) {\n",
    "    assigned_values[df[[column_name]] == unique_values[i]] <- mapping[names(mapping) == unique_values[i]]\n",
    "  }\n",
    "  \n",
    "  # Add the assigned values to the original dataframe\n",
    "  df[[new_col_name]] = assigned_values\n",
    "  \n",
    "  return(df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4568c025-fffa-45df-bfff-086d67ad2a53",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using name, family_group, length as id variables\n",
      "\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method (SM + ML and Bigtree) divided per group (by sequence length)\n",
    "#\n",
    "\n",
    "# assign to each family_group a mapping values  (as new column). basically insert the length of the sequences in each family as an additional column in the dataframe.\n",
    "# Remember each family in the same group have the same length of sequence generated, the MSA though might differ in length due to gaps.\n",
    "mapping = c(\"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\" = \"group1\\nlen = 32\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\" = \"group2\\nlen = 46\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\" = \"group3\\nlen = 50\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\" = \"group4\\nlen = 60\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\" = \"group5\\nlen = 82\")\n",
    "fin_tracefile_mod_reassigned = assign_mapped_values(fin_tracefile_mod_wide, \"family_group\", mapping, \"length\")\n",
    "\n",
    "# reorder the columns to have the numerical values at the end\n",
    "fin_tracefile_mod_reassigned = fin_tracefile_mod_reassigned[, c(\"name\", \"family_group\", \"length\", \"Big_Tree\", \"Supermatrix_ML\")]\n",
    "\n",
    "# It takes a table of all families as rows (500) and it goes to a table of 1000 (500 BigTree + 500 SM + ML) where the running times instead of being separated into two different coluns now they are in the same one.\n",
    "# name + group columns will work as unique ID.  group is the directory name of the main.nf run/s, while name is the name of the family. \n",
    "fin_tracefile_mod_melted           = melt(fin_tracefile_mod_reassigned)\n",
    "\n",
    "\n",
    "# transform running times from millisecond to seconds. And rename the column names.\n",
    "fin_tracefile_mod_melted$value     = fin_tracefile_mod_melted$value / 1000\n",
    "colnames(fin_tracefile_mod_melted) = c(\"name\", \"family_group\", \"length\", \"Method\", \"Time\")\n",
    "\n",
    "# create a table with 5 pairs of boxplot. Group1, 2.. ecc.. each group has 100 families and two running time (realtime) mesuraments for the BigTree and SM + ML approaches.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel.\n",
    "p = ggplot(fin_tracefile_mod_melted, aes(x=length, y=Time, fill=Method)) +\n",
    "    geom_boxplot(show.legend = TRUE) +\n",
    "    theme_light() +\n",
    "    labs(x = \"sequence length\",\n",
    "       y = \"CPU time (sec)\",\n",
    "       fill = \"Method\")\n",
    "ggsave(filename=\"figures/CPU_time_per_group_BigTree_SM_ML.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83903c34-3f32-431b-abc5-156cb2abeaf8",
   "metadata": {},
   "source": [
    "### Suplementary Fig todo put number of supplementary figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "81650244-8771-4d70-a7c6-57d5218f928d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method ratio (Bigtree / SM + ML ) boxplot of each family ratio by group (x axis groups y ratio Big/SM) \n",
    "# \n",
    "\n",
    "# start by taking the already present info for the families and their grouping from fig 2 cell and compute ratio Bigtree / SM + ML\n",
    "# Create a new column with the division of Big_Tree by Supermatrix_ML running times for each family (500)\n",
    "# and then remove the original columns Big_Tree Supermatrix_ML\n",
    "fin_tracefile_mod_ratio = fin_tracefile_mod_reassigned %>%\n",
    "  mutate(fold_ratio = Big_Tree / Supermatrix_ML) %>%\n",
    "  select(-Big_Tree, -Supermatrix_ML)\n",
    "\n",
    "# create a table with 5 boxplots. Group1, 2.. ecc.. each group has 100 families and  the ratio of two running time (realtime) mesuraments for the BigTree and SM + ML approaches.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel.\n",
    "p = ggplot(fin_tracefile_mod_ratio, aes(x=length, y=fold_ratio)) +\n",
    "    geom_boxplot(show.legend = FALSE) +\n",
    "    theme_light() +\n",
    "    labs(x = \"sequence length\",\n",
    "       y = \"CPU time fold ratio (BigTree / Supermatrix_ML)\")\n",
    "ggsave(filename=\"figures/CPU_time_per_group_BigTree_SM_ML_ratio.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "e1ff4f0e-ec70-4989-89c7-f007d09969c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using name, family_group as id variables\n",
      "\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method (SM + ML and Bigtree)   OLD VERSION\n",
    "#\n",
    "\n",
    "# It takes a table of all families as rows (500) and it goes to a table of 1000 (500 BigTree + 500 SM + ML) where the running times instead of being separated into two different coluns now they are in the same one.\n",
    "# name + group columns will work as unique ID.  group is the directory name of the main.nf run/s, while name is the name of the family. \n",
    "fin_tracefile_mod_wide_melted           = melt(fin_tracefile_mod_wide)\n",
    "\n",
    "# transform running times from millisecond to seconds. And rename the column names.\n",
    "fin_tracefile_mod_wide_melted$value     = fin_tracefile_mod_wide_melted$value / 1000\n",
    "colnames(fin_tracefile_mod_wide_melted) = c(\"name\", \"family_group\", \"Method\", \"Time\")\n",
    "\n",
    "# crate a table with two boxplots (x axis) with Method columns as values and on the y the running times.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel. \n",
    "p = ggplot(fin_tracefile_mod_wide_melted,aes(x=Method, y=Time, fill=Method)) +\n",
    "    geom_boxplot(show.legend = FALSE) +\n",
    "    theme_light() +\n",
    "    ylab(\"CPU time (sec)\") + xlab(\"Method\")\n",
    "ggsave(filename=\"figures/CPU_time_per_method.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92bfb310-9e7c-43cb-8a9d-dce71d69920d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "<br><br><br>\n",
    "# USEFULL STUFF BUT NOT INCLUDED IN PAPER\n",
    "<br><br><br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6df071e6-9d31-4158-ae52-3384e9e35a6b",
   "metadata": {},
   "source": [
    "### OLD version all families toghther"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53a77a3f-22e3-4136-b883-32db83aaa230",
   "metadata": {},
   "source": [
    "### TODO make the following better or remove it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "a73a53a9-f1a1-40c9-84e9-09965fe17f26",
   "metadata": {},
   "outputs": [],
   "source": [
    "# update the trace matrix with the length of the input msa\n",
    "library(\"Biostrings\") \n",
    "fin_tracefile_mod_with_len = fin_tracefile_mod_wide\n",
    "\n",
    "# for each family go and read the input MSA (input of main.nf) to get length of MSA (number of columns).\n",
    "# Iterate over each row of the dataframe\n",
    "for (i in 1:nrow(fin_tracefile_mod_with_len)) {\n",
    "    # Construct the full file path using group (directory) and name (filename) to read the input msa\n",
    "    # TODO make it not depend on the rerun_bla_bla direcrtory\n",
    "    input_msa_path = file.path(paste0(parent_of_all_dir, \"rerun_25_species_corrected_branches\"), fin_tracefile_mod_with_len$family_group[i], paste0(gsub(\"[()]\", \"\", fin_tracefile_mod_with_len$name[i]), \".ma\"))\n",
    "\n",
    "    # read the msa in fasta format and get the length of the first sequence (aka length of the alignment)\n",
    "    msa_file = readAAStringSet(input_msa_path)\n",
    "    msa_len  = length(msa_file[[1]])\n",
    "    \n",
    "    # add the length to the table \n",
    "    fin_tracefile_mod_with_len$length[i] = msa_len\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1b9d82dd-7856-44ec-99f5-60086ee8b8d1",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error: object 'fin_tracefile_mod_with_len' not found\n",
     "output_type": "error",
     "traceback": [
      "Error: object 'fin_tracefile_mod_with_len' not found\nTraceback:\n",
      "1. select(., -Big_Tree, -Supermatrix_ML)",
      "2. mutate(., fold_ratio = Big_Tree/Supermatrix_ML)",
      "3. .handleSimpleError(function (cnd) \n . {\n .     watcher$capture_plot_and_output()\n .     cnd <- sanitize_call(cnd)\n .     watcher$push(cnd)\n .     switch(on_error, continue = invokeRestart(\"eval_continue\"), \n .         stop = invokeRestart(\"eval_stop\"), error = invokeRestart(\"eval_error\", \n .             cnd))\n . }, \"object 'fin_tracefile_mod_with_len' not found\", base::quote(eval(expr, \n .     envir)))"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method (SM + ML and Bigtree) ratio plot of each family (x axis number columns BigMSA) \n",
    "# \n",
    "\n",
    "# Create a new column with the division of Big_Tree by Supermatrix_ML running times for each family (500)\n",
    "# and then remove the original columns Big_Tree Supermatrix_ML\n",
    "fin_tracefile_mod_ratio = fin_tracefile_mod_with_len %>%\n",
    "  mutate(fold_ratio = Big_Tree / Supermatrix_ML) %>%\n",
    "  select(-Big_Tree, -Supermatrix_ML)\n",
    "\n",
    "# create a table with on x axis the length of the MSA (n columns)  and on y the fold ratio (BigTree / SM + ML) of running times. The plot will have now 500 points.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel.\n",
    "p = ggplot(fin_tracefile_mod_ratio, aes(x=length, y=fold_ratio)) +\n",
    "    geom_point(color = 'blue', size = 2, alpha = 0.6) +  # Add points\n",
    "    theme_light() +\n",
    "    labs(x = \"MSA length (n# columns)\",\n",
    "       y = \"fold ratio (BigTree / Supermatrix_ML)\")\n",
    "ggsave(filename=\"figures/CPU_time_scatterplot_msa_len_BigTree_SM_ML.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1a03b1b-cfef-4abf-a0f0-d0fd195c15bf",
   "metadata": {},
   "source": [
    "### TODO the below was just a 10 familoy trial make it go away "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "873d9c1c-ab54-40b6-aed9-90caa7d43601",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO make this into the section of custom functions.\n",
    "# Function to convert to milliseconds\n",
    "convert_to_ms <- function(time_str) {\n",
    "    # Extract components using regular expressions\n",
    "    millis = \"0\"\n",
    "    seconds = \"0\"\n",
    "    minutes = \"0\"\n",
    "    if (str_detect(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=ms)\")) {\n",
    "        millis <- str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=ms)\")\n",
    "    } else {\n",
    "        seconds <- str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=s)\")\n",
    "        minutes <- str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=m)\")\n",
    "    }\n",
    "    \n",
    "    # Convert each component to milliseconds (coercing to numeric where needed)\n",
    "    seconds_ms <- as.numeric(seconds) * 1000\n",
    "    minutes_ms <- as.numeric(minutes) * 60 * 1000\n",
    "    millis_ms <- as.numeric(millis)\n",
    "    \n",
    "    # Sum all components and handle NAs by treating them as 0\n",
    "    total_ms <- sum(c(seconds_ms, minutes_ms, millis_ms), na.rm = TRUE)\n",
    "    return(total_ms)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "9a84bc61-7c7b-4805-a7dc-47071e51fc39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"results\"\n"
     ]
    }
   ],
   "source": [
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    # each main.nf output run has a trace file repoprting various information on the run itself\n",
    "    tracefile_path = file.path(run, \"pipeline_info/execution_trace_2024-09-27_17-39-02.txt\")\n",
    "    tracefile      = read.table(tracefile_path, sep = \"\\t\", header = TRUE)\n",
    "\n",
    "    # here we parse the trace file to retain only the processes that we want to take the running time (realtime) from\n",
    "    tracefile_mod = tracefile[grep(\"simulated_data:run_phylo_ML_supermatrix_aln_sim|simulated_data:only_concatenate_aln_sim|simulated_data:run_phylo_ML_full_aln_sim\", tracefile$name), c(\"name\", \"realtime\")]\n",
    "\n",
    "    # Use strsplit to split the 'name' column by whitespace\n",
    "    split_name    = strsplit(as.character(tracefile_mod$name), \" \")\n",
    "\n",
    "    # Create new columns\n",
    "    tracefile_mod$name       = sapply(split_name, `[`, 2) # First part \n",
    "    tracefile_mod$native_id  = sapply(split_name, `[`, 1) # Second part \n",
    "\n",
    "    # Reorder columns so 'native_id' becomes the second column\n",
    "    tracefile_mod = tracefile_mod[, c(\"name\", \"native_id\", \"realtime\")]\n",
    "\n",
    "    # trasform from human readable time to millisecond time\n",
    "    tracefile_mod$realtime = sapply(tracefile_mod$realtime, convert_to_ms)\n",
    "\n",
    "    # transforming the above table to one that have the process names (simulated_data:run_phylo_ML_supermatrix_aln_sim ecc..) as column names, \n",
    "    # the rows will be still the family names (working as ID) and the values in each column will be the running times associated to the column name (process name) for that given family.\n",
    "    tracefile_mod_wide = spread(tracefile_mod, native_id, realtime)\n",
    "    \n",
    "    # creating a new last column called Supermatrix_ML that has aas values the running times of concatenation plus ML tree computation (on concatenated MSA) \n",
    "    tracefile_mod_wide$Supermatrix_ML = as.numeric(tracefile_mod_wide$`simulated_data:only_concatenate_aln_sim`) + as.numeric(tracefile_mod_wide$`simulated_data:run_phylo_ML_supermatrix_aln_sim`)\n",
    "\n",
    "    # retain only the family name, BIgTree ML running time and SM + ML running time\n",
    "    tracefile_mod_wide = tracefile_mod_wide[,c(1,3,5)]\n",
    "\n",
    "    # call the second column name Big_Tree and set their value to numeric type\n",
    "    colnames(tracefile_mod_wide)[2] = \"Big_Tree\"\n",
    "    tracefile_mod_wide$Big_Tree = as.numeric(tracefile_mod_wide$Big_Tree)\n",
    "\n",
    "    # adding a first column that groups families by the run of the main.nf\n",
    "    tracefile_mod_wide$group = rep(basename(run), 10)\n",
    "\n",
    "    # add tables toghether raw-wise\n",
    "    if(run == run_dirs[1]) {\n",
    "        fin_tracefile_mod_wide = tracefile_mod_wide\n",
    "    } else {\n",
    "        fin_tracefile_mod_wide = rbind(fin_tracefile_mod_wide, tracefile_mod_wide)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "ee66b8ad-a4df-4ad9-bdf2-7f8d7a13c8c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>Big_Tree</dt><dd>195.4</dd><dt>Supermatrix_ML</dt><dd>23.47</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[Big\\textbackslash{}\\_Tree] 195.4\n",
       "\\item[Supermatrix\\textbackslash{}\\_ML] 23.47\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "Big_Tree\n",
       ":   195.4Supermatrix_ML\n",
       ":   23.47\n",
       "\n"
      ],
      "text/plain": [
       "      Big_Tree Supermatrix_ML \n",
       "        195.40          23.47 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "196"
      ],
      "text/latex": [
       "196"
      ],
      "text/markdown": [
       "196"
      ],
      "text/plain": [
       "[1] 196"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "21.87"
      ],
      "text/latex": [
       "21.87"
      ],
      "text/markdown": [
       "21.87"
      ],
      "text/plain": [
       "[1] 21.87"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# make the means of all (500) family running time for the BigTree and ML + SM. Bring them to seconds (/1000) and round the digits to 2\n",
    "round(colMeans(fin_tracefile_mod_wide[, c(2, 3)])/1000,digits = 2)\n",
    "\n",
    "# compute the median for the BigTree approach and the  ML + SM.\n",
    "round(median(fin_tracefile_mod_wide$Big_Tree)/1000,digits = 2)\n",
    "round(median(fin_tracefile_mod_wide$Supermatrix_ML)/1000,digits = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "ed1275aa-e0bd-4628-bab2-5eb5a54271ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using name, group as id variables\n",
      "\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "\n",
    "fin_tracefile_mod_wide_melted           = melt(fin_tracefile_mod_wide)\n",
    "\n",
    "fin_tracefile_mod_wide_melted$value     = fin_tracefile_mod_wide_melted$value / 1000\n",
    "colnames(fin_tracefile_mod_wide_melted) = c(\"name\",\"group\",\"Method\",\"Time\")\n",
    "\n",
    "p = ggplot(fin_tracefile_mod_wide_melted,aes(x=Method,y=Time,fill=Method)) +\n",
    "    geom_boxplot(show.legend = FALSE) +\n",
    "    theme_light() +\n",
    "    ylab(\"CPU time (sec)\") + xlab(\"Method\")\n",
    "ggsave(filename=\"CPU_time_per_method_10_fam.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31c3465f-5bf7-465b-bf62-e004f9441080",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
