{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8d6c408c-3b82-4cbb-b94a-cb9728c78fe6",
   "metadata": {},
   "source": [
    "# Analysis of simulated data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "606b5d7a-8d48-4caf-8c72-e89dce7c1208",
   "metadata": {},
   "source": [
    "\n",
    "TODO add here a brief description of your project or analysis. Explain the purpose, goals, or context of the notebook here.\n",
    "\n",
    "## Table of Contents\n",
    "- [1. Chapter 1 - Introduction](#chapter-1-introduction)\n",
    "- [2. Chapter 2 - R packages](#chapter-2-R-packages)\n",
    "- [3. Chapter 3 - Input specification](#chapter-3-Input-specification)\n",
    "- [4. Chapter 4 - Supermatrix and Supertree trees VS true topologies](#chapter-4-Supermatrix-and-Supertree-trees-VS-true-topologies)\n",
    "- [5. Chapter 5 - Running times](#chapter-5-Running-times)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d203093e-da42-45a7-8e98-853d4403630a",
   "metadata": {},
   "source": [
    "<a id=\"chapter-1-introduction\"></a>\n",
    "\n",
    "## Chapter 1 - Introduction\n",
    "\n",
    "TODO add description of notebook and how to use it.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3912a30d-42ec-4e90-99e6-8bacfeb50512",
   "metadata": {},
   "source": [
    "<a id=\"chapter-2-R-packages\"></a>\n",
    "\n",
    "## Chapter 2 - R packages\n",
    "\n",
    "Here is the list of packages needed for this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "67058cec-ea5e-4ce3-be29-faf428919086",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"phangorn\")\n",
    "library(\"ggplot2\")\n",
    "library(\"reshape2\")\n",
    "library(\"geiger\")\n",
    "library(\"adephylo\")\n",
    "library(\"phytools\")\n",
    "library(\"rlist\")\n",
    "library(\"plotrix\")\n",
    "library(\"plyr\")\n",
    "library(\"dplyr\")\n",
    "library(\"hrbrthemes\")\n",
    "library(\"rstatix\")\n",
    "library(\"gplots\")\n",
    "library(\"viridis\")\n",
    "library('gtools')\n",
    "library(gridExtra)\n",
    "library(grid)\n",
    "library(\"ggforce\")\n",
    "library(\"scales\")\n",
    "library(\"Biostrings\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b609768d-d957-4b11-83ce-c27152a96fb2",
   "metadata": {},
   "source": [
    "TODO make a chapter for aqll custom functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20bb020e-4f1d-4338-ac5a-abecc8a06121",
   "metadata": {},
   "source": [
    "<a id=\"chapter-3-Input-specification\"></a>\n",
    "\n",
    "## Chapter 3 - Input specification\n",
    "\n",
    "Change the parent_of_all_dir variable to give the location of the input. All input files are expected to be under it. Otherwise change the single input location to you liking.<br>\n",
    "All input files or directory paths are listed here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "04187777-5e20-43b3-9f39-62523fc0fc0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the location of the directory that contains all inputs for analyses \n",
    "parent_of_all_dir = \"/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/\"\n",
    "\n",
    "# The directory (one or more) that contain all the results from the main.nf pipeline run. It can be a glob path.\n",
    "#dir_containing_all_pipe_results = \"../results\"\n",
    "dir_containing_all_pipe_results = paste0(parent_of_all_dir, \"rerun_25_species_corrected_branches/rerun_sim_b_factor*\")\n",
    "\n",
    "#  TODO describe what this file does\n",
    "reference_tree_codefile = file.path(parent_of_all_dir, \"ref_tree.code\")\n",
    "\n",
    "# get the species/organism list of names used by the main.nf pipeline\n",
    "species_name_file = file.path(parent_of_all_dir, \"orthologs_org_ids_to_concatenate\")\n",
    "\n",
    "\n",
    "# each run of main.nf has one output directory. here we get that or multiple runs outputs directories.\n",
    "run_dirs = Sys.glob(dir_containing_all_pipe_results)\n",
    "run_dirs = run_dirs[file.info(run_dirs)$isdir]\n",
    "\n",
    "# Create the 'figures' directory if it doesn't exist. Created where this notebook is executed.\n",
    "if (!dir.exists(\"figures\")) {\n",
    "    dir.create(\"figures\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "566da14f-4b2d-4610-9a9a-6f892dfc9964",
   "metadata": {},
   "source": [
    "<a id=\"chapter-4-Supermatrix-and-Supertree-trees-VS-true-topologies\"></a>\n",
    "\n",
    "## Chapter 4 - Supermatrix and Supertree trees VS true topologies\n",
    "\n",
    "TODO add description "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "c707c5c4-51d2-4b25-a6cc-b8a9ccb9e0aa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \".nextflow\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in file(file, \"r\"):\n",
      "“cannot open file '/users/cn/abaltzis/projects/1_paralogs/simulated_datasets/simulated_sequences/rerun_25_species_corrected_branches/rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species/.nextflow/results/rerun/all_units.nwk': No such file or directory”\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in file(file, \"r\"): cannot open the connection\n",
     "output_type": "error",
     "traceback": [
      "Error in file(file, \"r\"): cannot open the connection\nTraceback:\n",
      "1. scan(file = file, what = \"\", sep = \"\\n\", quiet = TRUE, skip = skip, \n .     comment.char = comment.char, ...)",
      "2. file(file, \"r\")",
      "3. .handleSimpleError(function (cnd) \n . {\n .     watcher$capture_plot_and_output()\n .     cnd <- sanitize_call(cnd)\n .     watcher$push(cnd)\n .     switch(on_error, continue = invokeRestart(\"eval_continue\"), \n .         stop = invokeRestart(\"eval_stop\"), error = invokeRestart(\"eval_error\", \n .             cnd))\n . }, \"cannot open the connection\", base::quote(file(file, \"r\")))"
     ]
    }
   ],
   "source": [
    "RF_avg_list = list()\n",
    "RF_se_list = list()\n",
    "RF_avg_norm_list = list()\n",
    "collect_branch_len_supermatrix_ml = c()\n",
    "collect_diameter_supermatrix_ml = c()\n",
    "\n",
    "# each run of main.nf has one output directory. here we get that or multiple runs outputs directories.\n",
    "run_dirs = Sys.glob(dir_containing_all_pipe_results)\n",
    "run_dirs = run_dirs[file.info(run_dirs)$isdir]\n",
    "\n",
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    ref_codefile = read.table(file = reference_tree_codefile )\n",
    "    \n",
    "    # get al the family subdirectories in each maqin.nf output run dir\n",
    "    fam_dirs = grep(\"avg_branchlen*\", list.dirs(run, recursive = FALSE), value = TRUE)\n",
    "    count_fam = 0\n",
    "\n",
    "    # this will contain the mean branch length for all true/reference paralog trees for each gene family\n",
    "    avg_branch_length = c()\n",
    "\n",
    "    # read the names of the ortholog species.\n",
    "    orgs = read.table(file = species_name_file)\n",
    "\n",
    "    # get all units trees for all combination of methods \n",
    "    for (m in fam_dirs) {\n",
    "        print(basename(m))\n",
    "        count_fam = count_fam + 1\n",
    "\n",
    "        # Get all units trees (250 = 25 organism * 10 samples/statistical repeats) for the combination Minimun Evolution (ME) + supermatrix method (SM). \n",
    "        # We unroot every tree and fully resolve it to use the RF.dist package and others. \n",
    "        all_units_file  = file.path(m, \"results/rerun\", \"all_units.nwk\")\n",
    "        all_units_trees = read.tree(all_units_file, keep.multi=TRUE)\n",
    "        all_units_trees = unroot(all_units_trees)\n",
    "        \n",
    "        # check if there any non-bynary (consensus trees) among the 250 trees present in the file above\n",
    "        if (length(which(is.binary(all_units_trees) == FALSE)) != 0) {\n",
    "            \n",
    "            # if there are non-binary trees, iterate through them and transformed into a binary one using multi2di package\n",
    "            for (tree in which(is.binary(all_units_trees) == FALSE)) {\n",
    "                all_units_trees[[tree]] = multi2di(all_units_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250) ME + supertree (ST) method and handle non-binary trees as above\n",
    "        all_superfine_file  = file.path(m, \"results/rerun\", \"all_units_superfine_trees.nwk\")\n",
    "        all_superfine_trees = read.tree(all_superfine_file, keep.multi=TRUE)\n",
    "        all_superfine_trees = unroot(all_superfine_trees)\n",
    "        if (length(which(is.binary(all_superfine_trees) == FALSE)) != 0) {\n",
    "            for (tree in which(is.binary(all_superfine_trees) == FALSE)) {\n",
    "                all_superfine_trees[[tree]] = multi2di(all_superfine_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250) Maximum Likelihood (ML) + SM \n",
    "        all_units_raxml_file  = file.path(m, \"results/rerun\", \"all_units_RAxML_trees.nwk\")\n",
    "        all_units_raxml_trees = read.tree(all_units_raxml_file, keep.multi=TRUE)\n",
    "        all_units_raxml_trees = unroot(all_units_raxml_trees)\n",
    "        if (length(which(is.binary(all_units_raxml_trees) == FALSE)) != 0) {\n",
    "            for (tree in which(is.binary(all_units_raxml_trees) == FALSE)) {\n",
    "                all_units_raxml_trees[[tree]] = multi2di(all_units_raxml_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # take only the last 10 of the 250 above trees (unit25's 10 samples/replicates)(unit25 = all 25 single units concatenated) and collect branch lengths (each branch in tree, 27 in total 2*15-3) and diameter (using patristic distance)\n",
    "        for(mytree in 241:length(all_units_raxml_trees)) {\n",
    "            collect_branch_len_supermatrix_ml = c(collect_branch_len_supermatrix_ml, all_units_raxml_trees[[mytree]]$edge.length)\n",
    "            collect_diameter_supermatrix_ml = c(collect_diameter_supermatrix_ml, max(distTips(all_units_raxml_trees[[mytree]], method = 'patristic')))\n",
    "        }\n",
    "\n",
    "        # Get all units trees (250) ML + ST\n",
    "        all_raxml_superfine_file  = file.path(m, \"results/rerun\", \"all_units_RAxML_superfine_trees.nwk\")\n",
    "        all_raxml_superfine_trees = read.tree(all_raxml_superfine_file, keep.multi=TRUE)\n",
    "        all_raxml_superfine_trees = unroot(all_raxml_superfine_trees)\n",
    "        if (length(which(is.binary(all_raxml_superfine_trees) == FALSE)) != 0) {\n",
    "            for (tree in which(is.binary(all_raxml_superfine_trees) == FALSE)) {\n",
    "                all_raxml_superfine_trees[[tree]] = multi2di(all_raxml_superfine_trees[[tree]])\n",
    "            }\n",
    "        }\n",
    "\n",
    "        # Reference tree block, each gene family has it's own reference tree that is the one generated by the simulation step. Along this tree all sequences are generated and then alligned and passed as input for main.nf pipeline.\n",
    "        ref_tree_file = file.path(run, paste0(basename(m), \".nwk\"))\n",
    "        ref_tree      = read.tree(ref_tree_file, keep.multi=FALSE)\n",
    "\n",
    "        # Since each gene family has a predifined gene tree (ref_tree) composed of a fixed species tree pasted as leaf of the paralog tree. The tree is designed like this by choic4. Each gene families varies only in the paralog tree that it has that is unique (15 tips).\n",
    "        # Here we extract the paralog tree and to do so we just need to extract the subtree with all the sequences of omne species. Since each species have the same paralog tree. So we can extract the subtree of whichever species it does not matter.\n",
    "        ref_subtree   = keep.tip(ref_tree,as.character(ref_codefile$V1))\n",
    "\n",
    "        # we then rename the tips of the paralog tree (ref_subtree) to a more generic keyname. Effectively going from rat1 to seq1. \n",
    "        for (j in 1:length(ref_subtree$tip.label)) { \n",
    "            ref_subtree$tip.label[j] = as.character(ref_codefile$V2[which(ref_codefile$V1 == ref_subtree[[\"tip.label\"]][j])])\n",
    "        }\n",
    "\n",
    "        # Unroot the tree and perform aother operation and then save the paralog tree to the specifc gene faimily directory\n",
    "        ref_subtree        = unroot(ref_subtree)\n",
    "        ref_subtree        = list(ref_subtree)\n",
    "        class(ref_subtree) = \"multiPhylo\"\n",
    "        subtree_out_file   = file.path(m, paste0(\"ref_subtree_\", basename(m), \".nwk\"))\n",
    "        write.tree(ref_subtree, file=subtree_out_file)\n",
    "\n",
    "        # update on the average branch length with the mean of the true/refernce paralog tree of the given gene family\n",
    "        avg_branch_length = c(avg_branch_length, mean(ref_subtree[[1]]$edge.length))\n",
    "\n",
    "        # Make a list containing all trees gathered so far and compute all possible pair comparisons among trees. There are 1001*1001-1001 unique pairewise comparisons ( a square matrix of size 1001 with 0 in diagonal ).\n",
    "        # On this matrix (plotdata) all other measuramnts are done. This is a symmetric matrix (distance matrix)\n",
    "        #all_trees        = c(ref_subtree, all_units_trees, all_superfine_trees, all_units_raxml_trees, all_raxml_superfine_trees)\n",
    "        all_trees        = c(ref_subtree, all_units_trees[seq(1, 250, 10)], all_superfine_trees[seq(1, 250, 10)], all_units_raxml_trees[seq(1, 250, 10)], all_raxml_superfine_trees[seq(1, 250, 10)])\n",
    "        class(all_trees) = \"multiPhylo\"\n",
    "        all_vs_all_rf    = RF.dist(all_trees,normalize=TRUE)\n",
    "        plotdata         = data.matrix(all_vs_all_rf)\n",
    "\n",
    "        # all ME metrics to be computed\n",
    "        mean_rf           = c()\n",
    "        se_rf             = c()\n",
    "        mean_rf_norm      = c()\n",
    "        mean_superfine_rf = c()\n",
    "        se_superfine_rf   = c()\n",
    "\n",
    "        # all ML metrics to be computed\n",
    "        mean_raxml_rf           = c()\n",
    "        se_raxml_rf             = c()\n",
    "        mean_raxml_rf_norm      = c()\n",
    "        mean_raxml_superfine_rf = c()\n",
    "        se_raxml_superfine_rf   = c()\n",
    "\n",
    "        # proceed throught the plotdata matrix taking from the second element one every ten up to 242. Basically proceed at chunkcs of 10 (samples). On those chuncks compute the mean and the Standard error. \n",
    "        # average all sample \"values\" whithin a unit. going from 250 value to 25 averages\n",
    "        # This iterate only through the values of the comparison between the reference paralofg tree (ref_subtree)(first column/row of matrix) and the 250 paraogs trees of ME + SM (all_units_trees)\n",
    "        for (i in seq(2, (length(orgs$V1)*10)+1, by=10)) {\n",
    "            mean_rf = c(mean_rf, mean(plotdata[1, i:(i+9)]))\n",
    "            se_rf   = c(se_rf, (sd(plotdata[1, i:(i+9)])/sqrt(10)))\n",
    "        }\n",
    "\n",
    "        # This instead goes from row 252 to 492 of the plotdata matrix. It still goes in chuncks of 10 (samples). \n",
    "        # So this effectively does what is done above but between the reference paralofg tree and the 250 paraogs trees of ME + ST (all_superfine_trees)\n",
    "        for (i in seq( (length(orgs$V1)*10)+2, (2*(length(orgs$V1)*10)+1), by=10 )) {\n",
    "            mean_superfine_rf = c(mean_superfine_rf, mean(plotdata[1,i:(i+9)]))\n",
    "            se_superfine_rf   = c(se_superfine_rf, (sd(plotdata[1,i:(i+9)])/sqrt(10)))\n",
    "        }\n",
    "\n",
    "        # This for loop goes from 502 to 742. It still goes in chuncks of 10 (samples).\n",
    "        # So this effectively does what is done above but between the reference paralofg tree and the 250 paraogs trees of ML + SM (all_units_raxml_trees)\n",
    "        for (i in seq( (2*(length(orgs$V1)*10))+2, (3*(length(orgs$V1)*10)+1) ,by=10 )) {\n",
    "            mean_raxml_rf = c(mean_raxml_rf, mean(plotdata[1,i:(i+9)]))\n",
    "            se_raxml_rf   = c(se_raxml_rf, (sd(plotdata[1,i:(i+9)])/sqrt(10)))\n",
    "        }\n",
    "\n",
    "        # This for loop goes from 752 to 992. It still goes in chuncks of 10 (samples).\n",
    "        # So this effectively does what is done above but between the reference paralofg tree and the 250 paraogs trees of ML + ST (all_raxml_superfine_trees)\n",
    "        for (i in seq( (3*(length(orgs$V1)*10))+2, (4*(length(orgs$V1)*10)+1), by=10)) {\n",
    "            mean_raxml_superfine_rf = c(mean_raxml_superfine_rf, mean(plotdata[1,i:(i+9)]))\n",
    "            se_raxml_superfine_rf   = c(se_raxml_superfine_rf, (sd(plotdata[1,i:(i+9)])/sqrt(10)))\n",
    "        }\n",
    "\n",
    "        # TODO finish this analysis\n",
    "        \n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "7c4e667f-714b-4b1c-9cc1-b270727b7397",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 41 × 41 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>1</th><th scope=col>2</th><th scope=col>3</th><th scope=col>4</th><th scope=col>5</th><th scope=col>6</th><th scope=col>7</th><th scope=col>8</th><th scope=col>9</th><th scope=col>10</th><th scope=col>⋯</th><th scope=col>32</th><th scope=col>33</th><th scope=col>34</th><th scope=col>35</th><th scope=col>36</th><th scope=col>37</th><th scope=col>38</th><th scope=col>39</th><th scope=col>40</th><th scope=col>41</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>0.0000000</td><td>0.3333333</td><td>0.41666667</td><td>0.33333333</td><td>0.25000000</td><td>0.25000000</td><td>0.33333333</td><td>0.33333333</td><td>0.25000000</td><td>0.33333333</td><td>⋯</td><td>0.5000000</td><td>0.25000000</td><td>0.25000000</td><td>0.25000000</td><td>0.33333333</td><td>0.25000000</td><td>0.33333333</td><td>0.16666667</td><td>0.33333333</td><td>0.16666667</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>0.3333333</td><td>0.0000000</td><td>0.41666667</td><td>0.33333333</td><td>0.25000000</td><td>0.25000000</td><td>0.33333333</td><td>0.33333333</td><td>0.25000000</td><td>0.33333333</td><td>⋯</td><td>0.3333333</td><td>0.25000000</td><td>0.33333333</td><td>0.33333333</td><td>0.41666667</td><td>0.25000000</td><td>0.41666667</td><td>0.16666667</td><td>0.41666667</td><td>0.16666667</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>0.4166667</td><td>0.4166667</td><td>0.00000000</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.08333333</td><td>⋯</td><td>0.4166667</td><td>0.16666667</td><td>0.33333333</td><td>0.25000000</td><td>0.33333333</td><td>0.16666667</td><td>0.41666667</td><td>0.25000000</td><td>0.41666667</td><td>0.25000000</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>0.3333333</td><td>0.3333333</td><td>0.08333333</td><td>0.00000000</td><td>0.08333333</td><td>0.08333333</td><td>0.00000000</td><td>0.00000000</td><td>0.08333333</td><td>0.00000000</td><td>⋯</td><td>0.4166667</td><td>0.08333333</td><td>0.25000000</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.33333333</td><td>0.16666667</td><td>0.33333333</td><td>0.16666667</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>0.2500000</td><td>0.2500000</td><td>0.16666667</td><td>0.08333333</td><td>0.00000000</td><td>0.00000000</td><td>0.08333333</td><td>0.08333333</td><td>0.00000000</td><td>0.08333333</td><td>⋯</td><td>0.4166667</td><td>0.00000000</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>0.00000000</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.08333333</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>0.2500000</td><td>0.2500000</td><td>0.16666667</td><td>0.08333333</td><td>0.00000000</td><td>0.00000000</td><td>0.08333333</td><td>0.08333333</td><td>0.00000000</td><td>0.08333333</td><td>⋯</td><td>0.4166667</td><td>0.00000000</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>0.00000000</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.08333333</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>0.3333333</td><td>0.3333333</td><td>0.08333333</td><td>0.00000000</td><td>0.08333333</td><td>0.08333333</td><td>0.00000000</td><td>0.00000000</td><td>0.08333333</td><td>0.00000000</td><td>⋯</td><td>0.4166667</td><td>0.08333333</td><td>0.25000000</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.33333333</td><td>0.16666667</td><td>0.33333333</td><td>0.16666667</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>0.3333333</td><td>0.3333333</td><td>0.08333333</td><td>0.00000000</td><td>0.08333333</td><td>0.08333333</td><td>0.00000000</td><td>0.00000000</td><td>0.08333333</td><td>0.00000000</td><td>⋯</td><td>0.4166667</td><td>0.08333333</td><td>0.25000000</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.33333333</td><td>0.16666667</td><td>0.33333333</td><td>0.16666667</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>0.2500000</td><td>0.2500000</td><td>0.16666667</td><td>0.08333333</td><td>0.00000000</td><td>0.00000000</td><td>0.08333333</td><td>0.08333333</td><td>0.00000000</td><td>0.08333333</td><td>⋯</td><td>0.4166667</td><td>0.00000000</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>0.00000000</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.08333333</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>0.3333333</td><td>0.3333333</td><td>0.08333333</td><td>0.00000000</td><td>0.08333333</td><td>0.08333333</td><td>0.00000000</td><td>0.00000000</td><td>0.08333333</td><td>0.00000000</td><td>⋯</td><td>0.4166667</td><td>0.08333333</td><td>0.25000000</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.33333333</td><td>0.16666667</td><td>0.33333333</td><td>0.16666667</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>0.3333333</td><td>0.3333333</td><td>0.08333333</td><td>0.00000000</td><td>0.08333333</td><td>0.08333333</td><td>0.00000000</td><td>0.00000000</td><td>0.08333333</td><td>0.00000000</td><td>⋯</td><td>0.4166667</td><td>0.08333333</td><td>0.25000000</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.33333333</td><td>0.16666667</td><td>0.33333333</td><td>0.16666667</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>0.3333333</td><td>0.0000000</td><td>0.41666667</td><td>0.33333333</td><td>0.25000000</td><td>0.25000000</td><td>0.33333333</td><td>0.33333333</td><td>0.25000000</td><td>0.33333333</td><td>⋯</td><td>0.3333333</td><td>0.25000000</td><td>0.33333333</td><td>0.33333333</td><td>0.41666667</td><td>0.25000000</td><td>0.41666667</td><td>0.16666667</td><td>0.41666667</td><td>0.16666667</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>0.4166667</td><td>0.5000000</td><td>0.16666667</td><td>0.25000000</td><td>0.25000000</td><td>0.25000000</td><td>0.25000000</td><td>0.25000000</td><td>0.25000000</td><td>0.25000000</td><td>⋯</td><td>0.5000000</td><td>0.25000000</td><td>0.33333333</td><td>0.16666667</td><td>0.33333333</td><td>0.25000000</td><td>0.41666667</td><td>0.33333333</td><td>0.41666667</td><td>0.33333333</td></tr>\n",
       "\t<tr><th scope=row>14</th><td>0.3333333</td><td>0.2500000</td><td>0.33333333</td><td>0.25000000</td><td>0.33333333</td><td>0.33333333</td><td>0.25000000</td><td>0.25000000</td><td>0.33333333</td><td>0.25000000</td><td>⋯</td><td>0.3333333</td><td>0.33333333</td><td>0.41666667</td><td>0.41666667</td><td>0.50000000</td><td>0.33333333</td><td>0.50000000</td><td>0.25000000</td><td>0.50000000</td><td>0.25000000</td></tr>\n",
       "\t<tr><th scope=row>15</th><td>0.2500000</td><td>0.2500000</td><td>0.25000000</td><td>0.25000000</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.25000000</td><td>0.16666667</td><td>0.25000000</td><td>⋯</td><td>0.3333333</td><td>0.16666667</td><td>0.25000000</td><td>0.25000000</td><td>0.33333333</td><td>0.16666667</td><td>0.33333333</td><td>0.08333333</td><td>0.33333333</td><td>0.08333333</td></tr>\n",
       "\t<tr><th scope=row>16</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><th scope=row>17</th><td>0.2500000</td><td>0.2500000</td><td>0.16666667</td><td>0.08333333</td><td>0.00000000</td><td>0.00000000</td><td>0.08333333</td><td>0.08333333</td><td>0.00000000</td><td>0.08333333</td><td>⋯</td><td>0.4166667</td><td>0.00000000</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>0.00000000</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.08333333</td></tr>\n",
       "\t<tr><th scope=row>18</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><th scope=row>19</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><th scope=row>20</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><th scope=row>21</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><th scope=row>22</th><td>0.5000000</td><td>0.3333333</td><td>0.41666667</td><td>0.41666667</td><td>0.41666667</td><td>0.41666667</td><td>0.41666667</td><td>0.41666667</td><td>0.41666667</td><td>0.41666667</td><td>⋯</td><td>0.0000000</td><td>0.41666667</td><td>0.50000000</td><td>0.50000000</td><td>0.41666667</td><td>0.41666667</td><td>0.41666667</td><td>0.33333333</td><td>0.41666667</td><td>0.33333333</td></tr>\n",
       "\t<tr><th scope=row>23</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><th scope=row>24</th><td>0.2500000</td><td>0.3333333</td><td>0.33333333</td><td>0.25000000</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.25000000</td><td>0.16666667</td><td>0.25000000</td><td>⋯</td><td>0.5000000</td><td>0.16666667</td><td>0.00000000</td><td>0.16666667</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td></tr>\n",
       "\t<tr><th scope=row>25</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><th scope=row>26</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><th scope=row>27</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><th scope=row>28</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><th scope=row>29</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><th scope=row>30</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><th scope=row>31</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><th scope=row>32</th><td>0.5000000</td><td>0.3333333</td><td>0.41666667</td><td>0.41666667</td><td>0.41666667</td><td>0.41666667</td><td>0.41666667</td><td>0.41666667</td><td>0.41666667</td><td>0.41666667</td><td>⋯</td><td>0.0000000</td><td>0.41666667</td><td>0.50000000</td><td>0.50000000</td><td>0.41666667</td><td>0.41666667</td><td>0.41666667</td><td>0.33333333</td><td>0.41666667</td><td>0.33333333</td></tr>\n",
       "\t<tr><th scope=row>33</th><td>0.2500000</td><td>0.2500000</td><td>0.16666667</td><td>0.08333333</td><td>0.00000000</td><td>0.00000000</td><td>0.08333333</td><td>0.08333333</td><td>0.00000000</td><td>0.08333333</td><td>⋯</td><td>0.4166667</td><td>0.00000000</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>0.00000000</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.08333333</td></tr>\n",
       "\t<tr><th scope=row>34</th><td>0.2500000</td><td>0.3333333</td><td>0.33333333</td><td>0.25000000</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.25000000</td><td>0.16666667</td><td>0.25000000</td><td>⋯</td><td>0.5000000</td><td>0.16666667</td><td>0.00000000</td><td>0.16666667</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td></tr>\n",
       "\t<tr><th scope=row>35</th><td>0.2500000</td><td>0.3333333</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.5000000</td><td>0.08333333</td><td>0.16666667</td><td>0.00000000</td><td>0.16666667</td><td>0.08333333</td><td>0.25000000</td><td>0.16666667</td><td>0.25000000</td><td>0.16666667</td></tr>\n",
       "\t<tr><th scope=row>36</th><td>0.3333333</td><td>0.4166667</td><td>0.33333333</td><td>0.25000000</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.25000000</td><td>0.16666667</td><td>0.25000000</td><td>⋯</td><td>0.4166667</td><td>0.16666667</td><td>0.16666667</td><td>0.16666667</td><td>0.00000000</td><td>0.16666667</td><td>0.08333333</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td></tr>\n",
       "\t<tr><th scope=row>37</th><td>0.2500000</td><td>0.2500000</td><td>0.16666667</td><td>0.08333333</td><td>0.00000000</td><td>0.00000000</td><td>0.08333333</td><td>0.08333333</td><td>0.00000000</td><td>0.08333333</td><td>⋯</td><td>0.4166667</td><td>0.00000000</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>0.00000000</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.08333333</td></tr>\n",
       "\t<tr><th scope=row>38</th><td>0.3333333</td><td>0.4166667</td><td>0.41666667</td><td>0.33333333</td><td>0.25000000</td><td>0.25000000</td><td>0.33333333</td><td>0.33333333</td><td>0.25000000</td><td>0.33333333</td><td>⋯</td><td>0.4166667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td></tr>\n",
       "\t<tr><th scope=row>39</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "\t<tr><th scope=row>40</th><td>0.3333333</td><td>0.4166667</td><td>0.41666667</td><td>0.33333333</td><td>0.25000000</td><td>0.25000000</td><td>0.33333333</td><td>0.33333333</td><td>0.25000000</td><td>0.33333333</td><td>⋯</td><td>0.4166667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td></tr>\n",
       "\t<tr><th scope=row>41</th><td>0.1666667</td><td>0.1666667</td><td>0.25000000</td><td>0.16666667</td><td>0.08333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.08333333</td><td>0.16666667</td><td>⋯</td><td>0.3333333</td><td>0.08333333</td><td>0.16666667</td><td>0.16666667</td><td>0.25000000</td><td>0.08333333</td><td>0.25000000</td><td>0.00000000</td><td>0.25000000</td><td>0.00000000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 41 × 41 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & 1 & 2 & 3 & 4 & 5 & 6 & 7 & 8 & 9 & 10 & ⋯ & 32 & 33 & 34 & 35 & 36 & 37 & 38 & 39 & 40 & 41\\\\\n",
       "\\hline\n",
       "\t1 & 0.0000000 & 0.3333333 & 0.41666667 & 0.33333333 & 0.25000000 & 0.25000000 & 0.33333333 & 0.33333333 & 0.25000000 & 0.33333333 & ⋯ & 0.5000000 & 0.25000000 & 0.25000000 & 0.25000000 & 0.33333333 & 0.25000000 & 0.33333333 & 0.16666667 & 0.33333333 & 0.16666667\\\\\n",
       "\t2 & 0.3333333 & 0.0000000 & 0.41666667 & 0.33333333 & 0.25000000 & 0.25000000 & 0.33333333 & 0.33333333 & 0.25000000 & 0.33333333 & ⋯ & 0.3333333 & 0.25000000 & 0.33333333 & 0.33333333 & 0.41666667 & 0.25000000 & 0.41666667 & 0.16666667 & 0.41666667 & 0.16666667\\\\\n",
       "\t3 & 0.4166667 & 0.4166667 & 0.00000000 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.08333333 & ⋯ & 0.4166667 & 0.16666667 & 0.33333333 & 0.25000000 & 0.33333333 & 0.16666667 & 0.41666667 & 0.25000000 & 0.41666667 & 0.25000000\\\\\n",
       "\t4 & 0.3333333 & 0.3333333 & 0.08333333 & 0.00000000 & 0.08333333 & 0.08333333 & 0.00000000 & 0.00000000 & 0.08333333 & 0.00000000 & ⋯ & 0.4166667 & 0.08333333 & 0.25000000 & 0.16666667 & 0.25000000 & 0.08333333 & 0.33333333 & 0.16666667 & 0.33333333 & 0.16666667\\\\\n",
       "\t5 & 0.2500000 & 0.2500000 & 0.16666667 & 0.08333333 & 0.00000000 & 0.00000000 & 0.08333333 & 0.08333333 & 0.00000000 & 0.08333333 & ⋯ & 0.4166667 & 0.00000000 & 0.16666667 & 0.08333333 & 0.16666667 & 0.00000000 & 0.25000000 & 0.08333333 & 0.25000000 & 0.08333333\\\\\n",
       "\t6 & 0.2500000 & 0.2500000 & 0.16666667 & 0.08333333 & 0.00000000 & 0.00000000 & 0.08333333 & 0.08333333 & 0.00000000 & 0.08333333 & ⋯ & 0.4166667 & 0.00000000 & 0.16666667 & 0.08333333 & 0.16666667 & 0.00000000 & 0.25000000 & 0.08333333 & 0.25000000 & 0.08333333\\\\\n",
       "\t7 & 0.3333333 & 0.3333333 & 0.08333333 & 0.00000000 & 0.08333333 & 0.08333333 & 0.00000000 & 0.00000000 & 0.08333333 & 0.00000000 & ⋯ & 0.4166667 & 0.08333333 & 0.25000000 & 0.16666667 & 0.25000000 & 0.08333333 & 0.33333333 & 0.16666667 & 0.33333333 & 0.16666667\\\\\n",
       "\t8 & 0.3333333 & 0.3333333 & 0.08333333 & 0.00000000 & 0.08333333 & 0.08333333 & 0.00000000 & 0.00000000 & 0.08333333 & 0.00000000 & ⋯ & 0.4166667 & 0.08333333 & 0.25000000 & 0.16666667 & 0.25000000 & 0.08333333 & 0.33333333 & 0.16666667 & 0.33333333 & 0.16666667\\\\\n",
       "\t9 & 0.2500000 & 0.2500000 & 0.16666667 & 0.08333333 & 0.00000000 & 0.00000000 & 0.08333333 & 0.08333333 & 0.00000000 & 0.08333333 & ⋯ & 0.4166667 & 0.00000000 & 0.16666667 & 0.08333333 & 0.16666667 & 0.00000000 & 0.25000000 & 0.08333333 & 0.25000000 & 0.08333333\\\\\n",
       "\t10 & 0.3333333 & 0.3333333 & 0.08333333 & 0.00000000 & 0.08333333 & 0.08333333 & 0.00000000 & 0.00000000 & 0.08333333 & 0.00000000 & ⋯ & 0.4166667 & 0.08333333 & 0.25000000 & 0.16666667 & 0.25000000 & 0.08333333 & 0.33333333 & 0.16666667 & 0.33333333 & 0.16666667\\\\\n",
       "\t11 & 0.3333333 & 0.3333333 & 0.08333333 & 0.00000000 & 0.08333333 & 0.08333333 & 0.00000000 & 0.00000000 & 0.08333333 & 0.00000000 & ⋯ & 0.4166667 & 0.08333333 & 0.25000000 & 0.16666667 & 0.25000000 & 0.08333333 & 0.33333333 & 0.16666667 & 0.33333333 & 0.16666667\\\\\n",
       "\t12 & 0.3333333 & 0.0000000 & 0.41666667 & 0.33333333 & 0.25000000 & 0.25000000 & 0.33333333 & 0.33333333 & 0.25000000 & 0.33333333 & ⋯ & 0.3333333 & 0.25000000 & 0.33333333 & 0.33333333 & 0.41666667 & 0.25000000 & 0.41666667 & 0.16666667 & 0.41666667 & 0.16666667\\\\\n",
       "\t13 & 0.4166667 & 0.5000000 & 0.16666667 & 0.25000000 & 0.25000000 & 0.25000000 & 0.25000000 & 0.25000000 & 0.25000000 & 0.25000000 & ⋯ & 0.5000000 & 0.25000000 & 0.33333333 & 0.16666667 & 0.33333333 & 0.25000000 & 0.41666667 & 0.33333333 & 0.41666667 & 0.33333333\\\\\n",
       "\t14 & 0.3333333 & 0.2500000 & 0.33333333 & 0.25000000 & 0.33333333 & 0.33333333 & 0.25000000 & 0.25000000 & 0.33333333 & 0.25000000 & ⋯ & 0.3333333 & 0.33333333 & 0.41666667 & 0.41666667 & 0.50000000 & 0.33333333 & 0.50000000 & 0.25000000 & 0.50000000 & 0.25000000\\\\\n",
       "\t15 & 0.2500000 & 0.2500000 & 0.25000000 & 0.25000000 & 0.16666667 & 0.16666667 & 0.25000000 & 0.25000000 & 0.16666667 & 0.25000000 & ⋯ & 0.3333333 & 0.16666667 & 0.25000000 & 0.25000000 & 0.33333333 & 0.16666667 & 0.33333333 & 0.08333333 & 0.33333333 & 0.08333333\\\\\n",
       "\t16 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\t17 & 0.2500000 & 0.2500000 & 0.16666667 & 0.08333333 & 0.00000000 & 0.00000000 & 0.08333333 & 0.08333333 & 0.00000000 & 0.08333333 & ⋯ & 0.4166667 & 0.00000000 & 0.16666667 & 0.08333333 & 0.16666667 & 0.00000000 & 0.25000000 & 0.08333333 & 0.25000000 & 0.08333333\\\\\n",
       "\t18 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\t19 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\t20 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\t21 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\t22 & 0.5000000 & 0.3333333 & 0.41666667 & 0.41666667 & 0.41666667 & 0.41666667 & 0.41666667 & 0.41666667 & 0.41666667 & 0.41666667 & ⋯ & 0.0000000 & 0.41666667 & 0.50000000 & 0.50000000 & 0.41666667 & 0.41666667 & 0.41666667 & 0.33333333 & 0.41666667 & 0.33333333\\\\\n",
       "\t23 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\t24 & 0.2500000 & 0.3333333 & 0.33333333 & 0.25000000 & 0.16666667 & 0.16666667 & 0.25000000 & 0.25000000 & 0.16666667 & 0.25000000 & ⋯ & 0.5000000 & 0.16666667 & 0.00000000 & 0.16666667 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & 0.08333333 & 0.16666667\\\\\n",
       "\t25 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\t26 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\t27 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\t28 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\t29 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\t30 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\t31 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\t32 & 0.5000000 & 0.3333333 & 0.41666667 & 0.41666667 & 0.41666667 & 0.41666667 & 0.41666667 & 0.41666667 & 0.41666667 & 0.41666667 & ⋯ & 0.0000000 & 0.41666667 & 0.50000000 & 0.50000000 & 0.41666667 & 0.41666667 & 0.41666667 & 0.33333333 & 0.41666667 & 0.33333333\\\\\n",
       "\t33 & 0.2500000 & 0.2500000 & 0.16666667 & 0.08333333 & 0.00000000 & 0.00000000 & 0.08333333 & 0.08333333 & 0.00000000 & 0.08333333 & ⋯ & 0.4166667 & 0.00000000 & 0.16666667 & 0.08333333 & 0.16666667 & 0.00000000 & 0.25000000 & 0.08333333 & 0.25000000 & 0.08333333\\\\\n",
       "\t34 & 0.2500000 & 0.3333333 & 0.33333333 & 0.25000000 & 0.16666667 & 0.16666667 & 0.25000000 & 0.25000000 & 0.16666667 & 0.25000000 & ⋯ & 0.5000000 & 0.16666667 & 0.00000000 & 0.16666667 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & 0.08333333 & 0.16666667\\\\\n",
       "\t35 & 0.2500000 & 0.3333333 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.5000000 & 0.08333333 & 0.16666667 & 0.00000000 & 0.16666667 & 0.08333333 & 0.25000000 & 0.16666667 & 0.25000000 & 0.16666667\\\\\n",
       "\t36 & 0.3333333 & 0.4166667 & 0.33333333 & 0.25000000 & 0.16666667 & 0.16666667 & 0.25000000 & 0.25000000 & 0.16666667 & 0.25000000 & ⋯ & 0.4166667 & 0.16666667 & 0.16666667 & 0.16666667 & 0.00000000 & 0.16666667 & 0.08333333 & 0.25000000 & 0.08333333 & 0.25000000\\\\\n",
       "\t37 & 0.2500000 & 0.2500000 & 0.16666667 & 0.08333333 & 0.00000000 & 0.00000000 & 0.08333333 & 0.08333333 & 0.00000000 & 0.08333333 & ⋯ & 0.4166667 & 0.00000000 & 0.16666667 & 0.08333333 & 0.16666667 & 0.00000000 & 0.25000000 & 0.08333333 & 0.25000000 & 0.08333333\\\\\n",
       "\t38 & 0.3333333 & 0.4166667 & 0.41666667 & 0.33333333 & 0.25000000 & 0.25000000 & 0.33333333 & 0.33333333 & 0.25000000 & 0.33333333 & ⋯ & 0.4166667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000 & 0.25000000\\\\\n",
       "\t39 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\t40 & 0.3333333 & 0.4166667 & 0.41666667 & 0.33333333 & 0.25000000 & 0.25000000 & 0.33333333 & 0.33333333 & 0.25000000 & 0.33333333 & ⋯ & 0.4166667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000 & 0.25000000\\\\\n",
       "\t41 & 0.1666667 & 0.1666667 & 0.25000000 & 0.16666667 & 0.08333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.08333333 & 0.16666667 & ⋯ & 0.3333333 & 0.08333333 & 0.16666667 & 0.16666667 & 0.25000000 & 0.08333333 & 0.25000000 & 0.00000000 & 0.25000000 & 0.00000000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 41 × 41 of type dbl\n",
       "\n",
       "| <!--/--> | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | ⋯ | 32 | 33 | 34 | 35 | 36 | 37 | 38 | 39 | 40 | 41 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 0.0000000 | 0.3333333 | 0.41666667 | 0.33333333 | 0.25000000 | 0.25000000 | 0.33333333 | 0.33333333 | 0.25000000 | 0.33333333 | ⋯ | 0.5000000 | 0.25000000 | 0.25000000 | 0.25000000 | 0.33333333 | 0.25000000 | 0.33333333 | 0.16666667 | 0.33333333 | 0.16666667 |\n",
       "| 2 | 0.3333333 | 0.0000000 | 0.41666667 | 0.33333333 | 0.25000000 | 0.25000000 | 0.33333333 | 0.33333333 | 0.25000000 | 0.33333333 | ⋯ | 0.3333333 | 0.25000000 | 0.33333333 | 0.33333333 | 0.41666667 | 0.25000000 | 0.41666667 | 0.16666667 | 0.41666667 | 0.16666667 |\n",
       "| 3 | 0.4166667 | 0.4166667 | 0.00000000 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.08333333 | ⋯ | 0.4166667 | 0.16666667 | 0.33333333 | 0.25000000 | 0.33333333 | 0.16666667 | 0.41666667 | 0.25000000 | 0.41666667 | 0.25000000 |\n",
       "| 4 | 0.3333333 | 0.3333333 | 0.08333333 | 0.00000000 | 0.08333333 | 0.08333333 | 0.00000000 | 0.00000000 | 0.08333333 | 0.00000000 | ⋯ | 0.4166667 | 0.08333333 | 0.25000000 | 0.16666667 | 0.25000000 | 0.08333333 | 0.33333333 | 0.16666667 | 0.33333333 | 0.16666667 |\n",
       "| 5 | 0.2500000 | 0.2500000 | 0.16666667 | 0.08333333 | 0.00000000 | 0.00000000 | 0.08333333 | 0.08333333 | 0.00000000 | 0.08333333 | ⋯ | 0.4166667 | 0.00000000 | 0.16666667 | 0.08333333 | 0.16666667 | 0.00000000 | 0.25000000 | 0.08333333 | 0.25000000 | 0.08333333 |\n",
       "| 6 | 0.2500000 | 0.2500000 | 0.16666667 | 0.08333333 | 0.00000000 | 0.00000000 | 0.08333333 | 0.08333333 | 0.00000000 | 0.08333333 | ⋯ | 0.4166667 | 0.00000000 | 0.16666667 | 0.08333333 | 0.16666667 | 0.00000000 | 0.25000000 | 0.08333333 | 0.25000000 | 0.08333333 |\n",
       "| 7 | 0.3333333 | 0.3333333 | 0.08333333 | 0.00000000 | 0.08333333 | 0.08333333 | 0.00000000 | 0.00000000 | 0.08333333 | 0.00000000 | ⋯ | 0.4166667 | 0.08333333 | 0.25000000 | 0.16666667 | 0.25000000 | 0.08333333 | 0.33333333 | 0.16666667 | 0.33333333 | 0.16666667 |\n",
       "| 8 | 0.3333333 | 0.3333333 | 0.08333333 | 0.00000000 | 0.08333333 | 0.08333333 | 0.00000000 | 0.00000000 | 0.08333333 | 0.00000000 | ⋯ | 0.4166667 | 0.08333333 | 0.25000000 | 0.16666667 | 0.25000000 | 0.08333333 | 0.33333333 | 0.16666667 | 0.33333333 | 0.16666667 |\n",
       "| 9 | 0.2500000 | 0.2500000 | 0.16666667 | 0.08333333 | 0.00000000 | 0.00000000 | 0.08333333 | 0.08333333 | 0.00000000 | 0.08333333 | ⋯ | 0.4166667 | 0.00000000 | 0.16666667 | 0.08333333 | 0.16666667 | 0.00000000 | 0.25000000 | 0.08333333 | 0.25000000 | 0.08333333 |\n",
       "| 10 | 0.3333333 | 0.3333333 | 0.08333333 | 0.00000000 | 0.08333333 | 0.08333333 | 0.00000000 | 0.00000000 | 0.08333333 | 0.00000000 | ⋯ | 0.4166667 | 0.08333333 | 0.25000000 | 0.16666667 | 0.25000000 | 0.08333333 | 0.33333333 | 0.16666667 | 0.33333333 | 0.16666667 |\n",
       "| 11 | 0.3333333 | 0.3333333 | 0.08333333 | 0.00000000 | 0.08333333 | 0.08333333 | 0.00000000 | 0.00000000 | 0.08333333 | 0.00000000 | ⋯ | 0.4166667 | 0.08333333 | 0.25000000 | 0.16666667 | 0.25000000 | 0.08333333 | 0.33333333 | 0.16666667 | 0.33333333 | 0.16666667 |\n",
       "| 12 | 0.3333333 | 0.0000000 | 0.41666667 | 0.33333333 | 0.25000000 | 0.25000000 | 0.33333333 | 0.33333333 | 0.25000000 | 0.33333333 | ⋯ | 0.3333333 | 0.25000000 | 0.33333333 | 0.33333333 | 0.41666667 | 0.25000000 | 0.41666667 | 0.16666667 | 0.41666667 | 0.16666667 |\n",
       "| 13 | 0.4166667 | 0.5000000 | 0.16666667 | 0.25000000 | 0.25000000 | 0.25000000 | 0.25000000 | 0.25000000 | 0.25000000 | 0.25000000 | ⋯ | 0.5000000 | 0.25000000 | 0.33333333 | 0.16666667 | 0.33333333 | 0.25000000 | 0.41666667 | 0.33333333 | 0.41666667 | 0.33333333 |\n",
       "| 14 | 0.3333333 | 0.2500000 | 0.33333333 | 0.25000000 | 0.33333333 | 0.33333333 | 0.25000000 | 0.25000000 | 0.33333333 | 0.25000000 | ⋯ | 0.3333333 | 0.33333333 | 0.41666667 | 0.41666667 | 0.50000000 | 0.33333333 | 0.50000000 | 0.25000000 | 0.50000000 | 0.25000000 |\n",
       "| 15 | 0.2500000 | 0.2500000 | 0.25000000 | 0.25000000 | 0.16666667 | 0.16666667 | 0.25000000 | 0.25000000 | 0.16666667 | 0.25000000 | ⋯ | 0.3333333 | 0.16666667 | 0.25000000 | 0.25000000 | 0.33333333 | 0.16666667 | 0.33333333 | 0.08333333 | 0.33333333 | 0.08333333 |\n",
       "| 16 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "| 17 | 0.2500000 | 0.2500000 | 0.16666667 | 0.08333333 | 0.00000000 | 0.00000000 | 0.08333333 | 0.08333333 | 0.00000000 | 0.08333333 | ⋯ | 0.4166667 | 0.00000000 | 0.16666667 | 0.08333333 | 0.16666667 | 0.00000000 | 0.25000000 | 0.08333333 | 0.25000000 | 0.08333333 |\n",
       "| 18 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "| 19 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "| 20 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "| 21 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "| 22 | 0.5000000 | 0.3333333 | 0.41666667 | 0.41666667 | 0.41666667 | 0.41666667 | 0.41666667 | 0.41666667 | 0.41666667 | 0.41666667 | ⋯ | 0.0000000 | 0.41666667 | 0.50000000 | 0.50000000 | 0.41666667 | 0.41666667 | 0.41666667 | 0.33333333 | 0.41666667 | 0.33333333 |\n",
       "| 23 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "| 24 | 0.2500000 | 0.3333333 | 0.33333333 | 0.25000000 | 0.16666667 | 0.16666667 | 0.25000000 | 0.25000000 | 0.16666667 | 0.25000000 | ⋯ | 0.5000000 | 0.16666667 | 0.00000000 | 0.16666667 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | 0.08333333 | 0.16666667 |\n",
       "| 25 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "| 26 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "| 27 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "| 28 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "| 29 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "| 30 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "| 31 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "| 32 | 0.5000000 | 0.3333333 | 0.41666667 | 0.41666667 | 0.41666667 | 0.41666667 | 0.41666667 | 0.41666667 | 0.41666667 | 0.41666667 | ⋯ | 0.0000000 | 0.41666667 | 0.50000000 | 0.50000000 | 0.41666667 | 0.41666667 | 0.41666667 | 0.33333333 | 0.41666667 | 0.33333333 |\n",
       "| 33 | 0.2500000 | 0.2500000 | 0.16666667 | 0.08333333 | 0.00000000 | 0.00000000 | 0.08333333 | 0.08333333 | 0.00000000 | 0.08333333 | ⋯ | 0.4166667 | 0.00000000 | 0.16666667 | 0.08333333 | 0.16666667 | 0.00000000 | 0.25000000 | 0.08333333 | 0.25000000 | 0.08333333 |\n",
       "| 34 | 0.2500000 | 0.3333333 | 0.33333333 | 0.25000000 | 0.16666667 | 0.16666667 | 0.25000000 | 0.25000000 | 0.16666667 | 0.25000000 | ⋯ | 0.5000000 | 0.16666667 | 0.00000000 | 0.16666667 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | 0.08333333 | 0.16666667 |\n",
       "| 35 | 0.2500000 | 0.3333333 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.5000000 | 0.08333333 | 0.16666667 | 0.00000000 | 0.16666667 | 0.08333333 | 0.25000000 | 0.16666667 | 0.25000000 | 0.16666667 |\n",
       "| 36 | 0.3333333 | 0.4166667 | 0.33333333 | 0.25000000 | 0.16666667 | 0.16666667 | 0.25000000 | 0.25000000 | 0.16666667 | 0.25000000 | ⋯ | 0.4166667 | 0.16666667 | 0.16666667 | 0.16666667 | 0.00000000 | 0.16666667 | 0.08333333 | 0.25000000 | 0.08333333 | 0.25000000 |\n",
       "| 37 | 0.2500000 | 0.2500000 | 0.16666667 | 0.08333333 | 0.00000000 | 0.00000000 | 0.08333333 | 0.08333333 | 0.00000000 | 0.08333333 | ⋯ | 0.4166667 | 0.00000000 | 0.16666667 | 0.08333333 | 0.16666667 | 0.00000000 | 0.25000000 | 0.08333333 | 0.25000000 | 0.08333333 |\n",
       "| 38 | 0.3333333 | 0.4166667 | 0.41666667 | 0.33333333 | 0.25000000 | 0.25000000 | 0.33333333 | 0.33333333 | 0.25000000 | 0.33333333 | ⋯ | 0.4166667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 | 0.25000000 |\n",
       "| 39 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "| 40 | 0.3333333 | 0.4166667 | 0.41666667 | 0.33333333 | 0.25000000 | 0.25000000 | 0.33333333 | 0.33333333 | 0.25000000 | 0.33333333 | ⋯ | 0.4166667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 | 0.25000000 |\n",
       "| 41 | 0.1666667 | 0.1666667 | 0.25000000 | 0.16666667 | 0.08333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.08333333 | 0.16666667 | ⋯ | 0.3333333 | 0.08333333 | 0.16666667 | 0.16666667 | 0.25000000 | 0.08333333 | 0.25000000 | 0.00000000 | 0.25000000 | 0.00000000 |\n",
       "\n"
      ],
      "text/plain": [
       "   1         2         3          4          5          6          7         \n",
       "1  0.0000000 0.3333333 0.41666667 0.33333333 0.25000000 0.25000000 0.33333333\n",
       "2  0.3333333 0.0000000 0.41666667 0.33333333 0.25000000 0.25000000 0.33333333\n",
       "3  0.4166667 0.4166667 0.00000000 0.08333333 0.16666667 0.16666667 0.08333333\n",
       "4  0.3333333 0.3333333 0.08333333 0.00000000 0.08333333 0.08333333 0.00000000\n",
       "5  0.2500000 0.2500000 0.16666667 0.08333333 0.00000000 0.00000000 0.08333333\n",
       "6  0.2500000 0.2500000 0.16666667 0.08333333 0.00000000 0.00000000 0.08333333\n",
       "7  0.3333333 0.3333333 0.08333333 0.00000000 0.08333333 0.08333333 0.00000000\n",
       "8  0.3333333 0.3333333 0.08333333 0.00000000 0.08333333 0.08333333 0.00000000\n",
       "9  0.2500000 0.2500000 0.16666667 0.08333333 0.00000000 0.00000000 0.08333333\n",
       "10 0.3333333 0.3333333 0.08333333 0.00000000 0.08333333 0.08333333 0.00000000\n",
       "11 0.3333333 0.3333333 0.08333333 0.00000000 0.08333333 0.08333333 0.00000000\n",
       "12 0.3333333 0.0000000 0.41666667 0.33333333 0.25000000 0.25000000 0.33333333\n",
       "13 0.4166667 0.5000000 0.16666667 0.25000000 0.25000000 0.25000000 0.25000000\n",
       "14 0.3333333 0.2500000 0.33333333 0.25000000 0.33333333 0.33333333 0.25000000\n",
       "15 0.2500000 0.2500000 0.25000000 0.25000000 0.16666667 0.16666667 0.25000000\n",
       "16 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "17 0.2500000 0.2500000 0.16666667 0.08333333 0.00000000 0.00000000 0.08333333\n",
       "18 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "19 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "20 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "21 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "22 0.5000000 0.3333333 0.41666667 0.41666667 0.41666667 0.41666667 0.41666667\n",
       "23 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "24 0.2500000 0.3333333 0.33333333 0.25000000 0.16666667 0.16666667 0.25000000\n",
       "25 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "26 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "27 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "28 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "29 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "30 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "31 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "32 0.5000000 0.3333333 0.41666667 0.41666667 0.41666667 0.41666667 0.41666667\n",
       "33 0.2500000 0.2500000 0.16666667 0.08333333 0.00000000 0.00000000 0.08333333\n",
       "34 0.2500000 0.3333333 0.33333333 0.25000000 0.16666667 0.16666667 0.25000000\n",
       "35 0.2500000 0.3333333 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "36 0.3333333 0.4166667 0.33333333 0.25000000 0.16666667 0.16666667 0.25000000\n",
       "37 0.2500000 0.2500000 0.16666667 0.08333333 0.00000000 0.00000000 0.08333333\n",
       "38 0.3333333 0.4166667 0.41666667 0.33333333 0.25000000 0.25000000 0.33333333\n",
       "39 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "40 0.3333333 0.4166667 0.41666667 0.33333333 0.25000000 0.25000000 0.33333333\n",
       "41 0.1666667 0.1666667 0.25000000 0.16666667 0.08333333 0.08333333 0.16666667\n",
       "   8          9          10         ⋯ 32        33         34        \n",
       "1  0.33333333 0.25000000 0.33333333 ⋯ 0.5000000 0.25000000 0.25000000\n",
       "2  0.33333333 0.25000000 0.33333333 ⋯ 0.3333333 0.25000000 0.33333333\n",
       "3  0.08333333 0.16666667 0.08333333 ⋯ 0.4166667 0.16666667 0.33333333\n",
       "4  0.00000000 0.08333333 0.00000000 ⋯ 0.4166667 0.08333333 0.25000000\n",
       "5  0.08333333 0.00000000 0.08333333 ⋯ 0.4166667 0.00000000 0.16666667\n",
       "6  0.08333333 0.00000000 0.08333333 ⋯ 0.4166667 0.00000000 0.16666667\n",
       "7  0.00000000 0.08333333 0.00000000 ⋯ 0.4166667 0.08333333 0.25000000\n",
       "8  0.00000000 0.08333333 0.00000000 ⋯ 0.4166667 0.08333333 0.25000000\n",
       "9  0.08333333 0.00000000 0.08333333 ⋯ 0.4166667 0.00000000 0.16666667\n",
       "10 0.00000000 0.08333333 0.00000000 ⋯ 0.4166667 0.08333333 0.25000000\n",
       "11 0.00000000 0.08333333 0.00000000 ⋯ 0.4166667 0.08333333 0.25000000\n",
       "12 0.33333333 0.25000000 0.33333333 ⋯ 0.3333333 0.25000000 0.33333333\n",
       "13 0.25000000 0.25000000 0.25000000 ⋯ 0.5000000 0.25000000 0.33333333\n",
       "14 0.25000000 0.33333333 0.25000000 ⋯ 0.3333333 0.33333333 0.41666667\n",
       "15 0.25000000 0.16666667 0.25000000 ⋯ 0.3333333 0.16666667 0.25000000\n",
       "16 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "17 0.08333333 0.00000000 0.08333333 ⋯ 0.4166667 0.00000000 0.16666667\n",
       "18 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "19 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "20 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "21 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "22 0.41666667 0.41666667 0.41666667 ⋯ 0.0000000 0.41666667 0.50000000\n",
       "23 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "24 0.25000000 0.16666667 0.25000000 ⋯ 0.5000000 0.16666667 0.00000000\n",
       "25 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "26 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "27 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "28 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "29 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "30 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "31 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "32 0.41666667 0.41666667 0.41666667 ⋯ 0.0000000 0.41666667 0.50000000\n",
       "33 0.08333333 0.00000000 0.08333333 ⋯ 0.4166667 0.00000000 0.16666667\n",
       "34 0.25000000 0.16666667 0.25000000 ⋯ 0.5000000 0.16666667 0.00000000\n",
       "35 0.16666667 0.08333333 0.16666667 ⋯ 0.5000000 0.08333333 0.16666667\n",
       "36 0.25000000 0.16666667 0.25000000 ⋯ 0.4166667 0.16666667 0.16666667\n",
       "37 0.08333333 0.00000000 0.08333333 ⋯ 0.4166667 0.00000000 0.16666667\n",
       "38 0.33333333 0.25000000 0.33333333 ⋯ 0.4166667 0.25000000 0.08333333\n",
       "39 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "40 0.33333333 0.25000000 0.33333333 ⋯ 0.4166667 0.25000000 0.08333333\n",
       "41 0.16666667 0.08333333 0.16666667 ⋯ 0.3333333 0.08333333 0.16666667\n",
       "   35         36         37         38         39         40         41        \n",
       "1  0.25000000 0.33333333 0.25000000 0.33333333 0.16666667 0.33333333 0.16666667\n",
       "2  0.33333333 0.41666667 0.25000000 0.41666667 0.16666667 0.41666667 0.16666667\n",
       "3  0.25000000 0.33333333 0.16666667 0.41666667 0.25000000 0.41666667 0.25000000\n",
       "4  0.16666667 0.25000000 0.08333333 0.33333333 0.16666667 0.33333333 0.16666667\n",
       "5  0.08333333 0.16666667 0.00000000 0.25000000 0.08333333 0.25000000 0.08333333\n",
       "6  0.08333333 0.16666667 0.00000000 0.25000000 0.08333333 0.25000000 0.08333333\n",
       "7  0.16666667 0.25000000 0.08333333 0.33333333 0.16666667 0.33333333 0.16666667\n",
       "8  0.16666667 0.25000000 0.08333333 0.33333333 0.16666667 0.33333333 0.16666667\n",
       "9  0.08333333 0.16666667 0.00000000 0.25000000 0.08333333 0.25000000 0.08333333\n",
       "10 0.16666667 0.25000000 0.08333333 0.33333333 0.16666667 0.33333333 0.16666667\n",
       "11 0.16666667 0.25000000 0.08333333 0.33333333 0.16666667 0.33333333 0.16666667\n",
       "12 0.33333333 0.41666667 0.25000000 0.41666667 0.16666667 0.41666667 0.16666667\n",
       "13 0.16666667 0.33333333 0.25000000 0.41666667 0.33333333 0.41666667 0.33333333\n",
       "14 0.41666667 0.50000000 0.33333333 0.50000000 0.25000000 0.50000000 0.25000000\n",
       "15 0.25000000 0.33333333 0.16666667 0.33333333 0.08333333 0.33333333 0.08333333\n",
       "16 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000\n",
       "17 0.08333333 0.16666667 0.00000000 0.25000000 0.08333333 0.25000000 0.08333333\n",
       "18 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000\n",
       "19 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000\n",
       "20 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000\n",
       "21 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000\n",
       "22 0.50000000 0.41666667 0.41666667 0.41666667 0.33333333 0.41666667 0.33333333\n",
       "23 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000\n",
       "24 0.16666667 0.16666667 0.16666667 0.08333333 0.16666667 0.08333333 0.16666667\n",
       "25 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000\n",
       "26 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000\n",
       "27 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000\n",
       "28 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000\n",
       "29 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000\n",
       "30 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000\n",
       "31 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000\n",
       "32 0.50000000 0.41666667 0.41666667 0.41666667 0.33333333 0.41666667 0.33333333\n",
       "33 0.08333333 0.16666667 0.00000000 0.25000000 0.08333333 0.25000000 0.08333333\n",
       "34 0.16666667 0.16666667 0.16666667 0.08333333 0.16666667 0.08333333 0.16666667\n",
       "35 0.00000000 0.16666667 0.08333333 0.25000000 0.16666667 0.25000000 0.16666667\n",
       "36 0.16666667 0.00000000 0.16666667 0.08333333 0.25000000 0.08333333 0.25000000\n",
       "37 0.08333333 0.16666667 0.00000000 0.25000000 0.08333333 0.25000000 0.08333333\n",
       "38 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000 0.25000000\n",
       "39 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000\n",
       "40 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000 0.25000000\n",
       "41 0.16666667 0.25000000 0.08333333 0.25000000 0.00000000 0.25000000 0.00000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plotdata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfa19d80-b20f-4a24-af34-ea65b39aa0f2",
   "metadata": {},
   "source": [
    "<a id=\"chapter-5-Running-times\"></a>\n",
    "\n",
    "## Chapter 5 - Running times\n",
    "\n",
    "TODO add description "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "062ad02a-baca-4fce-b3ee-e8ea3b3bb039",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\"\n",
      "[1] \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\"\n"
     ]
    }
   ],
   "source": [
    "# read the trace files for running times\n",
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    # each main.nf output run has a trace file repoprting various information on the run itself\n",
    "    tracefile_path = file.path(run, \"trace_cpu_time.txt\")\n",
    "    tracefile      = read.table(tracefile_path, header = TRUE)\n",
    "    \n",
    "    # here we parse the trace file to retain only the processes that we want to take the running time (realtime) from\n",
    "    tracefile_mod  = tracefile[grep(\"simulated_data:run_phylo_ML_supermatrix_aln_sim|simulated_data:only_concatenate_aln_sim|simulated_data:run_phylo_ML_full_aln_sim\", tracefile$native_id), c(\"name\", \"native_id\", \"realtime\")]\n",
    "    \n",
    "    # transforming the above table to one that have the process names (simulated_data:run_phylo_ML_supermatrix_aln_sim ecc..) as column names, \n",
    "    # the rows will be still the family names (working as ID) and the values in each column will be the running times associated to the column name (process name) for that given family.\n",
    "    tracefile_mod_wide = spread(tracefile_mod, native_id, realtime)\n",
    "\n",
    "    # creating a new last column called Supermatrix_ML that has aas values the running times of concatenation plus ML tree computation (on concatenated MSA) \n",
    "    tracefile_mod_wide$Supermatrix_ML = as.numeric(tracefile_mod_wide$`simulated_data:only_concatenate_aln_sim`) + as.numeric(tracefile_mod_wide$`simulated_data:run_phylo_ML_supermatrix_aln_sim`)\n",
    "\n",
    "    # retain only the family name, BIgTree ML running time and SM + ML running time\n",
    "    tracefile_mod_wide = tracefile_mod_wide[,c(1,3,5)]\n",
    "\n",
    "    # call the second column name Big_Tree and set their value to numeric type\n",
    "    colnames(tracefile_mod_wide)[2] = \"Big_Tree\"\n",
    "    tracefile_mod_wide$Big_Tree = as.numeric(tracefile_mod_wide$Big_Tree)\n",
    "\n",
    "    # adding a first column that groups families by the run of the main.nf\n",
    "    tracefile_mod_wide$family_group = rep(basename(run), 100)\n",
    "\n",
    "    # add tables toghether raw-wise\n",
    "    if(run == run_dirs[1]) {\n",
    "        fin_tracefile_mod_wide = tracefile_mod_wide\n",
    "    } else {\n",
    "        fin_tracefile_mod_wide = rbind(fin_tracefile_mod_wide, tracefile_mod_wide)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "d6a00cd4-31ce-440c-8528-94805af06047",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>Big_Tree</dt><dd>638.12</dd><dt>Supermatrix_ML</dt><dd>67.76</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[Big\\textbackslash{}\\_Tree] 638.12\n",
       "\\item[Supermatrix\\textbackslash{}\\_ML] 67.76\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "Big_Tree\n",
       ":   638.12Supermatrix_ML\n",
       ":   67.76\n",
       "\n"
      ],
      "text/plain": [
       "      Big_Tree Supermatrix_ML \n",
       "        638.12          67.76 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "588.3"
      ],
      "text/latex": [
       "588.3"
      ],
      "text/markdown": [
       "588.3"
      ],
      "text/plain": [
       "[1] 588.3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "63.85"
      ],
      "text/latex": [
       "63.85"
      ],
      "text/markdown": [
       "63.85"
      ],
      "text/plain": [
       "[1] 63.85"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# TODO make nicer and update with whatever is changed above\n",
    "\n",
    "# make the means of all (500) family running time for the BigTree and ML + SM. Bring them to seconds (/1000) and round the digits to 2\n",
    "round(colMeans(fin_tracefile_mod_wide[, c(2, 3)])/1000,digits = 2)\n",
    "\n",
    "# compute the median for the BigTree approach and the  ML + SM.\n",
    "round(median(fin_tracefile_mod_wide$Big_Tree)/1000,digits = 2)\n",
    "round(median(fin_tracefile_mod_wide$Supermatrix_ML)/1000,digits = 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b79de7a-f8b9-4dc7-8fc6-232cf86be231",
   "metadata": {},
   "source": [
    "### Figure 2\n",
    "\n",
    "TODO add chapter link and description if necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "e1ff4f0e-ec70-4989-89c7-f007d09969c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using name, family_group as id variables\n",
      "\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method (SM + ML and Bigtree)\n",
    "#\n",
    "\n",
    "# It takes a table of all families as rows (500) and it goes to a table of 1000 (500 BigTree + 500 SM + ML) where the running times instead of being separated into two different coluns now they are in the same one.\n",
    "# name + group columns will work as unique ID.  group is the directory name of the main.nf run/s, while name is the name of the family. \n",
    "fin_tracefile_mod_wide_melted           = melt(fin_tracefile_mod_wide)\n",
    "\n",
    "# transform running times from millisecond to seconds. And rename the column names.\n",
    "fin_tracefile_mod_wide_melted$value     = fin_tracefile_mod_wide_melted$value / 1000\n",
    "colnames(fin_tracefile_mod_wide_melted) = c(\"name\", \"family_group\", \"Method\", \"Time\")\n",
    "\n",
    "# crate a table with two boxplots (x axis) with Method columns as values and on the y the running times.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel. \n",
    "p = ggplot(fin_tracefile_mod_wide_melted,aes(x=Method, y=Time, fill=Method)) +\n",
    "    geom_boxplot(show.legend = FALSE) +\n",
    "    theme_light() +\n",
    "    ylab(\"CPU time (sec)\") + xlab(\"Method\")\n",
    "ggsave(filename=\"figures/CPU_time_per_method.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1291310-3aea-4f2c-b836-0f4587c40f8c",
   "metadata": {},
   "source": [
    "### todo make the follewing plot block either disappear or better."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ba993fba-7f4b-48aa-9b03-775093172898",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO make this into the section of custom functions.\n",
    "# Function to assign new values based on a mapping provided by the user\n",
    "assign_mapped_values <- function(df, column_name, mapping, new_col_name) {\n",
    "  # Check if mapping is a named vector or a list\n",
    "  if (!is.vector(mapping) || is.null(names(mapping))) {\n",
    "    stop(\"Mapping must be a named vector or list.\")\n",
    "  }\n",
    "  \n",
    "  # Get the unique values from the specified column\n",
    "  unique_values = unique(df[[column_name]])\n",
    "  \n",
    "  # Check if the mapping contains all unique values\n",
    "  if (!all(unique_values %in% names(mapping))) {\n",
    "    stop(\"Mapping must include all unique values from the specified column.\")\n",
    "  }\n",
    "  \n",
    "  # Initialize a vector to store the assigned values\n",
    "  assigned_values = vector(\"character\", nrow(df))\n",
    "  \n",
    "  # Assign values based on the mapping\n",
    "  for (i in seq_along(unique_values)) {\n",
    "    assigned_values[df[[column_name]] == unique_values[i]] <- mapping[names(mapping) == unique_values[i]]\n",
    "  }\n",
    "  \n",
    "  # Add the assigned values to the original dataframe\n",
    "  df[[new_col_name]] = assigned_values\n",
    "  \n",
    "  return(df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4568c025-fffa-45df-bfff-086d67ad2a53",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using name, family_group, length as id variables\n",
      "\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method (SM + ML and Bigtree) divided per group (by sequence length)\n",
    "#\n",
    "\n",
    "# assign to each family_group a mapping values  (as new column). basically insert the length of the sequences in each family as an additional column in the dataframe.\n",
    "# Remember each family in the same group have the same length of sequence generated, the MSA though might differ in length due to gaps.\n",
    "mapping = c(\"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species\" = \"group1\\nlen = 32\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_22th_and_23th_50_fams_25_species\" = \"group2\\nlen = 46\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_20th_22th_and_23th_50_fams_25_species\" = \"group3\\nlen = 50\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_17th_20th_22th_and_23th_50_fams_25_species\" = \"group4\\nlen = 60\",\n",
    "            \"rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\" = \"group5\\nlen = 82\")\n",
    "fin_tracefile_mod_reassigned = assign_mapped_values(fin_tracefile_mod_wide, \"family_group\", mapping, \"length\")\n",
    "\n",
    "# reorder the columns to have the numerical values at the end\n",
    "fin_tracefile_mod_reassigned = fin_tracefile_mod_reassigned[, c(\"name\", \"family_group\", \"length\", \"Big_Tree\", \"Supermatrix_ML\")]\n",
    "\n",
    "# It takes a table of all families as rows (500) and it goes to a table of 1000 (500 BigTree + 500 SM + ML) where the running times instead of being separated into two different coluns now they are in the same one.\n",
    "# name + group columns will work as unique ID.  group is the directory name of the main.nf run/s, while name is the name of the family. \n",
    "fin_tracefile_mod_melted           = melt(fin_tracefile_mod_reassigned)\n",
    "\n",
    "\n",
    "# transform running times from millisecond to seconds. And rename the column names.\n",
    "fin_tracefile_mod_melted$value     = fin_tracefile_mod_melted$value / 1000\n",
    "colnames(fin_tracefile_mod_melted) = c(\"name\", \"family_group\", \"length\", \"Method\", \"Time\")\n",
    "\n",
    "# create a table with 5 pairs of boxplot. Group1, 2.. ecc.. each group has 100 families and two running time (realtime) mesuraments for the BigTree and SM + ML approaches.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel.\n",
    "p = ggplot(fin_tracefile_mod_melted, aes(x=length, y=Time, fill=Method)) +\n",
    "    geom_boxplot(show.legend = TRUE) +\n",
    "    theme_light() +\n",
    "    labs(x = \"sequence length\",\n",
    "       y = \"CPU time (sec)\",\n",
    "       fill = \"Method\")\n",
    "ggsave(filename=\"figures/CPU_time_per_group_BigTree_SM_ML.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53a77a3f-22e3-4136-b883-32db83aaa230",
   "metadata": {},
   "source": [
    "### TODO make the following better or remove it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "a73a53a9-f1a1-40c9-84e9-09965fe17f26",
   "metadata": {},
   "outputs": [],
   "source": [
    "# update the trace matrix with the length of the input msa\n",
    "\n",
    "fin_tracefile_mod_with_len = fin_tracefile_mod_wide\n",
    "\n",
    "# for each family go and read the input MSA (input of main.nf) to get length of MSA (number of columns).\n",
    "# Iterate over each row of the dataframe\n",
    "for (i in 1:nrow(fin_tracefile_mod_with_len)) {\n",
    "    # Construct the full file path using group (directory) and name (filename) to read the input msa\n",
    "    # TODO make it not depend on the rerun_bla_bla direcrtory\n",
    "    input_msa_path = file.path(paste0(parent_of_all_dir, \"rerun_25_species_corrected_branches\"), fin_tracefile_mod_with_len$family_group[i], paste0(gsub(\"[()]\", \"\", fin_tracefile_mod_with_len$name[i]), \".ma\"))\n",
    "\n",
    "    # read the msa in fasta format and get the length of the first sequence (aka length of the alignment)\n",
    "    msa_file = readAAStringSet(input_msa_path)\n",
    "    msa_len  = length(msa_file[[1]])\n",
    "    \n",
    "    # add the length to the table \n",
    "    fin_tracefile_mod_with_len$length[i] = msa_len\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "1b9d82dd-7856-44ec-99f5-60086ee8b8d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 500 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>name</th><th scope=col>family_group</th><th scope=col>length</th><th scope=col>fold_ratio</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_001.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>87</td><td> 9.424765</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_002.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>90</td><td> 6.662016</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_003.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>89</td><td>13.213900</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_004.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>95</td><td> 8.258667</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_005.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>87</td><td>12.687788</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_006.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>87</td><td> 8.113131</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_007.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>86</td><td> 8.950375</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_008.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>87</td><td> 9.403475</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_009.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>87</td><td> 7.888725</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_010.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>85</td><td>10.139597</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_011.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>88</td><td> 8.198670</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_012.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>85</td><td> 7.604240</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_013.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>87</td><td> 9.985599</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_014.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>88</td><td>10.144638</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_015.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>90</td><td>11.592057</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_016.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>86</td><td> 6.645899</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_017.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>94</td><td> 9.688913</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_018.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>90</td><td> 7.567856</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_019.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>89</td><td>12.502437</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_020.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>92</td><td> 8.820057</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_021.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>87</td><td> 5.386698</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_022.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>87</td><td> 7.052036</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_023.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>90</td><td>12.585050</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_024.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>86</td><td> 7.787928</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_025.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>89</td><td>10.458884</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_026.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>90</td><td> 6.884656</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_027.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>87</td><td> 5.919583</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_028.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>97</td><td> 8.038591</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_029.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>84</td><td> 8.874304</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_030.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species</td><td>88</td><td> 9.217389</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_071.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>10.576043</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_072.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td> 7.840660</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_073.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td> 6.745972</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_074.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>10.126736</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_075.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td> 8.249280</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_076.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>12.240216</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_077.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td> 7.810175</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_078.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>12.059140</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_079.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>11.410401</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_080.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>10.793529</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_081.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>14.853009</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_082.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>14.766811</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_083.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td> 9.990763</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_084.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>10.008237</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_085.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>12.210939</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_086.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td> 7.173738</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_087.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>10.241020</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_088.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>10.609300</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_089.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td> 5.875330</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_090.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>10.152796</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_091.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>12.802404</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_092.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td> 7.955354</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_093.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td> 9.076086</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_094.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>10.801513</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_095.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td> 9.462174</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_096.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td> 8.487253</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_097.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td>12.404810</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_098.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td> 7.811825</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_099.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td> 4.518833</td></tr>\n",
       "\t<tr><td>(avg_branchlen_0.7_protein_family_100.sub)</td><td>rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species</td><td>32</td><td> 9.332632</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 500 × 4\n",
       "\\begin{tabular}{llll}\n",
       " name & family\\_group & length & fold\\_ratio\\\\\n",
       " <chr> & <chr> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_001.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 87 &  9.424765\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_002.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 90 &  6.662016\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_003.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 89 & 13.213900\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_004.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 95 &  8.258667\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_005.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 87 & 12.687788\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_006.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 87 &  8.113131\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_007.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 86 &  8.950375\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_008.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 87 &  9.403475\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_009.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 87 &  7.888725\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_010.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 85 & 10.139597\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_011.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 88 &  8.198670\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_012.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 85 &  7.604240\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_013.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 87 &  9.985599\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_014.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 88 & 10.144638\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_015.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 90 & 11.592057\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_016.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 86 &  6.645899\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_017.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 94 &  9.688913\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_018.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 90 &  7.567856\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_019.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 89 & 12.502437\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_020.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 92 &  8.820057\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_021.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 87 &  5.386698\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_022.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 87 &  7.052036\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_023.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 90 & 12.585050\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_024.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 86 &  7.787928\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_025.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 89 & 10.458884\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_026.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 90 &  6.884656\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_027.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 87 &  5.919583\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_028.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 97 &  8.038591\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_029.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 84 &  8.874304\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_030.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_11th\\_17th\\_20th\\_22th\\_and\\_23th\\_50\\_fams\\_25\\_species & 88 &  9.217389\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_071.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 10.576043\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_072.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 &  7.840660\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_073.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 &  6.745972\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_074.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 10.126736\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_075.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 &  8.249280\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_076.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 12.240216\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_077.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 &  7.810175\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_078.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 12.059140\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_079.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 11.410401\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_080.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 10.793529\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_081.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 14.853009\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_082.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 14.766811\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_083.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 &  9.990763\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_084.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 10.008237\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_085.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 12.210939\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_086.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 &  7.173738\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_087.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 10.241020\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_088.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 10.609300\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_089.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 &  5.875330\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_090.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 10.152796\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_091.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 12.802404\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_092.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 &  7.955354\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_093.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 &  9.076086\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_094.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 10.801513\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_095.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 &  9.462174\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_096.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 &  8.487253\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_097.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 & 12.404810\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_098.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 &  7.811825\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_099.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 &  4.518833\\\\\n",
       "\t (avg\\_branchlen\\_0.7\\_protein\\_family\\_100.sub) & rerun\\_sim\\_b\\_factor\\_3.5\\_avg\\_branchlen\\_0.7\\_from\\_23th\\_50\\_fams\\_25\\_species & 32 &  9.332632\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 500 × 4\n",
       "\n",
       "| name &lt;chr&gt; | family_group &lt;chr&gt; | length &lt;int&gt; | fold_ratio &lt;dbl&gt; |\n",
       "|---|---|---|---|\n",
       "| (avg_branchlen_0.7_protein_family_001.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 87 |  9.424765 |\n",
       "| (avg_branchlen_0.7_protein_family_002.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 90 |  6.662016 |\n",
       "| (avg_branchlen_0.7_protein_family_003.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 89 | 13.213900 |\n",
       "| (avg_branchlen_0.7_protein_family_004.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 95 |  8.258667 |\n",
       "| (avg_branchlen_0.7_protein_family_005.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 87 | 12.687788 |\n",
       "| (avg_branchlen_0.7_protein_family_006.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 87 |  8.113131 |\n",
       "| (avg_branchlen_0.7_protein_family_007.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 86 |  8.950375 |\n",
       "| (avg_branchlen_0.7_protein_family_008.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 87 |  9.403475 |\n",
       "| (avg_branchlen_0.7_protein_family_009.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 87 |  7.888725 |\n",
       "| (avg_branchlen_0.7_protein_family_010.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 85 | 10.139597 |\n",
       "| (avg_branchlen_0.7_protein_family_011.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 88 |  8.198670 |\n",
       "| (avg_branchlen_0.7_protein_family_012.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 85 |  7.604240 |\n",
       "| (avg_branchlen_0.7_protein_family_013.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 87 |  9.985599 |\n",
       "| (avg_branchlen_0.7_protein_family_014.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 88 | 10.144638 |\n",
       "| (avg_branchlen_0.7_protein_family_015.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 90 | 11.592057 |\n",
       "| (avg_branchlen_0.7_protein_family_016.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 86 |  6.645899 |\n",
       "| (avg_branchlen_0.7_protein_family_017.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 94 |  9.688913 |\n",
       "| (avg_branchlen_0.7_protein_family_018.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 90 |  7.567856 |\n",
       "| (avg_branchlen_0.7_protein_family_019.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 89 | 12.502437 |\n",
       "| (avg_branchlen_0.7_protein_family_020.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 92 |  8.820057 |\n",
       "| (avg_branchlen_0.7_protein_family_021.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 87 |  5.386698 |\n",
       "| (avg_branchlen_0.7_protein_family_022.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 87 |  7.052036 |\n",
       "| (avg_branchlen_0.7_protein_family_023.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 90 | 12.585050 |\n",
       "| (avg_branchlen_0.7_protein_family_024.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 86 |  7.787928 |\n",
       "| (avg_branchlen_0.7_protein_family_025.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 89 | 10.458884 |\n",
       "| (avg_branchlen_0.7_protein_family_026.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 90 |  6.884656 |\n",
       "| (avg_branchlen_0.7_protein_family_027.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 87 |  5.919583 |\n",
       "| (avg_branchlen_0.7_protein_family_028.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 97 |  8.038591 |\n",
       "| (avg_branchlen_0.7_protein_family_029.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 84 |  8.874304 |\n",
       "| (avg_branchlen_0.7_protein_family_030.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species | 88 |  9.217389 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| (avg_branchlen_0.7_protein_family_071.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 10.576043 |\n",
       "| (avg_branchlen_0.7_protein_family_072.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 |  7.840660 |\n",
       "| (avg_branchlen_0.7_protein_family_073.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 |  6.745972 |\n",
       "| (avg_branchlen_0.7_protein_family_074.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 10.126736 |\n",
       "| (avg_branchlen_0.7_protein_family_075.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 |  8.249280 |\n",
       "| (avg_branchlen_0.7_protein_family_076.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 12.240216 |\n",
       "| (avg_branchlen_0.7_protein_family_077.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 |  7.810175 |\n",
       "| (avg_branchlen_0.7_protein_family_078.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 12.059140 |\n",
       "| (avg_branchlen_0.7_protein_family_079.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 11.410401 |\n",
       "| (avg_branchlen_0.7_protein_family_080.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 10.793529 |\n",
       "| (avg_branchlen_0.7_protein_family_081.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 14.853009 |\n",
       "| (avg_branchlen_0.7_protein_family_082.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 14.766811 |\n",
       "| (avg_branchlen_0.7_protein_family_083.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 |  9.990763 |\n",
       "| (avg_branchlen_0.7_protein_family_084.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 10.008237 |\n",
       "| (avg_branchlen_0.7_protein_family_085.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 12.210939 |\n",
       "| (avg_branchlen_0.7_protein_family_086.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 |  7.173738 |\n",
       "| (avg_branchlen_0.7_protein_family_087.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 10.241020 |\n",
       "| (avg_branchlen_0.7_protein_family_088.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 10.609300 |\n",
       "| (avg_branchlen_0.7_protein_family_089.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 |  5.875330 |\n",
       "| (avg_branchlen_0.7_protein_family_090.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 10.152796 |\n",
       "| (avg_branchlen_0.7_protein_family_091.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 12.802404 |\n",
       "| (avg_branchlen_0.7_protein_family_092.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 |  7.955354 |\n",
       "| (avg_branchlen_0.7_protein_family_093.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 |  9.076086 |\n",
       "| (avg_branchlen_0.7_protein_family_094.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 10.801513 |\n",
       "| (avg_branchlen_0.7_protein_family_095.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 |  9.462174 |\n",
       "| (avg_branchlen_0.7_protein_family_096.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 |  8.487253 |\n",
       "| (avg_branchlen_0.7_protein_family_097.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 | 12.404810 |\n",
       "| (avg_branchlen_0.7_protein_family_098.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 |  7.811825 |\n",
       "| (avg_branchlen_0.7_protein_family_099.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 |  4.518833 |\n",
       "| (avg_branchlen_0.7_protein_family_100.sub) | rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species | 32 |  9.332632 |\n",
       "\n"
      ],
      "text/plain": [
       "    name                                      \n",
       "1   (avg_branchlen_0.7_protein_family_001.sub)\n",
       "2   (avg_branchlen_0.7_protein_family_002.sub)\n",
       "3   (avg_branchlen_0.7_protein_family_003.sub)\n",
       "4   (avg_branchlen_0.7_protein_family_004.sub)\n",
       "5   (avg_branchlen_0.7_protein_family_005.sub)\n",
       "6   (avg_branchlen_0.7_protein_family_006.sub)\n",
       "7   (avg_branchlen_0.7_protein_family_007.sub)\n",
       "8   (avg_branchlen_0.7_protein_family_008.sub)\n",
       "9   (avg_branchlen_0.7_protein_family_009.sub)\n",
       "10  (avg_branchlen_0.7_protein_family_010.sub)\n",
       "11  (avg_branchlen_0.7_protein_family_011.sub)\n",
       "12  (avg_branchlen_0.7_protein_family_012.sub)\n",
       "13  (avg_branchlen_0.7_protein_family_013.sub)\n",
       "14  (avg_branchlen_0.7_protein_family_014.sub)\n",
       "15  (avg_branchlen_0.7_protein_family_015.sub)\n",
       "16  (avg_branchlen_0.7_protein_family_016.sub)\n",
       "17  (avg_branchlen_0.7_protein_family_017.sub)\n",
       "18  (avg_branchlen_0.7_protein_family_018.sub)\n",
       "19  (avg_branchlen_0.7_protein_family_019.sub)\n",
       "20  (avg_branchlen_0.7_protein_family_020.sub)\n",
       "21  (avg_branchlen_0.7_protein_family_021.sub)\n",
       "22  (avg_branchlen_0.7_protein_family_022.sub)\n",
       "23  (avg_branchlen_0.7_protein_family_023.sub)\n",
       "24  (avg_branchlen_0.7_protein_family_024.sub)\n",
       "25  (avg_branchlen_0.7_protein_family_025.sub)\n",
       "26  (avg_branchlen_0.7_protein_family_026.sub)\n",
       "27  (avg_branchlen_0.7_protein_family_027.sub)\n",
       "28  (avg_branchlen_0.7_protein_family_028.sub)\n",
       "29  (avg_branchlen_0.7_protein_family_029.sub)\n",
       "30  (avg_branchlen_0.7_protein_family_030.sub)\n",
       "⋮   ⋮                                         \n",
       "471 (avg_branchlen_0.7_protein_family_071.sub)\n",
       "472 (avg_branchlen_0.7_protein_family_072.sub)\n",
       "473 (avg_branchlen_0.7_protein_family_073.sub)\n",
       "474 (avg_branchlen_0.7_protein_family_074.sub)\n",
       "475 (avg_branchlen_0.7_protein_family_075.sub)\n",
       "476 (avg_branchlen_0.7_protein_family_076.sub)\n",
       "477 (avg_branchlen_0.7_protein_family_077.sub)\n",
       "478 (avg_branchlen_0.7_protein_family_078.sub)\n",
       "479 (avg_branchlen_0.7_protein_family_079.sub)\n",
       "480 (avg_branchlen_0.7_protein_family_080.sub)\n",
       "481 (avg_branchlen_0.7_protein_family_081.sub)\n",
       "482 (avg_branchlen_0.7_protein_family_082.sub)\n",
       "483 (avg_branchlen_0.7_protein_family_083.sub)\n",
       "484 (avg_branchlen_0.7_protein_family_084.sub)\n",
       "485 (avg_branchlen_0.7_protein_family_085.sub)\n",
       "486 (avg_branchlen_0.7_protein_family_086.sub)\n",
       "487 (avg_branchlen_0.7_protein_family_087.sub)\n",
       "488 (avg_branchlen_0.7_protein_family_088.sub)\n",
       "489 (avg_branchlen_0.7_protein_family_089.sub)\n",
       "490 (avg_branchlen_0.7_protein_family_090.sub)\n",
       "491 (avg_branchlen_0.7_protein_family_091.sub)\n",
       "492 (avg_branchlen_0.7_protein_family_092.sub)\n",
       "493 (avg_branchlen_0.7_protein_family_093.sub)\n",
       "494 (avg_branchlen_0.7_protein_family_094.sub)\n",
       "495 (avg_branchlen_0.7_protein_family_095.sub)\n",
       "496 (avg_branchlen_0.7_protein_family_096.sub)\n",
       "497 (avg_branchlen_0.7_protein_family_097.sub)\n",
       "498 (avg_branchlen_0.7_protein_family_098.sub)\n",
       "499 (avg_branchlen_0.7_protein_family_099.sub)\n",
       "500 (avg_branchlen_0.7_protein_family_100.sub)\n",
       "    family_group                                                                                 \n",
       "1   rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "2   rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "3   rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "4   rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "5   rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "6   rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "7   rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "8   rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "9   rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "10  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "11  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "12  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "13  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "14  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "15  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "16  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "17  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "18  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "19  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "20  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "21  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "22  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "23  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "24  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "25  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "26  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "27  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "28  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "29  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "30  rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_11th_17th_20th_22th_and_23th_50_fams_25_species\n",
       "⋮   ⋮                                                                                            \n",
       "471 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "472 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "473 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "474 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "475 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "476 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "477 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "478 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "479 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "480 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "481 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "482 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "483 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "484 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "485 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "486 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "487 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "488 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "489 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "490 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "491 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "492 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "493 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "494 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "495 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "496 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "497 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "498 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "499 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "500 rerun_sim_b_factor_3.5_avg_branchlen_0.7_from_23th_50_fams_25_species                        \n",
       "    length fold_ratio\n",
       "1   87      9.424765 \n",
       "2   90      6.662016 \n",
       "3   89     13.213900 \n",
       "4   95      8.258667 \n",
       "5   87     12.687788 \n",
       "6   87      8.113131 \n",
       "7   86      8.950375 \n",
       "8   87      9.403475 \n",
       "9   87      7.888725 \n",
       "10  85     10.139597 \n",
       "11  88      8.198670 \n",
       "12  85      7.604240 \n",
       "13  87      9.985599 \n",
       "14  88     10.144638 \n",
       "15  90     11.592057 \n",
       "16  86      6.645899 \n",
       "17  94      9.688913 \n",
       "18  90      7.567856 \n",
       "19  89     12.502437 \n",
       "20  92      8.820057 \n",
       "21  87      5.386698 \n",
       "22  87      7.052036 \n",
       "23  90     12.585050 \n",
       "24  86      7.787928 \n",
       "25  89     10.458884 \n",
       "26  90      6.884656 \n",
       "27  87      5.919583 \n",
       "28  97      8.038591 \n",
       "29  84      8.874304 \n",
       "30  88      9.217389 \n",
       "⋮   ⋮      ⋮         \n",
       "471 32     10.576043 \n",
       "472 32      7.840660 \n",
       "473 32      6.745972 \n",
       "474 32     10.126736 \n",
       "475 32      8.249280 \n",
       "476 32     12.240216 \n",
       "477 32      7.810175 \n",
       "478 32     12.059140 \n",
       "479 32     11.410401 \n",
       "480 32     10.793529 \n",
       "481 32     14.853009 \n",
       "482 32     14.766811 \n",
       "483 32      9.990763 \n",
       "484 32     10.008237 \n",
       "485 32     12.210939 \n",
       "486 32      7.173738 \n",
       "487 32     10.241020 \n",
       "488 32     10.609300 \n",
       "489 32      5.875330 \n",
       "490 32     10.152796 \n",
       "491 32     12.802404 \n",
       "492 32      7.955354 \n",
       "493 32      9.076086 \n",
       "494 32     10.801513 \n",
       "495 32      9.462174 \n",
       "496 32      8.487253 \n",
       "497 32     12.404810 \n",
       "498 32      7.811825 \n",
       "499 32      4.518833 \n",
       "500 32      9.332632 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cpu time vs Method (SM + ML and Bigtree) log ratio plot of each family (x axis number columns BigMSA) \n",
    "# \n",
    "\n",
    "# Create a new column with the division of Big_Tree by Supermatrix_ML running times for each family (500)\n",
    "# and then remove the original columns Big_Tree Supermatrix_ML\n",
    "fin_tracefile_mod_ratio = fin_tracefile_mod_with_len %>%\n",
    "  mutate(fold_ratio = Big_Tree / Supermatrix_ML) %>%\n",
    "  select(-Big_Tree, -Supermatrix_ML)\n",
    "\n",
    "# create a table with on x axis the length of the MSA (n columns)  and on y the fold ratio (BigTree / SM + ML) of running times. The plot will have now 500 points.\n",
    "# the plot is not showed because this file was executed on a cluster and rendering of plots creates conflicts with the kernel.\n",
    "p = ggplot(fin_tracefile_mod_ratio, aes(x=length, y=fold_ratio)) +\n",
    "    geom_point(color = 'blue', size = 2, alpha = 0.6) +  # Add points\n",
    "    theme_light() +\n",
    "    labs(x = \"MSA length (n# columns)\",\n",
    "       y = \"fold ratio (BigTree / Supermatrix_ML)\")\n",
    "ggsave(filename=\"figures/CPU_time_scatterplot_msa_len_BigTree_SM_ML.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1a03b1b-cfef-4abf-a0f0-d0fd195c15bf",
   "metadata": {},
   "source": [
    "### TODO the below was just a 10 familoy trial make it go away "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "873d9c1c-ab54-40b6-aed9-90caa7d43601",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO make this into the section of custom functions.\n",
    "# Function to convert to milliseconds\n",
    "convert_to_ms <- function(time_str) {\n",
    "    # Extract components using regular expressions\n",
    "    millis = \"0\"\n",
    "    seconds = \"0\"\n",
    "    minutes = \"0\"\n",
    "    if (str_detect(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=ms)\")) {\n",
    "        millis <- str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=ms)\")\n",
    "    } else {\n",
    "        seconds <- str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=s)\")\n",
    "        minutes <- str_extract(time_str, \"\\\\d+(\\\\.\\\\d+)?(?=m)\")\n",
    "    }\n",
    "    \n",
    "    # Convert each component to milliseconds (coercing to numeric where needed)\n",
    "    seconds_ms <- as.numeric(seconds) * 1000\n",
    "    minutes_ms <- as.numeric(minutes) * 60 * 1000\n",
    "    millis_ms <- as.numeric(millis)\n",
    "    \n",
    "    # Sum all components and handle NAs by treating them as 0\n",
    "    total_ms <- sum(c(seconds_ms, minutes_ms, millis_ms), na.rm = TRUE)\n",
    "    return(total_ms)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "9a84bc61-7c7b-4805-a7dc-47071e51fc39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"results\"\n"
     ]
    }
   ],
   "source": [
    "for (run in run_dirs) {\n",
    "    print(basename(run))\n",
    "\n",
    "    # each main.nf output run has a trace file repoprting various information on the run itself\n",
    "    tracefile_path = file.path(run, \"pipeline_info/execution_trace_2024-09-27_17-39-02.txt\")\n",
    "    tracefile      = read.table(tracefile_path, sep = \"\\t\", header = TRUE)\n",
    "\n",
    "    # here we parse the trace file to retain only the processes that we want to take the running time (realtime) from\n",
    "    tracefile_mod = tracefile[grep(\"simulated_data:run_phylo_ML_supermatrix_aln_sim|simulated_data:only_concatenate_aln_sim|simulated_data:run_phylo_ML_full_aln_sim\", tracefile$name), c(\"name\", \"realtime\")]\n",
    "\n",
    "    # Use strsplit to split the 'name' column by whitespace\n",
    "    split_name    = strsplit(as.character(tracefile_mod$name), \" \")\n",
    "\n",
    "    # Create new columns\n",
    "    tracefile_mod$name       = sapply(split_name, `[`, 2) # First part \n",
    "    tracefile_mod$native_id  = sapply(split_name, `[`, 1) # Second part \n",
    "\n",
    "    # Reorder columns so 'native_id' becomes the second column\n",
    "    tracefile_mod = tracefile_mod[, c(\"name\", \"native_id\", \"realtime\")]\n",
    "\n",
    "    # trasform from human readable time to millisecond time\n",
    "    tracefile_mod$realtime = sapply(tracefile_mod$realtime, convert_to_ms)\n",
    "\n",
    "    # transforming the above table to one that have the process names (simulated_data:run_phylo_ML_supermatrix_aln_sim ecc..) as column names, \n",
    "    # the rows will be still the family names (working as ID) and the values in each column will be the running times associated to the column name (process name) for that given family.\n",
    "    tracefile_mod_wide = spread(tracefile_mod, native_id, realtime)\n",
    "    \n",
    "    # creating a new last column called Supermatrix_ML that has aas values the running times of concatenation plus ML tree computation (on concatenated MSA) \n",
    "    tracefile_mod_wide$Supermatrix_ML = as.numeric(tracefile_mod_wide$`simulated_data:only_concatenate_aln_sim`) + as.numeric(tracefile_mod_wide$`simulated_data:run_phylo_ML_supermatrix_aln_sim`)\n",
    "\n",
    "    # retain only the family name, BIgTree ML running time and SM + ML running time\n",
    "    tracefile_mod_wide = tracefile_mod_wide[,c(1,3,5)]\n",
    "\n",
    "    # call the second column name Big_Tree and set their value to numeric type\n",
    "    colnames(tracefile_mod_wide)[2] = \"Big_Tree\"\n",
    "    tracefile_mod_wide$Big_Tree = as.numeric(tracefile_mod_wide$Big_Tree)\n",
    "\n",
    "    # adding a first column that groups families by the run of the main.nf\n",
    "    tracefile_mod_wide$group = rep(basename(run), 10)\n",
    "\n",
    "    # add tables toghether raw-wise\n",
    "    if(run == run_dirs[1]) {\n",
    "        fin_tracefile_mod_wide = tracefile_mod_wide\n",
    "    } else {\n",
    "        fin_tracefile_mod_wide = rbind(fin_tracefile_mod_wide, tracefile_mod_wide)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "ee66b8ad-a4df-4ad9-bdf2-7f8d7a13c8c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>Big_Tree</dt><dd>195.4</dd><dt>Supermatrix_ML</dt><dd>23.47</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[Big\\textbackslash{}\\_Tree] 195.4\n",
       "\\item[Supermatrix\\textbackslash{}\\_ML] 23.47\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "Big_Tree\n",
       ":   195.4Supermatrix_ML\n",
       ":   23.47\n",
       "\n"
      ],
      "text/plain": [
       "      Big_Tree Supermatrix_ML \n",
       "        195.40          23.47 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "196"
      ],
      "text/latex": [
       "196"
      ],
      "text/markdown": [
       "196"
      ],
      "text/plain": [
       "[1] 196"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "21.87"
      ],
      "text/latex": [
       "21.87"
      ],
      "text/markdown": [
       "21.87"
      ],
      "text/plain": [
       "[1] 21.87"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# make the means of all (500) family running time for the BigTree and ML + SM. Bring them to seconds (/1000) and round the digits to 2\n",
    "round(colMeans(fin_tracefile_mod_wide[, c(2, 3)])/1000,digits = 2)\n",
    "\n",
    "# compute the median for the BigTree approach and the  ML + SM.\n",
    "round(median(fin_tracefile_mod_wide$Big_Tree)/1000,digits = 2)\n",
    "round(median(fin_tracefile_mod_wide$Supermatrix_ML)/1000,digits = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "ed1275aa-e0bd-4628-bab2-5eb5a54271ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using name, group as id variables\n",
      "\n",
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "\n",
    "fin_tracefile_mod_wide_melted           = melt(fin_tracefile_mod_wide)\n",
    "\n",
    "fin_tracefile_mod_wide_melted$value     = fin_tracefile_mod_wide_melted$value / 1000\n",
    "colnames(fin_tracefile_mod_wide_melted) = c(\"name\",\"group\",\"Method\",\"Time\")\n",
    "\n",
    "p = ggplot(fin_tracefile_mod_wide_melted,aes(x=Method,y=Time,fill=Method)) +\n",
    "    geom_boxplot(show.legend = FALSE) +\n",
    "    theme_light() +\n",
    "    ylab(\"CPU time (sec)\") + xlab(\"Method\")\n",
    "ggsave(filename=\"CPU_time_per_method_10_fam.png\", plot=p, dpi = \"retina\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31c3465f-5bf7-465b-bf62-e004f9441080",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
